
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Custom Bi-Objective Germplasm Panel Reduction Pareto Frontier Visualization &#8212; pybrops 0.9.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/wider_borders.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/examples/pareto_frontier_visualization/germplasm_panel_reduction';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Single-Trait Phenotypic Selection" href="../breeding_program_simulation/single_trait_phenotypic_selection.html" />
    <link rel="prev" title="Tri-Objective Optimal Contribution Selection Pareto Frontier Visualization" href="triobjective_OCS_pareto_frontier.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">pybrops 0.9.0 documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item"><nav class="bd-links" id="bd-docs-nav" aria-label="Section navigation">
  <p class="bd-links__title" role="heading" aria-level="1">
    Section Navigation
  </p>
  <div class="bd-toc-item navbar-nav">
    <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../examples.html">Examples</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../breeding_program_simulation/single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../breeding_program_simulation/constrained_single_trait_phenotypic_selection.html">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../breeding_program_simulation/constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>

  </div>
</nav>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="custom-bi-objective-germplasm-panel-reduction-pareto-frontier-visualization">
<h1>Custom Bi-Objective Germplasm Panel Reduction Pareto Frontier Visualization<a class="headerlink" href="#custom-bi-objective-germplasm-panel-reduction-pareto-frontier-visualization" title="Permalink to this heading">#</a></h1>
<p>In some breeding and/or germplasm maintenance scenarios, it may be desirable to reduce the number of individuals in a panel to a representative subset and discard the rest. In these types of scenarios we may seek to minimize the number of selected individuals while minimizing the loss of genetic diversity. This type of optimization can be classified as a binary optimization problem where our decision vectors can be represented as <code class="docutils literal notranslate"><span class="pre">1</span></code> or <code class="docutils literal notranslate"><span class="pre">0</span></code> for whether the individual is selected or not selected, respectively. We may desire to see a trade-off frontier of options from which we can make germplasm selection decisions. In the following two examples, we demonstrate how to specify custom optimization problems and visualize the results.</p>
<section id="preliminaries">
<h2>Preliminaries<a class="headerlink" href="#preliminaries" title="Permalink to this heading">#</a></h2>
<section id="loading-required-modules-and-seeding-the-global-prng">
<h3>Loading Required Modules and Seeding the global PRNG<a class="headerlink" href="#loading-required-modules-and-seeding-the-global-prng" title="Permalink to this heading">#</a></h3>
<p>To begin, we import the modules we will be using into the Python namespace. To make our simulation replicable, we set the seed for the simulation using the <code class="docutils literal notranslate"><span class="pre">seed</span></code> function in the <code class="docutils literal notranslate"><span class="pre">pybrops.core.random.prng</span></code> module. This seeds the Python <code class="docutils literal notranslate"><span class="pre">random</span></code> and NumPy <code class="docutils literal notranslate"><span class="pre">numpy.random</span></code> modules with a single seed. Weâ€™ll also want to import  <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> to create images.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import libraries</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span>
<span class="kn">import</span> <span class="nn">pybrops</span>
<span class="kn">from</span> <span class="nn">pybrops.opt.algo.NSGA2BinaryGeneticAlgorithm</span> <span class="kn">import</span> <span class="n">NSGA2BinaryGeneticAlgorithm</span>
<span class="kn">from</span> <span class="nn">pybrops.opt.prob.BinaryProblem</span> <span class="kn">import</span> <span class="n">BinaryProblem</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DenseGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DenseGenotypeMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.DenseMolecularCoancestryMatrix</span> <span class="kn">import</span> <span class="n">DenseMolecularCoancestryMatrix</span>

<span class="c1"># seed python random and numpy random</span>
<span class="n">pybrops</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">prng</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">194711822</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-genotypic-data-from-a-vcf-file">
<h3>Loading Genotypic Data from a VCF File<a class="headerlink" href="#loading-genotypic-data-from-a-vcf-file" title="Permalink to this heading">#</a></h3>
<p>Next, we load our genotypic data from a VCF file named <code class="docutils literal notranslate"><span class="pre">&quot;widiv_2000SNPs.vcf.gz&quot;</span></code> using the <code class="docutils literal notranslate"><span class="pre">from_vcf</span></code> class method in the <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code> class. This creates an unphased genotype matrix with our SNPs coded as 0, 1, and 2. We automatically sort and group the genetic variants in the genotype matrix based on their chromosome assignments and physical positions using the <code class="docutils literal notranslate"><span class="pre">auto_group_vrnt</span> <span class="pre">=</span> <span class="pre">True</span></code> option. For these examples, we select the first 100 individuals to reduce the computational time to run this example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read unphased genetic markers from a vcf file</span>
<span class="n">gmat</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span>
    <span class="s2">&quot;widiv_2000SNPs.vcf.gz&quot;</span><span class="p">,</span> <span class="c1"># file name to load</span>
    <span class="n">auto_group_vrnt</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># automatically sort and group variants</span>
<span class="p">)</span>

<span class="c1"># get the first 100 taxa to keep the problem small</span>
<span class="n">gmat_reduced</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">100</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="scenario-1-minimize-number-of-individuals-and-relatedness">
<h2>Scenario 1: Minimize number of individuals and relatedness<a class="headerlink" href="#scenario-1-minimize-number-of-individuals-and-relatedness" title="Permalink to this heading">#</a></h2>
<p>In this first approach to the germplasm panel reduction problem, we propose two objectives: the number of selected individuals and the relatedness of those individuals. We seek to minimize both objectives.</p>
<section id="defining-a-custom-problem-class-for-scenario-1">
<h3>Defining a Custom Problem Class for Scenario 1<a class="headerlink" href="#defining-a-custom-problem-class-for-scenario-1" title="Permalink to this heading">#</a></h3>
<p>We define a germplasm panel reduction class which inherits from the <code class="docutils literal notranslate"><span class="pre">BinaryProblem</span></code> class since the germplasm panel reduction problem decision space is binary. We override the <code class="docutils literal notranslate"><span class="pre">evalfn</span></code> method and provide code which will calculate our objectives.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define germplasm panel reduction problem definition class</span>
<span class="k">class</span> <span class="nc">GermplasmPanelReduction1</span><span class="p">(</span><span class="n">BinaryProblem</span><span class="p">):</span>
    <span class="c1"># class constructor</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">C</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor for the germplasm panel reduction problem:</span>

<span class="sd">        Objective 1: minimize the number of individuals selected</span>
<span class="sd">        Objective 2: minimize the relatedness of individuals selected</span>
<span class="sd">        Inequality constraint 1: there must be at least one individual selected</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        C : numpy.ndarray</span>
<span class="sd">            Cholesky decomposition (upper triangle matrix) of a kinship matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">C</span> <span class="o">=</span> <span class="n">C</span>
        <span class="n">ndecn</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">C</span><span class="p">)</span>
        <span class="n">decn_space_lower</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ndecn</span><span class="p">)</span>
        <span class="n">decn_space_upper</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ndecn</span><span class="p">)</span>
        <span class="n">decn_space</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">stack</span><span class="p">([</span><span class="n">decn_space_lower</span><span class="p">,</span><span class="n">decn_space_upper</span><span class="p">])</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">ndecn</span> <span class="o">=</span> <span class="n">ndecn</span><span class="p">,</span>
            <span class="n">decn_space</span> <span class="o">=</span> <span class="n">decn_space</span><span class="p">,</span>
            <span class="n">decn_space_lower</span> <span class="o">=</span> <span class="n">decn_space_lower</span><span class="p">,</span>
            <span class="n">decn_space_upper</span> <span class="o">=</span> <span class="n">decn_space_upper</span><span class="p">,</span>
            <span class="n">nobj</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">nineqcv</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>
    <span class="c1">### method required by PyBrOpS interface ###</span>
    <span class="k">def</span> <span class="nf">evalfn</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
            <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluate a candidate solution for the given Problem.</span>

<span class="sd">        This calculates three vectors which are to be minimized:</span>

<span class="sd">        .. math::</span>

<span class="sd">            \\mathbf{v_{obj}} = \\mathbf{w_{obj} \\odot F_{obj}(x)} \\</span>
<span class="sd">            \\mathbf{v_{ineqcv}} = \\mathbf{w_{ineqcv} \\odot G_{ineqcv}(x)} \\</span>
<span class="sd">            \\mathbf{v_{eqcv}} = \\mathbf{w_{eqcv} \\odot H_{eqcv}(x)}</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        x : numpy.ndarray</span>
<span class="sd">            A candidate solution vector of shape ``(ndecn,)``.</span>
<span class="sd">        args : tuple</span>
<span class="sd">            Additional non-keyword arguments.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : tuple</span>
<span class="sd">            A tuple ``(obj, ineqcv, eqcv)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``obj`` is a numpy.ndarray of shape ``(nobj,)`` that contains</span>
<span class="sd">                objective function evaluations.</span>
<span class="sd">            - ``ineqcv`` is a numpy.ndarray of shape ``(nineqcv,)`` that contains</span>
<span class="sd">                inequality constraint violation values.</span>
<span class="sd">            - ``eqcv`` is a numpy.ndarray of shape ``(neqcv,)`` that contains</span>
<span class="sd">                equality constraint violation values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># calculate sum(x)</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="c1"># if sum(x) ~== 0, then set to 1</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="n">f1</span> <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">f1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mf">1e-10</span> <span class="k">else</span> <span class="mf">1.0</span>
        <span class="c1"># scale x to have a sum of 1 (contribution)</span>
        <span class="c1"># (n,) -&gt; (n,)</span>
        <span class="n">contrib</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">denom</span><span class="p">)</span> <span class="o">*</span> <span class="n">x</span>
        <span class="c1"># convert sum(x) to ndarray</span>
        <span class="c1"># scalar -&gt; (1,)</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">f1</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>
        <span class="c1"># calculate mean genomic contribution</span>
        <span class="c1"># (n,n) . (n,) -&gt; (n,)</span>
        <span class="c1"># scalar * (n,) -&gt; (n,)</span>
        <span class="c1"># norm2( (n,), keepdims=True ) -&gt; (1,)</span>
        <span class="n">f2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">C</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">contrib</span><span class="p">),</span> <span class="nb">ord</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">keepdims</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="c1"># concatenate objective function evaluations</span>
        <span class="c1"># (1,) concat (1,) -&gt; (2,)</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj_wt</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">f1</span><span class="p">,</span><span class="n">f2</span><span class="p">])</span>
        <span class="c1"># calculate inequality constraint violations</span>
        <span class="c1"># (1,)</span>
        <span class="n">ineqcv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ineqcv_wt</span> <span class="o">*</span> <span class="p">(</span><span class="n">f1</span> <span class="o">&lt;=</span> <span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="c1"># calculate equality constraint violations</span>
        <span class="n">eqcv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">eqcv_wt</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">neqcv</span><span class="p">)</span>
        <span class="c1"># return (2,), (1,), (0,)</span>
        <span class="k">return</span> <span class="n">obj</span><span class="p">,</span> <span class="n">ineqcv</span><span class="p">,</span> <span class="n">eqcv</span>
    <span class="c1">### method required by PyMOO interface ###</span>
    <span class="c1"># use default ``_evaluate`` method which uses the ``evalfn`` method</span>
</pre></div>
</div>
</section>
<section id="calculating-a-kinship-matrix-and-its-decomposition">
<h3>Calculating a Kinship Matrix and its Decomposition<a class="headerlink" href="#calculating-a-kinship-matrix-and-its-decomposition" title="Permalink to this heading">#</a></h3>
<p>Next, we calculate a kinship matrix for our panel and perform a Cholesky decomposition to get a triangle matrix, which is required by our <code class="docutils literal notranslate"><span class="pre">GermplasmPanelReduction1</span></code> class definition above.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate identity by state coancestry</span>
<span class="n">cmat</span> <span class="o">=</span> <span class="n">DenseMolecularCoancestryMatrix</span><span class="o">.</span><span class="n">from_gmat</span><span class="p">(</span><span class="n">gmat_reduced</span><span class="p">)</span>

<span class="c1"># extract kinship as numpy.ndarray</span>
<span class="n">K</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;kinship&quot;</span><span class="p">)</span>

<span class="c1"># calculate Cholesky decomposition of kinship matrix</span>
<span class="n">C</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">cholesky</span><span class="p">(</span><span class="n">K</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="constructing-a-germplasm-panel-reduction-problem-object">
<h3>Constructing a Germplasm Panel Reduction Problem Object<a class="headerlink" href="#constructing-a-germplasm-panel-reduction-problem-object" title="Permalink to this heading">#</a></h3>
<p>Next, we construct the germplasm panel reduction problem object using the decomposition of our kinship matrix.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct optimization problem</span>
<span class="c1"># use C.T since C is a lower triangle matrix and we need an upper triangle matrix</span>
<span class="n">prob</span> <span class="o">=</span> <span class="n">GermplasmPanelReduction1</span><span class="p">(</span><span class="n">C</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="constructing-a-custom-genetic-algorithm-object">
<h3>Constructing a Custom Genetic Algorithm Object<a class="headerlink" href="#constructing-a-custom-genetic-algorithm-object" title="Permalink to this heading">#</a></h3>
<p>Next, we construct a custom NSGA-II genetic algorithm object to solve this problem.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct NSGA-II object</span>
<span class="c1"># this problem is complex and requires many generations</span>
<span class="n">moea</span> <span class="o">=</span> <span class="n">NSGA2BinaryGeneticAlgorithm</span><span class="p">(</span>
    <span class="n">ngen</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
    <span class="n">pop_size</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="estimating-the-pareto-frontier">
<h3>Estimating the Pareto Frontier<a class="headerlink" href="#estimating-the-pareto-frontier" title="Permalink to this heading">#</a></h3>
<p>After creating an algorithm object, we minimize the germplasm panel reduction problem and receive a solution.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># minimize the optimization problem</span>
<span class="n">soln</span> <span class="o">=</span> <span class="n">moea</span><span class="o">.</span><span class="n">minimize</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="visualizing-the-pareto-frontier-with-matplotlib">
<h3>Visualizing the Pareto Frontier with <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code><a class="headerlink" href="#visualizing-the-pareto-frontier-with-matplotlib" title="Permalink to this heading">#</a></h3>
<p>After optimizing the objectives, we can use <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> or any other plotting packages to visualize the results of the optimization. The code below creates a figure to visualize the estimated Pareto frontier.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">soln</span><span class="o">.</span><span class="n">soln_obj</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">soln</span><span class="o">.</span><span class="n">soln_obj</span><span class="p">[:,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Pareto frontier for number of selected individuals and molecular diversity&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Number of individuals&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Mean molecular coancestry&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;germplasm_panel_reduction1.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">250</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>Below is the resulting figure:</p>
<img alt="../../../_images/germplasm_panel_reduction1.png" src="../../../_images/germplasm_panel_reduction1.png" />
</section>
</section>
<section id="scenario-2-minimize-number-of-individuals-and-loss-of-alleles">
<h2>Scenario 2: Minimize number of individuals and loss of alleles<a class="headerlink" href="#scenario-2-minimize-number-of-individuals-and-loss-of-alleles" title="Permalink to this heading">#</a></h2>
<p>In this second approach to the germplasm panel reduction problem, we propose two objectives: the number of selected individuals (as in the previous scenario) and the number of loci which have become fixed. We seek to minimize both objectives.</p>
<section id="defining-a-custom-problem-class-for-scenario-2">
<h3>Defining a Custom Problem Class for Scenario 2<a class="headerlink" href="#defining-a-custom-problem-class-for-scenario-2" title="Permalink to this heading">#</a></h3>
<p>Once again, we define a germplasm panel reduction class which inherits from the <code class="docutils literal notranslate"><span class="pre">BinaryProblem</span></code> class since the germplasm panel reduction problem decision space is binary. We override the <code class="docutils literal notranslate"><span class="pre">evalfn</span></code> method and provide code which will calculate our objectives.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define germplasm panel reduction problem definition class</span>
<span class="k">class</span> <span class="nc">GermplasmPanelReduction2</span><span class="p">(</span><span class="n">BinaryProblem</span><span class="p">):</span>
    <span class="c1"># class constructor</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor for the germplasm panel reduction problem:</span>

<span class="sd">        Objective 1: minimize the number of individuals selected</span>
<span class="sd">        Objective 2: minimize the number of fixed markers</span>
<span class="sd">        Inequality constraint 1: there must be at least one individual selected</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            Genotype matrix of shape (nindiv,nmarker).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">X</span>
        <span class="n">ndecn</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="n">decn_space_lower</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ndecn</span><span class="p">)</span>
        <span class="n">decn_space_upper</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ndecn</span><span class="p">)</span>
        <span class="n">decn_space</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">stack</span><span class="p">([</span><span class="n">decn_space_lower</span><span class="p">,</span><span class="n">decn_space_upper</span><span class="p">])</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">ndecn</span> <span class="o">=</span> <span class="n">ndecn</span><span class="p">,</span>
            <span class="n">decn_space</span> <span class="o">=</span> <span class="n">decn_space</span><span class="p">,</span>
            <span class="n">decn_space_lower</span> <span class="o">=</span> <span class="n">decn_space_lower</span><span class="p">,</span>
            <span class="n">decn_space_upper</span> <span class="o">=</span> <span class="n">decn_space_upper</span><span class="p">,</span>
            <span class="n">nobj</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">nineqcv</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>
    <span class="c1">### method required by PyBrOpS interface ###</span>
    <span class="k">def</span> <span class="nf">evalfn</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
            <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluate a candidate solution for the given Problem.</span>

<span class="sd">        This calculates three vectors which are to be minimized:</span>

<span class="sd">        .. math::</span>

<span class="sd">            \\mathbf{v_{obj}} = \\mathbf{w_{obj} \\odot F_{obj}(x)} \\</span>
<span class="sd">            \\mathbf{v_{ineqcv}} = \\mathbf{w_{ineqcv} \\odot G_{ineqcv}(x)} \\</span>
<span class="sd">            \\mathbf{v_{eqcv}} = \\mathbf{w_{eqcv} \\odot H_{eqcv}(x)}</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        x : numpy.ndarray</span>
<span class="sd">            A candidate solution vector of shape ``(ndecn,)``.</span>
<span class="sd">        args : tuple</span>
<span class="sd">            Additional non-keyword arguments.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : tuple</span>
<span class="sd">            A tuple ``(obj, ineqcv, eqcv)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``obj`` is a numpy.ndarray of shape ``(nobj,)`` that contains</span>
<span class="sd">                objective function evaluations.</span>
<span class="sd">            - ``ineqcv`` is a numpy.ndarray of shape ``(nineqcv,)`` that contains</span>
<span class="sd">                inequality constraint violation values.</span>
<span class="sd">            - ``eqcv`` is a numpy.ndarray of shape ``(neqcv,)`` that contains</span>
<span class="sd">                equality constraint violation values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># calculate sum(x)</span>
        <span class="n">xsum</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="c1"># if sum(x) ~== 0, then set to 1</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="n">xsum</span> <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">xsum</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mf">1e-10</span> <span class="k">else</span> <span class="mf">1.0</span>
        <span class="c1"># scale x to have a sum of 1 (contribution)</span>
        <span class="c1"># (n,) -&gt; (n,)</span>
        <span class="n">contrib</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">denom</span><span class="p">)</span> <span class="o">*</span> <span class="n">x</span>
        <span class="c1"># convert sum(x) to ndarray</span>
        <span class="c1"># scalar -&gt; (1,)</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">xsum</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>
        <span class="c1"># calculate allele frequencies</span>
        <span class="c1"># (n,) @ (n,p) -&gt; (p,)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">contrib</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">X</span>
        <span class="c1"># find alleles that are fixed approximately</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bitwise_or</span><span class="p">(</span><span class="n">p</span> <span class="o">&lt;</span> <span class="mf">1e-10</span><span class="p">,</span> <span class="n">p</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="o">-</span><span class="mf">1e-10</span><span class="p">)</span>
        <span class="c1"># sum the mask</span>
        <span class="c1"># (p,) -&gt; (1,)</span>
        <span class="n">f2</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">,</span> <span class="n">keepdims</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="c1"># concatenate objective function evaluations</span>
        <span class="c1"># (1,) concat (1,) -&gt; (2,)</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj_wt</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">f1</span><span class="p">,</span><span class="n">f2</span><span class="p">])</span>
        <span class="c1"># calculate inequality constraint violations</span>
        <span class="c1"># (1,)</span>
        <span class="n">ineqcv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ineqcv_wt</span> <span class="o">*</span> <span class="p">(</span><span class="n">f1</span> <span class="o">&lt;=</span> <span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="c1"># calculate equality constraint violations</span>
        <span class="n">eqcv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">eqcv_wt</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">neqcv</span><span class="p">)</span>
        <span class="c1"># return (2,), (1,), (0,)</span>
        <span class="k">return</span> <span class="n">obj</span><span class="p">,</span> <span class="n">ineqcv</span><span class="p">,</span> <span class="n">eqcv</span>
    <span class="c1">### method required by PyMOO interface ###</span>
    <span class="c1"># use default ``_evaluate`` method which uses the ``evalfn`` method</span>
</pre></div>
</div>
</section>
<section id="extracting-a-genotype-matrix-as-a-numpy-array">
<h3>Extracting a Genotype Matrix as a NumPy Array<a class="headerlink" href="#extracting-a-genotype-matrix-as-a-numpy-array" title="Permalink to this heading">#</a></h3>
<p>Next, we extract a genotype matrix as a raw NumPy array so we can construct a <code class="docutils literal notranslate"><span class="pre">GermplasmPanelReduction2</span></code> object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># get genotype matrix</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">gmat_reduced</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id1">
<h3>Constructing a Germplasm Panel Reduction Problem Object<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<p>Next, we construct the germplasm panel reduction problem object using the raw NumPy genotype matrix.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct optimization problem</span>
<span class="n">prob</span> <span class="o">=</span> <span class="n">GermplasmPanelReduction2</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id2">
<h3>Constructing a Custom Genetic Algorithm Object<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<p>Next, we construct a custom NSGA-II genetic algorithm object to solve this problem.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct NSGA-II object</span>
<span class="c1"># this problem is complex and requires many generations</span>
<span class="n">moea</span> <span class="o">=</span> <span class="n">NSGA2BinaryGeneticAlgorithm</span><span class="p">(</span>
    <span class="n">ngen</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
    <span class="n">pop_size</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id3">
<h3>Estimating the Pareto Frontier<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<p>After creating an algorithm object, we minimize the germplasm panel reduction problem and receive a solution.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># minimize the optimization problem</span>
<span class="n">soln</span> <span class="o">=</span> <span class="n">moea</span><span class="o">.</span><span class="n">minimize</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id4">
<h3>Visualizing the Pareto Frontier with <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code><a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<p>Finally, we can use <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> or any other plotting packages to visualize the results of the optimization. The code below creates a figure to visualize the estimated Pareto frontier.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">soln</span><span class="o">.</span><span class="n">soln_obj</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">soln</span><span class="o">.</span><span class="n">soln_obj</span><span class="p">[:,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Pareto frontier for number of selected individuals and number of fixed alleles&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Number of individuals&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of alleles fixed&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;germplasm_panel_reduction2.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">250</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>Below is the resulting figure:</p>
<img alt="../../../_images/germplasm_panel_reduction2.png" src="../../../_images/germplasm_panel_reduction2.png" />
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="triobjective_OCS_pareto_frontier.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Tri-Objective Optimal Contribution Selection Pareto Frontier Visualization</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="../breeding_program_simulation/single_trait_phenotypic_selection.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Single-Trait Phenotypic Selection</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preliminaries">
   Preliminaries
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-required-modules-and-seeding-the-global-prng">
     Loading Required Modules and Seeding the global PRNG
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-genotypic-data-from-a-vcf-file">
     Loading Genotypic Data from a VCF File
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#scenario-1-minimize-number-of-individuals-and-relatedness">
   Scenario 1: Minimize number of individuals and relatedness
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#defining-a-custom-problem-class-for-scenario-1">
     Defining a Custom Problem Class for Scenario 1
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-a-kinship-matrix-and-its-decomposition">
     Calculating a Kinship Matrix and its Decomposition
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#constructing-a-germplasm-panel-reduction-problem-object">
     Constructing a Germplasm Panel Reduction Problem Object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#constructing-a-custom-genetic-algorithm-object">
     Constructing a Custom Genetic Algorithm Object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#estimating-the-pareto-frontier">
     Estimating the Pareto Frontier
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualizing-the-pareto-frontier-with-matplotlib">
     Visualizing the Pareto Frontier with
     <code class="docutils literal notranslate">
      <span class="pre">
       matplotlib
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#scenario-2-minimize-number-of-individuals-and-loss-of-alleles">
   Scenario 2: Minimize number of individuals and loss of alleles
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#defining-a-custom-problem-class-for-scenario-2">
     Defining a Custom Problem Class for Scenario 2
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extracting-a-genotype-matrix-as-a-numpy-array">
     Extracting a Genotype Matrix as a NumPy Array
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Constructing a Germplasm Panel Reduction Problem Object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Constructing a Custom Genetic Algorithm Object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Estimating the Pareto Frontier
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id4">
     Visualizing the Pareto Frontier with
     <code class="docutils literal notranslate">
      <span class="pre">
       matplotlib
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2023, Robert Shrote.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>