
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Variance Matrices &#8212; pybrops 1.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/genomic_modeling/variance_matrices';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genotyping Protocols" href="../breeding_protocols/genotyping_protocols.html" />
    <link rel="prev" title="Genomic Models" href="genomic_models.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/examples.html">Examples</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/germplasm_panel_reduction.html">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_single_trait_phenotypic_selection.html">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/multiobjective_genomic_selection.html">Bi-Trait Genotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../standard_usage.html" class="nav-link">Standard PyBrOpS Usage</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Variance Matrices</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="variance-matrices">
<h1>Variance Matrices<a class="headerlink" href="#variance-matrices" title="Link to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> families of classes allow for representation of the expected progeny trait variances from a cross between individuals. The purpose of these families of classes is to calculate expected progeny trait variances assuming linkage for the <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> family of classe and no linkage for the <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> family of classes. Both families of variance matrices utilize genomic models to calculate variances and are designed to be agnostic of <code class="docutils literal notranslate"><span class="pre">GenomicModel</span></code> type. For additive linear genomic models, there exist deterministic equations to calculate progeny variance for two-, three-, and four-way crosses. <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> families which assume an additive linear genomic model have interfaces of <code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">AdditiveGenicVarianceMatrix</span></code>, respectfully.</p>
<p>Deriving from the <code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code> interface are several implemented classes which are useful. They are summarized below:</p>
</section>
<section id="summary-of-variance-matrix-classes">
<h2>Summary of Variance Matrix Classes<a class="headerlink" href="#summary-of-variance-matrix-classes" title="Link to this heading">#</a></h2>
<p>Genetic and genic variance matrices are found in the <code class="docutils literal notranslate"><span class="pre">pybrops.model.vmat</span></code> module. The two subsections below summarize the genetic and genic variance matrices</p>
<section id="genetic-variance-matrix-classes">
<h3>Genetic variance matrix classes<a class="headerlink" href="#genetic-variance-matrix-classes" title="Link to this heading">#</a></h3>
<p>Genetic variance matrices, which represent the expected progeny trait variances assuming linkage disequilibrium, are summarized in the table below.</p>
<table class="table" id="id1">
<caption><span class="caption-text">Summary of genetic variance classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.model.vmat</span></code> module</span><a class="headerlink" href="#id1" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 26.3%" />
<col style="width: 21.1%" />
<col style="width: 52.6%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all additive genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseGeneticVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense additive genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from two-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseThreeWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from three-way crosses.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseFourWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from four-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseDihybridDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from dihybrid crosses.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="genic-variance-matrix-classes">
<h3>Genic variance matrix classes<a class="headerlink" href="#genic-variance-matrix-classes" title="Link to this heading">#</a></h3>
<p>Genic variance matrices, which represent the expected progeny trait variances assuming linkage equilibrium, are summarized in the table below.</p>
<table class="table" id="id2">
<caption><span class="caption-text">Summary of genic variance classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.model.vmat</span></code> module</span><a class="headerlink" href="#id2" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 26.3%" />
<col style="width: 21.1%" />
<col style="width: 52.6%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">AdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all additive genic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseGenicVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense genic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense additive genic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from two-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseThreeWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from three-way crosses.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseFourWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from four-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseDihybridDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from dihybrid crosses.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="variance-matrix-properties">
<h2>Variance Matrix Properties<a class="headerlink" href="#variance-matrix-properties" title="Link to this heading">#</a></h2>
<p>Both genetic and genic variance matrix classes share a common set of properties. These properties can be grouped into four categories: general properties, taxa properties, trait properties, and square properties. The following four subsections summarize these property categories.</p>
<section id="general-properties">
<h3>General properties<a class="headerlink" href="#general-properties" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices share several shape properties that are common to all <code class="docutils literal notranslate"><span class="pre">Matrix</span></code> classes. In addition, genetic and genic variance matrices have an <code class="docutils literal notranslate"><span class="pre">epgc</span></code> property which is used to indicate the expected parental genomic contribution of each parent corresponding to each parental axis. The table below summarizes variance matrix general properties.</p>
<table class="table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> general properties</span><a class="headerlink" href="#id3" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat</span></code></p></td>
<td><p>The raw genetic or genic variance matrix pointer</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_ndim</span></code></p></td>
<td><p>The number of dimensions for the genetic or genic variance matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat_shape</span></code></p></td>
<td><p>The genetic or genic variance matrix shape</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">epgc</span></code></p></td>
<td><p>The expected parental genomic contribution corresponding to each parental axis.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="taxa-properties">
<h3>Taxa properties<a class="headerlink" href="#taxa-properties" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices have several taxa related properties including taxa names, taxa group identities, and sorting metadata, which can be used for quick group access and sorting. The table below summarizes variance matrix taxa properties.</p>
<table class="table" id="id4">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> taxa properties</span><a class="headerlink" href="#id4" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntaxa</span></code></p></td>
<td><p>The number of taxa represented by the genetic or genic variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></td>
<td><p>The names of the taxa</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_axis</span></code></p></td>
<td><p>The matrix axis along which taxa are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code></p></td>
<td><p>An optional taxa group label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_name</span></code></p></td>
<td><p>If taxa are sorted by group: get the names of the groups</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_stix</span></code></p></td>
<td><p>If taxa are sorted by group: get the start indices (inclusive) for each group</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_spix</span></code></p></td>
<td><p>If taxa are sorted by group: get the stop indices (exclusive) for each group</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_len</span></code></p></td>
<td><p>If taxa are sorted by group: get the length of each group</p></td>
</tr>
</tbody>
</table>
</section>
<section id="trait-properties">
<h3>Trait properties<a class="headerlink" href="#trait-properties" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices have several trait related properties of which the most important is the trait names.</p>
<table class="table" id="id5">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> trait properties</span><a class="headerlink" href="#id5" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntrait</span></code></p></td>
<td><p>The number of traits represented by the genetic or genic variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">trait</span></code></p></td>
<td><p>The names of the traits</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">trait_axis</span></code></p></td>
<td><p>The matrix axis along which traits are stored</p></td>
</tr>
</tbody>
</table>
</section>
<section id="square-matrix-properties">
<h3>Square matrix properties<a class="headerlink" href="#square-matrix-properties" title="Link to this heading">#</a></h3>
<p>Since genetic and genic matrices are square by nature, they also have several properties which extract data regarding their squareness. These properties are summarized below.</p>
<table class="table" id="id6">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> square matrix properties</span><a class="headerlink" href="#id6" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nsquare</span></code></p></td>
<td><p>The number of square axes for the genetic or genic variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes</span></code></p></td>
<td><p>The axes indices for the square axes for the genetic or genic variance matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes_len</span></code></p></td>
<td><p>The lengths of the square axes for the genetic or genic variance matrix</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="loading-variance-matrix-modules">
<h2>Loading Variance Matrix Modules<a class="headerlink" href="#loading-variance-matrix-modules" title="Link to this heading">#</a></h2>
<section id="loading-genetic-variance-matrix-modules">
<h3>Loading genetic variance matrix modules<a class="headerlink" href="#loading-genetic-variance-matrix-modules" title="Link to this heading">#</a></h3>
<p>Importing genetic variance matrix classes can be accomplished using the following import statements:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import abstract interface classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.GeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">GeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.AdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">AdditiveGeneticVarianceMatrix</span>

<span class="c1"># import semi-abstract classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseAdditiveGeneticVarianceMatrix</span>

<span class="c1"># import concrete implemented classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseTwoWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseThreeWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseThreeWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseFourWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseFourWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseDihybridDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseDihybridDHAdditiveGeneticVarianceMatrix</span>
</pre></div>
</div>
</section>
<section id="loading-genic-variance-matrix-modules">
<h3>Loading genic variance matrix modules<a class="headerlink" href="#loading-genic-variance-matrix-modules" title="Link to this heading">#</a></h3>
<p>Importing genic variance matrix classes can be accomplished using the following import statements:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import abstract interface classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.GenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">GenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.AdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">AdditiveGenicVarianceMatrix</span>

<span class="c1"># import semi-abstract classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseAdditiveGenicVarianceMatrix</span>

<span class="c1"># import concrete implemented classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseTwoWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseThreeWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseThreeWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseFourWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseFourWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseDihybridDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseDihybridDHAdditiveGenicVarianceMatrix</span>
</pre></div>
</div>
</section>
</section>
<section id="creating-variance-matrices">
<h2>Creating Variance Matrices<a class="headerlink" href="#creating-variance-matrices" title="Link to this heading">#</a></h2>
<p>Genetic and genic variance matrices can be created using several method including from raw NumPy arrays, from genotype matrices and genomic models, from Pandas DataFrames, from CSV files, and from HDF5 files. The following subsections detail the creation or loading of genetic or genic variance matrices from their corresponding sources.</p>
<section id="creating-variance-matrices-from-numpy-arrays">
<h3>Creating variance matrices from NumPy arrays<a class="headerlink" href="#creating-variance-matrices-from-numpy-arrays" title="Link to this heading">#</a></h3>
<p>Using the constructor of a <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> class, one can create genetic and genic variance matrices, respectively, from NumPy arrays. The example below demonstrates the creation of a <code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGenicVarianceMatrix`</span></code> objects from raw NumPy arrays.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># create random variance values</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create trait names</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;trait&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntrait</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create genetic variance matrix</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span>
<span class="p">)</span>

<span class="c1"># create genic variance matrix</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-variance-matrices-from-genomic-models">
<h3>Calculating variance matrices from genomic models<a class="headerlink" href="#calculating-variance-matrices-from-genomic-models" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices may also be calculated from a combination of <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenomicModel</span></code> objects. This can be accomplished using the <code class="docutils literal notranslate"><span class="pre">from_gmod</span></code> or <code class="docutils literal notranslate"><span class="pre">from_algmod</span></code> class methods. The code below demonstrates how to use this method to accomplish this task.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a dummy genomic model</span>
<span class="c1"># model parameters</span>
<span class="n">nfixed</span> <span class="o">=</span> <span class="mi">1</span>      <span class="c1"># number of fixed effects</span>
<span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span>      <span class="c1"># number of traits</span>
<span class="n">nmisc</span> <span class="o">=</span> <span class="mi">0</span>       <span class="c1"># number of miscellaneous random effects</span>
<span class="n">nadditive</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># number of additive marker effects</span>

<span class="c1"># create dummy values</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nfixed</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">u_misc</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nmisc</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">u_a</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nadditive</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;Trait&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntrait</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create additive linear genomic model</span>
<span class="n">algmod</span> <span class="o">=</span> <span class="n">DenseAdditiveLinearGenomicModel</span><span class="p">(</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">beta</span><span class="p">,</span>
    <span class="n">u_misc</span> <span class="o">=</span> <span class="n">u_misc</span><span class="p">,</span>
    <span class="n">u_a</span> <span class="o">=</span> <span class="n">u_a</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span><span class="p">,</span>
    <span class="n">model_name</span> <span class="o">=</span> <span class="s2">&quot;example&quot;</span><span class="p">,</span>
    <span class="n">params</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span>

<span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">nvrnt</span> <span class="o">=</span> <span class="n">nadditive</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">nchrom</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># create random genotypes</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ploidy</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">nvrnt</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;Taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant chromsome assignments</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nchrom</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker physical positions</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker genetic positions</span>
<span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_genpos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create a genotype matrix from scratch using NumPy arrays</span>
<span class="n">pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
    <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
<span class="p">)</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>

<span class="c1"># calculate genetic variance matrix from GenomicModel</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_gmod</span><span class="p">(</span>
    <span class="n">gmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">ncross</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
<span class="p">)</span>

<span class="c1"># calculate genetic variance matrix from AdditiveLinearGenomicModel</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_algmod</span><span class="p">(</span>
    <span class="n">algmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">ncross</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
<span class="p">)</span>

<span class="c1"># calculate genic variance matrix from GenomicModel</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_gmod</span><span class="p">(</span>
    <span class="n">gmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span>

<span class="c1"># calculate genic variance matrix from AdditiveLinearGenomicModel</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_algmod</span><span class="p">(</span>
    <span class="n">algmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="creating-variance-matrices-from-pandas-dataframes">
<h3>Creating variance matrices from Pandas DataFrames<a class="headerlink" href="#creating-variance-matrices-from-pandas-dataframes" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices may be read from Pandas DataFrames. The <code class="docutils literal notranslate"><span class="pre">from_pandas</span></code> class method may be used to read a <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> from a Pandas DataFrame. The code example below demonstrates this method’s usage.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create dummy pandas dataframe</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;Female&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Female Group&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,],</span>
    <span class="s2">&quot;Male&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Male Group&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,],</span>
    <span class="s2">&quot;Trait&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Variance&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span><span class="mf">0.3</span><span class="p">,</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">0.7</span><span class="p">],</span>
<span class="p">})</span>

<span class="c1"># construct genetic variance matrix from pandas dataframe</span>
<span class="n">tmp_vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
    <span class="n">female_col</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">,</span>
    <span class="n">female_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Female Group&quot;</span><span class="p">,</span>
    <span class="n">male_col</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">,</span>
    <span class="n">male_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Male Group&quot;</span><span class="p">,</span>
    <span class="n">trait_col</span> <span class="o">=</span> <span class="s2">&quot;Trait&quot;</span><span class="p">,</span>
    <span class="n">variance_col</span> <span class="o">=</span> <span class="s2">&quot;Variance&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># construct genic variance matrix from pandas dataframe</span>
<span class="n">tmp_gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
    <span class="n">female_col</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">,</span>
    <span class="n">female_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Female Group&quot;</span><span class="p">,</span>
    <span class="n">male_col</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">,</span>
    <span class="n">male_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Male Group&quot;</span><span class="p">,</span>
    <span class="n">trait_col</span> <span class="o">=</span> <span class="s2">&quot;Trait&quot;</span><span class="p">,</span>
    <span class="n">variance_col</span> <span class="o">=</span> <span class="s2">&quot;Variance&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-variance-matrices-from-csv-files">
<h3>Loading variance matrices from CSV files<a class="headerlink" href="#loading-variance-matrices-from-csv-files" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices may also be read from CSV files in a manner similar to Pandas DataFrames. The <code class="docutils literal notranslate"><span class="pre">from_csv</span></code> class method may be used to load genetic or genic variance matrices from CSV files. The following code block demonstrates the usage of this method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create dummy pandas dataframe and export as CSV</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;Female&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Female Group&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,],</span>
    <span class="s2">&quot;Male&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Male Group&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,],</span>
    <span class="s2">&quot;Trait&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,</span><span class="s2">&quot;Yield&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;Variance&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span><span class="mf">0.3</span><span class="p">,</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">0.7</span><span class="p">],</span>
<span class="p">})</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;saved_df.csv&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

<span class="c1"># construct genetic variance matrix from csv</span>
<span class="n">tmp_vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;saved_df.csv&quot;</span><span class="p">,</span>
    <span class="n">female_col</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">,</span>
    <span class="n">female_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Female Group&quot;</span><span class="p">,</span>
    <span class="n">male_col</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">,</span>
    <span class="n">male_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Male Group&quot;</span><span class="p">,</span>
    <span class="n">trait_col</span> <span class="o">=</span> <span class="s2">&quot;Trait&quot;</span><span class="p">,</span>
    <span class="n">variance_col</span> <span class="o">=</span> <span class="s2">&quot;Variance&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># construct genic variance matrix from csv</span>
<span class="n">tmp_gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;saved_df.csv&quot;</span><span class="p">,</span>
    <span class="n">female_col</span> <span class="o">=</span> <span class="s2">&quot;Female&quot;</span><span class="p">,</span>
    <span class="n">female_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Female Group&quot;</span><span class="p">,</span>
    <span class="n">male_col</span> <span class="o">=</span> <span class="s2">&quot;Male&quot;</span><span class="p">,</span>
    <span class="n">male_grp_col</span> <span class="o">=</span> <span class="s2">&quot;Male Group&quot;</span><span class="p">,</span>
    <span class="n">trait_col</span> <span class="o">=</span> <span class="s2">&quot;Trait&quot;</span><span class="p">,</span>
    <span class="n">variance_col</span> <span class="o">=</span> <span class="s2">&quot;Variance&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-variance-matrices-from-hdf5-files">
<h3>Loading variance matrices from HDF5 files<a class="headerlink" href="#loading-variance-matrices-from-hdf5-files" title="Link to this heading">#</a></h3>
<p>As with all classes in the <code class="docutils literal notranslate"><span class="pre">Matrix</span></code> family, <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> objects may be imported and exported to an HDF5 format. To read saved genetic or genic variance matrices from an HDF5 file, use the <code class="docutils literal notranslate"><span class="pre">from_hdf5</span></code> class method. The code below demonstrates the use of this method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read genetic variance matrix from HDF5 file</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_vmat.h5&quot;</span><span class="p">)</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="copying-variance-matrices">
<h2>Copying Variance Matrices<a class="headerlink" href="#copying-variance-matrices" title="Link to this heading">#</a></h2>
<p>Genetic and genic variance matrices may be copied using two methods: shallow copying and deep copying.</p>
<section id="shallow-copying">
<h3>Shallow copying<a class="headerlink" href="#shallow-copying" title="Link to this heading">#</a></h3>
<p>In shallow copying, references to a <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code>’s data are copied to a new genetic or genic variance matrix object. Copying is only one level deep which means that changes to the original object may affect data values in the copied object. The code below illustrates the use of the <code class="docutils literal notranslate"><span class="pre">copy</span></code> method bound to <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> objects and the base Python function <a class="reference external" href="https://docs.python.org/3/library/copy.html#copy.copy"><code class="docutils literal notranslate"><span class="pre">copy.copy</span></code></a> which can both be used to shallow copy a genetic or genic variance matrix object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy a genetic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># copy a genic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="deep-copying">
<h3>Deep copying<a class="headerlink" href="#deep-copying" title="Link to this heading">#</a></h3>
<p>In deep copying, data in a <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> is recursively copied to a new genetic or genic variance matrix object. Copying occurs down to the deepest levels so that changes to the original object will not affect data values in the copied object. The code below illustrates the use of the <code class="docutils literal notranslate"><span class="pre">deepcopy</span></code> method bound to <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> or <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> objects and the base Python function <a class="reference external" href="https://docs.python.org/3/library/copy.html#copy.deepcopy"><code class="docutils literal notranslate"><span class="pre">copy.deepcopy</span></code></a> which can both be used to deep copy a genetic or genic variance matrix object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy a genetic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># deep copy a genic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="copy-on-element-manipulation">
<h2>Copy-On Element Manipulation<a class="headerlink" href="#copy-on-element-manipulation" title="Link to this heading">#</a></h2>
<p>Genetic or genic variance matrices have several methods by which modifed copies of the original matrix can be made. These are called copy-on element manipulation methods. Matrices may have taxa and trait axes adjoined, deleted, inserted, or selected. The following sections demonstrate the use of these method families.</p>
<section id="adjoining-elements">
<h3>Adjoining elements<a class="headerlink" href="#adjoining-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">adjoin</span></code> family of methods allows for taxa and trait axes of a genetic or genic variance matrix to be adjoined together, creating a new matrix in the process. Use of the <code class="docutils literal notranslate"><span class="pre">adjoin</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new variance matrices to demonstrate</span>
<span class="n">newvmat</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">newgvmat</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># adjoin variance matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">)</span>

<span class="c1"># adjoin variance matrices along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin_trait</span><span class="p">(</span><span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin_trait</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="deleting-elements">
<h3>Deleting elements<a class="headerlink" href="#deleting-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">delete</span></code> family of methods allows for taxa and trait axes of a genetic or genic variance matrix to be removed in a copy of the original. Use of the <code class="docutils literal notranslate"><span class="pre">delete</span></code> method family is demonstrated in the code below.</p>
<section id="delete-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">delete</span></code> taxa<a class="headerlink" href="#delete-taxa" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># delete first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="delete-traits">
<h4><code class="docutils literal notranslate"><span class="pre">delete</span></code> traits<a class="headerlink" href="#delete-traits" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># delete first trait using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first two traits using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>

<span class="c1"># delete first two traits using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="inserting-elements">
<h3>Inserting elements<a class="headerlink" href="#inserting-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">insert</span></code> family of methods allows for taxa and trait axes of a genetic or genic variance matrix to be inserted into a copy of the original matrix. Use of the <code class="docutils literal notranslate"><span class="pre">insert</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new variance matrix to demonstrate</span>
<span class="n">newvmat</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">newgvmat</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># insert variance matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">)</span>

<span class="c1"># insert variance matrix along the trait axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="selecting-elements">
<h3>Selecting elements<a class="headerlink" href="#selecting-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">select</span></code> family of methods allows for taxa and trait axes of the genetic or genic variance matrix to be selected and extracted to a copy of the original matrix. Use of the <code class="docutils literal notranslate"><span class="pre">select</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># select first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1"># select first two traits using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="in-place-element-manipulation">
<h2>In-Place Element Manipulation<a class="headerlink" href="#in-place-element-manipulation" title="Link to this heading">#</a></h2>
<p>Genetic or genic variance matrices have several methods which execute in-place element manipulations. These are called in-place element manipulation methods. Genetic or genic variance matrices may have taxa and trait axes appended, removed, incorporated, or concatenated. The following sections demonstrate the use of these method families.</p>
<section id="appending-elements">
<h3>Appending elements<a class="headerlink" href="#appending-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">append</span></code> family of methods allows for new taxa and trait axes to be appended to the genetic or genic variance matrix. The code segment below demonstrates their use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># append variance matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span> <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>                  <span class="c1"># append original to copy</span>

<span class="c1"># append variance matrices along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span> <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span><span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_trait</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>                  <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_trait</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>                 <span class="c1"># append original to copy</span>
</pre></div>
</div>
</section>
<section id="removing-elements">
<h3>Removing elements<a class="headerlink" href="#removing-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">remove</span></code> family of methods allows for taxa and trait axes to be removed from a genetic or genic variance matrix. A demonstration of their use can be seen below.</p>
<section id="remove-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> taxa<a class="headerlink" href="#remove-taxa" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>           <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span> <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
<section id="remove-traits">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> traits<a class="headerlink" href="#remove-traits" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first trait using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>           <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>          <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                             <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                             <span class="c1"># remove from copy</span>

<span class="c1"># remove first trait using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span> <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>                    <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>                    <span class="c1"># remove from copy</span>

<span class="c1"># remove first trait using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>         <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>        <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span>                           <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span>                           <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
</section>
<section id="incorporating-elements">
<h3>Incorporating elements<a class="headerlink" href="#incorporating-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">incorp</span></code> family of methods allows for new taxa and trait axes to be inserted at specific locations in a genetic or genic variance matrix. Use of the <code class="docutils literal notranslate"><span class="pre">incorp</span></code> family is demonstrated in the code segment below below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># incorp variance matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>      <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>    <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">)</span>                       <span class="c1"># incorporate into copy</span>

<span class="c1"># incorp variance matrix along the trait axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>     <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>   <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">)</span>                       <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">)</span>                      <span class="c1"># incorporate into copy</span>
</pre></div>
</div>
</section>
<section id="concatenating-matrices">
<h3>Concatenating matrices<a class="headerlink" href="#concatenating-matrices" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">concat</span></code> family of methods allows for multiple genetic or genic variance matrices to be concatenated to each other along taxa or trait axes. The code segment below demonstrates their use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># concatenate along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">])</span>

<span class="c1"># concatenate along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat_trait</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat_trait</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping-and-sorting">
<h2>Grouping and Sorting<a class="headerlink" href="#grouping-and-sorting" title="Link to this heading">#</a></h2>
<p>Genetic or genic variance matrices in PyBrOpS have several sorting and grouping focused methods. Sorting methods can be used to reorder, sort, and group taxa axes alphanumerically and reorder and sort trait axes.. The following sections demonstrate the use of the <code class="docutils literal notranslate"><span class="pre">reorder</span></code>, <code class="docutils literal notranslate"><span class="pre">lexsort</span></code>, <code class="docutils literal notranslate"><span class="pre">sort</span></code>, and <code class="docutils literal notranslate"><span class="pre">group</span></code> method families.</p>
<section id="reordering-elements">
<h3>Reordering elements<a class="headerlink" href="#reordering-elements" title="Link to this heading">#</a></h3>
<p>Taxa and trait axes in a genetic or genic variance matrix can be reordered using the <code class="docutils literal notranslate"><span class="pre">reorder</span></code> family of methods. Demonstrations of this method family are below.</p>
<section id="reorder-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> taxa<a class="headerlink" href="#reorder-taxa" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>

<span class="c1"># reorder values along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reorder-traits">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> traits<a class="headerlink" href="#reorder-traits" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntrait</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>

<span class="c1"># reorder values along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_trait</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_trait</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="lexsorting-elements">
<h3>Lexsorting elements<a class="headerlink" href="#lexsorting-elements" title="Link to this heading">#</a></h3>
<p>An indirect stable sort - or lexsort - for taxa or trait axes can be performed using the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> family of methods. The code segment below illustrates the use of this family of methods.</p>
<section id="lexsort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> taxa<a class="headerlink" href="#lexsort-taxa" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for taxa</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the taxa axis</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="lexsort-traits">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> traits<a class="headerlink" href="#lexsort-traits" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for trait</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="sorting-elements">
<h3>Sorting elements<a class="headerlink" href="#sorting-elements" title="Link to this heading">#</a></h3>
<p>Alphanumeric sorting along taxa and trait axes can be done using the <code class="docutils literal notranslate"><span class="pre">sort</span></code> family of methods. Sorting examples are illustrated below.</p>
<section id="sort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> taxa<a class="headerlink" href="#sort-taxa" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="sort-traits">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> traits<a class="headerlink" href="#sort-traits" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_trait</span><span class="p">()</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_trait</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping-elements">
<h3>Grouping elements<a class="headerlink" href="#grouping-elements" title="Link to this heading">#</a></h3>
<p>Grouping along taxa axes can be done using the <code class="docutils literal notranslate"><span class="pre">group</span></code> family of methods. The following code illustrates the use of the <code class="docutils literal notranslate"><span class="pre">group</span></code> method family along the taxa axes of a genetic or genic variance matrix.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># group taxa</span>
<span class="c1"># ++++++++++</span>

<span class="c1"># sort genetic along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>

<span class="c1"># sort genic variance matrix along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="square-matrix-functions">
<h2>Square matrix functions<a class="headerlink" href="#square-matrix-functions" title="Link to this heading">#</a></h2>
<p>Since variance matrices are also inherently square, there are a methods dealing with the squareness of a variance matrix.</p>
<section id="determine-whether-all-square-axes-are-of-equal-length">
<h3>Determine whether all square axes are of equal length<a class="headerlink" href="#determine-whether-all-square-axes-are-of-equal-length" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">is_square</span></code> method for genetic and genic variance matrices can be used to determine if a variance matrix is square.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># boolean value</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">is_square</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">is_square</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="exporting-variance-matrices">
<h2>Exporting Variance Matrices<a class="headerlink" href="#exporting-variance-matrices" title="Link to this heading">#</a></h2>
<p>Genetic and genic variance matrices may be exported to multiple formats including Pandas DataFrames, CSV files, and HDF5 files. The following subsections provide export examples.</p>
<section id="exporting-to-pandas-dataframe">
<h3>Exporting to Pandas DataFrame<a class="headerlink" href="#exporting-to-pandas-dataframe" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">to_pandas</span></code> method can be used to export a genetic or genic variance matrix to a Pandas DataFrame. Column names may be optionally provided to override default column names.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write variance matrices to a pandas dataframe</span>
<span class="n">out_df</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="n">out_df</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="exporting-to-csv">
<h3>Exporting to CSV<a class="headerlink" href="#exporting-to-csv" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">to_csv</span></code> method can be used to export a genetic or genic variance matrix to a CSV file. Like the <code class="docutils literal notranslate"><span class="pre">to_pandas</span></code> method, column names may be optionally provided to override default column names.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write variance matrices to a CSV file</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;saved_vmat.csv&quot;</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="exporting-to-hdf5">
<h3>Exporting to HDF5<a class="headerlink" href="#exporting-to-hdf5" title="Link to this heading">#</a></h3>
<p>Genetic and genic variance matrices can be written to an HDF5 file using the <code class="docutils literal notranslate"><span class="pre">to_hdf5</span></code> method. The code below demonstrates the use of this method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write variance matrices to an HDF5 file</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_vmat.h5&quot;</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="genomic_models.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genomic Models</p>
      </div>
    </a>
    <a class="right-next"
       href="../breeding_protocols/genotyping_protocols.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotyping Protocols</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#class-family-overview">Class Family Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-variance-matrix-classes">Summary of Variance Matrix Classes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-variance-matrix-classes">Genetic variance matrix classes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genic-variance-matrix-classes">Genic variance matrix classes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variance-matrix-properties">Variance Matrix Properties</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#general-properties">General properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#taxa-properties">Taxa properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trait-properties">Trait properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#square-matrix-properties">Square matrix properties</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-variance-matrix-modules">Loading Variance Matrix Modules</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genetic-variance-matrix-modules">Loading genetic variance matrix modules</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genic-variance-matrix-modules">Loading genic variance matrix modules</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-variance-matrices">Creating Variance Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-variance-matrices-from-numpy-arrays">Creating variance matrices from NumPy arrays</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-variance-matrices-from-genomic-models">Calculating variance matrices from genomic models</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-variance-matrices-from-pandas-dataframes">Creating variance matrices from Pandas DataFrames</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-variance-matrices-from-csv-files">Loading variance matrices from CSV files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-variance-matrices-from-hdf5-files">Loading variance matrices from HDF5 files</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copying-variance-matrices">Copying Variance Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#shallow-copying">Shallow copying</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deep-copying">Deep copying</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copy-on-element-manipulation">Copy-On Element Manipulation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adjoining-elements">Adjoining elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deleting-elements">Deleting elements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#delete-taxa"><code class="docutils literal notranslate"><span class="pre">delete</span></code> taxa</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#delete-traits"><code class="docutils literal notranslate"><span class="pre">delete</span></code> traits</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inserting-elements">Inserting elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#selecting-elements">Selecting elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-place-element-manipulation">In-Place Element Manipulation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#appending-elements">Appending elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-elements">Removing elements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-taxa"><code class="docutils literal notranslate"><span class="pre">remove</span></code> taxa</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-traits"><code class="docutils literal notranslate"><span class="pre">remove</span></code> traits</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#incorporating-elements">Incorporating elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#concatenating-matrices">Concatenating matrices</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#grouping-and-sorting">Grouping and Sorting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reordering-elements">Reordering elements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reorder-taxa"><code class="docutils literal notranslate"><span class="pre">reorder</span></code> taxa</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reorder-traits"><code class="docutils literal notranslate"><span class="pre">reorder</span></code> traits</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lexsorting-elements">Lexsorting elements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#lexsort-taxa"><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> taxa</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#lexsort-traits"><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> traits</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sorting-elements">Sorting elements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sort-taxa"><code class="docutils literal notranslate"><span class="pre">sort</span></code> taxa</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sort-traits"><code class="docutils literal notranslate"><span class="pre">sort</span></code> traits</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#grouping-elements">Grouping elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#square-matrix-functions">Square matrix functions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#determine-whether-all-square-axes-are-of-equal-length">Determine whether all square axes are of equal length</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-variance-matrices">Exporting Variance Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-to-pandas-dataframe">Exporting to Pandas DataFrame</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-to-csv">Exporting to CSV</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-to-hdf5">Exporting to HDF5</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>