
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Variance Matrices &#8212; pybrops 1.0.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/genomic_modeling/variance_matrices';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genotyping Protocols" href="../breeding_simulation/genotyping_protocols.html" />
    <link rel="prev" title="Genomic Models" href="genomic_models.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">pybrops 1.0.0 documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../installation/installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../prerequisites/prerequisites.html">
                        Prerequisites
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../examples.html">
                        .. Examples
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../installation/installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../prerequisites/prerequisites.html">
                        Prerequisites
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../examples.html">
                        .. Examples
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item"><nav class="bd-links" id="bd-docs-nav" aria-label="Section navigation">
  <p class="bd-links__title" role="heading" aria-level="1">
    Section Navigation
  </p>
  <div class="bd-toc-item navbar-nav">
    <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../population_genetics/coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Simulation Construction:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_simulation/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_simulation/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_simulation/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_simulation/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_simulation/selection_protocols.html">Selection Protocols</a></li>
</ul>

  </div>
</nav>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="variance-matrices">
<h1>Variance Matrices<a class="headerlink" href="#variance-matrices" title="Permalink to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Permalink to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> families of classes allow for representation of the expected progeny trait variances from a cross between individuals. The purpose of these families of classes is to calculate expected progeny trait variances assuming linkage for the <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> family of classe and no linkage for the <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> family of classes. Both families of variance matrices utilize genomic models to calculate variances and are designed to be agnostic of <code class="docutils literal notranslate"><span class="pre">GenomicModel</span></code> type. For additive linear genomic models, there exist deterministic equations to calculate progeny variance for two-, three-, and four-way crosses. <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> families which assume an additive linear genomic model have interfaces of <code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">AdditiveGenicVarianceMatrix</span></code>, respectfully.</p>
<p>Deriving from the <code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code> interface are several implemented classes which are useful. They are summarized below:</p>
</section>
<section id="summary-of-variance-matrix-classes">
<h2>Summary of Variance Matrix Classes<a class="headerlink" href="#summary-of-variance-matrix-classes" title="Permalink to this heading">#</a></h2>
<section id="genetic-variance-matrix-classes">
<h3>Genetic variance matrix classes<a class="headerlink" href="#genetic-variance-matrix-classes" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id1">
<caption><span class="caption-text">Summary of genetic variance classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.model.vmat</span></code> module</span><a class="headerlink" href="#id1" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 26%" />
<col style="width: 21%" />
<col style="width: 53%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">AdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all additive genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseGeneticVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense additive genetic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from two-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseThreeWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from three-way crosses.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseFourWayDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from four-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseDihybridDHAdditiveGeneticVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic variance matrices calculated from dihybrid crosses.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="genic-variance-matrix-classes">
<h3>Genic variance matrix classes<a class="headerlink" href="#genic-variance-matrix-classes" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id2">
<caption><span class="caption-text">Summary of genic variance classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.model.vmat</span></code> module</span><a class="headerlink" href="#id2" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 26%" />
<col style="width: 21%" />
<col style="width: 53%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">AdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all additive genic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseGenicVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense genic variance matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense additive genic variance matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseTwoWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from two-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseThreeWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from three-way crosses.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseFourWayDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from four-way crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseDihybridDHAdditiveGenicVarianceMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genic variance matrices calculated from dihybrid crosses.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="loading-variance-matrix-modules">
<h2>Loading Variance Matrix Modules<a class="headerlink" href="#loading-variance-matrix-modules" title="Permalink to this heading">#</a></h2>
<section id="loading-genetic-variance-matrix-modules">
<h3>Loading genetic variance matrix modules<a class="headerlink" href="#loading-genetic-variance-matrix-modules" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import abstract interface classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.GeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">GeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.AdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">AdditiveGeneticVarianceMatrix</span>

<span class="c1"># import semi-abstract classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseAdditiveGeneticVarianceMatrix</span>

<span class="c1"># import concrete implemented classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseTwoWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseThreeWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseThreeWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseFourWayDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseFourWayDHAdditiveGeneticVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseDihybridDHAdditiveGeneticVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseDihybridDHAdditiveGeneticVarianceMatrix</span>
</pre></div>
</div>
</section>
<section id="loading-genic-variance-matrix-modules">
<h3>Loading genic variance matrix modules<a class="headerlink" href="#loading-genic-variance-matrix-modules" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import abstract interface classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.GenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">GenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.AdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">AdditiveGenicVarianceMatrix</span>

<span class="c1"># import semi-abstract classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseAdditiveGenicVarianceMatrix</span>

<span class="c1"># import concrete implemented classes</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseTwoWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseThreeWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseThreeWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseFourWayDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseFourWayDHAdditiveGenicVarianceMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.model.vmat.DenseDihybridDHAdditiveGenicVarianceMatrix</span> <span class="kn">import</span> <span class="n">DenseDihybridDHAdditiveGenicVarianceMatrix</span>
</pre></div>
</div>
</section>
</section>
<section id="creating-variance-matrices">
<h2>Creating Variance Matrices<a class="headerlink" href="#creating-variance-matrices" title="Permalink to this heading">#</a></h2>
<section id="creating-variance-matrices-from-numpy-arrays">
<h3>Creating variance matrices from NumPy arrays<a class="headerlink" href="#creating-variance-matrices-from-numpy-arrays" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># create random variance values</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create trait names</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;trait&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntrait</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create genetic variance matrix</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span>
<span class="p">)</span>

<span class="c1"># create genic variance matrix</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="creating-variance-matrices-from-genomic-models">
<h3>Creating variance matrices from genomic models<a class="headerlink" href="#creating-variance-matrices-from-genomic-models" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a dummy genomic model</span>
<span class="c1"># model parameters</span>
<span class="n">nfixed</span> <span class="o">=</span> <span class="mi">1</span>      <span class="c1"># number of fixed effects</span>
<span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span>      <span class="c1"># number of traits</span>
<span class="n">nmisc</span> <span class="o">=</span> <span class="mi">0</span>       <span class="c1"># number of miscellaneous random effects</span>
<span class="n">nadditive</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># number of additive marker effects</span>

<span class="c1"># create dummy values</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nfixed</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">u_misc</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nmisc</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">u_a</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">nadditive</span><span class="p">,</span><span class="n">ntrait</span><span class="p">))</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;Trait&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntrait</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create additive linear genomic model</span>
<span class="n">algmod</span> <span class="o">=</span> <span class="n">DenseAdditiveLinearGenomicModel</span><span class="p">(</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">beta</span><span class="p">,</span>
    <span class="n">u_misc</span> <span class="o">=</span> <span class="n">u_misc</span><span class="p">,</span>
    <span class="n">u_a</span> <span class="o">=</span> <span class="n">u_a</span><span class="p">,</span>
    <span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span><span class="p">,</span>
    <span class="n">model_name</span> <span class="o">=</span> <span class="s2">&quot;example&quot;</span><span class="p">,</span>
    <span class="n">params</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span>

<span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">nvrnt</span> <span class="o">=</span> <span class="n">nadditive</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">nchrom</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># create random genotypes</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ploidy</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">nvrnt</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;Taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant chromsome assignments</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nchrom</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker physical positions</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker genetic positions</span>
<span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_genpos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create a genotype matrix from scratch using NumPy arrays</span>
<span class="n">pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
    <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
<span class="p">)</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>

<span class="c1"># calculate genetic variance matrix from GenomicModel</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_gmod</span><span class="p">(</span>
    <span class="n">gmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">ncross</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
<span class="p">)</span>

<span class="c1"># calculate genetic variance matrix from AdditiveLinearGenomicModel</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_algmod</span><span class="p">(</span>
    <span class="n">algmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">ncross</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
<span class="p">)</span>

<span class="c1"># calculate genic variance matrix from GenomicModel</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_gmod</span><span class="p">(</span>
    <span class="n">gmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span>

<span class="c1"># calculate genic variance matrix from AdditiveLinearGenomicModel</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_algmod</span><span class="p">(</span>
    <span class="n">algmod</span> <span class="o">=</span> <span class="n">algmod</span><span class="p">,</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">pgmat</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-variance-matrices-from-hdf5-files">
<h3>Loading variance matrices from HDF5 files<a class="headerlink" href="#loading-variance-matrices-from-hdf5-files" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read genetic variance matrix from HDF5 file</span>
<span class="n">vmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGeneticVarianceMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_vmat.h5&quot;</span><span class="p">)</span>
<span class="n">gvmat</span> <span class="o">=</span> <span class="n">DenseTwoWayDHAdditiveGenicVarianceMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="variance-matrix-properties">
<h2>Variance matrix properties<a class="headerlink" href="#variance-matrix-properties" title="Permalink to this heading">#</a></h2>
<section id="general-properties">
<h3>General properties<a class="headerlink" href="#general-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> general properties</span><a class="headerlink" href="#id3" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat</span></code></p></td>
<td><p>The raw variance matrix pointer</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_ndim</span></code></p></td>
<td><p>The number of dimensions for the variance matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat_shape</span></code></p></td>
<td><p>The variance matrix shape</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">epgc</span></code></p></td>
<td><p>The expected parental genomic contribution for each parental axis</p></td>
</tr>
</tbody>
</table>
</section>
<section id="square-properties">
<h3>Square properties<a class="headerlink" href="#square-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id4">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> square matrix properties</span><a class="headerlink" href="#id4" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nsquare</span></code></p></td>
<td><p>The number of square axes for the variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes</span></code></p></td>
<td><p>The axes indices for the square axes for the variance matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes_len</span></code></p></td>
<td><p>The lengths of the square axes for the variance matrix</p></td>
</tr>
</tbody>
</table>
</section>
<section id="taxa-properties">
<h3>Taxa properties<a class="headerlink" href="#taxa-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id5">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> taxa properties</span><a class="headerlink" href="#id5" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntaxa</span></code></p></td>
<td><p>The number of taxa represented by the variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></td>
<td><p>The names of the taxa</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_axis</span></code></p></td>
<td><p>The matrix axis along which taxa are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code></p></td>
<td><p>An optional taxa group label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_name</span></code></p></td>
<td><p>If taxa are sorted by group: get the names of the groups</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_stix</span></code></p></td>
<td><p>If taxa are sorted by group: get the start indices (inclusive) for each group</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_spix</span></code></p></td>
<td><p>If taxa are sorted by group: get the stop indices (exclusive) for each group</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_len</span></code></p></td>
<td><p>If taxa are sorted by group: get the length of each group</p></td>
</tr>
</tbody>
</table>
</section>
<section id="trait-properties">
<h3>Trait properties<a class="headerlink" href="#trait-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id6">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticVarianceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">GenicVarianceMatrix</span></code> square matrix properties</span><a class="headerlink" href="#id6" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntrait</span></code></p></td>
<td><p>The number of square axes for the variance matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">trait</span></code></p></td>
<td><p>The names of the traits represented by the variance matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">trait_axis</span></code></p></td>
<td><p>The trait axis for the variance matrix</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="copying">
<h2>Copying<a class="headerlink" href="#copying" title="Permalink to this heading">#</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy a genetic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># copy a genic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># deep copy a genetic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># deep copy a genic variance matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="variance-matrix-element-copy-on-manipulation">
<h2>Variance Matrix Element Copy-On-Manipulation<a class="headerlink" href="#variance-matrix-element-copy-on-manipulation" title="Permalink to this heading">#</a></h2>
<section id="adjoining-elements">
<h3>Adjoining elements<a class="headerlink" href="#adjoining-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new variance matrices to demonstrate</span>
<span class="n">newvmat</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">newgvmat</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># adjoin variance matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">)</span>

<span class="c1"># adjoin variance matrices along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">adjoin_trait</span><span class="p">(</span><span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">adjoin_trait</span><span class="p">(</span><span class="n">newgvmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="deleting-elements">
<h3>Deleting elements<a class="headerlink" href="#deleting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># ``delete`` taxa examples</span>
<span class="c1"># ++++++++++++++++++++++++</span>

<span class="c1"># delete first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1">#</span>
<span class="c1"># ``delete`` traits examples</span>
<span class="c1"># ++++++++++++++++++++++++++</span>

<span class="c1"># delete first trait using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first two traits using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>

<span class="c1"># delete first two traits using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">delete_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="inserting-elements">
<h3>Inserting elements<a class="headerlink" href="#inserting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new variance matrix to demonstrate</span>
<span class="n">newvmat</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">newgvmat</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># insert variance matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">)</span>

<span class="c1"># insert variance matrix along the trait axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">insert_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newvmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">insert_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">newgvmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="selecting-elements">
<h3>Selecting elements<a class="headerlink" href="#selecting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># select first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1"># select first two traits using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">select_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">select_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="variance-matrix-element-in-place-manipulation">
<h2>Variance Matrix Element In-Place-Manipulation<a class="headerlink" href="#variance-matrix-element-in-place-manipulation" title="Permalink to this heading">#</a></h2>
<section id="appending-elements">
<h3>Appending elements<a class="headerlink" href="#appending-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># append variance matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span> <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>                  <span class="c1"># append original to copy</span>

<span class="c1"># append variance matrices along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span> <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span><span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_trait</span><span class="p">(</span><span class="n">vmat</span><span class="p">)</span>                  <span class="c1"># append original to copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                  <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_trait</span><span class="p">(</span><span class="n">gvmat</span><span class="p">)</span>                 <span class="c1"># append original to copy</span>
</pre></div>
</div>
</section>
<section id="removing-elements">
<h3>Removing elements<a class="headerlink" href="#removing-elements" title="Permalink to this heading">#</a></h3>
<section id="remove-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> taxa<a class="headerlink" href="#remove-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>           <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span> <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
<section id="remove-traits">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> traits<a class="headerlink" href="#remove-traits" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first trait using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>           <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>          <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                             <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                             <span class="c1"># remove from copy</span>

<span class="c1"># remove first trait using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span> <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>                    <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>                    <span class="c1"># remove from copy</span>

<span class="c1"># remove first trait using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>         <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>        <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span>                           <span class="c1"># remove from copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_trait</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span>                           <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
</section>
<section id="incorporating-elements">
<h3>Incorporating elements<a class="headerlink" href="#incorporating-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># incorp variance matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>      <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>    <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">)</span>                       <span class="c1"># incorporate into copy</span>

<span class="c1"># incorp variance matrix along the trait axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>     <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>   <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmat</span><span class="p">)</span>                       <span class="c1"># incorporate into copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                          <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_trait</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">)</span>                      <span class="c1"># incorporate into copy</span>
</pre></div>
</div>
</section>
<section id="concatenating-matrices">
<h3>Concatenating matrices<a class="headerlink" href="#concatenating-matrices" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># concatenate along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">])</span>

<span class="c1"># concatenate along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">concat_trait</span><span class="p">([</span><span class="n">vmat</span><span class="p">,</span> <span class="n">vmat</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">concat_trait</span><span class="p">([</span><span class="n">gvmat</span><span class="p">,</span> <span class="n">gvmat</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping-and-sorting">
<h2>Grouping and sorting<a class="headerlink" href="#grouping-and-sorting" title="Permalink to this heading">#</a></h2>
<section id="reordering">
<h3>Reordering<a class="headerlink" href="#reordering" title="Permalink to this heading">#</a></h3>
<section id="reorder-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> taxa<a class="headerlink" href="#reorder-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>

<span class="c1"># reorder values along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reorder-traits">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> traits<a class="headerlink" href="#reorder-traits" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntrait</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>

<span class="c1"># reorder values along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_trait</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_trait</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="lexsorting">
<h3>Lexsorting<a class="headerlink" href="#lexsorting" title="Permalink to this heading">#</a></h3>
<section id="lexsort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> taxa<a class="headerlink" href="#lexsort-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for taxa</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the taxa axis</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="lexsort-traits">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> traits<a class="headerlink" href="#lexsort-traits" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for trait</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">vmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="sorting">
<h3>Sorting<a class="headerlink" href="#sorting" title="Permalink to this heading">#</a></h3>
<section id="sort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> taxa<a class="headerlink" href="#sort-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="sort-traits">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> traits<a class="headerlink" href="#sort-traits" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along trait axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_trait</span><span class="p">()</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">trait_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_trait</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping">
<h3>Grouping<a class="headerlink" href="#grouping" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># group taxa</span>
<span class="c1"># ++++++++++</span>

<span class="c1"># sort genetic along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>

<span class="c1"># sort genic variance matrix along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="square-matrix-functions">
<h2>Square matrix functions<a class="headerlink" href="#square-matrix-functions" title="Permalink to this heading">#</a></h2>
<section id="determine-whether-all-square-axes-are-of-equal-length">
<h3>Determine whether all square axes are of equal length<a class="headerlink" href="#determine-whether-all-square-axes-are-of-equal-length" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># boolean value</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">vmat</span><span class="o">.</span><span class="n">is_square</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gvmat</span><span class="o">.</span><span class="n">is_square</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="saving-breeding-value-matrices">
<h2>Saving Breeding Value Matrices<a class="headerlink" href="#saving-breeding-value-matrices" title="Permalink to this heading">#</a></h2>
<section id="write-to-hdf5-file">
<h3>Write to HDF5 file<a class="headerlink" href="#write-to-hdf5-file" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write variance matrices to an HDF5 file</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_vmat.h5&quot;</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="write-to-csv-file">
<h3>Write to CSV file<a class="headerlink" href="#write-to-csv-file" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write variance matrices to a CSV file</span>
<span class="n">vmat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;saved_vmat.csv&quot;</span><span class="p">)</span>
<span class="n">gvmat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;saved_gvmat.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="genomic_models.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Genomic Models</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="../breeding_simulation/genotyping_protocols.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Genotyping Protocols</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#class-family-overview">
   Class Family Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-of-variance-matrix-classes">
   Summary of Variance Matrix Classes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#genetic-variance-matrix-classes">
     Genetic variance matrix classes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#genic-variance-matrix-classes">
     Genic variance matrix classes
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-variance-matrix-modules">
   Loading Variance Matrix Modules
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-genetic-variance-matrix-modules">
     Loading genetic variance matrix modules
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-genic-variance-matrix-modules">
     Loading genic variance matrix modules
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-variance-matrices">
   Creating Variance Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-variance-matrices-from-numpy-arrays">
     Creating variance matrices from NumPy arrays
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-variance-matrices-from-genomic-models">
     Creating variance matrices from genomic models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-variance-matrices-from-hdf5-files">
     Loading variance matrices from HDF5 files
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#variance-matrix-properties">
   Variance matrix properties
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#general-properties">
     General properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#square-properties">
     Square properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#taxa-properties">
     Taxa properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#trait-properties">
     Trait properties
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#copying">
   Copying
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#variance-matrix-element-copy-on-manipulation">
   Variance Matrix Element Copy-On-Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#adjoining-elements">
     Adjoining elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#deleting-elements">
     Deleting elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inserting-elements">
     Inserting elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#selecting-elements">
     Selecting elements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#variance-matrix-element-in-place-manipulation">
   Variance Matrix Element In-Place-Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#appending-elements">
     Appending elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#removing-elements">
     Removing elements
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#remove-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         remove
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#remove-traits">
       <code class="docutils literal notranslate">
        <span class="pre">
         remove
        </span>
       </code>
       traits
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#incorporating-elements">
     Incorporating elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#concatenating-matrices">
     Concatenating matrices
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-and-sorting">
   Grouping and sorting
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reordering">
     Reordering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reorder-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         reorder
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reorder-traits">
       <code class="docutils literal notranslate">
        <span class="pre">
         reorder
        </span>
       </code>
       traits
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#lexsorting">
     Lexsorting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#lexsort-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         lexsort
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#lexsort-traits">
       <code class="docutils literal notranslate">
        <span class="pre">
         lexsort
        </span>
       </code>
       traits
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sorting">
     Sorting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sort-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         sort
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sort-traits">
       <code class="docutils literal notranslate">
        <span class="pre">
         sort
        </span>
       </code>
       traits
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#grouping">
     Grouping
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#square-matrix-functions">
   Square matrix functions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#determine-whether-all-square-axes-are-of-equal-length">
     Determine whether all square axes are of equal length
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#saving-breeding-value-matrices">
   Saving Breeding Value Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#write-to-hdf5-file">
     Write to HDF5 file
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#write-to-csv-file">
     Write to CSV file
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2023, Robert Shrote.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>