
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pybrops.popgen.gmat.DenseGenotypeMatrix &#8212; pybrops 1.0.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../../_static/documentation_options.js?v=1ed6394b"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pybrops/popgen/gmat/DenseGenotypeMatrix';</script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.2 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pybrops.popg...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pybrops.popgen.gmat.DenseGenotypeMatrix</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module providing implementations of dense genotype matrices and associated error</span>
<span class="sd">checking routines.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;DenseGenotypeMatrix&quot;</span><span class="p">,</span>
    <span class="s2">&quot;check_is_DenseGenotypeMatrix&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Real</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Sequence</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span>
<span class="kn">import</span> <span class="nn">cyvcf2</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">numpy.typing</span> <span class="kn">import</span> <span class="n">DTypeLike</span>
<span class="kn">from</span> <span class="nn">numpy.typing</span> <span class="kn">import</span> <span class="n">ArrayLike</span>

<span class="kn">from</span> <span class="nn">pybrops.core.error.error_io_python</span> <span class="kn">import</span> <span class="n">check_file_exists</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_int</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_is_ndarray</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_dtype_is_int8</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_attr_python</span> <span class="kn">import</span> <span class="n">error_readonly</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_h5py</span> <span class="kn">import</span> <span class="n">check_h5py_File_has_group</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_h5py</span> <span class="kn">import</span> <span class="n">check_h5py_File_is_readable</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_h5py</span> <span class="kn">import</span> <span class="n">check_h5py_File_is_writable</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_ndim</span>
<span class="kn">from</span> <span class="nn">pybrops.core.mat.Matrix</span> <span class="kn">import</span> <span class="n">Matrix</span>
<span class="kn">from</span> <span class="nn">pybrops.core.mat.DenseTaxaVariantMatrix</span> <span class="kn">import</span> <span class="n">DenseTaxaVariantMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">h5py_File_read_int</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">h5py_File_read_ndarray</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">h5py_File_read_ndarray_int8</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">h5py_File_read_ndarray_utf8</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">h5py_File_write_dict</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.DenseGeneticMappableMatrix</span> <span class="kn">import</span> <span class="n">DenseGeneticMappableMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.GenotypeMatrix</span> <span class="kn">import</span> <span class="n">GenotypeMatrix</span>

<div class="viewcode-block" id="DenseGenotypeMatrix">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix">[docs]</a>
<span class="k">class</span> <span class="nc">DenseGenotypeMatrix</span><span class="p">(</span>
        <span class="n">DenseTaxaVariantMatrix</span><span class="p">,</span>
        <span class="n">DenseGeneticMappableMatrix</span><span class="p">,</span>
        <span class="n">GenotypeMatrix</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A concrete class for unphased genoypte matrix objects.</span>

<span class="sd">    The purpose of this concrete class is to implement functionality for:</span>
<span class="sd">        1) Genotype matrix ploidy and phase metadata.</span>
<span class="sd">        2) Genotype matrix format conversion.</span>
<span class="sd">        3) Genotype matrix allele counting routines.</span>
<span class="sd">        4) Genotype matrix genotype counting routines.</span>
<span class="sd">        5) Loading genotype matrices from VCF and HDF5.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">########################## Special Object Methods ##########################</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">mat</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">taxa</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">taxa_grp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">vrnt_xoprob</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapalt</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">vrnt_hapref</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_mask</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mat : numpy.ndarray</span>
<span class="sd">            An int8 haplotype matrix. Must be {0,1,2} format.</span>
<span class="sd">        </span>
<span class="sd">        taxa : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(n,)`` containing taxa names.</span>
<span class="sd">            If ``None``, do not store any taxa name information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``n`` is the number of taxa.</span>

<span class="sd">        taxa_grp : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(n,)`` containing taxa groupings.</span>
<span class="sd">            If ``None``, do not store any taxa group information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``n`` is the number of taxa.</span>

<span class="sd">        vrnt_chrgrp : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant chromosome</span>
<span class="sd">            group labels. If ``None``, do not store any variant chromosome group</span>
<span class="sd">            label information.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_phypos : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant chromosome</span>
<span class="sd">            physical positions. If ``None``, do not store any variant chromosome</span>
<span class="sd">            physical position information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_name : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant names.</span>
<span class="sd">            If ``None``, do not store any variant names.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_genpos : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant chromosome</span>
<span class="sd">            genetic positions. If ``None``, do not store any variant chromosome</span>
<span class="sd">            genetic position information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_xoprob : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant crossover</span>
<span class="sd">            probabilities. If ``None``, do not store any variant crossover</span>
<span class="sd">            probabilities.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_hapgrp : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant haplotype</span>
<span class="sd">            group labels. If ``None``, do not store any variant haplotype group</span>
<span class="sd">            label information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_hapalt : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant alternative</span>
<span class="sd">            alleles. If ``None``, do not store any variant alternative allele</span>
<span class="sd">            information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_hapref : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing variant reference</span>
<span class="sd">            alleles. If ``None``, do not store any variant reference allele</span>
<span class="sd">            information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        vrnt_mask : numpy.ndarray, None</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing a variant mask.</span>
<span class="sd">            If ``None``, do not store any variant mask information.</span>
<span class="sd">        </span>
<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of variants.</span>

<span class="sd">        ploidy : int</span>
<span class="sd">            The ploidy represented by the genotype matrix. This only represents</span>
<span class="sd">            ploidy of the reproductive habit. If the organism represented is an</span>
<span class="sd">            allopolyploid (e.g. hexaploid wheat), the ploidy is 2 since it</span>
<span class="sd">            reproduces in a diploid manner.</span>
<span class="sd">        </span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># set ploidy</span>
        <span class="n">check_is_int</span><span class="p">(</span><span class="n">ploidy</span><span class="p">,</span> <span class="s2">&quot;ploidy&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>

        <span class="c1"># TODO: Add a mat_format option to store as {0,1,2}, {-1,0,1}, etc.</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">vrnt_xoprob</span><span class="p">,</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">vrnt_hapgrp</span><span class="p">,</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">vrnt_hapalt</span><span class="p">,</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">vrnt_hapref</span><span class="p">,</span>
            <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="n">vrnt_mask</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

    <span class="c1">#################### Matrix copying ####################</span>
    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the the matrix.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A shallow copy of the matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># create new object</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mat</span><span class="p">),</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa</span><span class="p">),</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp</span><span class="p">),</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">),</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">),</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_name</span><span class="p">),</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">),</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_xoprob</span><span class="p">),</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapgrp</span><span class="p">),</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapalt</span><span class="p">),</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapref</span><span class="p">),</span>
            <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_mask</span><span class="p">),</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="p">)</span>
        <span class="c1"># copy taxa metadata</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_name</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_stix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_spix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_len</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_len</span><span class="p">)</span>

        <span class="c1"># copy variant metadata</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="k">def</span> <span class="nf">__deepcopy__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the matrix.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : Matrix</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># create new object</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mat</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_name</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_xoprob</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapgrp</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapalt</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapref</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_mask</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="p">)</span>
        <span class="c1"># copy taxa metadata</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_name</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_stix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_spix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_len</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_len</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>

        <span class="c1"># copy variant metadata</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="c1">########### Miscellaneous special functions ############</span>
    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return repr(self).</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : str</span>
<span class="sd">            A representation of the object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;&lt;</span><span class="si">{0}</span><span class="s2"> of shape (nphase = </span><span class="si">{1}</span><span class="s2">, ntaxa = </span><span class="si">{2}</span><span class="s2">, nvrnt = </span><span class="si">{3}</span><span class="s2">) with ploidy = </span><span class="si">{4}</span><span class="s2"> at </span><span class="si">{5}</span><span class="s2">&gt;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nphase</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="nb">hex</span><span class="p">(</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">)),</span>
        <span class="p">)</span>

    <span class="c1">############################ Object Properties #############################</span>

    <span class="c1">############## Genotype Data Properites ##############</span>
    <span class="nd">@DenseTaxaVariantMatrix</span><span class="o">.</span><span class="n">mat</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">mat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;mat&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_int8</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;mat&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;mat&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="c1"># check_ndarray_in_interval(value, &quot;mat&quot;, 0, self._ploidy + 1)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1">############## General matrix properties ###############</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">ploidy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Genome ploidy number represented by matrix.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span>
    <span class="nd">@ploidy</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">ploidy</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set matrix ploidy number&quot;&quot;&quot;</span>
        <span class="n">error_readonly</span><span class="p">(</span><span class="s2">&quot;ploidy&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">nphase</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of chromosome phases represented by the matrix.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="nd">@nphase</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">nphase</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set number of phases&quot;&quot;&quot;</span>
        <span class="n">error_readonly</span><span class="p">(</span><span class="s2">&quot;nphase&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mat_format</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Matrix representation format.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;{0,1,2}&quot;</span>
    <span class="nd">@mat_format</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">mat_format</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set matrix representation format&quot;&quot;&quot;</span>
        <span class="n">error_readonly</span><span class="p">(</span><span class="s2">&quot;mat_format&quot;</span><span class="p">)</span>

    <span class="c1">############### Taxa Metadata Properites ###############</span>
    <span class="nd">@DenseTaxaVariantMatrix</span><span class="o">.</span><span class="n">taxa_axis</span><span class="o">.</span><span class="n">getter</span>
    <span class="k">def</span> <span class="nf">taxa_axis</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get taxa axis number&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mi">0</span>

    <span class="c1">############# Variant Metadata Properites ##############</span>
    <span class="nd">@DenseTaxaVariantMatrix</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="o">.</span><span class="n">getter</span>
    <span class="k">def</span> <span class="nf">vrnt_axis</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get variant axis&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mi">1</span>

    <span class="c1">############################## Object Methods ##############################</span>

    <span class="c1">#################### Matrix copying ####################</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.copy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.copy">[docs]</a>
    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the DenseGenotypeMatrix.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A shallow copy of the original DenseGenotypeMatrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__copy__</span><span class="p">()</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.deepcopy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.deepcopy">[docs]</a>
    <span class="k">def</span> <span class="nf">deepcopy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the DenseGenotypeMatrix.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A deep copy of the original DenseGenotypeMatrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__deepcopy__</span><span class="p">(</span><span class="n">memo</span><span class="p">)</span></div>


    <span class="c1">######### Matrix element copy-on-manipulation ##########</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.adjoin_taxa">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.adjoin_taxa">[docs]</a>
    <span class="k">def</span> <span class="nf">adjoin_taxa</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">values</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Matrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">taxa</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">taxa_grp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add additional elements to the end of the Matrix along the taxa axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        values : Matrix, numpy.ndarray</span>
<span class="sd">            Values are appended to adjoin to the Matrix.</span>
<span class="sd">        taxa : numpy.ndarray</span>
<span class="sd">            Taxa names to adjoin to the Matrix.</span>
<span class="sd">            If values is a DenseHaplotypeMatrix that has a non-None</span>
<span class="sd">            taxa field, providing this argument overwrites the field.</span>
<span class="sd">        taxa_grp : numpy.ndarray</span>
<span class="sd">            Taxa groups to adjoin to the Matrix.</span>
<span class="sd">            If values is a DenseHaplotypeMatrix that has a non-None</span>
<span class="sd">            taxa_grp field, providing this argument overwrites the field.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : Matrix</span>
<span class="sd">            A copy of mat with values appended to axis. Note that adjoin does</span>
<span class="sd">            not occur in-place: a new Matrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">values</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.adjoin_vrnt">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.adjoin_vrnt">[docs]</a>
    <span class="k">def</span> <span class="nf">adjoin_vrnt</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">values</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Matrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_xoprob</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapalt</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapref</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_mask</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add additional elements to the end of the Matrix along the variant axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        values : Matrix, numpy.ndarray</span>
<span class="sd">            Values are appended to adjoin to the Matrix.</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            Variant chromosome groups to adjoin to the Matrix.</span>
<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            Variant chromosome physical positions to adjoin to the Matrix.</span>
<span class="sd">        vrnt_name : numpy.ndarray</span>
<span class="sd">            Variant names to adjoin to the Matrix.</span>
<span class="sd">        vrnt_genpos : numpy.ndarray</span>
<span class="sd">            Variant chromosome genetic positions to adjoin to the Matrix.</span>
<span class="sd">        vrnt_xoprob : numpy.ndarray</span>
<span class="sd">            Sequential variant crossover probabilities to adjoin to the Matrix.</span>
<span class="sd">        vrnt_hapgrp : numpy.ndarray</span>
<span class="sd">            Variant haplotype labels to adjoin to the Matrix.</span>
<span class="sd">        vrnt_hapalt : numpy.ndarray</span>
<span class="sd">            Variant haplotype sequence.</span>
<span class="sd">        vrnt_hapref : numpy.ndarray</span>
<span class="sd">            Variant haplotype reference sequence.</span>
<span class="sd">        vrnt_mask : numpy.ndarray</span>
<span class="sd">            Variant mask to adjoin to the Matrix.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : Matrix</span>
<span class="sd">            A copy of mat with values appended to axis. Note that adjoin does</span>
<span class="sd">            not occur in-place: a new Matrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">adjoin_vrnt</span><span class="p">(</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">values</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">vrnt_xoprob</span><span class="p">,</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">vrnt_hapgrp</span><span class="p">,</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">vrnt_hapalt</span><span class="p">,</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">vrnt_hapref</span><span class="p">,</span>
            <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="n">vrnt_mask</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.delete_taxa">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.delete_taxa">[docs]</a>
    <span class="k">def</span> <span class="nf">delete_taxa</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">obj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Delete sub-arrays along the taxa axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        obj : int, slice, or Sequence of ints</span>
<span class="sd">            Indicate indices of sub-arrays to remove along the specified axis.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A DenseGenotypeMatrix with deleted elements. Note that concat does not occur</span>
<span class="sd">            in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span>
            <span class="n">obj</span> <span class="o">=</span> <span class="n">obj</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.delete_vrnt">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.delete_vrnt">[docs]</a>
    <span class="k">def</span> <span class="nf">delete_vrnt</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">obj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Delete sub-arrays along the variant axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        obj : int, slice, or Sequence of ints</span>
<span class="sd">            Indicate indices of sub-arrays to remove along the specified axis.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A DenseGenotypeMatrix with deleted elements. Note that concat does not occur</span>
<span class="sd">            in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">(</span>
            <span class="n">obj</span> <span class="o">=</span> <span class="n">obj</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.insert_taxa">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.insert_taxa">[docs]</a>
    <span class="k">def</span> <span class="nf">insert_taxa</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">obj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="n">values</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Matrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">taxa</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">taxa_grp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Insert values along the taxa axis before the given indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        obj: int, slice, or Sequence of ints</span>
<span class="sd">            Object that defines the index or indices before which values is</span>
<span class="sd">            inserted.</span>
<span class="sd">        values : Matrix, numpy.ndarray</span>
<span class="sd">            Values to insert into the matrix.</span>
<span class="sd">        taxa : numpy.ndarray</span>
<span class="sd">            Taxa names to insert into the Matrix.</span>
<span class="sd">        taxa_grp : numpy.ndarray</span>
<span class="sd">            Taxa groups to insert into the Matrix.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A DenseGenotypeMatrix with values inserted. Note that insert does not occur</span>
<span class="sd">            in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># create output</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span>
            <span class="n">obj</span> <span class="o">=</span> <span class="n">obj</span><span class="p">,</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">values</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.insert_vrnt">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.insert_vrnt">[docs]</a>
    <span class="k">def</span> <span class="nf">insert_vrnt</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">obj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="n">values</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Matrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_xoprob</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapgrp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapalt</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_hapref</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">vrnt_mask</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Insert values along the variant axis before the given indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        obj: int, slice, or Sequence of ints</span>
<span class="sd">            Object that defines the index or indices before which values is</span>
<span class="sd">            inserted.</span>
<span class="sd">        values : array_like</span>
<span class="sd">            Values to insert into the matrix.</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            Variant chromosome groups to insert into the Matrix.</span>
<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            Variant chromosome physical positions to insert into the Matrix.</span>
<span class="sd">        vrnt_name : numpy.ndarray</span>
<span class="sd">            Variant names to insert into the Matrix.</span>
<span class="sd">        vrnt_genpos : numpy.ndarray</span>
<span class="sd">            Variant chromosome genetic positions to insert into the Matrix.</span>
<span class="sd">        vrnt_xoprob : numpy.ndarray</span>
<span class="sd">            Sequential variant crossover probabilities to insert into the Matrix.</span>
<span class="sd">        vrnt_hapgrp : numpy.ndarray</span>
<span class="sd">            Variant haplotype labels to insert into the Matrix.</span>
<span class="sd">        vrnt_hapalt : numpy.ndarray</span>
<span class="sd">            Variant alternative haplotype labels to insert into the Matrix.</span>
<span class="sd">        vrnt_hapref : numpy.ndarray</span>
<span class="sd">            Variant reference haplotype labels to insert into the Matrix.</span>
<span class="sd">        vrnt_mask : numpy.ndarray</span>
<span class="sd">            Variant mask to insert into the Matrix.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            A DenseGenotypeMatrix with values inserted. Note that insert does not occur</span>
<span class="sd">            in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># create output</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">insert_vrnt</span><span class="p">(</span>
            <span class="n">obj</span> <span class="o">=</span> <span class="n">obj</span><span class="p">,</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">values</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">vrnt_xoprob</span><span class="p">,</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">vrnt_hapgrp</span><span class="p">,</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">vrnt_hapalt</span><span class="p">,</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">vrnt_hapref</span><span class="p">,</span>
            <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="n">vrnt_mask</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.select_taxa">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.select_taxa">[docs]</a>
    <span class="k">def</span> <span class="nf">select_taxa</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">indices</span><span class="p">:</span> <span class="n">ArrayLike</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Select certain values from the Matrix along the taxa axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        indices : array_like (Nj, ...)</span>
<span class="sd">            The indices of the values to select.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            The output DenseGenotypeMatrix with values selected. Note that select does not</span>
<span class="sd">            occur in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">indices</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.select_vrnt">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.select_vrnt">[docs]</a>
    <span class="k">def</span> <span class="nf">select_vrnt</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">indices</span><span class="p">:</span> <span class="n">ArrayLike</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Select certain values from the Matrix along the variant axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        indices : array_like (Nj, ...)</span>
<span class="sd">            The indices of the values to select.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            The output DenseGenotypeMatrix with values selected. Note that select does not</span>
<span class="sd">            occur in-place: a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">select_vrnt</span><span class="p">(</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">indices</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.concat_taxa">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.concat_taxa">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">concat_taxa</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">mats</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Concatenate list of Matrix together along the taxa axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mats : Sequence of Matrix</span>
<span class="sd">            List of Matrix to concatenate. The matrices must have the same</span>
<span class="sd">            shape, except in the dimension corresponding to axis.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            The concatenated DenseGenotypeMatrix. Note that concat does not occur in-place:</span>
<span class="sd">            a new DenseGenotypeMatrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">cls</span><span class="p">)</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span>
            <span class="n">mats</span> <span class="o">=</span> <span class="n">mats</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">mats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.concat_vrnt">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.concat_vrnt">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">concat_vrnt</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">mats</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Concatenate list of Matrix together along the variant axis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mats : Sequence of Matrix</span>
<span class="sd">            List of Matrix to concatenate. The matrices must have the same</span>
<span class="sd">            shape, except in the dimension corresponding to axis.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : Matrix</span>
<span class="sd">            The concatenated matrix. Note that concat does not occur in-place:</span>
<span class="sd">            a new Matrix is allocated and filled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DenseGenotypeMatrix</span><span class="p">,</span> <span class="bp">cls</span><span class="p">)</span><span class="o">.</span><span class="n">concat_vrnt</span><span class="p">(</span>
            <span class="n">mats</span> <span class="o">=</span> <span class="n">mats</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">mats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">################## Matrix conversion ###################</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.mat_asformat">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.mat_asformat">[docs]</a>
    <span class="k">def</span> <span class="nf">mat_asformat</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get mat in a specific format type.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        format : str</span>
<span class="sd">            Desired output format. Options are &quot;{0,1,2}&quot;, &quot;{-1,0,1}&quot;, &quot;{-1,m,1}&quot;.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Matrix in the desired output format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;{0,1,2}&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;{-1,0,1}&quot;</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mat</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;{-1,m,1}&quot;</span><span class="p">:</span>
            <span class="c1"># OPTIMIZE: there&#39;s probably a matrix multiplication way to do this instead of if-else</span>
            <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mat</span> <span class="o">-</span> <span class="mf">1.0</span>    <span class="c1"># (n,p) float64</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="n">view</span> <span class="o">=</span> <span class="n">out</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span>
                <span class="n">mean</span> <span class="o">=</span> <span class="n">view</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
                <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">view</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">out</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean</span>
            <span class="k">return</span> <span class="n">out</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Format not recognized. Options are &quot;{0,1,2}&quot;, &quot;{-1,0,1}&quot;, &quot;{-1,m,1}&quot;.&#39;</span><span class="p">)</span></div>


    <span class="c1">############## Matrix summary statistics ###############</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.tacount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.tacount">[docs]</a>
    <span class="k">def</span> <span class="nf">tacount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allele count of the non-zero allele within each taxon.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the accumulator and returned array.</span>
<span class="sd">            If ``None``, use the native accumulator type (int or float).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(n,p)`` containing allele counts of the</span>
<span class="sd">            allele coded as ``1`` for all ``n`` individuals, for all ``p`` loci.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get accumulator type</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">issubdtype</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">integer</span><span class="p">):</span>
                <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span>
            <span class="k">elif</span> <span class="n">numpy</span><span class="o">.</span><span class="n">issubdtype</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">floating</span><span class="p">):</span>
                <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No default accumulator type for GenotypeMatrix dtype </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">dtype</span><span class="p">))</span>
        <span class="c1"># calculate the taxa allele sums</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.tafreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.tafreq">[docs]</a>
    <span class="k">def</span> <span class="nf">tafreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allele frequency of the non-zero allele within each taxon.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(n,p)`` containing allele frequencies of</span>
<span class="sd">            the allele coded as ``1`` for all ``n`` individuals, for all ``p``</span>
<span class="sd">            loci.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">rnphase</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span>         <span class="c1"># take the reciprocal of the ploidy number</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">rnphase</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span>           <span class="c1"># take sum across the phase axis (0) and divide by ploidy</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>               <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>      <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>          <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>       <span class="c1"># convert to correct dtype</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.acount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.acount">[docs]</a>
    <span class="k">def</span> <span class="nf">acount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allele count of the non-zero allele across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the accumulator and returned array. If ``None``, use the native integer type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing allele counts of the</span>
<span class="sd">            allele coded as ``1`` for all ``p`` loci.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># take sum across the taxa axis</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.afreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.afreq">[docs]</a>
    <span class="k">def</span> <span class="nf">afreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allele frequency of the non-zero allele across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing allele frequencies of</span>
<span class="sd">            the allele coded as ``1`` for all ``p`` loci.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>              <span class="c1"># get ploidy * ntaxa</span>
        <span class="n">rnphase</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">denom</span>                           <span class="c1"># take 1 / (ploidy * ntaxa)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">rnphase</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># take sum across the taxa axis (0) and divide by nphase</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>                           <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>                  <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>                      <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>                   <span class="c1"># convert to correct dtype</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.afixed">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.afixed">[docs]</a>
    <span class="k">def</span> <span class="nf">afixed</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine allele fixation for loci across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing indicator variables </span>
<span class="sd">            for whether a locus is fixed at a particular locus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get the allele frequency</span>
        <span class="n">afreq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>

        <span class="c1"># determine whether the allele frequency is equal to 0.0 or 1.0</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">afreq</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">afreq</span> <span class="o">==</span> <span class="mf">1.0</span><span class="p">)</span>

        <span class="c1"># convert to specific dtype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>                           <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>                  <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>                      <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>                   <span class="c1"># convert to correct dtype</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.apoly">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.apoly">[docs]</a>
    <span class="k">def</span> <span class="nf">apoly</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allele polymorphism presence or absense across all loci.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing indicator variables for</span>
<span class="sd">            whether the locus is polymorphic.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get the allele frequency</span>
        <span class="n">afreq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>

        <span class="c1"># determine whether the allele frequency is between 0.0 and 1.0</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">afreq</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">afreq</span> <span class="o">&lt;</span> <span class="mf">1.0</span><span class="p">)</span>

        <span class="c1"># convert to specific dtype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>                           <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>                  <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>                      <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>                   <span class="c1"># convert to correct dtype</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.maf">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.maf">[docs]</a>
    <span class="k">def</span> <span class="nf">maf</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Minor allele frequency across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` of shape ``(p,)`` containing allele frequencies for</span>
<span class="sd">            the minor allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">afreq</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>     <span class="c1"># get allele frequencies</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">out</span> <span class="o">&gt;</span> <span class="mf">0.5</span>            <span class="c1"># create mask of allele frequencies &gt; 0.5</span>
        <span class="n">out</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">out</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="c1"># take 1 - allele frequency</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.meh">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.meh">[docs]</a>
    <span class="k">def</span> <span class="nf">meh</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Real</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Mean expected heterozygosity across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : Real</span>
<span class="sd">            A number representing the mean expected heterozygous.</span>
<span class="sd">            If ``dtype`` is ``None``, then a native 64-bit floating point is</span>
<span class="sd">            returned. Otherwise, of type specified by ``dtype``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># OPTIMIZE: take dot product with allele counts, then divide?</span>
        <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>                    <span class="c1"># get haplotype frequency (p)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span>         <span class="c1"># take p*(1-p) across all loci and sum the products</span>
        <span class="n">rnphase</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">nvrnt</span>  <span class="c1"># ploidy / nvrnt</span>
        <span class="n">out</span> <span class="o">*=</span> <span class="n">rnphase</span>                      <span class="c1"># multiply summation by (nphase / nvrnt)</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>               <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>      <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>          <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>       <span class="c1"># convert to correct dtype</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.gtcount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.gtcount">[docs]</a>
    <span class="k">def</span> <span class="nf">gtcount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gather genotype counts for homozygous major, heterozygous, homozygous</span>
<span class="sd">        minor for all individuals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` array of shape ``(g,p)`` containing allele counts</span>
<span class="sd">            across all ``p`` loci for each of ``g`` genotype combinations.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``out[0]`` is the count of ``0`` genotype across all loci</span>
<span class="sd">            - ``out[1]`` is the count of ``1`` genotype across all loci</span>
<span class="sd">            - ``out[2]`` is the count of ``2`` genotype across all loci</span>
<span class="sd">            - ``...``</span>
<span class="sd">            - ``out[g-1]`` is the count of ``g-1`` genotype across all loci</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ngt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nphase</span> <span class="o">+</span> <span class="mi">1</span>   <span class="c1"># get number of genotype combos</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mat</span>         <span class="c1"># get matrix</span>

        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span>      <span class="c1"># allocate output array</span>
            <span class="p">(</span><span class="n">ngt</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">),</span>  <span class="c1"># shape (g, p)</span>
            <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int64&#39;</span>       <span class="c1"># int64 dtype</span>
        <span class="p">)</span>

        <span class="c1"># get correct axis to sum across</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_axis</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ngt</span><span class="p">):</span>                <span class="c1"># for each genotype combo</span>
            <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">mat</span> <span class="o">==</span> <span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="p">)</span>   <span class="c1"># record counts for genotype &#39;i&#39;</span>

        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>               <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>      <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>          <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>       <span class="c1"># convert to correct dtype</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.gtfreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.gtfreq">[docs]</a>
    <span class="k">def</span> <span class="nf">gtfreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DTypeLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gather genotype frequencies for homozygous major, heterozygous,</span>
<span class="sd">        homozygous minor across all individuals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dtype : dtype, None</span>
<span class="sd">            The data type of the returned array. </span>
<span class="sd">            If ``None``, use the native type.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A ``numpy.ndarray`` array of shape ``(g,p)`` containing haplotype counts</span>
<span class="sd">            across all ``p`` loci for all ``g`` genotype combinations.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``out[0]`` is the frequency of ``0`` genotype across all loci</span>
<span class="sd">            - ``out[1]`` is the frequency of ``1`` genotype across all loci</span>
<span class="sd">            - ``out[2]`` is the frequency of ``2`` genotype across all loci</span>
<span class="sd">            - ``...``</span>
<span class="sd">            - ``out[g-1]`` is the frequency of ``g-1`` genotype across all loci</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">recip</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">ntaxa</span>        <span class="c1"># get reciprocal of number of taxa</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">recip</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">gtcount</span><span class="p">()</span>    <span class="c1"># calculate genotype frequencies</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>           <span class="c1"># if dtype is specified</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>  <span class="c1"># ensure conversion to dtype class</span>
            <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">dtype</span><span class="p">:</span>      <span class="c1"># if output dtype and desired are different</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>   <span class="c1"># convert to correct dtype</span>
        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">################### Matrix File I/O ####################</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.to_hdf5">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.to_hdf5">[docs]</a>
    <span class="k">def</span> <span class="nf">to_hdf5</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">,</span><span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">],</span> 
            <span class="n">groupname</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write GenotypeMatrix to an HDF5 file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str, Path, h5py.File</span>
<span class="sd">            If ``str``, an HDF5 file name to which to write. File is closed after writing.</span>
<span class="sd">            If ``h5py.File``, an opened HDF5 file to which to write. File is not closed after writing.</span>

<span class="sd">        groupname : str, None</span>
<span class="sd">            If ``str``, an HDF5 group name under which GenotypeMatrix data is stored.</span>
<span class="sd">            If None, GenotypeMatrix is written to the base HDF5 group.</span>

<span class="sd">        overwrite : bool</span>
<span class="sd">            Whether to overwrite values in an HDF5 file if a field already exists.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">########################################################</span>
        <span class="c1">############ process ``filename`` argument #############</span>

        <span class="c1"># HDF5 file object</span>
        <span class="n">h5file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># if we have a string or Path, open HDF5 file in append (``r+``) mode</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">)):</span>
            <span class="n">h5file</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span>

        <span class="c1"># elif we have an h5py.File, make sure mode is writable, and copy pointer</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">check_h5py_File_is_writable</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="n">h5file</span> <span class="o">=</span> <span class="n">filename</span>
        
        <span class="c1"># else raise TypeError</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;``filename`` must be of type ``str``, ``Path``, or ``h5py.File`` but received type ``</span><span class="si">{0}</span><span class="s2">``&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">type</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1">########################################################</span>
        <span class="c1">############ process ``groupname`` argument ############</span>

        <span class="c1"># if we have a string</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">groupname</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c1"># if last character in string is not &#39;/&#39;, add &#39;/&#39; to end of string</span>
            <span class="k">if</span> <span class="n">groupname</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;/&#39;</span><span class="p">:</span>
                <span class="n">groupname</span> <span class="o">+=</span> <span class="s1">&#39;/&#39;</span>
        
        <span class="c1"># else if ``groupname`` is None, set ``groupname`` to empty string</span>
        <span class="k">elif</span> <span class="n">groupname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">groupname</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        
        <span class="c1"># else raise error</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;``groupname`` must be of type ``str`` or ``None`` but received type ``</span><span class="si">{0}</span><span class="s2">``&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">type</span><span class="p">(</span><span class="n">groupname</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1">########################################################</span>
        <span class="c1">#### write data to HDF5 file and (optionally) close ####</span>

        <span class="c1"># data dictionary</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mat&quot;</span>               <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mat</span><span class="p">,</span>
            <span class="s2">&quot;taxa&quot;</span>              <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp&quot;</span>          <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_phypos&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_name&quot;</span>         <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_name</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_genpos&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_xoprob&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_xoprob</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapgrp&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapgrp</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapalt&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapalt</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapref&quot;</span>       <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_hapref</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_mask&quot;</span>         <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_mask</span><span class="p">,</span>
            <span class="s2">&quot;ploidy&quot;</span>            <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
            <span class="c1"># metadata</span>
            <span class="s2">&quot;taxa_grp_name&quot;</span>     <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_name</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_stix&quot;</span>     <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_stix</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_spix&quot;</span>     <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_spix</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_len&quot;</span>      <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_grp_len</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_name&quot;</span>  <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span>  <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span>  <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_len&quot;</span>   <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="c1"># save data</span>
        <span class="n">h5py_File_write_dict</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">)</span>

        <span class="c1"># close the file, only if the provided filename was a string or Path and not a h5py.File.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">)):</span>
            <span class="n">h5file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


    <span class="c1">############################## Class Methods ###############################</span>

    <span class="c1">################### Matrix File I/O ####################</span>
<div class="viewcode-block" id="DenseGenotypeMatrix.from_hdf5">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.from_hdf5">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_hdf5</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">,</span><span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">],</span> 
            <span class="n">groupname</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read GenotypeMatrix from an HDF5 file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str, Path, h5py.File</span>
<span class="sd">            If ``str`` or ``Path``, an HDF5 file name from which to read. File is closed after reading.</span>
<span class="sd">            If ``h5py.File``, an opened HDF5 file from which to read. File is not closed after reading.</span>
<span class="sd">        groupname : str, None</span>
<span class="sd">            If ``str``, an HDF5 group name under which GenotypeMatrix data is stored.</span>
<span class="sd">            If ``None``, GenotypeMatrix is read from base HDF5 group.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            A genotype matrix read from file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">########################################################</span>
        <span class="c1">############ process ``filename`` argument #############</span>

        <span class="c1"># HDF5 file object</span>
        <span class="n">h5file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># if we have a string or Path, open HDF5 file in read (``r``) mode</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">)):</span>
            <span class="n">check_file_exists</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="n">h5file</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>

        <span class="c1"># elif we have an ``h5py.File``, make sure mode is in at least ``r`` mode, and copy pointer</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">check_h5py_File_is_readable</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="n">h5file</span> <span class="o">=</span> <span class="n">filename</span>
        
        <span class="c1"># else raise TypeError</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;``filename`` must be of type ``str``, ``Path``, or ``h5py.File`` but received type ``</span><span class="si">{0}</span><span class="s2">``&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">type</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1">########################################################</span>
        <span class="c1">############ process ``groupname`` argument ############</span>

        <span class="c1"># if we have a string</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">groupname</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c1"># FIXME: errors if groupname == &quot;&quot; or &quot;/&quot;</span>
            <span class="c1"># if the group does not exist in the file, close and raise error</span>
            <span class="n">check_h5py_File_has_group</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span><span class="p">)</span>

            <span class="c1"># if last character in string is not &#39;/&#39;, add &#39;/&#39; to end of string</span>
            <span class="k">if</span> <span class="n">groupname</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;/&#39;</span><span class="p">:</span>
                <span class="n">groupname</span> <span class="o">+=</span> <span class="s1">&#39;/&#39;</span>
        
        <span class="c1"># else if ``groupname`` is None, set ``groupname`` to empty string</span>
        <span class="k">elif</span> <span class="n">groupname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">groupname</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        
        <span class="c1"># else raise error</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;``groupname`` must be of type ``str`` or ``None`` but received type ``</span><span class="si">{0}</span><span class="s2">``&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">type</span><span class="p">(</span><span class="n">groupname</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1">########################################################</span>
        <span class="c1">######## check that we have all required fields ########</span>

        <span class="c1"># all required arguments</span>
        <span class="n">required_fields</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;mat&quot;</span><span class="p">]</span>

        <span class="c1"># for each required field, check if the field exists in the HDF5 file.</span>
        <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">required_fields</span><span class="p">:</span>
            <span class="n">check_h5py_File_has_group</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="n">field</span><span class="p">)</span>
        
        <span class="c1">########################################################</span>
        <span class="c1">### read data from HDF5 file and (optionally) close ####</span>
        
        <span class="c1"># output dictionary</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mat&quot;</span>               <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;taxa&quot;</span>              <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp&quot;</span>          <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_phypos&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_name&quot;</span>         <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_genpos&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_xoprob&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapgrp&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapalt&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_hapref&quot;</span>       <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_mask&quot;</span>         <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;ploidy&quot;</span>            <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="c1"># metadata</span>
            <span class="s2">&quot;taxa_grp_name&quot;</span>     <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_stix&quot;</span>     <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_spix&quot;</span>     <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;taxa_grp_len&quot;</span>      <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_name&quot;</span>  <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span>  <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span>  <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;vrnt_chrgrp_len&quot;</span>   <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="c1">##################################</span>
        <span class="c1">### read mandatory data fields ###</span>

        <span class="c1"># read mat array (ndarray dtype = int8)</span>
        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;mat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray_int8</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;mat&quot;</span><span class="p">)</span>
        
        <span class="c1">#################################</span>
        <span class="c1">### read optional data fields ###</span>

        <span class="c1"># read taxa array (ndarray dtype = unicode / object)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray_utf8</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa&quot;</span><span class="p">)</span>

        <span class="c1"># read taxa_grp array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp&quot;</span><span class="p">)</span>
        
        <span class="c1"># read vrnt_chrgrp array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_phypos array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_phypos&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_name array (ndarray dtype = unicode / object)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_name&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray_utf8</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_name&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_genpos array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_genpos&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_xoprob array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_xoprob&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_xoprob&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_xoprob&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_hapgrp array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapgrp&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapgrp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapgrp&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_hapalt array (ndarray dtype = unicode / object)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapalt&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapalt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray_utf8</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapalt&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_hapref array (ndarray dtype = unicode / object)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapref&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapref&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray_utf8</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_hapref&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_mask array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_mask&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_mask&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_mask&quot;</span><span class="p">)</span>

        <span class="c1"># read ploidy (dtype = int)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;ploidy&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ploidy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_int</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;ploidy&quot;</span><span class="p">)</span>

        <span class="c1">#####################################</span>
        <span class="c1">### read optional metadata fields ###</span>

        <span class="c1"># read taxa_grp_name array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_name&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_name&quot;</span><span class="p">)</span>

        <span class="c1"># read taxa_grp_stix array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_stix&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_stix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_stix&quot;</span><span class="p">)</span>

        <span class="c1"># read taxa_grp_spix array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_spix&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_spix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_spix&quot;</span><span class="p">)</span>

        <span class="c1"># read taxa_grp_len array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_len&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_len&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;taxa_grp_len&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_chrgrp_name array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_name&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_chrgrp_stix array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_chrgrp_spix array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">)</span>

        <span class="c1"># read vrnt_chrgrp_len array (ndarray dtype = any)</span>
        <span class="k">if</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_len&quot;</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5py_File_read_ndarray</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">)</span>

        <span class="c1">######################</span>
        <span class="c1">### close the file ###</span>

        <span class="c1"># close the file, only if the provided fieldname was a string or Path an not an h5py.File.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">Path</span><span class="p">)):</span>
            <span class="n">h5file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1">########################################################</span>
        <span class="c1">################### Object creation ####################</span>
        
        <span class="c1"># create object from read data</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span>
            <span class="n">mat</span>         <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;mat&quot;</span><span class="p">],</span>
            <span class="n">taxa</span>        <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa&quot;</span><span class="p">],</span>
            <span class="n">taxa_grp</span>    <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp&quot;</span><span class="p">],</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">],</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">],</span>
            <span class="n">vrnt_name</span>   <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_name&quot;</span><span class="p">],</span>
            <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">],</span>
            <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_xoprob&quot;</span><span class="p">],</span>
            <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapgrp&quot;</span><span class="p">],</span>
            <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapalt&quot;</span><span class="p">],</span>
            <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_hapref&quot;</span><span class="p">],</span>
            <span class="n">vrnt_mask</span>   <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_mask&quot;</span><span class="p">],</span> 
            <span class="n">ploidy</span>      <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ploidy&quot;</span><span class="p">],</span>
        <span class="p">)</span>

        <span class="c1"># copy metadata</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_name</span>    <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_name&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_stix</span>    <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_stix&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_spix</span>    <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_spix&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">taxa_grp_len</span>     <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;taxa_grp_len&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseGenotypeMatrix.from_vcf">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.DenseGenotypeMatrix.from_vcf">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_vcf</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">auto_group_vrnt</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseGenotypeMatrix&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a DenseGenotypeMatrix from a VCF file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            Path to VCF file.</span>
<span class="sd">        auto_group_vrnt : bool</span>
<span class="sd">            Whether to group variants in returned genotype matrix.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseGenotypeMatrix</span>
<span class="sd">            An unphased genotype matrix with associated metadata from VCF file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># make VCF iterator</span>
        <span class="n">vcf</span> <span class="o">=</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

        <span class="c1"># extract taxa names from vcf header</span>
        <span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

        <span class="c1"># make empty lists to store extracted values</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">vrnt_name</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># iterate through VCF file and accumulate variants</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
            <span class="c1"># append chromosome integer</span>
            <span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">CHROM</span><span class="p">))</span>

            <span class="c1"># append variant position coordinates</span>
            <span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="p">)</span>

            <span class="c1"># append marker name</span>
            <span class="n">vrnt_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">ID</span><span class="p">))</span>

            <span class="c1"># extract allele states + whether they are phased or not</span>
            <span class="n">phases</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int8</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">genotypes</span><span class="p">)</span>

            <span class="c1"># check that they are all phased</span>
            <span class="c1">#pybrops.util.check_matrix_all_value(phases[:,2], &quot;is_phased&quot;, True)</span>

            <span class="c1"># TODO: maybe modify shapes here to avoid transpose and copy below?</span>
            <span class="c1"># append genotype states</span>
            <span class="n">mat</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">phases</span><span class="p">[:,</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>

        <span class="c1"># convert to compact int8 type and transpose genotype matrix</span>
        <span class="c1"># (p,n,m) --transpose-&gt; (m,n,p)</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int8</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># extract ploidy of sample</span>
        <span class="n">ploidy</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># sum across phases</span>
        <span class="c1"># (m,n,p).sum(0) -&gt; (n,p)</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s1">&#39;int8&#39;</span><span class="p">)</span>

        <span class="c1"># convert to numpy.ndarray</span>
        <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int64</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">)</span>              <span class="c1"># convert to int64 array</span>
        <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int64</span><span class="p">(</span><span class="n">vrnt_phypos</span><span class="p">)</span>              <span class="c1"># convert to int64 array</span>
        <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vrnt_name</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>  <span class="c1"># convert to object array</span>

        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">auto_group_vrnt</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">out</span></div>
</div>




<span class="c1">################################## Utilities ###################################</span>
<div class="viewcode-block" id="check_is_DenseGenotypeMatrix">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmat.DenseGenotypeMatrix.check_is_DenseGenotypeMatrix.html#pybrops.popgen.gmat.DenseGenotypeMatrix.check_is_DenseGenotypeMatrix">[docs]</a>
<span class="k">def</span> <span class="nf">check_is_DenseGenotypeMatrix</span><span class="p">(</span><span class="n">v</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> <span class="n">vname</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if object is of type ``DenseGenotypeMatrix``. Otherwise raise ``TypeError``.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    v : object</span>
<span class="sd">        Any Python object to test.</span>
<span class="sd">    vname : str</span>
<span class="sd">        Name of variable to print in ``TypeError`` message.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">DenseGenotypeMatrix</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;</span><span class="si">{0}</span><span class="s2">&#39; must be a DenseGenotypeMatrix.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">vname</span><span class="p">))</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>