
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pybrops.model.gmod.DenseLinearGenomicModel &#8212; pybrops 0.9.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../../_static/documentation_options.js?v=39bb1c6d"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pybrops/model/gmod/DenseLinearGenomicModel';</script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 0.9.0 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pybrops.mode...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pybrops.model.gmod.DenseLinearGenomicModel</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module implementing classes and error checking routines for genomic prediction</span>
<span class="sd">models that incorporate linear genomic effects.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Integral</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_io_python</span> <span class="kn">import</span> <span class="n">check_file_exists</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_dict</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_is_ndarray</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_str</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_dtype_is_float64</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_dtype_is_object</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_h5py</span> <span class="kn">import</span> <span class="n">check_h5py_File_has_group</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_ndim</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_attr_python</span> <span class="kn">import</span> <span class="n">error_readonly</span>
<span class="kn">from</span> <span class="nn">pybrops.core.util.h5py</span> <span class="kn">import</span> <span class="n">save_dict_to_hdf5</span>
<span class="kn">from</span> <span class="nn">pybrops.model.gmod.LinearGenomicModel</span> <span class="kn">import</span> <span class="n">LinearGenomicModel</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.bvmat.BreedingValueMatrix</span> <span class="kn">import</span> <span class="n">BreedingValueMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.bvmat.DenseGenomicEstimatedBreedingValueMatrix</span> <span class="kn">import</span> <span class="n">DenseGenomicEstimatedBreedingValueMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.GenotypeMatrix</span> <span class="kn">import</span> <span class="n">GenotypeMatrix</span>

<div class="viewcode-block" id="DenseLinearGenomicModel">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel">[docs]</a>
<span class="k">class</span> <span class="nc">DenseLinearGenomicModel</span><span class="p">(</span><span class="n">LinearGenomicModel</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The DenseLinearGenomicModel class represents a Multivariate Multiple</span>
<span class="sd">    Linear Regression model.</span>

<span class="sd">    A Multivariate Multiple Linear Regression model is defined as:</span>

<span class="sd">    .. math::</span>
<span class="sd">        Y = X \\beta + Zu + e</span>

<span class="sd">    Where:</span>

<span class="sd">    - :math:`Y` is a matrix of response variables of shape ``(n,t)``.</span>
<span class="sd">    - :math:`X` is a matrix of fixed effect predictors of shape ``(n,q)``.</span>
<span class="sd">    - :math:`\\beta` is a matrix of fixed effect regression coefficients of</span>
<span class="sd">      shape ``(q,t)``.</span>
<span class="sd">    - :math:`Z` is a matrix of random effect predictors of shape ``(n,p)``.</span>
<span class="sd">    - :math:`u` is a matrix of random effect regression coefficients of shape</span>
<span class="sd">      ``(p,t)``.</span>
<span class="sd">    - :math:`e` is a matrix of error terms of shape ``(n,t)``.</span>

<span class="sd">    Shape definitions:</span>

<span class="sd">    - ``n`` is the number of individuals</span>
<span class="sd">    - ``q`` is the number of fixed effect predictors (e.g. environments)</span>
<span class="sd">    - ``p`` is the number of random effect predictors (e.g. genomic markers)</span>
<span class="sd">    - ``t`` is the number of traits</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">########################## Special Object Methods ##########################</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">beta</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">u</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">trait</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">model_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">hyperparams</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor for DenseLinearGenomicModel class.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        beta : numpy.ndarray</span>
<span class="sd">            A ``float64`` fixed effect regression coefficient matrix of shape</span>
<span class="sd">            ``(q,t)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``q`` is the number of fixed effect predictors (e.g.</span>
<span class="sd">              environments).</span>
<span class="sd">            - ``t`` is the number of individuals.</span>
<span class="sd">        u : numpy.ndarray</span>
<span class="sd">            A ``float64`` random effect regression coefficient matrix of shape</span>
<span class="sd">            ``(p,t)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``p`` is the number of random effect predictors (e.g. genomic</span>
<span class="sd">              markers).</span>
<span class="sd">            - ``t`` is the number of individuals.</span>
<span class="sd">        trait : numpy.ndarray, None</span>
<span class="sd">            An ``object_`` array of shape ``(t,)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        model_name : str, None</span>
<span class="sd">            Name of the model.</span>
<span class="sd">        hyperparams : dict, None</span>
<span class="sd">            Model parameters.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Used for cooperative inheritance. Dictionary passing unused</span>
<span class="sd">            arguments to the parent class constructor.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DenseLinearGenomicModel</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># set variables</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">beta</span> <span class="o">=</span> <span class="n">beta</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">=</span> <span class="n">u</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trait</span> <span class="o">=</span> <span class="n">trait</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model_name</span> <span class="o">=</span> <span class="n">model_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hyperparams</span> <span class="o">=</span> <span class="n">hyperparams</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return repr(self).</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : str</span>
<span class="sd">            A representation of the object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;&lt;</span><span class="si">{0}</span><span class="s2"> of shape [B = </span><span class="si">{1}</span><span class="s2">, U = </span><span class="si">{2}</span><span class="s2">] at </span><span class="si">{3}</span><span class="s2">&gt;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
            <span class="nb">hex</span><span class="p">(</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseLinearGenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the GenomicModel.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseLinearGenomicModel</span>
<span class="sd">            A shallow copy of the model.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span>
            <span class="n">beta</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="p">),</span>
            <span class="n">u</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">),</span>
            <span class="n">trait</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">trait</span><span class="p">),</span>
            <span class="n">model_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_name</span><span class="p">),</span>
            <span class="n">hyperparams</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hyperparams</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="k">def</span> <span class="nf">__deepcopy__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseLinearGenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the GenomicModel.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : DenseLinearGenomicModel</span>
<span class="sd">            A deep copy of the model.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span>
            <span class="n">beta</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="p">),</span>
            <span class="n">u</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">),</span>
            <span class="n">trait</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">trait</span><span class="p">),</span>
            <span class="n">model_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_name</span><span class="p">),</span>
            <span class="n">hyperparams</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hyperparams</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="c1">############################ Object Properties #############################</span>

    <span class="c1">########### Linear Genomic Model Parameters ############</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">nparam_beta</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Integral</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of fixed effect parameters.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_beta</span><span class="o">.</span><span class="n">size</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">beta</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Fixed effect regression coefficients.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_beta</span>
    <span class="nd">@beta</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">beta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set fixed effect regression coefficients&quot;&quot;&quot;</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_float64</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_beta</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">nparam_u</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Integral</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of random effect parameters.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_u</span><span class="o">.</span><span class="n">size</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">u</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Random effect regression coefficients.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_u</span>
    <span class="nd">@u</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">u</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set random effect regression coefficients&quot;&quot;&quot;</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;u&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;u&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_float64</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;u&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_u</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1">################## Genomic Model Data ##################</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">model_name</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Name of the model.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_model_name</span>
    <span class="nd">@model_name</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">model_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set name of the model.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="s2">&quot;unnamed&quot;</span>
        <span class="n">check_is_str</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;model_name&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_model_name</span> <span class="o">=</span> <span class="n">value</span>
    
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">hyperparams</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Model parameters.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_params</span>
    <span class="nd">@hyperparams</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">hyperparams</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set model parameters.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">check_is_dict</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;hyperparams&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_params</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">trait</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Names of the traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_trait</span>
    <span class="nd">@trait</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">trait</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set names of the traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;trait&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;trait&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">check_ndarray_dtype_is_object</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;trait&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_trait</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">ntrait</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_trait</span><span class="p">)</span>
    <span class="nd">@ntrait</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">ntrait</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set number of traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="n">error_readonly</span><span class="p">(</span><span class="s2">&quot;ntrait&quot;</span><span class="p">)</span>

    <span class="c1">############################## Object Methods ##############################</span>

    <span class="c1">####### methods for model fitting and prediction #######</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.fit_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.fit_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">fit_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Y</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fit a dense, linear genomic model.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Y : numpy.ndarray</span>
<span class="sd">            A phenotype matrix of shape (n,t).</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A covariate matrix of shape (n,q).</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A genotypes matrix of shape (n,p).</span>
<span class="sd">        trait : numpy.ndarray</span>
<span class="sd">            A trait name array of shape (t,).</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span><span class="s2">&quot;DenseLinearGenomicModel is read-only&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.fit">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.fit">[docs]</a>
    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">ptobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">BreedingValueMatrix</span><span class="p">,</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fit a dense, linear genomic model.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptobj : BreedingValueMatrix, pandas.DataFrame, numpy.ndarray</span>
<span class="sd">            An object containing phenotype data. Must be a matrix of breeding</span>
<span class="sd">            values or a phenotype data frame.</span>
<span class="sd">        cvobj : numpy.ndarray</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        trait : numpy.ndarray, None</span>
<span class="sd">            A trait name array of shape (t,).</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span><span class="s2">&quot;DenseLinearGenomicModel is read-only&quot;</span><span class="p">)</span></div>


    <span class="c1">######## methods for estimated breeding values #########</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.predict_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.predict_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">predict_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predict breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict_numpy&#39; and &#39;gebv_numpy&#39; is that</span>
<span class="sd">        &#39;predict_numpy&#39; can incorporate other factors (e.g., fixed effects) to</span>
<span class="sd">        provide prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A matrix of covariates.</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotype values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Y_hat : numpy.ndarray</span>
<span class="sd">            A matrix of estimated breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Y = Xβ + Zu</span>
        <span class="n">Y_hat</span> <span class="o">=</span> <span class="p">(</span><span class="n">X</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">Z</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Y_hat</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.predict">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.predict">[docs]</a>
    <span class="k">def</span> <span class="nf">predict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BreedingValueMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predict breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict&#39; and &#39;gebv&#39; is that &#39;predict&#39;</span>
<span class="sd">        can incorporate other factors (e.g., fixed effects) to provide</span>
<span class="sd">        prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cvobj : numpy.ndarray</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix,</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : BreedingValueMatrix</span>
<span class="sd">            Estimated breeding values matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process cvobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cvobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">X</span> <span class="o">=</span> <span class="n">cvobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;accepted types are numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">taxa</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">taxa_grp</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;accepted types are GenotypeMatrix, numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># make predictions</span>
        <span class="n">Y_hat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict_numpy</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># create output breeding value matrix</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">DenseGenomicEstimatedBreedingValueMatrix</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">Y_hat</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
            <span class="n">trait</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">trait</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.score_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.score_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">score_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Y</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the coefficient of determination R**2 of the prediction.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Y : numpy.ndarray</span>
<span class="sd">            A matrix of phenotypes.</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A matrix of covariates.</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Rsq : numpy.ndarray</span>
<span class="sd">            A coefficient of determination array of shape ``(t,)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: array shape checks</span>

        <span class="c1"># calculate predictions</span>
        <span class="c1"># (n,q) @ (q,t) -&gt; (n,t)</span>
        <span class="c1"># (n,p) @ (p,t) -&gt; (n,t)</span>
        <span class="c1"># (n,t) + (n,t) -&gt; (n,t)</span>
        <span class="n">Y_hat</span> <span class="o">=</span> <span class="p">(</span><span class="n">X</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">Z</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">)</span>

        <span class="c1"># calculate sum of squares error</span>
        <span class="c1"># (n,t) - (n,t) -&gt; (n,t)</span>
        <span class="c1"># (n,t)**2 -&gt; (n,t)</span>
        <span class="c1"># (n,t).sum(0) -&gt; (t,)</span>
        <span class="n">SSE</span> <span class="o">=</span> <span class="p">((</span><span class="n">Y</span> <span class="o">-</span> <span class="n">Y_hat</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># calculate means for each trait</span>
        <span class="c1"># (n,t).mean(0) -&gt; (t,)</span>
        <span class="n">Y_mean</span> <span class="o">=</span> <span class="n">Y</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># calculate sum of squares total</span>
        <span class="c1"># (n,t) - (t,) -&gt; (n,t) - (1,t)</span>
        <span class="c1"># (n,t) - (1,t) -&gt; (n,t)</span>
        <span class="c1"># (n,t).sum(0) -&gt; (t,)</span>
        <span class="n">SST</span> <span class="o">=</span> <span class="p">((</span><span class="n">Y</span> <span class="o">-</span> <span class="n">Y_mean</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># calculate R**2</span>
        <span class="c1"># (t,) / (t,) -&gt; (t,)</span>
        <span class="c1"># scalar - (t,) -&gt; (t,)</span>
        <span class="n">Rsq</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="n">SSE</span><span class="o">/</span><span class="n">SST</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Rsq</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.score">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.score">[docs]</a>
    <span class="k">def</span> <span class="nf">score</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">ptobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">BreedingValueMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the coefficient of determination R**2 of the prediction.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptobj : BreedingValueMatrix, pandas.DataFrame, numpy.ndarray</span>
<span class="sd">            An object containing phenotype data. Must be a matrix of breeding</span>
<span class="sd">            values or a phenotype data frame.</span>
<span class="sd">        cvobj : numpy.ndarray</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Rsq : numpy.ndarray</span>
<span class="sd">            A coefficient of determination array of shape ``(t,)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process ptobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ptobj</span><span class="p">,</span> <span class="n">BreedingValueMatrix</span><span class="p">):</span>
            <span class="n">Y</span> <span class="o">=</span> <span class="n">ptobj</span><span class="o">.</span><span class="n">unscale</span><span class="p">()</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ptobj</span><span class="p">,</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;not implmented yet&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ptobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Y</span> <span class="o">=</span> <span class="n">ptobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be BreedingValueMatrix, pandas.DataFrame, numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># process cvobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cvobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">X</span> <span class="o">=</span> <span class="n">cvobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># calculate coefficient of determination</span>
        <span class="n">Rsq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">score_numpy</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Rsq</span></div>


    <span class="c1">######## methods for estimated breeding values #########</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.gebv_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.gebv_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">gebv_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict_numpy&#39; and &#39;gebv_numpy&#39; is that</span>
<span class="sd">        &#39;predict_numpy&#39; can incorporate other factors (e.g., fixed effects) to</span>
<span class="sd">        provide prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotype values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        gebv_hat : numpy.ndarray</span>
<span class="sd">            A matrix of genomic estimated breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Y = Zu</span>
        <span class="n">gebv_hat</span> <span class="o">=</span> <span class="p">(</span><span class="n">Z</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gebv_hat</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.gebv">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.gebv">[docs]</a>
    <span class="k">def</span> <span class="nf">gebv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BreedingValueMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict&#39; and &#39;gebv&#39; is that &#39;predict&#39;</span>
<span class="sd">        can incorporate other factors (e.g., fixed effects) to provide</span>
<span class="sd">        prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : BreedingValueMatrix</span>
<span class="sd">            Genomic estimated breeding values matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">taxa</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">taxa_grp</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;accepted types are GenotypeMatrix, numpy.ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># make predictions</span>
        <span class="n">gebv_hat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gebv_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># construct contrast (X*) to take cell means assuming that the first</span>
        <span class="c1"># column in X is a corner (first fixed effect is the intercept)</span>
        <span class="c1"># construct a contrast to calculate the GEBV center</span>
        <span class="n">nfixed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">Xstar</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="n">nfixed</span><span class="p">),</span> <span class="n">dtype</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
        <span class="n">Xstar</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">Xstar</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="n">nfixed</span>

        <span class="c1"># calculate intercepts (location)</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">Xstar</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span>

        <span class="c1"># add location to gebv_hat</span>
        <span class="n">gebv_hat</span> <span class="o">+=</span> <span class="n">location</span>

        <span class="c1"># create output breeding value matrix</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">DenseGenomicEstimatedBreedingValueMatrix</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">gebv_hat</span><span class="p">,</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
            <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
            <span class="n">trait</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">trait</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">###### methods for population variance prediction ######</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.var_G_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_G_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">var_G_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population genetic variance.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population genetic variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># estimate breeding values (n,t)</span>
        <span class="n">gebv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gebv_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># calculate variance</span>
        <span class="c1"># (n,t).var(0) -&gt; (t,)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">gebv</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.var_G">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_G">[docs]</a>
    <span class="k">def</span> <span class="nf">var_G</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population genetic variance.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population genetic variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="n">out</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">var_G_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.var_A_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_A_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">var_A_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genetic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population additive genetic </span>
<span class="sd">            variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># estimate breeding values (n,t)</span>
        <span class="n">gebv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gebv_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># calculate variance</span>
        <span class="c1"># (n,t).var(0) -&gt; (t,)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">gebv</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.var_A">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_A">[docs]</a>
    <span class="k">def</span> <span class="nf">var_A</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genetic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population additive genetic </span>
<span class="sd">            variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="n">out</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">var_A_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.var_a_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_a_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">var_a_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Integral</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : Integral</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population additive genic </span>
<span class="sd">            variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># change shape to (p,1)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">p</span><span class="p">[:,</span><span class="kc">None</span><span class="p">]</span>

        <span class="c1"># calculate additive genic variance</span>
        <span class="c1"># (p,t)**2 * (p,1) * (p,1) -&gt; (p,t)</span>
        <span class="c1"># (p,t).sum[0] -&gt; (t,)</span>
        <span class="c1"># scalar * (t,) -&gt; (t,)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">ploidy</span><span class="o">**</span><span class="mf">2.0</span><span class="p">)</span> <span class="o">*</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="n">p</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="n">p</span><span class="p">))</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.var_a">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.var_a">[docs]</a>
    <span class="k">def</span> <span class="nf">var_a</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>

<span class="sd">            If ploidy is None:</span>

<span class="sd">            - If gtobj is a GenotypeMatrix, then get ploidy from GenotypeMatrix.</span>
<span class="sd">            - If gtobj is a numpy.ndarray, then assumed to be 2 (diploid).</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population additive genic </span>
<span class="sd">            variances.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">ploidy</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>    <span class="c1"># get allele frequencies</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># calculate genic variance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">var_a_numpy</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.bulmer_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.bulmer_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">bulmer_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Integral</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the Bulmer effect.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : Integral</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population Bulmer effect statistics.</span>
<span class="sd">            In the event that additive genic variance is zero, NaN&#39;s are produced.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sigma_A</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">var_A_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">)</span>           <span class="c1"># calculate additive genetic variance</span>
        <span class="n">sigma_a</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">var_a_numpy</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">)</span>   <span class="c1"># calculate additive genetic variance</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">sigma_a</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">)</span>                 <span class="c1"># determine where division by zero occurs</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="n">sigma_a</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>                  <span class="c1"># copy array</span>
        <span class="n">denom</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>                       <span class="c1"># substitute non-zero value</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">sigma_A</span> <span class="o">/</span> <span class="n">denom</span>                   <span class="c1"># calculate Bulmer effect</span>
        <span class="n">out</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span>                   <span class="c1"># add NaN&#39;s (avoids div by zero warning)</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.bulmer">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.bulmer">[docs]</a>
    <span class="k">def</span> <span class="nf">bulmer</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the Bulmer effect.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">            If ploidy is None:</span>

<span class="sd">            - If gtobj is a GenotypeMatrix, then get ploidy from GenotypeMatrix.</span>
<span class="sd">            - If gtobj is a numpy.ndarray, then assumed to be 2 (diploid).</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population Bulmer effect statistics.</span>
<span class="sd">            In the event that additive genic variance is zero, NaN&#39;s are produced.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="s2">&quot;{0,1,2}&quot;</span><span class="p">)</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">Z</span> <span class="o">=</span> <span class="n">gtobj</span>       <span class="c1"># get genotypes</span>
            <span class="k">if</span> <span class="n">ploidy</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>                  <span class="c1"># if ploidy not provided</span>
                <span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>                      <span class="c1"># assume diploid</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>    <span class="c1"># get allele frequencies</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># calculate Bulmer effect</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bulmer_numpy</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">############# methods for selection limits #############</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.usl_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.usl_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">usl_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Integral</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the upper selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population upper selection </span>
<span class="sd">            limit statistics.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># reshape allele frequencies</span>
        <span class="c1"># (p,) -&gt; (p,1)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">p</span><span class="p">[:,</span><span class="kc">None</span><span class="p">]</span>

        <span class="c1"># get maximum attainable genotype</span>
        <span class="c1"># (p,t) ? (p,1) : (p,1) -&gt; (p,t)</span>
        <span class="n">uslgeno</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">,</span>       <span class="c1"># if the allele effect is positive</span>
            <span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">,</span>            <span class="c1"># +allele: 1 if we have at least one +allele</span>
            <span class="n">p</span> <span class="o">&gt;=</span> <span class="mf">1.0</span>            <span class="c1"># -allele: 1 if we have fixation for -allele</span>
        <span class="p">)</span>

        <span class="c1"># calculate usl value</span>
        <span class="c1"># scalar * (p,t) * (p,t) -&gt; (p,t)</span>
        <span class="c1"># (p,t).sum[0] -&gt; (t,)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">ploidy</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">*</span> <span class="n">uslgeno</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.usl">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.usl">[docs]</a>
    <span class="k">def</span> <span class="nf">usl</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the upper selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population upper selection </span>
<span class="sd">            limit statistics.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">ploidy</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>    <span class="c1"># get allele frequencies</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># calculate genic variance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">usl_numpy</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.lsl_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.lsl_numpy">[docs]</a>
    <span class="k">def</span> <span class="nf">lsl_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Integral</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the lower selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population lower selection </span>
<span class="sd">            limit statistics.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># reshape allele frequencies</span>
        <span class="c1"># (p,) -&gt; (p,1)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">p</span><span class="p">[:,</span><span class="kc">None</span><span class="p">]</span>

        <span class="c1"># get minimum attainable genotype</span>
        <span class="c1"># (p,t) ? (p,1) : (p,1) -&gt; (p,t)</span>
        <span class="n">lslgeno</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">,</span>       <span class="c1"># if the allele effect is positive</span>
            <span class="n">p</span> <span class="o">&gt;=</span> <span class="mf">1.0</span><span class="p">,</span>           <span class="c1"># +allele: 1 if we have fixation for +allele</span>
            <span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.0</span>             <span class="c1"># -allele: 1 if we have at least one -allele</span>
        <span class="p">)</span>

        <span class="c1"># calculate lsl value</span>
        <span class="c1"># scalar * (p,t) * (p,t) -&gt; (p,t)</span>
        <span class="c1"># (p,t).sum[0] -&gt; (t,)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">ploidy</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">*</span> <span class="n">lslgeno</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.lsl">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.lsl">[docs]</a>
    <span class="k">def</span> <span class="nf">lsl</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the lower selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape ``(t,)`` containing population lower selection </span>
<span class="sd">            limit statistics.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process gtobj</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">):</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">ploidy</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtobj</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">ploidy</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="n">gtobj</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>    <span class="c1"># get allele frequencies</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;must be GenotypeMatrix, ndarray&quot;</span><span class="p">)</span>

        <span class="c1"># calculate genic variance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lsl_numpy</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">############ methods for allele attributes #############</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.facount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.facount">[docs]</a>
    <span class="k">def</span> <span class="nf">facount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Favorable allele count across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to count favorable alleles.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing allele counts of the favorable allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># construct mask for beneficial alleles</span>
        <span class="c1"># (p,t)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">)</span>

        <span class="c1"># get allele counts for the genotype matrix</span>
        <span class="c1"># (p,) -&gt; (p,1)</span>
        <span class="n">acount</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">acount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">)[:,</span><span class="kc">None</span><span class="p">]</span>

        <span class="c1"># get maximum number of favorable alleles</span>
        <span class="c1"># scalar</span>
        <span class="n">maxfav</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>

        <span class="c1"># calculate favorable allele counts</span>
        <span class="c1"># (p,t)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">acount</span><span class="p">,</span> <span class="n">maxfav</span> <span class="o">-</span> <span class="n">acount</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.fafreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.fafreq">[docs]</a>
    <span class="k">def</span> <span class="nf">fafreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Favorable allele frequency across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing allele frequencies of the favorable allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get favorable allele frequencies</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span>
            <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">facount</span><span class="p">(</span><span class="n">gmat</span><span class="p">),</span> <span class="c1"># favorable allele counts</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.faavail">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.faavail">[docs]</a>
    <span class="k">def</span> <span class="nf">faavail</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a favorable allele is available in the present taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing whether a favorable allele is available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">bool</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get favorable allele counts</span>
        <span class="n">facount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">facount</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>

        <span class="c1"># get boolean mask of favorable alleles that are available</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">facount</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1"># convert datatype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="o">!=</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.fafixed">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.fafixed">[docs]</a>
    <span class="k">def</span> <span class="nf">fafixed</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a favorable allele is fixed across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing whether a favorable allele is fixed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">bool</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get favorable allele counts</span>
        <span class="n">facount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">facount</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>

        <span class="c1"># get maximum number of favorable alleles</span>
        <span class="n">maxfav</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span>

        <span class="c1"># get boolean mask of favorable alleles that are fixed</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">facount</span> <span class="o">==</span> <span class="n">maxfav</span><span class="p">)</span>

        <span class="c1"># convert datatype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="o">!=</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.dacount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.dacount">[docs]</a>
    <span class="k">def</span> <span class="nf">dacount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Deleterious allele count across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to count deleterious alleles.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing allele counts of the deleterious allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># convert data type</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># construct mask for deleterious alleles</span>
        <span class="c1"># (p,t)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">u</span> <span class="o">&lt;</span> <span class="mf">0.0</span><span class="p">)</span>

        <span class="c1"># get allele counts for the genotype matrix</span>
        <span class="c1"># (p,) -&gt; (p,1)</span>
        <span class="n">acount</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">acount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span><span class="p">)[:,</span><span class="kc">None</span><span class="p">]</span>

        <span class="c1"># get maximum number of favorable alleles</span>
        <span class="c1"># scalar</span>
        <span class="n">maxfav</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>

        <span class="c1"># calculate favorable allele counts</span>
        <span class="c1"># (p,t)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">acount</span><span class="p">,</span> <span class="n">maxfav</span> <span class="o">-</span> <span class="n">acount</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.dafreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.dafreq">[docs]</a>
    <span class="k">def</span> <span class="nf">dafreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Deleterious allele frequency across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing allele frequencies of the deleterious allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get favorable allele frequencies</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span>
            <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dacount</span><span class="p">(</span><span class="n">gmat</span><span class="p">),</span> <span class="c1"># favorable allele counts</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>

    
<div class="viewcode-block" id="DenseLinearGenomicModel.daavail">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.daavail">[docs]</a>
    <span class="k">def</span> <span class="nf">daavail</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a deleterious allele is available in the present taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native boolean type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing whether a deleterious allele is available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">bool</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get deleterious allele counts</span>
        <span class="n">facount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dacount</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>

        <span class="c1"># get boolean mask of deleterious alleles that are available</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">facount</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1"># convert datatype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="o">!=</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="DenseLinearGenomicModel.dafixed">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.dafixed">[docs]</a>
    <span class="k">def</span> <span class="nf">dafixed</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a deleterious allele is fixed across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing whether a deleterious allele is fixed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># process dtype</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dtype</span> <span class="o">=</span> <span class="nb">bool</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>

        <span class="c1"># get deleterious allele counts</span>
        <span class="n">dacount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dacount</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>

        <span class="c1"># get maximum number of deleterious alleles</span>
        <span class="n">maxdel</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ploidy</span> <span class="o">*</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span>

        <span class="c1"># get boolean mask of deleterious alleles that are fixed</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span><span class="n">dacount</span> <span class="o">==</span> <span class="n">maxdel</span><span class="p">)</span>

        <span class="c1"># convert datatype if needed</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="o">!=</span> <span class="n">out</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">dtype</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">################### File I/O methods ###################</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.to_hdf5">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.to_hdf5">[docs]</a>
    <span class="k">def</span> <span class="nf">to_hdf5</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> 
            <span class="n">groupname</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write GenotypeMatrix to an HDF5 file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            HDF5 file name to which to write.</span>
<span class="sd">        groupname : str or None</span>
<span class="sd">            HDF5 group name under which GenotypeMatrix data is stored.</span>
<span class="sd">            If ``None``, GenotypeMatrix is written to the base HDF5 group.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">h5file</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span>                       <span class="c1"># open HDF5 in write mode</span>
        <span class="c1">######################################################### process groupname argument</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">groupname</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>                          <span class="c1"># if we have a string</span>
            <span class="k">if</span> <span class="n">groupname</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;/&#39;</span><span class="p">:</span>                            <span class="c1"># if last character in string is not &#39;/&#39;</span>
                <span class="n">groupname</span> <span class="o">+=</span> <span class="s1">&#39;/&#39;</span>                                <span class="c1"># add &#39;/&#39; to end of string</span>
        <span class="k">elif</span> <span class="n">groupname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>                                 <span class="c1"># else if groupname is None</span>
            <span class="n">groupname</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>                                      <span class="c1"># empty string</span>
        <span class="k">else</span><span class="p">:</span>                                                   <span class="c1"># else raise error</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;groupname&#39; must be of type str or None&quot;</span><span class="p">)</span>
        <span class="c1">######################################################### populate HDF5 file</span>
        <span class="n">data_dict</span> <span class="o">=</span> <span class="p">{</span>                                           <span class="c1"># data dictionary</span>
            <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta</span><span class="p">,</span>
            <span class="s2">&quot;u&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">u</span><span class="p">,</span>
            <span class="s2">&quot;trait&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">trait</span><span class="p">,</span>
            <span class="s2">&quot;model_name&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_name</span><span class="p">,</span>
            <span class="s2">&quot;hyperparams&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">hyperparams</span>
        <span class="p">}</span>
        <span class="n">save_dict_to_hdf5</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">groupname</span><span class="p">,</span> <span class="n">data_dict</span><span class="p">)</span>         <span class="c1"># write data</span>
        <span class="c1">######################################################### write conclusion</span>
        <span class="n">h5file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>                                          <span class="c1"># close the file</span></div>


    <span class="c1">############################## Class Methods ###############################</span>

    <span class="c1">################### File I/O methods ###################</span>
<div class="viewcode-block" id="DenseLinearGenomicModel.from_hdf5">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.DenseLinearGenomicModel.from_hdf5">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_hdf5</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> 
            <span class="n">groupname</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DenseLinearGenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read GenotypeMatrix from an HDF5 file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            HDF5 file name which to read.</span>
<span class="sd">        groupname : str or None</span>
<span class="sd">            HDF5 group name under which GenotypeMatrix data is stored.</span>
<span class="sd">            If ``None``, GenotypeMatrix is read from base HDF5 group.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            A genotype matrix read from file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">check_file_exists</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>                             <span class="c1"># check file exists</span>
        <span class="n">h5file</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>                       <span class="c1"># open HDF5 in read only</span>
        <span class="c1">######################################################### process groupname argument</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">groupname</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>                          <span class="c1"># if we have a string</span>
            <span class="n">check_h5py_File_has_group</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">groupname</span><span class="p">)</span>    <span class="c1"># check that group exists</span>
            <span class="k">if</span> <span class="n">groupname</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;/&#39;</span><span class="p">:</span>                            <span class="c1"># if last character in string is not &#39;/&#39;</span>
                <span class="n">groupname</span> <span class="o">+=</span> <span class="s1">&#39;/&#39;</span>                                <span class="c1"># add &#39;/&#39; to end of string</span>
        <span class="k">elif</span> <span class="n">groupname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>                                 <span class="c1"># else if groupname is None</span>
            <span class="n">groupname</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>                                      <span class="c1"># empty string</span>
        <span class="k">else</span><span class="p">:</span>                                                   <span class="c1"># else raise error</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;groupname&#39; must be of type str or None&quot;</span><span class="p">)</span>
        <span class="c1">######################################################### check that we have all required fields</span>
        <span class="n">required_fields</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="s2">&quot;u&quot;</span><span class="p">]</span>                         <span class="c1"># all required arguments</span>
        <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">required_fields</span><span class="p">:</span>                           <span class="c1"># for each required field</span>
            <span class="n">fieldname</span> <span class="o">=</span> <span class="n">groupname</span> <span class="o">+</span> <span class="n">field</span>                       <span class="c1"># concatenate base groupname and field</span>
            <span class="n">check_h5py_File_has_group</span><span class="p">(</span><span class="n">h5file</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">fieldname</span><span class="p">)</span>    <span class="c1"># check that group exists</span>
        <span class="c1">######################################################### read data</span>
        <span class="n">data_dict</span> <span class="o">=</span> <span class="p">{</span>                                           <span class="c1"># output dictionary</span>
            <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;u&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;trait&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;model_name&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;hyperparams&quot;</span><span class="p">:</span> <span class="kc">None</span>
        <span class="p">}</span>
        <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5file</span><span class="p">[</span><span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;beta&quot;</span><span class="p">][()]</span>      <span class="c1"># read beta array</span>
        <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;u&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5file</span><span class="p">[</span><span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;u&quot;</span><span class="p">][()]</span>            <span class="c1"># read u array</span>
        <span class="n">fieldname</span> <span class="o">=</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;trait&quot;</span>                         <span class="c1"># construct &quot;groupname/trait&quot;</span>
        <span class="k">if</span> <span class="n">fieldname</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>                                 <span class="c1"># if &quot;groupname/trait&quot; in hdf5</span>
            <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;trait&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5file</span><span class="p">[</span><span class="n">fieldname</span><span class="p">][()]</span>          <span class="c1"># read trait array</span>
            <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;trait&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>                   <span class="c1"># convert trait string from byte to utf-8</span>
                <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;trait&quot;</span><span class="p">]],</span>
                <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
            <span class="p">)</span>
        <span class="n">fieldname</span> <span class="o">=</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;model_name&quot;</span>                    <span class="c1"># construct &quot;groupname/model_name&quot;</span>
        <span class="k">if</span> <span class="n">fieldname</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>                                 <span class="c1"># if &quot;groupname/model_name&quot; in hdf5</span>
            <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;model_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">h5file</span><span class="p">[</span><span class="n">fieldname</span><span class="p">][()]</span>     <span class="c1"># read string (as bytes); convert to utf-8</span>
            <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;model_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;model_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span>
        <span class="n">fieldname</span> <span class="o">=</span> <span class="n">groupname</span> <span class="o">+</span> <span class="s2">&quot;hyperparams&quot;</span>                        <span class="c1"># construct &quot;groupname/hyperparams&quot;</span>
        <span class="k">if</span> <span class="n">fieldname</span> <span class="ow">in</span> <span class="n">h5file</span><span class="p">:</span>                                 <span class="c1"># if &quot;groupname/hyperparams&quot; in hdf5</span>
            <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;hyperparams&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>                            <span class="c1"># create empty dictionary</span>
            <span class="n">view</span> <span class="o">=</span> <span class="n">h5file</span><span class="p">[</span><span class="n">fieldname</span><span class="p">]</span>                            <span class="c1"># get view of dataset</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">view</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>                             <span class="c1"># for each field</span>
                <span class="n">data_dict</span><span class="p">[</span><span class="s2">&quot;hyperparams&quot;</span><span class="p">][</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">view</span><span class="p">[</span><span class="n">key</span><span class="p">][()]</span>        <span class="c1"># extract data</span>
        <span class="c1">######################################################### read conclusion</span>
        <span class="n">h5file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>                                          <span class="c1"># close file</span>
        <span class="c1">######################################################### create object</span>
        <span class="n">glgmod</span> <span class="o">=</span> <span class="n">DenseLinearGenomicModel</span><span class="p">(</span><span class="o">**</span><span class="n">data_dict</span><span class="p">)</span>         <span class="c1"># create object from read data</span>
        <span class="k">return</span> <span class="n">glgmod</span></div>
</div>




<span class="c1">################################## Utilities ###################################</span>
<div class="viewcode-block" id="check_is_DenseLinearGenomicModel">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.DenseLinearGenomicModel.check_is_DenseLinearGenomicModel.html#pybrops.model.gmod.DenseLinearGenomicModel.check_is_DenseLinearGenomicModel">[docs]</a>
<span class="k">def</span> <span class="nf">check_is_DenseLinearGenomicModel</span><span class="p">(</span><span class="n">v</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> <span class="n">vname</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if object is of type DenseLinearGenomicModel. Otherwise raise TypeError.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    v : object</span>
<span class="sd">        Any Python object to test.</span>
<span class="sd">    vname : str</span>
<span class="sd">        Name of variable to print in TypeError message.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">DenseLinearGenomicModel</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;variable &#39;</span><span class="si">{0}</span><span class="s2">&#39; must be a DenseLinearGenomicModel&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">vname</span><span class="p">))</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, Robert Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>