
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pybrops.model.gmod.GenomicModel &#8212; pybrops 1.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pybrops/model/gmod/GenomicModel';</script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pybrops.mode...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pybrops.model.gmod.GenomicModel</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module defining basal interfaces and error checking routines for genomic models.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">ABCMeta</span><span class="p">,</span> <span class="n">abstractmethod</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Integral</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">from</span> <span class="nn">pybrops.core.io.HDF5InputOutput</span> <span class="kn">import</span> <span class="n">HDF5InputOutput</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.bvmat.BreedingValueMatrix</span> <span class="kn">import</span> <span class="n">BreedingValueMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.GenotypeMatrix</span> <span class="kn">import</span> <span class="n">GenotypeMatrix</span>

<div class="viewcode-block" id="GenomicModel">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel">[docs]</a>
<span class="k">class</span> <span class="nc">GenomicModel</span><span class="p">(</span><span class="n">HDF5InputOutput</span><span class="p">,</span><span class="n">metaclass</span><span class="o">=</span><span class="n">ABCMeta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    An abstract class for genomic models.</span>

<span class="sd">    The purpose for this abstract interface is to provide base functionality for:</span>
<span class="sd">        1) Model metadata storage.</span>
<span class="sd">        2) Model fitting.</span>
<span class="sd">        3) Model prediction.</span>
<span class="sd">        4) Model scoring.</span>
<span class="sd">        5) Genetic variance estimation using the model.</span>
<span class="sd">        6) Estimation of upper and lower selection limits using the model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">########################## Special Object Methods ##########################</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the ``GenomicModel``.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : GenomicModel</span>
<span class="sd">            A shallow copy of the ``GenomicModel``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">__deepcopy__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the ``GenomicModel``.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : GenomicModel</span>
<span class="sd">            A deep copy of the ``GenomicModel``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span>

    <span class="c1">############################ Object Properties #############################</span>

    <span class="c1">############### Genomic Model Parameters ###############</span>
    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">nexplan</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Integral</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of explanatory variables required by the model.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">nparam</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Integral</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of model parameters.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="c1">################## Genomic Model Data ##################</span>
    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">model_name</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Name of the model.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>
    <span class="nd">@model_name</span><span class="o">.</span><span class="n">setter</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">model_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the name of the model&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">hyperparams</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Model parameters.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>
    <span class="nd">@hyperparams</span><span class="o">.</span><span class="n">setter</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">hyperparams</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the model parameters&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">trait</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Names of the traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>
    <span class="nd">@trait</span><span class="o">.</span><span class="n">setter</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">trait</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the names of the traits predicted by the model&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">ntrait</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of traits predicted by the model.&quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;property is abstract&quot;</span><span class="p">)</span>

    <span class="c1">############################## Object Methods ##############################</span>

    <span class="c1">#################### Model copying #####################</span>
<div class="viewcode-block" id="GenomicModel.copy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.copy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the GenomicModel.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : GenomicModel</span>
<span class="sd">            A shallow copy of the original GenomicModel</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.deepcopy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.deepcopy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">deepcopy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">memo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicModel&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the GenomicModel.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : GenomicModel</span>
<span class="sd">            A deep copy of the original GenomicModel</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">####### methods for model fitting and prediction #######</span>
<div class="viewcode-block" id="GenomicModel.fit_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.fit_numpy">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">fit_numpy</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">Y</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fit the model.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Y : numpy.ndarray</span>
<span class="sd">            A phenotype matrix of shape (n,t).</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A covariate matrix of shape (n,q).</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A genotypes matrix of shape (n,p).</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.fit">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.fit">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">ptobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">BreedingValueMatrix</span><span class="p">,</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fit the model.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptobj : BreedingValueMatrix, pandas.DataFrame, numpy.ndarray</span>
<span class="sd">            An object containing phenotype data. Must be a matrix of breeding</span>
<span class="sd">            values or a phenotype data frame.</span>
<span class="sd">        cvobj : numpy.ndarray</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">######## methods for estimated breeding values #########</span>
<div class="viewcode-block" id="GenomicModel.predict_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.predict_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">predict_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predict breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict_numpy&#39; and &#39;gebv_numpy&#39; is that</span>
<span class="sd">        &#39;predict_numpy&#39; can incorporate other factors (e.g., fixed effects) to</span>
<span class="sd">        provide prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A matrix of covariates.</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotype values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Y_hat : numpy.ndarray</span>
<span class="sd">            A matrix of predicted breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.predict">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.predict">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">predict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BreedingValueMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predict breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict&#39; and &#39;gebv&#39; is that &#39;predict&#39;</span>
<span class="sd">        can incorporate other factors (e.g., fixed effects) to provide</span>
<span class="sd">        prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cvobj : numpy.ndarray</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bvmat_hat : BreedingValueMatrix</span>
<span class="sd">            Estimated breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.score_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.score_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">score_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Y</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">X</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the coefficient of determination R**2 of the prediction.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Y : numpy.ndarray</span>
<span class="sd">            A matrix of phenotypes.</span>
<span class="sd">        X : numpy.ndarray</span>
<span class="sd">            A matrix of covariates.</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Rsq : numpy.ndarray</span>
<span class="sd">            A coefficient of determination array of shape ``(t,)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.score">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.score">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">score</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">ptobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">BreedingValueMatrix</span><span class="p">,</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span> 
            <span class="n">cvobj</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the coefficient of determination R**2 of the prediction.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptobj : BreedingValueMatrix or pandas.DataFrame</span>
<span class="sd">            An object containing phenotype data. Must be a matrix of breeding</span>
<span class="sd">            values or a phenotype data frame.</span>
<span class="sd">        cvobj : object</span>
<span class="sd">            An object containing covariate data.</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Rsq : numpy.ndarray</span>
<span class="sd">            A coefficient of determination array of shape ``(t,)``.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``t`` is the number of traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">######## methods for estimated breeding values #########</span>
<div class="viewcode-block" id="GenomicModel.gebv_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.gebv_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">gebv_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict_numpy&#39; and &#39;gebv_numpy&#39; is that</span>
<span class="sd">        &#39;predict_numpy&#39; can incorporate other factors (e.g., fixed effects) to</span>
<span class="sd">        provide prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotype values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        gebv_hat : numpy.ndarray</span>
<span class="sd">            A matrix of genomic estimated breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.gebv">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.gebv">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">gebv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BreedingValueMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated breeding values.</span>

<span class="sd">        Remark: The difference between &#39;predict&#39; and &#39;gebv&#39; is that &#39;predict&#39;</span>
<span class="sd">        can incorporate other factors (e.g., fixed effects) to provide</span>
<span class="sd">        prediction estimates.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        gebvmat_hat : BreedingValueMatrix</span>
<span class="sd">            Genomic estimated breeding values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">######## methods for estimated genotypic value #########</span>
<div class="viewcode-block" id="GenomicModel.gegv_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.gegv_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">gegv_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated genotypic values.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypic markers.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A matrix of genomic estimated genotypic values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>

    
<div class="viewcode-block" id="GenomicModel.gegv">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.gegv">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">gegv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BreedingValueMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate genomic estimated genotypic values.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypic markers.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A matrix of genomic estimated genotypic values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">###### methods for population variance prediction ######</span>
<div class="viewcode-block" id="GenomicModel.var_G_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_G_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_G_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population genetic variance.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing genetic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.var_G">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_G">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_G</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population genetic variance.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing genetic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.var_A_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_A_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_A_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genetic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing additive genetic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.var_A">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_A">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_A</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genetic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing additive genetic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.var_a_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_a_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_a_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing additive genic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.var_a">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.var_a">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">var_a</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenotypeMatrix</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the population additive genic variance</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix, numpy.ndarray</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing additive genic variances for each trait.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.bulmer_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.bulmer_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">bulmer_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">Z</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the Bulmer effect.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Z : numpy.ndarray</span>
<span class="sd">            A matrix of genotypes.</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing Bulmer effects for each trait.</span>
<span class="sd">            In the event that additive genic variance is zero, NaN&#39;s are produced.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.bulmer">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.bulmer">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">bulmer</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the Bulmer effect.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Array of shape ``(t,)`` contianing Bulmer effects for each trait.</span>
<span class="sd">            In the event that additive genic variance is zero, NaN&#39;s are produced.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">############# methods for selection limits #############</span>
<div class="viewcode-block" id="GenomicModel.usl_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.usl_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">usl_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="n">unscale</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the upper selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        unscale : bool</span>
<span class="sd">            If ``True``, then apply the mean of the fixed effects to the output.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape (t,) containing upper selection limits for each of ``t`` traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.usl">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.usl">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">usl</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="n">unscale</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the upper selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        unscale : bool</span>
<span class="sd">            If ``True``, then apply the mean of the fixed effects to the output.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape (t,) containing upper selection limits for each of ``t`` traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.lsl_numpy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.lsl_numpy">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">lsl_numpy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">p</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="n">unscale</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the lower selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.ndarray</span>
<span class="sd">            A vector of genotype allele frequencies of shape (p,).</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        unscale : bool</span>
<span class="sd">            If ``True``, then apply the mean of the fixed effects to the output.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape (t,) containing lower selection limits for each of ``t`` traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.lsl">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.lsl">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">lsl</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gtobj</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
            <span class="n">unscale</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the lower selection limit for a population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gtobj : GenotypeMatrix</span>
<span class="sd">            An object containing genotype data. Must be a matrix of genotype</span>
<span class="sd">            values.</span>
<span class="sd">        ploidy : int</span>
<span class="sd">            Ploidy of the species.</span>
<span class="sd">        unscale : bool</span>
<span class="sd">            If ``True``, then apply the mean of the fixed effects to the output.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            An array of shape (t,) containing lower selection limits for each of ``t`` traits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


    <span class="c1">############ methods for allele attributes #############</span>
<div class="viewcode-block" id="GenomicModel.facount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.facount">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">facount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Favorable allele count across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to count favorable alleles.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing allele counts of the favorable allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.fafreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.fafreq">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">fafreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Favorable allele frequency across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing allele frequencies of the favorable allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>

    
<div class="viewcode-block" id="GenomicModel.faavail">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.faavail">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">faavail</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a favorable allele is available in the present taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing whether a favorable </span>
<span class="sd">            allele is available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.fafixed">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.fafixed">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">fafixed</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a favorable allele is fixed across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine favorable allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing whether a favorable allele is fixed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.dacount">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.dacount">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">dacount</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Deleterious allele count across all taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to count deleterious alleles.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing allele counts of the deleterious allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.dafreq">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.dafreq">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">dafreq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Deleterious allele frequency across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing allele frequencies of the deleterious allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>

    
<div class="viewcode-block" id="GenomicModel.daavail">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.daavail">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">daavail</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a deleterious allele is available in the present taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele </span>
<span class="sd">            frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native boolean </span>
<span class="sd">            type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,t)`` containing whether a deleterious </span>
<span class="sd">            allele is available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicModel.dafixed">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.GenomicModel.html#pybrops.model.gmod.GenomicModel.GenomicModel.dafixed">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">dafixed</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">gmat</span><span class="p">:</span> <span class="n">GenotypeMatrix</span><span class="p">,</span> 
            <span class="n">dtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether a deleterious allele is fixed across all taxa.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gmat : GenotypeMatrix</span>
<span class="sd">            Genotype matrix for which to determine deleterious allele frequencies.</span>
<span class="sd">        dtype : numpy.dtype, None</span>
<span class="sd">            Datatype of the returned array. If ``None``, use the native type.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A numpy.ndarray of shape ``(p,)`` containing whether a deleterious allele is fixed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;method is abstract&quot;</span><span class="p">)</span></div>
</div>




<span class="c1">################################## Utilities ###################################</span>
<div class="viewcode-block" id="check_is_GenomicModel">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.model.gmod.GenomicModel.check_is_GenomicModel.html#pybrops.model.gmod.GenomicModel.check_is_GenomicModel">[docs]</a>
<span class="k">def</span> <span class="nf">check_is_GenomicModel</span><span class="p">(</span><span class="n">v</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> <span class="n">vname</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if object is of type GenomicModel. Otherwise raise TypeError.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    v : object</span>
<span class="sd">        Any Python object to test.</span>
<span class="sd">    vname : str</span>
<span class="sd">        Name of variable to print in TypeError message.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">GenomicModel</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;variable &#39;</span><span class="si">{0}</span><span class="s2">&#39; must be a GenomicModel&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">vname</span><span class="p">))</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
       Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>