
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pybrops.popgen.gmap.StandardGeneticMap &#8212; pybrops 0.9.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../../_static/documentation_options.js?v=39bb1c6d"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pybrops/popgen/gmap/StandardGeneticMap';</script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 0.9.0 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pybrops.popg...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pybrops.popgen.gmap.StandardGeneticMap</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module implementing a standard genetic map format and associated error checking</span>
<span class="sd">routines.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;StandardGeneticMap&quot;</span><span class="p">,</span>
    <span class="s2">&quot;check_is_StandardGeneticMap&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Integral</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">from</span> <span class="nn">scipy.interpolate</span> <span class="kn">import</span> <span class="n">interp1d</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_is_ndarray</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_is_str_or_ndarray</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_dtype_is_floating</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_dtype_is_integer</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_pandas</span> <span class="kn">import</span> <span class="n">check_is_pandas_DataFrame</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_len_eq</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_numpy</span> <span class="kn">import</span> <span class="n">check_ndarray_ndim</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_dict</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_str</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_type_python</span> <span class="kn">import</span> <span class="n">check_is_str_or_Integral</span>
<span class="kn">from</span> <span class="nn">pybrops.core.error.error_value_python</span> <span class="kn">import</span> <span class="n">check_tuple_len_eq</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.GeneticMap</span> <span class="kn">import</span> <span class="n">GeneticMap</span>


<div class="viewcode-block" id="StandardGeneticMap">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap">[docs]</a>
<span class="k">class</span> <span class="nc">StandardGeneticMap</span><span class="p">(</span><span class="n">GeneticMap</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A concrete class for representing a standard genetic map format.</span>

<span class="sd">    The purpose of this concrete class is to implement functionality for:</span>
<span class="sd">        1) Genetic map representation.</span>
<span class="sd">        2) Genetic map metadata.</span>
<span class="sd">        3) Genetic map routines.</span>
<span class="sd">        4) Genetic map interpolation spline construction.</span>
<span class="sd">        5) Genetic map spline interpolation.</span>
<span class="sd">        6) Import and export of genetic maps.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">########################## Special Object Methods ##########################</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">spline</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">spline_kind</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span><span class="p">,</span>
            <span class="n">spline_fill_value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;extrapolate&quot;</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">auto_build_spline</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructor for creating a standard genetic map object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            Chromosome or linkage group assignment array of shape ``(n,)`` where </span>
<span class="sd">            ``n`` is the number of markers.</span>
<span class="sd">        </span>
<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            Physical positions array of shape ``(n,)`` where ``n`` is the number </span>
<span class="sd">            of markers. This array contains the physical positions on the </span>
<span class="sd">            chromosome or linkage group for each marker.</span>
<span class="sd">        </span>
<span class="sd">        vrnt_genpos : numpy.ndarray</span>
<span class="sd">            Genetic positions array of shape ``(n,)`` where ``n`` is the number </span>
<span class="sd">            of markers. This array contains the genetic positions on the </span>
<span class="sd">            chromosome or linkage group for each marker.</span>

<span class="sd">        spline : dict, None, default = None</span>
<span class="sd">            Pre-built interpolation spline to associate with the genetic map.</span>
<span class="sd">        </span>
<span class="sd">        spline_kind : str, default = &quot;linear&quot;</span>
<span class="sd">            In automatic building of splines, the spline kind to be built.</span>
<span class="sd">            Specifies the kind of interpolation as a string (&#39;linear&#39;,</span>
<span class="sd">            &#39;nearest&#39;, &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39;, &#39;cubic&#39;, &#39;previous&#39;,</span>
<span class="sd">            &#39;next&#39;, where &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39; and &#39;cubic&#39; refer to a</span>
<span class="sd">            spline interpolation of zeroth, first, second or third order;</span>
<span class="sd">            &#39;previous&#39; and &#39;next&#39; simply return the previous or next value of</span>
<span class="sd">            the point) or as an integer specifying the order of the spline</span>
<span class="sd">            interpolator to use.</span>
<span class="sd">            </span>
<span class="sd">        spline_fill_value : str, numpy.ndarray, default = &quot;extrapolate&quot;</span>
<span class="sd">            In automatic building of splines, the spline fill value to use.</span>
<span class="sd">            If &#39;extrapolate&#39;, then points outside the data range will be</span>
<span class="sd">            extrapolated.</span>
<span class="sd">            If a ndarray (or float), this value will be used to fill in for</span>
<span class="sd">            requested points outside of the data range. If not provided, then</span>
<span class="sd">            the default is NaN. The array-like must broadcast properly to the</span>
<span class="sd">            dimensions of the non-interpolation axes.</span>
<span class="sd">            If a two-element tuple, then the first element is used as a fill</span>
<span class="sd">            value for x_new &lt; x[0] and the second element is used for</span>
<span class="sd">            x_new &gt; x[-1]. Anything that is not a 2-element tuple (e.g., list</span>
<span class="sd">            or ndarray, regardless of shape) is taken to be a single array-like</span>
<span class="sd">            argument meant to be used for both bounds as below,</span>
<span class="sd">            above = fill_value, fill_value.</span>

<span class="sd">        vrnt_genpos_units : str, default = &quot;M&quot;</span>
<span class="sd">            Units in which genetic positions in the ``vrnt_genpos`` array are </span>
<span class="sd">            stored. Options are listed below and are case-sensitive:</span>
<span class="sd">            </span>
<span class="sd">            - ``&quot;M&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;Morgans&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;cM&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            - ``&quot;centiMorgans&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            </span>
<span class="sd">            Internally, all genetic positions are stored in Morgans. Providing </span>
<span class="sd">            the units of the input  </span>

<span class="sd">        auto_group : bool</span>
<span class="sd">            Whether to automatically sort and group variants into chromosome groups.</span>
<span class="sd">        </span>
<span class="sd">        auto_build_spline : bool</span>
<span class="sd">            Whether to automatically construct a spline on object construction.</span>
<span class="sd">            If ``spline`` is provided, then this spline is overwritten.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># order dependent assignments!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span>       <span class="o">=</span> <span class="n">vrnt_chrgrp</span> <span class="c1"># must be assigned first!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span>       <span class="o">=</span> <span class="n">vrnt_phypos</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span>       <span class="o">=</span> <span class="p">(</span><span class="n">vrnt_genpos</span><span class="p">,</span><span class="n">vrnt_genpos_units</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline</span>            <span class="o">=</span> <span class="n">spline</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span>       <span class="o">=</span> <span class="n">spline_kind</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">spline_fill_value</span>

        <span class="c1"># set variant metadata to None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># automatically group if requested</span>
        <span class="k">if</span> <span class="n">auto_group</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        
        <span class="c1"># automatically build spline if requested</span>
        <span class="k">if</span> <span class="n">auto_build_spline</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">build_spline</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the number of markers in the genetic map.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span><span class="p">)</span>

    <span class="c1">################## GeneticMap copying ##################</span>
    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the StandardGeneticMap.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A shallow copy of the original StandardGeneticMap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get the class of self</span>
        <span class="bp">cls</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span>

        <span class="c1"># create a new object instance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>

        <span class="c1"># initialize the object instance</span>
        <span class="n">out</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">vrnt_chrgrp</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">),</span> 
            <span class="n">vrnt_phypos</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">),</span> 
            <span class="n">vrnt_genpos</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">),</span>
            <span class="n">spline</span>            <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline</span><span class="p">),</span>
            <span class="n">spline_kind</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span><span class="p">),</span>
            <span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span><span class="p">),</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span>        <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># copy variant metadata to the new object</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="k">def</span> <span class="nf">__deepcopy__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the StandardGeneticMap.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A deep copy of the original StandardGeneticMap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get the class of self</span>
        <span class="bp">cls</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span>

        <span class="c1"># create a new object instance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>

        <span class="c1"># initialize the object instance</span>
        <span class="n">out</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">vrnt_chrgrp</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>       <span class="n">memo</span><span class="p">),</span> 
            <span class="n">vrnt_phypos</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span>       <span class="n">memo</span><span class="p">),</span> 
            <span class="n">vrnt_genpos</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">,</span>       <span class="n">memo</span><span class="p">),</span>
            <span class="n">spline</span>            <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline</span><span class="p">,</span>            <span class="n">memo</span><span class="p">),</span>
            <span class="n">spline_kind</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span><span class="p">,</span>       <span class="n">memo</span><span class="p">),</span>
            <span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span><span class="p">,</span> <span class="n">memo</span><span class="p">),</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span>        <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># copy variant metadata to the new object</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">,</span> <span class="n">memo</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">,</span>  <span class="n">memo</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span>

    <span class="c1">############################ Object Properties #############################</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">nvrnt</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Integral</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Number of variants in the GeneticMap.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant chromosome group label.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp</span>
    <span class="nd">@vrnt_chrgrp</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant chromosome group label array&quot;&quot;&quot;</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_phypos</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant physical position.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_phypos</span>
    <span class="nd">@vrnt_phypos</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_phypos</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant physical position array&quot;&quot;&quot;</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">check_ndarray_len_eq</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_phypos</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_genpos</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant genetic position in Morgans.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span>
    <span class="nd">@vrnt_genpos</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_genpos</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="n">Tuple</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant genetic position array&quot;&quot;&quot;</span>
        <span class="n">array</span> <span class="o">=</span> <span class="n">value</span>
        <span class="n">units</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
            <span class="n">check_tuple_len_eq</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;vrnt_genpos[0]&quot;</span><span class="p">)</span>
            <span class="n">check_is_str</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;vrnt_genpos[1]&quot;</span><span class="p">)</span>
            <span class="n">array</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">units</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">units</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;M&quot;</span><span class="p">,</span><span class="s2">&quot;Morgans&quot;</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="n">units</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;cM&quot;</span><span class="p">,</span><span class="s2">&quot;centiMorgans&quot;</span><span class="p">):</span>
            <span class="n">array</span> <span class="o">=</span> <span class="mf">0.01</span> <span class="o">*</span> <span class="n">array</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;vrnt_genpos units must be &#39;M&#39;, &#39;Morgans&#39;, &#39;cM&#39;, or &#39;centiMorgans&#39; but received value &#39;</span><span class="si">{0}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">units</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_floating</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">check_ndarray_len_eq</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span> <span class="o">=</span> <span class="n">array</span>

    <span class="c1">################# Metadata Properites ##################</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_name</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant chromosome group names.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_name</span>
    <span class="nd">@vrnt_chrgrp_name</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant chromosome group name array&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_name&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_stix</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant chromosome group start indices.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span>
    <span class="nd">@vrnt_chrgrp_stix</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_stix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant chromosome group start indices array&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_stix&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_spix</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant chromosome group stop indices.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_spix</span>
    <span class="nd">@vrnt_chrgrp_spix</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_spix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant chromosome group stop indices array&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_spix&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Variant chromosome group length.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_len</span>
    <span class="nd">@vrnt_chrgrp_len</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">vrnt_chrgrp_len</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set variant chromosome group length array&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">)</span>
            <span class="n">check_ndarray_ndim</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_len&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_len</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1">################## Spline Properites ###################</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">spline</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Interpolation spline(s).&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_spline</span>
    <span class="nd">@spline</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">spline</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set interpolation spline(s)&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_dict</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;spline&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_spline</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1">############# Spline Metadata Properites ###############</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">spline_kind</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Spline kind.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_spline_kind</span>
    <span class="nd">@spline_kind</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">spline_kind</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the spline kind&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_str</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;spline_kind&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_spline_kind</span> <span class="o">=</span> <span class="n">value</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">spline_fill_value</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Default spline fill value.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_spline_fill_value</span>
    <span class="nd">@spline_fill_value</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">spline_fill_value</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set the default spline fill value&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">check_is_str_or_ndarray</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;spline_fill_value&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_spline_fill_value</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c1">############################## Object Methods ##############################</span>

    <span class="c1">################## GeneticMap copying ##################</span>
<div class="viewcode-block" id="StandardGeneticMap.copy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.copy">[docs]</a>
    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a shallow copy of the StandardGeneticMap.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A shallow copy of the original StandardGeneticMap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__copy__</span><span class="p">()</span></div>


<div class="viewcode-block" id="StandardGeneticMap.deepcopy">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.deepcopy">[docs]</a>
    <span class="k">def</span> <span class="nf">deepcopy</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">memo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make a deep copy of the StandardGeneticMap.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        memo : dict, None</span>
<span class="sd">            Dictionary of memo metadata.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A deep copy of the original StandardGeneticMap.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__deepcopy__</span><span class="p">(</span><span class="n">memo</span><span class="p">)</span></div>


    <span class="c1">################### Sorting Methods ####################</span>
<div class="viewcode-block" id="StandardGeneticMap.lexsort">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.lexsort">[docs]</a>
    <span class="k">def</span> <span class="nf">lexsort</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">keys</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">tuple</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform an indirect stable sort using a sequence of keys.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        keys : (k, N) array or tuple containing k (N,)-shaped sequences</span>
<span class="sd">            The k different columns to be sorted. The last column (or row if</span>
<span class="sd">            keys is a 2D array) is the primary sort key.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        indices : (N,) ndarray of ints</span>
<span class="sd">            Array of indices that sort the keys along the specified axis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># if no keys were provided, set a default</span>
        <span class="k">if</span> <span class="n">keys</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># key priority: (1) vrnt_chrgrp (2) vrnt_phypos (3) vrnt_genpos</span>
            <span class="n">keys</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">)</span>
        
        <span class="c1"># if a tuple of keys was provided, check that all keys are ndarrays of </span>
        <span class="c1"># acceptable length, then remove any None</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">keys</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">keys</span><span class="p">):</span>
                <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="s2">&quot;keys[</span><span class="si">{0}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
                <span class="n">check_ndarray_len_eq</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="s2">&quot;keys[</span><span class="si">{0}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
        
        <span class="c1"># if a ndarray was provided, check that its length is acceptable, </span>
        <span class="c1"># then convert to tuple</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">keys</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">check_ndarray_len_eq</span><span class="p">(</span><span class="n">keys</span><span class="p">,</span> <span class="s2">&quot;keys&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
            <span class="n">keys</span> <span class="o">=</span> <span class="p">(</span><span class="n">keys</span><span class="p">,)</span>
        
        <span class="c1"># otherwise raise type error</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;argument &#39;</span><span class="si">{0}</span><span class="s2">&#39; must be of type &#39;</span><span class="si">{1}</span><span class="s2">&#39;, &#39;</span><span class="si">{2}</span><span class="s2">&#39;, or &#39;</span><span class="si">{3}</span><span class="s2">&#39; but received type &#39;</span><span class="si">{4}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;keys&quot;</span><span class="p">,</span>
                    <span class="nb">tuple</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                    <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                    <span class="s2">&quot;None&quot;</span><span class="p">,</span>
                    <span class="nb">type</span><span class="p">(</span><span class="n">keys</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># get indices</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">lexsort</span><span class="p">(</span><span class="n">keys</span><span class="p">)</span>

        <span class="c1"># return indices</span>
        <span class="k">return</span> <span class="n">indices</span></div>


<div class="viewcode-block" id="StandardGeneticMap.reorder">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.reorder">[docs]</a>
    <span class="k">def</span> <span class="nf">reorder</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">indices</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Sequence</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reorder markers in-place in the GeneticMap using an array of indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        indices : (N,) ndarray of ints</span>
<span class="sd">            Array of indices that reorder the matrix along the specified axis.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># sort internal self</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>

        <span class="c1"># reset variant chromosome group name, stix, spix, len</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span> <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="StandardGeneticMap.sort">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.sort">[docs]</a>
    <span class="k">def</span> <span class="nf">sort</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">keys</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">tuple</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span><span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sort slements of the GeneticMap using a sequence of keys.</span>
<span class="sd">        Note this modifies the GeneticMap in-place.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        keys : (k, N) array or tuple containing k (N,)-shaped sequences</span>
<span class="sd">            The k different columns to be sorted. The last column (or row if</span>
<span class="sd">            keys is a 2D array) is the primary sort key.</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get indices for sort</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lexsort</span><span class="p">(</span><span class="n">keys</span><span class="p">)</span>

        <span class="c1"># reorder internals</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span></div>


<div class="viewcode-block" id="StandardGeneticMap.group">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.group">[docs]</a>
    <span class="k">def</span> <span class="nf">group</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sort the GeneticMap jointly by chromosome group and physical position, </span>
<span class="sd">        then populate grouping indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># sort the genetic map using default keys</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

        <span class="c1"># get unique group names, start indices, group lengths</span>
        <span class="n">uniq</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp</span><span class="p">,</span> <span class="n">return_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">return_counts</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1"># make assignments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_len</span> <span class="o">=</span> <span class="n">uniq</span>

        <span class="c1"># calculate chr_grp_spix (stop indices)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_len</span></div>


<div class="viewcode-block" id="StandardGeneticMap.ungroup">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.ungroup">[docs]</a>
    <span class="k">def</span> <span class="nf">ungroup</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove grouping metadata from the GeneticMap.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># set grouping metadat to None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="StandardGeneticMap.is_grouped">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.is_grouped">[docs]</a>
    <span class="k">def</span> <span class="nf">is_grouped</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine whether the GeneticMap has been sorted and grouped.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        grouped : bool</span>
<span class="sd">            True or False indicating whether the GeneticMap has been sorted and</span>
<span class="sd">            grouped.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_spix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_len</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
        <span class="p">)</span></div>


    <span class="c1">################ Insert/Delete Methods #################</span>
<div class="viewcode-block" id="StandardGeneticMap.remove">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.remove">[docs]</a>
    <span class="k">def</span> <span class="nf">remove</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">indices</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove indices from the GeneticMap. If the GeneticMap was grouped </span>
<span class="sd">        beforehand, then re-sort and re-group internal arrays after removing </span>
<span class="sd">        indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        indices : int, slice, Sequence</span>
<span class="sd">            Array of shape ``(a,)``, ``slice`` or ``int`` of item(s) to remove.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``a`` is the number of indices to remove.</span>
<span class="sd">        </span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># delete indices from self</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">indices</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span> <span class="n">indices</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">,</span> <span class="n">indices</span><span class="p">)</span>

        <span class="c1"># if GeneticMap was previously grouped, re-sort and re-group</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">group</span><span class="p">()</span></div>


<div class="viewcode-block" id="StandardGeneticMap.select">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.select">[docs]</a>
    <span class="k">def</span> <span class="nf">select</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">indices</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">slice</span><span class="p">,</span><span class="n">Sequence</span><span class="p">],</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Keep only selected markers, removing all others from the GeneticMap.</span>
<span class="sd">        If the GeneticMap was grouped beforehand, then re-sort and re-group </span>
<span class="sd">        internal arrays after removing indices.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        indices : int, slice, Sequence</span>
<span class="sd">            Array of shape ``(a,)``, ``slice`` or ``int`` of item(s) to remove.</span>

<span class="sd">            Where:</span>

<span class="sd">            - ``a`` is the number of indices to remove.</span>
<span class="sd">        </span>
<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># keep only selected markers.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>

        <span class="c1"># if GeneticMap was previously grouped, re-sort and re-group</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">group</span><span class="p">()</span></div>


    <span class="c1">################## Integrity Methods ###################</span>
<div class="viewcode-block" id="StandardGeneticMap.congruence">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.congruence">[docs]</a>
    <span class="k">def</span> <span class="nf">congruence</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Assess physical and genetic map site congruency. If the genetic map is </span>
<span class="sd">        not grouped, it will be grouped. </span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A boolean matrix of map concordancies where:</span>

<span class="sd">            - ``True`` = the current marker has a map_pos &gt;= the previous </span>
<span class="sd">              position</span>
<span class="sd">            - ``False`` = the current marker has a map_pos &lt; the previous </span>
<span class="sd">              position</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This assumes high contiguity between physical and genetic maps</span>
<span class="sd">        (i.e. a high quality reference genome). This assumption may cause</span>
<span class="sd">        major issues if there are incorrect markers at the beginning of the</span>
<span class="sd">        chromosome. This also assumes the first marker on the chromosome is </span>
<span class="sd">        placed correctly.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># group as a prerequisite</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>

        <span class="c1"># make empty bool vector</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_phypos</span><span class="p">),</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s1">&#39;bool&#39;</span><span class="p">)</span>

        <span class="c1"># for each linkage group</span>
        <span class="k">for</span> <span class="n">st</span><span class="p">,</span><span class="n">sp</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_stix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp_spix</span><span class="p">):</span>
            <span class="c1"># we assume the first element is correctly placed</span>
            <span class="n">out</span><span class="p">[</span><span class="n">st</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># test if the current element is &lt;= the previous element</span>
            <span class="n">out</span><span class="p">[</span><span class="n">st</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">sp</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span><span class="p">[</span><span class="n">st</span><span class="p">:</span><span class="n">sp</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span><span class="p">[</span><span class="n">st</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">sp</span><span class="p">]</span>

        <span class="c1"># return results</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.is_congruent">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.is_congruent">[docs]</a>
    <span class="k">def</span> <span class="nf">is_congruent</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine if all sites in the genetic map demonstrate congruence with</span>
<span class="sd">        their supposed physical and genetic positions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : bool</span>
<span class="sd">            Whether all genetic map loci demonstrate congruence between </span>
<span class="sd">            physical and genetic positions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># determine if all sites are congruent</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">congruence</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.remove_discrepancies">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.remove_discrepancies">[docs]</a>
    <span class="k">def</span> <span class="nf">remove_discrepancies</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove discrepancies between the physical map and the genetic map.</span>
<span class="sd">        In instances of conflict, assume that the physical map is correct.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This assumption may cause major issues if there are incorrect markers </span>
<span class="sd">        at the beginning of the chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get boolean mask for concordant map positions</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">congruence</span><span class="p">()</span>

        <span class="c1"># if all are concordant, exit this function</span>
        <span class="k">if</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span> <span class="p">):</span>
            <span class="k">return</span>

        <span class="c1"># select (and group) only sites that are congruent</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span></div>


    <span class="c1">################ Interpolation Methods #################</span>
<div class="viewcode-block" id="StandardGeneticMap.build_spline">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.build_spline">[docs]</a>
    <span class="k">def</span> <span class="nf">build_spline</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">kind</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;linear&#39;</span><span class="p">,</span> 
            <span class="n">fill_value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;extrapolate&#39;</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build a spline for estimating genetic map distances.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kind : str, default = &#39;linear&#39;</span>
<span class="sd">            Specifies the kind of interpolation as a string (&#39;linear&#39;,</span>
<span class="sd">            &#39;nearest&#39;, &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39;, &#39;cubic&#39;, &#39;previous&#39;,</span>
<span class="sd">            &#39;next&#39;, where &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39; and &#39;cubic&#39; refer to a</span>
<span class="sd">            spline interpolation of zeroth, first, second or third order;</span>
<span class="sd">            &#39;previous&#39; and &#39;next&#39; simply return the previous or next value of</span>
<span class="sd">            the point) or as an integer specifying the order of the spline</span>
<span class="sd">            interpolator to use.</span>

<span class="sd">        fill_value : array-like, {&#39;extrapolate&#39;}, default = &#39;extrapolate&#39;</span>
<span class="sd">            If &#39;extrapolate&#39;, then points outside the data range will be</span>
<span class="sd">            extrapolated.</span>
<span class="sd">            If a ndarray (or float), this value will be used to fill in for</span>
<span class="sd">            requested points outside of the data range. If not provided, then</span>
<span class="sd">            the default is NaN. The array-like must broadcast properly to the</span>
<span class="sd">            dimensions of the non-interpolation axes.</span>
<span class="sd">            If a two-element tuple, then the first element is used as a fill</span>
<span class="sd">            value for x_new &lt; x[0] and the second element is used for</span>
<span class="sd">            x_new &gt; x[-1]. Anything that is not a 2-element tuple (e.g., list</span>
<span class="sd">            or ndarray, regardless of shape) is taken to be a single array-like</span>
<span class="sd">            argument meant to be used for both bounds as below,</span>
<span class="sd">            above = fill_value, fill_value.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get unique chromosome group labels</span>
        <span class="n">uniq</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp</span><span class="p">)</span>

        <span class="c1"># make an empty dictionary for map splines</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># set the spline type and fill value</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span> <span class="o">=</span> <span class="n">kind</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">fill_value</span>

        <span class="c1"># iterate through unique groups and build spline</span>
        <span class="k">for</span> <span class="n">grp</span> <span class="ow">in</span> <span class="n">uniq</span><span class="p">:</span>
            <span class="c1"># get mask of elements belonging to grp</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_chrgrp</span> <span class="o">==</span> <span class="n">grp</span><span class="p">)</span>

            <span class="c1"># build spline and assign to dictionary</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_spline</span><span class="p">[</span><span class="n">grp</span><span class="p">]</span> <span class="o">=</span> <span class="n">interp1d</span><span class="p">(</span>
                <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_phypos</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>    <span class="c1"># chromosome positions</span>
                <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_vrnt_genpos</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>    <span class="c1"># map positions</span>
                <span class="n">kind</span> <span class="o">=</span> <span class="n">kind</span><span class="p">,</span>                    <span class="c1"># type of interpolation</span>
                <span class="n">fill_value</span> <span class="o">=</span> <span class="n">fill_value</span><span class="p">,</span>        <span class="c1"># default fill value</span>
                <span class="n">assume_sorted</span> <span class="o">=</span> <span class="kc">False</span>           <span class="c1"># arrays are not sorted</span>
            <span class="p">)</span></div>


<div class="viewcode-block" id="StandardGeneticMap.has_spline">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.has_spline">[docs]</a>
    <span class="k">def</span> <span class="nf">has_spline</span><span class="p">(</span>
            <span class="bp">self</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return whether or not the GeneticMap has a built spline.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : bool</span>
<span class="sd">            Whether the GeneticMap has a spline built.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_spline</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_spline_kind</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_spline_fill_value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="StandardGeneticMap.interp_genpos">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.interp_genpos">[docs]</a>
    <span class="k">def</span> <span class="nf">interp_genpos</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Interpolate genetic positions given variant physical positions.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            Chromosome/linkage group labels for each marker variant.</span>
<span class="sd">        </span>
<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            Chromosome/linkage group physical positions for each marker variant.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            Interpolated genetic positions for each marker variant.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># raise error if no spline is found</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_spline</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;interpolation spline not built&quot;</span><span class="p">)</span>

        <span class="c1"># raise warning if map is not congruent</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_congruent</span><span class="p">():</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;genetic map is not congruent: markers are out of order&quot;</span><span class="p">,</span> <span class="ne">RuntimeWarning</span><span class="p">)</span>

        <span class="c1"># allocate empty memory</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>

        <span class="c1"># for each chromosome-position pair</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,(</span><span class="n">chrgrp</span><span class="p">,</span><span class="n">phypos</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_phypos</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>                                <span class="c1"># try to index dict</span>
                <span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_spline</span><span class="p">[</span><span class="n">chrgrp</span><span class="p">]</span>    <span class="c1"># try to get model</span>
                <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">phypos</span><span class="p">)</span>          <span class="c1"># interpolate genetic position</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>                    <span class="c1"># if dict key not found</span>
                <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span>              <span class="c1"># set to NaN</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.interp_gmap">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.interp_gmap">[docs]</a>
    <span class="k">def</span> <span class="nf">interp_gmap</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Interpolate a new genetic map from the current genetic map. Associate </span>
<span class="sd">        spline of current GeneticMap with new GeneticMap.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            Chromosome/linkage group labels for each marker variant.</span>
<span class="sd">        </span>
<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            Chromosome/linkage group physical positions for each marker variant.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            An interpolated genetic map sharing a copy of the spline from the </span>
<span class="sd">            original genetic map.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># interpolate genetic positions</span>
        <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interp_genpos</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_phypos</span><span class="p">)</span>

        <span class="c1"># get the class of self</span>
        <span class="bp">cls</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span>

        <span class="c1"># create a new object instance</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>

        <span class="c1"># initialize the object instance</span>
        <span class="n">out</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">vrnt_chrgrp</span>       <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span>       <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_genpos</span>       <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="n">spline</span>            <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline</span><span class="p">),</span>
            <span class="n">spline_kind</span>       <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_kind</span><span class="p">),</span>
            <span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spline_fill_value</span><span class="p">),</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span>        <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># copy variant metadata to the new object</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_name</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_stix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_spix</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span>  <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp_len</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">############### Genetic Distance Methods ###############</span>
<div class="viewcode-block" id="StandardGeneticMap.gdist1g">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.gdist1g">[docs]</a>
    <span class="k">def</span> <span class="nf">gdist1g</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ast</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">asp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate sequential genetic distances using genetic positions.</span>
<span class="sd">        Requires ``vrnt_chrgrp`` and ``vrnt_genpos`` to have been sorted </span>
<span class="sd">        jointly in ascending order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            A 1D array of variant chromosome groups.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_genpos``.</span>

<span class="sd">        vrnt_genpos : numpy.ndarray</span>
<span class="sd">            A 1D array of variant genetic positions.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_chrgrp``.</span>

<span class="sd">        ast : Integral, None</span>
<span class="sd">            Optional array start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all array elements are to be used for </span>
<span class="sd">            sequential genetic distance calculations.</span>

<span class="sd">        asp : Integral, None</span>
<span class="sd">            Optional array stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all array elements are to be used for </span>
<span class="sd">            sequential genetic distance calculations.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A 1D array of distances between the marker prior.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Sequential distance arrays will start every chromosome with numpy.inf!</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># type checks</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">vrnt_genpos</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_floating</span><span class="p">(</span><span class="n">vrnt_genpos</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos&quot;</span><span class="p">)</span>

        <span class="c1"># get views of only sections we&#39;ll be looking at</span>
        <span class="n">view_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">[</span><span class="n">ast</span><span class="p">:</span><span class="n">asp</span><span class="p">]</span>
        <span class="n">view_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">[</span><span class="n">ast</span><span class="p">:</span><span class="n">asp</span><span class="p">]</span>

        <span class="c1"># get chromosome groups, start index, length of region</span>
        <span class="n">uniq</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">view_chrgrp</span><span class="p">,</span> <span class="n">return_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">return_counts</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1"># get stop index</span>
        <span class="n">stop</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">counts</span>

        <span class="c1"># allocate empty array</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">view_genpos</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>

        <span class="c1"># for each chromosome group</span>
        <span class="k">for</span> <span class="n">st</span><span class="p">,</span><span class="n">sp</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">):</span>
            <span class="c1"># infinite distance between chromosomes</span>
            <span class="n">out</span><span class="p">[</span><span class="n">st</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">inf</span>
            <span class="c1"># subtract distances</span>
            <span class="n">out</span><span class="p">[</span><span class="n">st</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">sp</span><span class="p">]</span> <span class="o">=</span> <span class="n">view_genpos</span><span class="p">[</span><span class="n">st</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">sp</span><span class="p">]</span> <span class="o">-</span> <span class="n">view_genpos</span><span class="p">[</span><span class="n">st</span><span class="p">:</span><span class="n">sp</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.gdist2g">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.gdist2g">[docs]</a>
    <span class="k">def</span> <span class="nf">gdist2g</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">rst</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">rsp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">cst</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">csp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate pairwise genetic distances using genetic positions.</span>
<span class="sd">        Requires ``vrnt_chrgrp`` and ``vrnt_genpos`` to have been sorted </span>
<span class="sd">        jointly in ascending order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            A 1D array of variant chromosome groups.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_genpos``.</span>

<span class="sd">        vrnt_genpos : numpy.ndarray</span>
<span class="sd">            A 1D array of variant genetic positions.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_chrgrp``.</span>

<span class="sd">        rst : Integral, None</span>
<span class="sd">            Optional row start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all rows are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        rsp : Integral, None</span>
<span class="sd">            Optional row stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all rows are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        cst : Integral, None</span>
<span class="sd">            Optional column start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all columns are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        csp : Integral, None</span>
<span class="sd">            Optional column stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all columns are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A 2D array of distances between marker pairs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: # OPTIMIZE: fill with inf, then add distances? - assumes sort</span>
        <span class="c1"># create sparse meshgrid indicating where chromosomes differ</span>
        <span class="n">mi</span><span class="p">,</span> <span class="n">mj</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">[</span><span class="n">rst</span><span class="p">:</span><span class="n">rsp</span><span class="p">],</span> <span class="n">vrnt_chrgrp</span><span class="p">[</span><span class="n">cst</span><span class="p">:</span><span class="n">csp</span><span class="p">],</span> <span class="n">indexing</span><span class="o">=</span><span class="s1">&#39;ij&#39;</span><span class="p">,</span> <span class="n">sparse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># create sparse meshgrid indicating where genetic positions are</span>
        <span class="n">gi</span><span class="p">,</span> <span class="n">gj</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">vrnt_genpos</span><span class="p">[</span><span class="n">rst</span><span class="p">:</span><span class="n">rsp</span><span class="p">],</span> <span class="n">vrnt_genpos</span><span class="p">[</span><span class="n">cst</span><span class="p">:</span><span class="n">csp</span><span class="p">],</span> <span class="n">indexing</span><span class="o">=</span><span class="s1">&#39;ij&#39;</span><span class="p">,</span> <span class="n">sparse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># calculate absolute distances</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">gi</span> <span class="o">-</span> <span class="n">gj</span><span class="p">)</span>

        <span class="c1"># where chromosomes are different, set to numpy.inf</span>
        <span class="n">out</span><span class="p">[</span><span class="n">mi</span> <span class="o">!=</span> <span class="n">mj</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">inf</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.gdist1p">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.gdist1p">[docs]</a>
    <span class="k">def</span> <span class="nf">gdist1p</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">ast</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">asp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate sequential genetic distances using physical positions.</span>
<span class="sd">        Requires ``vrnt_chrgrp`` and ``vrnt_phypos`` to have been sorted </span>
<span class="sd">        jointly in ascending order. Requires an interpolation spline to have </span>
<span class="sd">        been built beforehand.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            A 1D array of variant chromosome groups.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_phypos``.</span>

<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            A 1D array of variant genetic positions.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_chrgrp``.</span>

<span class="sd">        ast : Integral, None</span>
<span class="sd">            Optional array start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all array elements are to be used for </span>
<span class="sd">            sequential genetic distance calculations.</span>

<span class="sd">        asp : Integral, None</span>
<span class="sd">            Optional array stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all array elements are to be used for </span>
<span class="sd">            sequential genetic distance calculations.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A 1D array of distances between the marker prior.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Sequential distance arrays will start every chromosome with numpy.inf!</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># type checks</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp&quot;</span><span class="p">)</span>
        <span class="n">check_is_ndarray</span><span class="p">(</span><span class="n">vrnt_phypos</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">)</span>
        <span class="n">check_ndarray_dtype_is_integer</span><span class="p">(</span><span class="n">vrnt_phypos</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos&quot;</span><span class="p">)</span>

        <span class="c1"># TODO: # OPTIMIZE: don&#39;t interpolate all markers</span>
        <span class="c1"># interpolate genetic positions</span>
        <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interp_genpos</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_phypos</span><span class="p">)</span>

        <span class="c1"># calculate genetic distances</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gdist1g</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_genpos</span><span class="p">,</span> <span class="n">ast</span><span class="p">,</span> <span class="n">asp</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.gdist2p">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.gdist2p">[docs]</a>
    <span class="k">def</span> <span class="nf">gdist2p</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">vrnt_phypos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
            <span class="n">rst</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">rsp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">cst</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">csp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate pairwise genetic distances using physical positions.</span>
<span class="sd">        Requires ``vrnt_chrgrp`` and ``vrnt_phypos`` to have been sorted </span>
<span class="sd">        jointly in ascending order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp : numpy.ndarray</span>
<span class="sd">            A 1D array of variant chromosome groups.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_phypos``.</span>

<span class="sd">        vrnt_phypos : numpy.ndarray</span>
<span class="sd">            A 1D array of variant genetic positions.</span>
<span class="sd">            Must be sorted in ascending order jointly with ``vrnt_chrgrp``.</span>

<span class="sd">        rst : Integral, None</span>
<span class="sd">            Optional row start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all rows are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        rsp : Integral, None</span>
<span class="sd">            Optional row stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all rows are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        cst : Integral, None</span>
<span class="sd">            Optional column start index (inclusive).</span>
<span class="sd">            If ``None``, assume that all columns are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        csp : Integral, None</span>
<span class="sd">            Optional column stop index (exclusive).</span>
<span class="sd">            If ``None``, assume that all columns are to be calculated in the </span>
<span class="sd">            pairwise genetic distance matrix are to be calculated.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : numpy.ndarray</span>
<span class="sd">            A 2D array of distances between marker pairs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: # OPTIMIZE: don&#39;t interpolate all markers</span>
        <span class="c1"># interpolate genetic positions</span>
        <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interp_genpos</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_phypos</span><span class="p">)</span>

        <span class="c1"># calculate genetic distances</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gdist2g</span><span class="p">(</span><span class="n">vrnt_chrgrp</span><span class="p">,</span> <span class="n">vrnt_genpos</span><span class="p">,</span> <span class="n">rst</span><span class="p">,</span> <span class="n">rsp</span><span class="p">,</span> <span class="n">cst</span><span class="p">,</span> <span class="n">csp</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


    <span class="c1">#################### Export Methods ####################</span>
<div class="viewcode-block" id="StandardGeneticMap.to_pandas">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.to_pandas">[docs]</a>
    <span class="k">def</span> <span class="nf">to_pandas</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">vrnt_chrgrp_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_phypos_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_units</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Export a GeneticMap to a pandas.DataFrame.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vrnt_chrgrp_col : str, default = &quot;chr&quot;</span>
<span class="sd">            Name of the chromosome/linkage group name column to which to export.</span>

<span class="sd">        vrnt_phypos_col : str, default = &quot;pos&quot;</span>
<span class="sd">            Name of the physical position column to which to export.</span>

<span class="sd">        vrnt_genpos_col : str, default = &quot;cM&quot;</span>
<span class="sd">            Name of the genetic position column to which to export.</span>

<span class="sd">        vrnt_genpos_units : str, default = &quot;cM&quot;</span>
<span class="sd">            Units of the genetic position column to which to export.</span>
<span class="sd">            Options are listed below and are case-sensitive:</span>
<span class="sd">            </span>
<span class="sd">            - ``&quot;M&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;Morgans&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;cM&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            - ``&quot;centiMorgans&quot;`` - genetic position units are in centiMorgans</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments to use for dictating export to a </span>
<span class="sd">            pandas.DataFrame.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : pandas.DataFrame</span>
<span class="sd">            An output dataframe.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># calculate genetic positions</span>
        <span class="n">vrnt_genpos_convert</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_genpos</span>
        <span class="k">if</span> <span class="n">vrnt_genpos_units</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;M&quot;</span><span class="p">,</span><span class="s2">&quot;Morgans&quot;</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="n">vrnt_genpos_units</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;cM&quot;</span><span class="p">,</span> <span class="s2">&quot;centiMorgans&quot;</span><span class="p">):</span>
            <span class="n">vrnt_genpos_convert</span> <span class="o">=</span> <span class="mf">100.0</span> <span class="o">*</span> <span class="n">vrnt_genpos_convert</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;vrnt_genpos units must be &#39;M&#39;, &#39;Morgans&#39;, &#39;cM&#39;, or &#39;centiMorgans&#39; but received value &#39;</span><span class="si">{0}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">vrnt_genpos_units</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># construct output dataframe</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="n">vrnt_chrgrp_col</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos_col</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_genpos_col</span><span class="p">:</span> <span class="n">vrnt_genpos_convert</span><span class="p">,</span>
        <span class="p">})</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.to_csv">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.to_csv">[docs]</a>
    <span class="k">def</span> <span class="nf">to_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_phypos_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_units</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>
            <span class="n">sep</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="p">,</span> 
            <span class="n">header</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> 
            <span class="n">index</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write a StandardGeneticMap to a CSV file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            CSV file name to which to write.</span>

<span class="sd">        vrnt_chrgrp_col : str, default = &quot;chr&quot;</span>
<span class="sd">            Name of the chromosome/linkage group name column to which to export.</span>

<span class="sd">        vrnt_phypos_col : str, default = &quot;pos&quot;</span>
<span class="sd">            Name of the physical position column to which to export.</span>

<span class="sd">        vrnt_genpos_col : str, default = &quot;cM&quot;</span>
<span class="sd">            Name of the genetic position column to which to export.</span>

<span class="sd">        vrnt_genpos_units : str, default = &quot;cM&quot;</span>
<span class="sd">            Units of the genetic position column to which to export.</span>
<span class="sd">            Options are listed below and are case-sensitive:</span>
<span class="sd">            </span>
<span class="sd">            - ``&quot;M&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;Morgans&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;cM&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            - ``&quot;centiMorgans&quot;`` - genetic position units are in centiMorgans</span>

<span class="sd">        sep : str, default = &quot;,&quot;</span>
<span class="sd">            Separator to use in the exported CSV file.</span>
<span class="sd">        </span>
<span class="sd">        header : bool, default = True</span>
<span class="sd">            Whether to save header names.</span>
<span class="sd">        </span>
<span class="sd">        index : bool, default = False</span>
<span class="sd">            Whether to save a row index in the exported CSV file.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments to use for dictating export to a CSV.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># convert GeneticMap to pandas.DataFrame</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">(</span>
            <span class="n">vrnt_chrgrp_col</span> <span class="o">=</span> <span class="n">vrnt_chrgrp_col</span><span class="p">,</span>
            <span class="n">vrnt_phypos_col</span> <span class="o">=</span> <span class="n">vrnt_phypos_col</span><span class="p">,</span>
            <span class="n">vrnt_genpos_col</span> <span class="o">=</span> <span class="n">vrnt_genpos_col</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="n">vrnt_genpos_units</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># use pandas to save to csv</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">path_or_buf</span> <span class="o">=</span> <span class="n">filename</span><span class="p">,</span>
            <span class="n">sep</span> <span class="o">=</span> <span class="n">sep</span><span class="p">,</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">header</span><span class="p">,</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">index</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span></div>


    <span class="c1">############################## Class Methods ###############################</span>

    <span class="c1">#################### Import Methods ####################</span>
<div class="viewcode-block" id="StandardGeneticMap.from_pandas">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.from_pandas">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_pandas</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span>
            <span class="n">df</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_phypos_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span> 
            <span class="n">spline</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">spline_kind</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span><span class="p">,</span>
            <span class="n">spline_fill_value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;extrapolate&quot;</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">auto_build_spline</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read a StandardGeneticMap from a pandas.DataFrame.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        df : pandas.DataFrame</span>
<span class="sd">            Pandas dataframe from which to read.</span>

<span class="sd">        vrnt_chrgrp_col : str, Integral, default = &quot;chr&quot;</span>
<span class="sd">            Name or number of the chromosome/linkage group name column from which to import.</span>

<span class="sd">        vrnt_phypos_col : str, Integral, default = &quot;pos&quot;</span>
<span class="sd">            Name or number of the physical position column from which to import.</span>

<span class="sd">        vrnt_genpos_col : str, Integral, default = &quot;cM&quot;</span>
<span class="sd">            Name or number of the genetic position column from which to import.</span>

<span class="sd">        spline : dict, None, default = None</span>
<span class="sd">            Pre-built interpolation spline to associate with the genetic map.</span>
<span class="sd">        </span>
<span class="sd">        spline_kind : str, default = &quot;linear&quot;</span>
<span class="sd">            In automatic building of splines, the spline kind to be built.</span>
<span class="sd">            Specifies the kind of interpolation as a string (&#39;linear&#39;,</span>
<span class="sd">            &#39;nearest&#39;, &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39;, &#39;cubic&#39;, &#39;previous&#39;,</span>
<span class="sd">            &#39;next&#39;, where &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39; and &#39;cubic&#39; refer to a</span>
<span class="sd">            spline interpolation of zeroth, first, second or third order;</span>
<span class="sd">            &#39;previous&#39; and &#39;next&#39; simply return the previous or next value of</span>
<span class="sd">            the point) or as an integer specifying the order of the spline</span>
<span class="sd">            interpolator to use.</span>
<span class="sd">            </span>
<span class="sd">        spline_fill_value : str, numpy.ndarray, default = &quot;extrapolate&quot;</span>
<span class="sd">            In automatic building of splines, the spline fill value to use.</span>
<span class="sd">            If &#39;extrapolate&#39;, then points outside the data range will be</span>
<span class="sd">            extrapolated.</span>
<span class="sd">            If a ndarray (or float), this value will be used to fill in for</span>
<span class="sd">            requested points outside of the data range. If not provided, then</span>
<span class="sd">            the default is NaN. The array-like must broadcast properly to the</span>
<span class="sd">            dimensions of the non-interpolation axes.</span>
<span class="sd">            If a two-element tuple, then the first element is used as a fill</span>
<span class="sd">            value for x_new &lt; x[0] and the second element is used for</span>
<span class="sd">            x_new &gt; x[-1]. Anything that is not a 2-element tuple (e.g., list</span>
<span class="sd">            or ndarray, regardless of shape) is taken to be a single array-like</span>
<span class="sd">            argument meant to be used for both bounds as below,</span>
<span class="sd">            above = fill_value, fill_value.</span>

<span class="sd">        auto_group : bool</span>
<span class="sd">            Whether to automatically sort and group variants into chromosome groups.</span>
<span class="sd">        </span>
<span class="sd">        auto_build_spline : bool</span>
<span class="sd">            Whether to automatically construct a spline on object construction.</span>
<span class="sd">            If ``spline`` is provided, then this spline is overwritten.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments to use for dictating importing from a </span>
<span class="sd">            pandas.DataFrame.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A StandardGeneticMap read from a pandas.DataFrame.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># data type checks</span>
        <span class="n">check_is_pandas_DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;df&quot;</span><span class="p">)</span>
        <span class="n">check_is_str_or_Integral</span><span class="p">(</span><span class="n">vrnt_chrgrp_col</span><span class="p">,</span> <span class="s2">&quot;vrnt_chrgrp_col&quot;</span><span class="p">)</span>
        <span class="n">check_is_str_or_Integral</span><span class="p">(</span><span class="n">vrnt_phypos_col</span><span class="p">,</span> <span class="s2">&quot;vrnt_phypos_col&quot;</span><span class="p">)</span>
        <span class="n">check_is_str_or_Integral</span><span class="p">(</span><span class="n">vrnt_genpos_col</span><span class="p">,</span> <span class="s2">&quot;vrnt_genpos_col&quot;</span><span class="p">)</span>

        <span class="c1"># extract data (pandas.Series data type)</span>
        <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">vrnt_chrgrp_col</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vrnt_chrgrp_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="n">vrnt_chrgrp_col</span><span class="p">]</span>
        <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">vrnt_phypos_col</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vrnt_phypos_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="n">vrnt_phypos_col</span><span class="p">]</span>
        <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">vrnt_genpos_col</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">vrnt_genpos_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="n">vrnt_genpos_col</span><span class="p">]</span>

        <span class="c1"># convert to numpy</span>
        <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>
        <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>
        <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>
        
        <span class="c1"># construct object</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span>
            <span class="n">vrnt_chrgrp</span>       <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
            <span class="n">vrnt_phypos</span>       <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
            <span class="n">vrnt_genpos</span>       <span class="o">=</span> <span class="n">vrnt_genpos</span><span class="p">,</span>
            <span class="n">spline</span>            <span class="o">=</span> <span class="n">spline</span><span class="p">,</span>
            <span class="n">spline_kind</span>       <span class="o">=</span> <span class="n">spline_kind</span><span class="p">,</span>
            <span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">spline_fill_value</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="n">vrnt_genpos_units</span><span class="p">,</span>
            <span class="n">auto_group</span>        <span class="o">=</span> <span class="n">auto_group</span><span class="p">,</span>
            <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="n">auto_build_spline</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>


<div class="viewcode-block" id="StandardGeneticMap.from_csv">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.StandardGeneticMap.from_csv">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_csv</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> 
            <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_phypos_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> 
            <span class="n">vrnt_genpos_col</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Integral</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span> 
            <span class="n">spline</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">spline_kind</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span><span class="p">,</span>
            <span class="n">spline_fill_value</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;extrapolate&quot;</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
            <span class="n">auto_group</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">auto_build_spline</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">sep</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="p">,</span> 
            <span class="n">header</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> 
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;StandardGeneticMap&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read a StandardGeneticMap from a CSV file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str, path object, or file-like object</span>
<span class="sd">            Any valid string path, including URLs. Valid URL schemes include </span>
<span class="sd">            http, ftp, s3, and file. For file URLs, a host is expected (see </span>
<span class="sd">            pandas docs).</span>
<span class="sd">        </span>
<span class="sd">        vrnt_chrgrp_col : str, Integral, default = &quot;chr&quot;</span>
<span class="sd">            Name or number of the chromosome/linkage group name column from </span>
<span class="sd">            which to import.</span>

<span class="sd">        vrnt_phypos_col : str, Integral, default = &quot;pos&quot;</span>
<span class="sd">            Name or number of the physical position column from which to import.</span>

<span class="sd">        vrnt_genpos_col : str, Integral, default = &quot;cM&quot;</span>
<span class="sd">            Name or number of the genetic position column from which to import.</span>

<span class="sd">        spline : dict, None, default = None</span>
<span class="sd">            Pre-built interpolation spline to associate with the genetic map.</span>
<span class="sd">        </span>
<span class="sd">        spline_kind : str, default = &quot;linear&quot;</span>
<span class="sd">            In automatic building of splines, the spline kind to be built.</span>
<span class="sd">            Specifies the kind of interpolation as a string (&#39;linear&#39;,</span>
<span class="sd">            &#39;nearest&#39;, &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39;, &#39;cubic&#39;, &#39;previous&#39;,</span>
<span class="sd">            &#39;next&#39;, where &#39;zero&#39;, &#39;slinear&#39;, &#39;quadratic&#39; and &#39;cubic&#39; refer to a</span>
<span class="sd">            spline interpolation of zeroth, first, second or third order;</span>
<span class="sd">            &#39;previous&#39; and &#39;next&#39; simply return the previous or next value of</span>
<span class="sd">            the point) or as an integer specifying the order of the spline</span>
<span class="sd">            interpolator to use.</span>
<span class="sd">            </span>
<span class="sd">        spline_fill_value : str, numpy.ndarray, default = &quot;extrapolate&quot;</span>
<span class="sd">            In automatic building of splines, the spline fill value to use.</span>
<span class="sd">            If &#39;extrapolate&#39;, then points outside the data range will be</span>
<span class="sd">            extrapolated.</span>
<span class="sd">            If a ndarray (or float), this value will be used to fill in for</span>
<span class="sd">            requested points outside of the data range. If not provided, then</span>
<span class="sd">            the default is NaN. The array-like must broadcast properly to the</span>
<span class="sd">            dimensions of the non-interpolation axes.</span>
<span class="sd">            If a two-element tuple, then the first element is used as a fill</span>
<span class="sd">            value for x_new &lt; x[0] and the second element is used for</span>
<span class="sd">            x_new &gt; x[-1]. Anything that is not a 2-element tuple (e.g., list</span>
<span class="sd">            or ndarray, regardless of shape) is taken to be a single array-like</span>
<span class="sd">            argument meant to be used for both bounds as below,</span>
<span class="sd">            above = fill_value, fill_value.</span>

<span class="sd">        vrnt_genpos_units : str, default = &quot;M&quot;</span>
<span class="sd">            Units in which genetic positions in the ``vrnt_genpos`` array are </span>
<span class="sd">            stored. Options are listed below and are case-sensitive:</span>
<span class="sd">            </span>
<span class="sd">            - ``&quot;M&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;Morgans&quot;`` - genetic position units are in Morgans</span>
<span class="sd">            - ``&quot;cM&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            - ``&quot;centiMorgans&quot;`` - genetic position units are in centiMorgans</span>
<span class="sd">            </span>
<span class="sd">            Internally, all genetic positions are stored in Morgans. Providing </span>
<span class="sd">            the units of the input  </span>

<span class="sd">        auto_group : bool</span>
<span class="sd">            Whether to automatically sort and group variants into chromosome groups.</span>
<span class="sd">        </span>
<span class="sd">        auto_build_spline : bool</span>
<span class="sd">            Whether to automatically construct a spline on object construction.</span>
<span class="sd">            If ``spline`` is provided, then this spline is overwritten.</span>

<span class="sd">        sep : str, default = &#39;,&#39;</span>
<span class="sd">            CSV delimiter to use.</span>
<span class="sd">        </span>
<span class="sd">        header : int, list of int, default=0</span>
<span class="sd">            Row number(s) to use as the column names, and the start of the data.</span>

<span class="sd">        kwargs : dict</span>
<span class="sd">            Additional keyword arguments to use for dictating importing from a CSV.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        out : StandardGeneticMap</span>
<span class="sd">            A StandardGeneticMap read from a CSV file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># read file using pandas</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">filename</span><span class="p">,</span>
            <span class="n">sep</span> <span class="o">=</span> <span class="n">sep</span><span class="p">,</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">header</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="c1"># construct genetic map from pandas.DataFrame</span>
        <span class="n">out</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
            <span class="n">vrnt_chrgrp_col</span> <span class="o">=</span> <span class="n">vrnt_chrgrp_col</span><span class="p">,</span>
            <span class="n">vrnt_phypos_col</span> <span class="o">=</span> <span class="n">vrnt_phypos_col</span><span class="p">,</span>
            <span class="n">vrnt_genpos_col</span> <span class="o">=</span> <span class="n">vrnt_genpos_col</span><span class="p">,</span>
            <span class="n">spline</span> <span class="o">=</span> <span class="n">spline</span><span class="p">,</span>
            <span class="n">spline_kind</span> <span class="o">=</span> <span class="n">spline_kind</span><span class="p">,</span>
            <span class="n">spline_fill_value</span> <span class="o">=</span> <span class="n">spline_fill_value</span><span class="p">,</span>
            <span class="n">vrnt_genpos_units</span> <span class="o">=</span> <span class="n">vrnt_genpos_units</span><span class="p">,</span>
            <span class="n">auto_group</span> <span class="o">=</span> <span class="n">auto_group</span><span class="p">,</span>
            <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="n">auto_build_spline</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">out</span></div>
</div>




<span class="c1">################################## Utilities ###################################</span>
<div class="viewcode-block" id="check_is_StandardGeneticMap">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.popgen.gmap.StandardGeneticMap.check_is_StandardGeneticMap.html#pybrops.popgen.gmap.StandardGeneticMap.check_is_StandardGeneticMap">[docs]</a>
<span class="k">def</span> <span class="nf">check_is_StandardGeneticMap</span><span class="p">(</span><span class="n">v</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> <span class="n">vname</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if object is of type StandardGeneticMap. Otherwise raise TypeError.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    v : object</span>
<span class="sd">        Any Python object to test.</span>
<span class="sd">    vname : str</span>
<span class="sd">        Name of variable to print in TypeError message.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">StandardGeneticMap</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
            <span class="s2">&quot;variable &#39;</span><span class="si">{0}</span><span class="s2">&#39; must be of type &#39;</span><span class="si">{1}</span><span class="s2">&#39; but received type &#39;</span><span class="si">{2}</span><span class="s2">&#39;.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">vname</span><span class="p">,</span>
                <span class="n">StandardGeneticMap</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span>
                <span class="nb">type</span><span class="p">(</span><span class="n">v</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span>
            <span class="p">)</span>
        <span class="p">)</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
       Copyright 2023, Robert Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>