
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Coancestry Matrices &#8212; pybrops 1.0.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/population_genetics/coancestry_matrices';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Optimization Problems" href="../optimization/optimization_problems.html" />
    <link rel="prev" title="Breeding Value Matrices" href="breeding_value_matrices.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">pybrops 1.0.0 documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item"><nav class="bd-links" id="bd-docs-nav" aria-label="Section navigation">
  <p class="bd-links__title" role="heading" aria-level="1">
    Section Navigation
  </p>
  <div class="bd-toc-item navbar-nav">
    <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../examples/examples.html">Examples</a></li>
</ul>

  </div>
</nav>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="coancestry-matrices">
<h1>Coancestry Matrices<a class="headerlink" href="#coancestry-matrices" title="Permalink to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Permalink to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code> family of classes is used to represent coancestry relationships between individuals. <code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code> objects can be used in the estimation of genomic prediction models and to make selection decisions.</p>
</section>
<section id="summary-of-coancestry-matrix-classes">
<h2>Summary of Coancestry Matrix Classes<a class="headerlink" href="#summary-of-coancestry-matrix-classes" title="Permalink to this heading">#</a></h2>
<table class="colwidths-given table" id="id1">
<caption><span class="caption-text">Summary of classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.cmat</span></code> module</span><a class="headerlink" href="#id1" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 26%" />
<col style="width: 21%" />
<col style="width: 53%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all coancestry matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseCoancestryMatrix</span></code></p></td>
<td><p>Semi-Abstract</p></td>
<td><p>Semi-implemented class for deriving new dense coancestry matrix child classes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseMolecularCoancestryMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing dense molecular coancestry matrices.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseVanRadenCoancestryMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing a genomic relationship matrix defined by VanRaden (2008).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DenseYangCoancestryMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing a genomic relationship matrix defined by Yang.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="loading-coancestry-matrix-modules">
<h2>Loading Coancestry Matrix Modules<a class="headerlink" href="#loading-coancestry-matrix-modules" title="Permalink to this heading">#</a></h2>
<p>Importing coancestry matrix classes can be accomplished using the following import statements:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import the CoancestryMatrix class (an abstract interface class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.CoancestryMatrix</span> <span class="kn">import</span> <span class="n">CoancestryMatrix</span>

<span class="c1"># import the DenseCoancestryMatrix class (a semi-abstract class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.DenseCoancestryMatrix</span> <span class="kn">import</span> <span class="n">DenseCoancestryMatrix</span>

<span class="c1"># import the DenseMolecularCoancestryMatrix class (a concrete implemented class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.DenseMolecularCoancestryMatrix</span> <span class="kn">import</span> <span class="n">DenseMolecularCoancestryMatrix</span>

<span class="c1"># import the DenseVanRadenCoancestryMatrix class (a concrete implemented class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.DenseVanRadenCoancestryMatrix</span> <span class="kn">import</span> <span class="n">DenseVanRadenCoancestryMatrix</span>

<span class="c1"># import the DenseYangCoancestryMatrix class (a concrete implemented class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.DenseYangCoancestryMatrix</span> <span class="kn">import</span> <span class="n">DenseYangCoancestryMatrix</span>
</pre></div>
</div>
</section>
<section id="creating-coancestry-matrices">
<h2>Creating Coancestry Matrices<a class="headerlink" href="#creating-coancestry-matrices" title="Permalink to this heading">#</a></h2>
<section id="creating-coancestry-matrices-from-numpy-arrays">
<h3>Creating coancestry matrices from NumPy arrays<a class="headerlink" href="#creating-coancestry-matrices-from-numpy-arrays" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># shape parameters</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># create random coancestries</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">))</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span>
    <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
<span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create a coancestry matrix from NumPy arrays</span>
<span class="n">cmat</span> <span class="o">=</span> <span class="n">DenseMolecularCoancestryMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="creating-coancestry-matrices-from-genotypematrix-objects">
<h3>Creating coancestry matrices from GenotypeMatrix objects<a class="headerlink" href="#creating-coancestry-matrices-from-genotypematrix-objects" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">nvrnt</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">nchrom</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">ploidy</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># create random genotypes</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">nvrnt</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span>
    <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
<span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant chromsome assignments</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nchrom</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker physical positions</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span>
    <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
<span class="p">)</span>

<span class="c1"># create a genotype matrix from scratch using NumPy arrays</span>
<span class="n">gmat</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
<span class="p">)</span>

<span class="c1"># group taxa and variants</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>

<span class="c1"># construct Coancestry Matrix from a Genotype Matrix</span>
<span class="n">cmat</span> <span class="o">=</span> <span class="n">DenseMolecularCoancestryMatrix</span><span class="o">.</span><span class="n">from_gmat</span><span class="p">(</span><span class="n">gmat</span> <span class="o">=</span> <span class="n">gmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-coancestry-matrices-from-hdf5-files">
<h3>Loading coancestry matrices from HDF5 files<a class="headerlink" href="#loading-coancestry-matrices-from-hdf5-files" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read from file</span>
<span class="n">cmat</span> <span class="o">=</span> <span class="n">DenseMolecularCoancestryMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;sample_coancestry_matrix.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="coancestry-matrix-properties">
<h2>Coancestry Matrix Properties<a class="headerlink" href="#coancestry-matrix-properties" title="Permalink to this heading">#</a></h2>
<section id="general-properties">
<h3>General properties<a class="headerlink" href="#general-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id2">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code> general properties</span><a class="headerlink" href="#id2" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat</span></code></p></td>
<td><p>The raw coancestry matrix pointer</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_ndim</span></code></p></td>
<td><p>The number of dimensions for the coancestry matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat_shape</span></code></p></td>
<td><p>The coancestry matrix shape</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">location</span></code></p></td>
<td><p>The location of the coancestry matrix if it has been transformed</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">scale</span></code></p></td>
<td><p>The scale of the coancestry matrix if it has been transformed</p></td>
</tr>
</tbody>
</table>
</section>
<section id="taxa-related-properties">
<h3>Taxa-related properties<a class="headerlink" href="#taxa-related-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code> taxa properties</span><a class="headerlink" href="#id3" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntaxa</span></code></p></td>
<td><p>The number of taxa represented by the coancestry matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></td>
<td><p>The names of the taxa</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_axis</span></code></p></td>
<td><p>The matrix axis along which taxa are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code></p></td>
<td><p>An optional taxa group label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_name</span></code></p></td>
<td><p>If taxa are sorted by group: get the names of the groups</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_stix</span></code></p></td>
<td><p>If taxa are sorted by group: get the start indices (inclusive) for each group</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_spix</span></code></p></td>
<td><p>If taxa are sorted by group: get the stop indices (exclusive) for each group</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_len</span></code></p></td>
<td><p>If taxa are sorted by group: get the length of each group</p></td>
</tr>
</tbody>
</table>
</section>
<section id="square-matrix-related-properties">
<h3>Square matrix-related properties<a class="headerlink" href="#square-matrix-related-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id4">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">CoancestryMatrix</span></code> square matrix properties</span><a class="headerlink" href="#id4" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nsquare</span></code></p></td>
<td><p>The number of square axes for the coancestry matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes</span></code></p></td>
<td><p>The axes indices for the square axes for the coancestry matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">square_axes_len</span></code></p></td>
<td><p>The lengths of the square axes for the coancestry matrix</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="copying-coancestry-matrices">
<h2>Copying Coancestry Matrices<a class="headerlink" href="#copying-coancestry-matrices" title="Permalink to this heading">#</a></h2>
<section id="shallow-copying">
<h3>Shallow copying<a class="headerlink" href="#shallow-copying" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy a coancestry matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">cmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="deep-copying">
<h3>Deep copying<a class="headerlink" href="#deep-copying" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy a coancestry matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">cmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="copy-on-element-manipulation">
<h2>Copy-On Element Manipulation<a class="headerlink" href="#copy-on-element-manipulation" title="Permalink to this heading">#</a></h2>
<section id="adjoin-elements">
<h3>Adjoin elements<a class="headerlink" href="#adjoin-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new coancestry matrix to demonstrate</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># adjoin coancestry matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="delete-elements">
<h3>Delete elements<a class="headerlink" href="#delete-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># delete first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="insert-elements">
<h3>Insert elements<a class="headerlink" href="#insert-elements" title="Permalink to this heading">#</a></h3>
</section>
<section id="select-elements">
<h3>Select elements<a class="headerlink" href="#select-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># select first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="in-place-element-manipulation">
<h2>In-Place Element Manipulation<a class="headerlink" href="#in-place-element-manipulation" title="Permalink to this heading">#</a></h2>
<section id="append-elements">
<h3>Append elements<a class="headerlink" href="#append-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># append coancestry matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">cmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>
</pre></div>
</div>
</section>
<section id="remove-elements">
<h3>Remove elements<a class="headerlink" href="#remove-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
<section id="incorporate-elements">
<h3>Incorporate elements<a class="headerlink" href="#incorporate-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># incorp coancestry matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">cmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>      <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">cmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>
</pre></div>
</div>
</section>
<section id="concatenate-elements">
<h3>Concatenate elements<a class="headerlink" href="#concatenate-elements" title="Permalink to this heading">#</a></h3>
</section>
</section>
<section id="grouping-and-sorting">
<h2>Grouping and Sorting<a class="headerlink" href="#grouping-and-sorting" title="Permalink to this heading">#</a></h2>
<section id="reordering-elements">
<h3>Reordering elements<a class="headerlink" href="#reordering-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">cmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values along the taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="lexsorting-elements">
<h3>Lexsorting elements<a class="headerlink" href="#lexsorting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for taxa</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">cmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">cmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the taxa axis</span>
<span class="n">cmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">cmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="sorting-elements">
<h3>Sorting elements<a class="headerlink" href="#sorting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="grouping-elements">
<h3>Grouping elements<a class="headerlink" href="#grouping-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>

<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="coancestry-and-kinship-methods">
<h2>Coancestry and Kinship Methods<a class="headerlink" href="#coancestry-and-kinship-methods" title="Permalink to this heading">#</a></h2>
<section id="retrieving-coancestry-values">
<h3>Retrieving coancestry values<a class="headerlink" href="#retrieving-coancestry-values" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the coancestry at a specific matrix coordinate</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">coancestry</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="retrieving-kinship-values">
<h3>Retrieving kinship values<a class="headerlink" href="#retrieving-kinship-values" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the kinship at a specific matrix coordinate</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">kinship</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="retrieving-the-coancestry-matrix-as-a-specific-format">
<h3>Retrieving the coancestry matrix as a specific format<a class="headerlink" href="#retrieving-the-coancestry-matrix-as-a-specific-format" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the coancestry matrix as a specific format</span>
<span class="n">cmat</span><span class="o">.</span><span class="n">mat_asformat</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;kinship&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="determining-if-the-coancestry-matrix-is-positive-semidefinite">
<h3>Determining if the coancestry matrix is positive semidefinite<a class="headerlink" href="#determining-if-the-coancestry-matrix-is-positive-semidefinite" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine if the coancestry matrix is positive semidefinite (convex)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">is_positive_semidefinite</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="applying-jitter-values-along-the-diagonal">
<h3>Applying jitter values along the diagonal<a class="headerlink" href="#applying-jitter-values-along-the-diagonal" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply a jitter along the diagonal to try to make the matrix positive semidefinite</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">apply_jitter</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="calculating-the-matrix-inverse">
<h3>Calculating the matrix inverse<a class="headerlink" href="#calculating-the-matrix-inverse" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the inverse of the coancestry matrix</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">inverse</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">inverse</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;kinship&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-maximum-attainable-inbreeding">
<h3>Calculating maximum attainable inbreeding<a class="headerlink" href="#calculating-maximum-attainable-inbreeding" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the maximum attainable inbreeding after 1 generation</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">max_inbreeding</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">min_inbreeding</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;kinship&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-minimum-attainable-inbreeding">
<h3>Calculating minimum attainable inbreeding<a class="headerlink" href="#calculating-minimum-attainable-inbreeding" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the minimum attainable inbreeding after 1 generation</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">min_inbreeding</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">min_inbreeding</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;kinship&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="summary-statistics">
<h2>Summary Statistics<a class="headerlink" href="#summary-statistics" title="Permalink to this heading">#</a></h2>
<section id="maximum-coancestry">
<h3>Maximum coancestry<a class="headerlink" href="#maximum-coancestry" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># get the max for the whole coancestry matrix</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="mean-coancestry">
<h3>Mean coancestry<a class="headerlink" href="#mean-coancestry" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># get the mean for the whole coancestry matrix</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="minimum-coancestry">
<h3>Minimum coancestry<a class="headerlink" href="#minimum-coancestry" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># get the min for the whole coancestry matrix</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">cmat</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="saving-coancestry-matrices">
<h2>Saving Coancestry Matrices<a class="headerlink" href="#saving-coancestry-matrices" title="Permalink to this heading">#</a></h2>
<section id="exporting-to-hdf5">
<h3>Exporting to HDF5<a class="headerlink" href="#exporting-to-hdf5" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write a coancestry matrix to an HDF5 file</span>
<span class="n">cmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_coancestry_matrix.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="breeding_value_matrices.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Breeding Value Matrices</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="../optimization/optimization_problems.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Optimization Problems</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#class-family-overview">
   Class Family Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-of-coancestry-matrix-classes">
   Summary of Coancestry Matrix Classes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-coancestry-matrix-modules">
   Loading Coancestry Matrix Modules
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-coancestry-matrices">
   Creating Coancestry Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-coancestry-matrices-from-numpy-arrays">
     Creating coancestry matrices from NumPy arrays
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-coancestry-matrices-from-genotypematrix-objects">
     Creating coancestry matrices from GenotypeMatrix objects
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-coancestry-matrices-from-hdf5-files">
     Loading coancestry matrices from HDF5 files
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#coancestry-matrix-properties">
   Coancestry Matrix Properties
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#general-properties">
     General properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#taxa-related-properties">
     Taxa-related properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#square-matrix-related-properties">
     Square matrix-related properties
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#copying-coancestry-matrices">
   Copying Coancestry Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#shallow-copying">
     Shallow copying
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#deep-copying">
     Deep copying
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#copy-on-element-manipulation">
   Copy-On Element Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#adjoin-elements">
     Adjoin elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#delete-elements">
     Delete elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#insert-elements">
     Insert elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#select-elements">
     Select elements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#in-place-element-manipulation">
   In-Place Element Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#append-elements">
     Append elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remove-elements">
     Remove elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#incorporate-elements">
     Incorporate elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#concatenate-elements">
     Concatenate elements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-and-sorting">
   Grouping and Sorting
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reordering-elements">
     Reordering elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#lexsorting-elements">
     Lexsorting elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sorting-elements">
     Sorting elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#grouping-elements">
     Grouping elements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#coancestry-and-kinship-methods">
   Coancestry and Kinship Methods
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#retrieving-coancestry-values">
     Retrieving coancestry values
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#retrieving-kinship-values">
     Retrieving kinship values
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#retrieving-the-coancestry-matrix-as-a-specific-format">
     Retrieving the coancestry matrix as a specific format
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#determining-if-the-coancestry-matrix-is-positive-semidefinite">
     Determining if the coancestry matrix is positive semidefinite
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#applying-jitter-values-along-the-diagonal">
     Applying jitter values along the diagonal
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-the-matrix-inverse">
     Calculating the matrix inverse
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-maximum-attainable-inbreeding">
     Calculating maximum attainable inbreeding
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-minimum-attainable-inbreeding">
     Calculating minimum attainable inbreeding
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-statistics">
   Summary Statistics
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#maximum-coancestry">
     Maximum coancestry
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mean-coancestry">
     Mean coancestry
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#minimum-coancestry">
     Minimum coancestry
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#saving-coancestry-matrices">
   Saving Coancestry Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#exporting-to-hdf5">
     Exporting to HDF5
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2023, Robert Shrote.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>