
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Genetic Map Functions &#8212; pybrops 1.0.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/population_genetics/genetic_map_functions';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genotype Matrices" href="genotype_matrices.html" />
    <link rel="prev" title="Genetic Maps" href="genetic_maps.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">pybrops 1.0.0 documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../installation/installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../prerequisites/prerequisites.html">
                        Prerequisites
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../examples.html">
                        .. Examples
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../installation/installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../prerequisites/prerequisites.html">
                        Prerequisites
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../examples.html">
                        .. Examples
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item"><nav class="bd-links" id="bd-docs-nav" aria-label="Section navigation">
  <p class="bd-links__title" role="heading" aria-level="1">
    Section Navigation
  </p>
  <div class="bd-toc-item navbar-nav">
    <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>

  </div>
</nav>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="genetic-map-functions">
<h1>Genetic Map Functions<a class="headerlink" href="#genetic-map-functions" title="Permalink to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Permalink to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">GeneticMapFunction</span></code> family of classes allows for the representation of genetic map functions within PyBrOpS. Genetic map functions are used to convert genetic map positions into recombination probabilities, which are useful in the simulation of genetic recombination and the estimation of expected progeny variance. Popular genetic map functions like the Haldane and Kosambi map functions are provided by PyBrOpS out-of-box.</p>
</section>
<section id="summary-of-genetic-map-function-classes">
<h2>Summary of Genetic Map Function Classes<a class="headerlink" href="#summary-of-genetic-map-function-classes" title="Permalink to this heading">#</a></h2>
<p>Genetic map function support in PyBrOpS is found in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmap</span></code> module. Contained in this submodule are several <code class="docutils literal notranslate"><span class="pre">GeneticMapFunction</span></code> class type definitions. These classes are summarized in the table below.</p>
<table class="colwidths-given table" id="id1">
<caption><span class="caption-text">Summary of genetic map function classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmap</span></code> module</span><a class="headerlink" href="#id1" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 28%" />
<col style="width: 17%" />
<col style="width: 56%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GeneticMapFunction</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genetic map function child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">HaldaneMapFunction</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing the Haldane genetic map function.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">KosambiMapFunction</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing the Kosambi genetic map function.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="loading-genetic-map-function-modules">
<h2>Loading Genetic Map Function Modules<a class="headerlink" href="#loading-genetic-map-function-modules" title="Permalink to this heading">#</a></h2>
<p>Genetic map function support in PyBrOpS is found in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmap</span></code> submodule. The <code class="docutils literal notranslate"><span class="pre">GeneticMapFunction</span></code> abstract class is the basal interface for all genetic map functions. PyBrOpS supports both the Haldane and Kosambi map functions, two of the most popular genetic map functions. Classes can be imported as so:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import the GeneticMapFunction class (an abstract interface class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.GeneticMapFunction</span> <span class="kn">import</span> <span class="n">GeneticMapFunction</span>

<span class="c1"># import the HaldaneMapFunction class (a concrete class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.HaldaneMapFunction</span> <span class="kn">import</span> <span class="n">HaldaneMapFunction</span>

<span class="c1"># import the KosambiMapFunction class (a concrete class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.KosambiMapFunction</span> <span class="kn">import</span> <span class="n">KosambiMapFunction</span>
</pre></div>
</div>
</section>
<section id="constructing-genetic-map-function-objects">
<h2>Constructing Genetic Map Function Objects<a class="headerlink" href="#constructing-genetic-map-function-objects" title="Permalink to this heading">#</a></h2>
<p>Haldane and Kosambi map function objects can be created as seen in the example below. Object construction requires no arguments.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a Haldane map function object</span>
<span class="n">haldane</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>

<span class="c1"># create a Kosambi map function object</span>
<span class="n">kosambi</span> <span class="o">=</span> <span class="n">KosambiMapFunction</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="calculating-recombination-probabilities">
<h2>Calculating Recombination Probabilities<a class="headerlink" href="#calculating-recombination-probabilities" title="Permalink to this heading">#</a></h2>
<p>Recombination probabilities for arbitrary genetic map distances can be calculated using the <code class="docutils literal notranslate"><span class="pre">mapfn</span></code> method. An example of its usage is below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create an array of map distances in Morgans</span>
<span class="n">d</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.2</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">1.4</span><span class="p">])</span>

<span class="c1"># calculate recombination probability using the Haldane map function</span>
<span class="n">r_haldane</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">mapfn</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>

<span class="c1"># calculate recombination probability using the Kosambi map function</span>
<span class="n">r_kosambi</span> <span class="o">=</span> <span class="n">kosambi</span><span class="o">.</span><span class="n">mapfn</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-genetic-map-distances">
<h2>Calculating Genetic Map Distances<a class="headerlink" href="#calculating-genetic-map-distances" title="Permalink to this heading">#</a></h2>
<p>Genetic map distances for arbitrary recombination probabilities can be calculated by using the <code class="docutils literal notranslate"><span class="pre">invmapfn</span></code> method, the inverse of the <code class="docutils literal notranslate"><span class="pre">mapfn</span></code> function. An example of its usage is below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># invert the recombination probability calculated previously</span>
<span class="n">d_haldane</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">invmapfn</span><span class="p">(</span><span class="n">r_haldane</span><span class="p">)</span>

<span class="c1"># invert the recombination probability calculated previously</span>
<span class="n">d_kosambi</span> <span class="o">=</span> <span class="n">kosambi</span><span class="o">.</span><span class="n">invmapfn</span><span class="p">(</span><span class="n">r_kosambi</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-sequential-and-pairwise-recombination-probabilities">
<h2>Calculating Sequential and Pairwise Recombination Probabilities<a class="headerlink" href="#calculating-sequential-and-pairwise-recombination-probabilities" title="Permalink to this heading">#</a></h2>
<p>In some instances, it may be useful to calculate sequential and/or pairwise recombination probabilities to simulate recombination. Genetic map function objects have four methods which allow for the calculation of these probabilities: <code class="docutils literal notranslate"><span class="pre">rprob1g</span></code>, <code class="docutils literal notranslate"><span class="pre">rprob1p</span></code>, <code class="docutils literal notranslate"><span class="pre">rprob2g</span></code>, and <code class="docutils literal notranslate"><span class="pre">rprob2p</span></code>.</p>
<section id="calculating-sequential-recombination-probabilities">
<h3>Calculating sequential recombination probabilities<a class="headerlink" href="#calculating-sequential-recombination-probabilities" title="Permalink to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">rprob1g</span></code> and <code class="docutils literal notranslate"><span class="pre">rprob1p</span></code> methods are used to calculate sequential recombination probabilities between successive markers, provided chromosome group assignments, and genetic positions or physical positions, respectively. Chromosome boundaries always start with probabilities of 0.5.</p>
<p>The code segment below demonstrates the use of the <code class="docutils literal notranslate"><span class="pre">rprob1g</span></code> and <code class="docutils literal notranslate"><span class="pre">rprob1p</span></code> methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate from genetic positions</span>
<span class="n">xoprob</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">rprob1g</span><span class="p">(</span>
    <span class="n">gmap</span> <span class="o">=</span> <span class="n">gmap</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_genpos</span>
<span class="p">)</span>

<span class="c1"># calculate from physical positions</span>
<span class="n">xoprob</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">rprob1p</span><span class="p">(</span>
    <span class="n">gmap</span> <span class="o">=</span> <span class="n">gmap</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_phypos</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-pairwise-recombination-probabilities">
<h3>Calculating pairwise recombination probabilities<a class="headerlink" href="#calculating-pairwise-recombination-probabilities" title="Permalink to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">rprob2g</span></code> and <code class="docutils literal notranslate"><span class="pre">rprob2p</span></code> methods are used to calculate pairwise recombination probabilities between markers, provided chromosome group assignments, and genetic positions or physical positions, respectively. Inter-chromosome pairings are always assigned probabilities of 0.5.</p>
<p>The code segment below demonstrates the use of the <code class="docutils literal notranslate"><span class="pre">rprob2g</span></code> and <code class="docutils literal notranslate"><span class="pre">rprob2p</span></code> methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate from genetic positions</span>
<span class="n">xoprob</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">rprob2g</span><span class="p">(</span>
    <span class="n">gmap</span> <span class="o">=</span> <span class="n">gmap</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_genpos</span>
<span class="p">)</span>

<span class="c1"># calculate from physical positions</span>
<span class="n">xoprob</span> <span class="o">=</span> <span class="n">haldane</span><span class="o">.</span><span class="n">rprob2p</span><span class="p">(</span>
    <span class="n">gmap</span> <span class="o">=</span> <span class="n">gmap</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">vrnt_phypos</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="genetic_maps.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Genetic Maps</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="genotype_matrices.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Genotype Matrices</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#class-family-overview">
   Class Family Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-of-genetic-map-function-classes">
   Summary of Genetic Map Function Classes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-genetic-map-function-modules">
   Loading Genetic Map Function Modules
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#constructing-genetic-map-function-objects">
   Constructing Genetic Map Function Objects
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculating-recombination-probabilities">
   Calculating Recombination Probabilities
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculating-genetic-map-distances">
   Calculating Genetic Map Distances
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculating-sequential-and-pairwise-recombination-probabilities">
   Calculating Sequential and Pairwise Recombination Probabilities
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-sequential-recombination-probabilities">
     Calculating sequential recombination probabilities
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculating-pairwise-recombination-probabilities">
     Calculating pairwise recombination probabilities
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2023, Robert Shrote.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>