
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genetic Maps &#8212; pybrops 1.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/population_genetics/genetic_maps';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genetic Map Functions" href="genetic_map_functions.html" />
    <link rel="prev" title="Random Number Generation" href="../core_infrastructure/random_number_generation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/examples.html">Examples</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/germplasm_panel_reduction.html">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_single_trait_phenotypic_selection.html">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/multiobjective_genomic_selection.html">Bi-Trait Genotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../standard_usage.html" class="nav-link">Standard PyBrOpS Usage</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Genetic Maps</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="genetic-maps">
<h1>Genetic Maps<a class="headerlink" href="#genetic-maps" title="Link to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Link to this heading">#</a></h2>
<p>Genetic maps are used to estimate marker-wise genetic map positions in mating simulations. PyBrOpS provides functionality for the reading and writing of genetic maps and the interpolation of genetic map positions via <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> classes. The ability to read a genetic map from a file allows for real genetic recombination data to be used in a breeding simulation, adding realism.</p>
</section>
<section id="summary-of-genetic-map-classes">
<h2>Summary of Genetic Map Classes<a class="headerlink" href="#summary-of-genetic-map-classes" title="Link to this heading">#</a></h2>
<p>Genetic map support for PyBrOpS is found in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmap</span></code> module. Contained in this module are several <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> class type definitions. These classes are summarized in the table below.</p>
<table class="table" id="id1">
<caption><span class="caption-text">Summary of genetic map classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmap</span></code> module</span><a class="headerlink" href="#id1" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 27.8%" />
<col style="width: 16.7%" />
<col style="width: 55.6%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genetic map child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ExtendedGeneticMap</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing genetic maps with additional genetic map metadata.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="loading-genetic-map-modules">
<h2>Loading Genetic Map Modules<a class="headerlink" href="#loading-genetic-map-modules" title="Link to this heading">#</a></h2>
<p>The various genetic map classes summarized above can be imported into a Python scope as demonstrated in the code example below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import the GeneticMap class (an abstract interface class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.GeneticMap</span> <span class="kn">import</span> <span class="n">GeneticMap</span>

<span class="c1"># import the ExtendedGeneticMap class (a concrete class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.ExtendedGeneticMap</span> <span class="kn">import</span> <span class="n">ExtendedGeneticMap</span>
</pre></div>
</div>
</section>
<section id="creating-genetic-maps">
<h2>Creating Genetic Maps<a class="headerlink" href="#creating-genetic-maps" title="Link to this heading">#</a></h2>
<section id="creating-genetic-maps-from-numpy-arrays">
<h3>Creating genetic maps from NumPy arrays<a class="headerlink" href="#creating-genetic-maps-from-numpy-arrays" title="Link to this heading">#</a></h3>
<p>Genetic map objects can be constructed from raw NumPy arrays, which may be useful for creating simulated genetic maps. The code below demonstrates the construction of an <code class="docutils literal notranslate"><span class="pre">ExtendedGeneticMap</span></code> from NumPy arrays.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define number of variants</span>
<span class="n">nvrnt</span> <span class="o">=</span> <span class="mi">100</span>

<span class="c1"># create random chromosome groups 1-9</span>
<span class="n">chroms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;123456789&quot;</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">chroms</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">object</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create variant physical positions in range [1, 2**28]</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">20</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create variant genetic positions in range [0,1]</span>
<span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_genpos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>

<span class="c1"># construct genetic map</span>
<span class="n">gmap</span> <span class="o">=</span> <span class="n">ExtendedGeneticMap</span><span class="p">(</span>
    <span class="n">vrnt_chrgrp</span><span class="o">=</span><span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span><span class="o">=</span><span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_stop</span><span class="o">=</span><span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span><span class="o">=</span><span class="n">vrnt_genpos</span><span class="p">,</span>
    <span class="n">vrnt_name</span><span class="o">=</span><span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">vrnt_fncode</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="c1"># not needed</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reading-genetic-maps-from-a-file">
<h3>Reading genetic maps from a file<a class="headerlink" href="#reading-genetic-maps-from-a-file" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> objects can also be read from a file. This may be useful for users who have their own, empirically determined genetic maps from an organism of their choosing. The example below illustrates the reading of an <code class="docutils literal notranslate"><span class="pre">ExtendedGeneticMap</span></code> from a file. Specifically, this is the maize genetic map constructed from the US NAM population as published by McMullen et al. (2009).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read genetic map from file</span>
<span class="c1"># for the purpose of this example, do not automatically group markers</span>
<span class="c1"># and build an interpolation spline after reading genetic map data.</span>
<span class="n">gmap</span> <span class="o">=</span> <span class="n">ExtendedGeneticMap</span><span class="o">.</span><span class="n">from_egmap</span><span class="p">(</span>
    <span class="s2">&quot;McMullen_2009_US_NAM.egmap&quot;</span><span class="p">,</span>
    <span class="n">auto_group</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">auto_build_spline</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="genetic-map-properties">
<h2>Genetic Map Properties<a class="headerlink" href="#genetic-map-properties" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> objects have a set of properties shared by all genetic maps. These properties can be grouped into two categories: marker variant properties and spline properties. The former set of properties contain information about the marker set constituting the genetic map, while the latter set of properties contain spline model information necessary for the interpolation of genetic map positions. Marker variant and spline properties are summarized in the tables below.</p>
<section id="marker-variant-properties">
<h3>Marker variant properties<a class="headerlink" href="#marker-variant-properties" title="Link to this heading">#</a></h3>
<table class="table" id="id2">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> marker variant properties</span><a class="headerlink" href="#id2" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nvrnt</span></code></p></td>
<td><p>Number of variants in the Genetic Map</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp</span></code></p></td>
<td><p>Marker variant chromosome group labels</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_phypos</span></code></p></td>
<td><p>Marker variant chromosome physical positions</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_genpos</span></code></p></td>
<td><p>Marker variant chromosome genetic positions</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_name</span></code></p></td>
<td><p>Marker variant names</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_name</span></code></p></td>
<td><p>Names of chromosome groups</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_stix</span></code></p></td>
<td><p>Chromosome group start indices</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_spix</span></code></p></td>
<td><p>Chromosome group stop indices</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_len</span></code></p></td>
<td><p>Number of marker variants on each chromosome group</p></td>
</tr>
</tbody>
</table>
</section>
<section id="spline-properties">
<h3>Spline properties<a class="headerlink" href="#spline-properties" title="Link to this heading">#</a></h3>
<table class="table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> spline properties</span><a class="headerlink" href="#id3" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">spline</span></code></p></td>
<td><p>Interpolation splines</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">spline_kind</span></code></p></td>
<td><p>Interpolation spline type</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">spline_fill_value</span></code></p></td>
<td><p>Interpolation spline default fill value</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="copying-genetic-maps">
<h2>Copying Genetic Maps<a class="headerlink" href="#copying-genetic-maps" title="Link to this heading">#</a></h2>
<p>At times, it may be necessary to copy a genetic map. There are two methods of copying genetic maps: shallow copying and deep copying.</p>
<section id="shallow-copying">
<h3>Shallow copying<a class="headerlink" href="#shallow-copying" title="Link to this heading">#</a></h3>
<p>In shallow copying, references to a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code>’s variant and spline data are copied to a new genetic map object. Copying is only one level deep and changes to the data in the original object may affect data values in the copied object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy the genetic map</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">gmap</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="deep-copying">
<h3>Deep copying<a class="headerlink" href="#deep-copying" title="Link to this heading">#</a></h3>
<p>In deep copying, data in a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code>’s variant and spline data are recursively copied. Copying occurs down to the deepest level making it so that changes to the data in the original object will not affect data values in the copied object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy the genetic map</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gmap</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="sorting-and-grouping-genetic-maps">
<h2>Sorting and Grouping Genetic Maps<a class="headerlink" href="#sorting-and-grouping-genetic-maps" title="Link to this heading">#</a></h2>
<section id="reordering-map-elements">
<h3>Reordering map elements<a class="headerlink" href="#reordering-map-elements" title="Link to this heading">#</a></h3>
<p>In some instances, it may be useful to manually reorder genetic map elements. This may be accomplished by providing an array of reordering indices to the <code class="docutils literal notranslate"><span class="pre">reorder</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gmap</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="lexsorting-map-elements">
<h3>Lexsorting map elements<a class="headerlink" href="#lexsorting-map-elements" title="Link to this heading">#</a></h3>
<p>An indirect stable sort may be performed using the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> method. If the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> method is not provided a set of <code class="docutils literal notranslate"><span class="pre">keys</span></code>, it defaults to utilizing marker variant chromosome group assignments, marker variant chromosome physical positions, and marker variant chromosome genetic positions in that order of priority.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">gmap</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">gmap</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span> <span class="n">gmap</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

<span class="c1"># lexsort using keys</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="sorting-map-elements">
<h3>Sorting map elements<a class="headerlink" href="#sorting-map-elements" title="Link to this heading">#</a></h3>
<p>In-place sorting of marker variants in a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> object can be accomplished using the <code class="docutils literal notranslate"><span class="pre">sort</span></code> method. The <code class="docutils literal notranslate"><span class="pre">sort</span></code> method optionally accepts a set of <code class="docutils literal notranslate"><span class="pre">keys</span></code> which can be used to sort the marker variants in the genetic map. If a set of <code class="docutils literal notranslate"><span class="pre">keys</span></code> is not provided, the keys are the same defaults as those in the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort the genetic map</span>
<span class="n">gmap</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="grouping-map-elements">
<h3>Grouping map elements<a class="headerlink" href="#grouping-map-elements" title="Link to this heading">#</a></h3>
<p>On import from a file, genetic map data needs to be sorted and grouped into marker linkage groups so that an interpolation spline can be built. Often, the constructor or an import method for a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> object will automatically sort and group data, but occasionally it may be necessary to manually sort and group marker data. This can be accomplished using the <code class="docutils literal notranslate"><span class="pre">group</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># group markers based on their chromosome/linkage group</span>
<span class="n">gmap</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
</pre></div>
</div>
<p>To test whether a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code>’s data have been sorted and grouped into linkage groups, the <code class="docutils literal notranslate"><span class="pre">is_grouped</span></code> method can be used:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># determine whether a GeneticMap is grouped using the ``is_grouped`` method</span>
<span class="n">value</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="genetic-map-congruency">
<h2>Genetic Map Congruency<a class="headerlink" href="#genetic-map-congruency" title="Link to this heading">#</a></h2>
<section id="checking-for-congruency">
<h3>Checking for congruency<a class="headerlink" href="#checking-for-congruency" title="Link to this heading">#</a></h3>
<p>Sometimes physical positions and genetic map positions are in disagreement as to their orderings. This may be caused by errors made in genome assemblies and/or genetic map assemblies. Unfortunately, these disagreements cause issues for interpolation spline construction and need to be removed or corrected. An elementwise physical position-genetic position congruence test can be conducted using the <code class="docutils literal notranslate"><span class="pre">congruence</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># elementwise test of marker congruence</span>
<span class="n">value</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">congruence</span><span class="p">()</span>
</pre></div>
</div>
<p>If one desires to test whether all of the markers in the <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> are congruent, one can use the <code class="docutils literal notranslate"><span class="pre">is_congruent</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># whole genetic map congruence test</span>
<span class="n">value</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">is_congruent</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="removing-map-discrepancies">
<h3>Removing map discrepancies<a class="headerlink" href="#removing-map-discrepancies" title="Link to this heading">#</a></h3>
<p>Loci where the physical positions and genetic positions are not in agreement may be automatically removed using the <code class="docutils literal notranslate"><span class="pre">remove_discrepancies</span></code>. Be mindful that a manual inspection and correction of a genetic map’s discrepancies may be superior to this automatic method.</p>
<p>The code below demonstrates automatic discrepancy removal:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># automatically remove discrepancies</span>
<span class="n">gmap</span><span class="o">.</span><span class="n">remove_discrepancies</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="building-interpolation-splines">
<h2>Building Interpolation Splines<a class="headerlink" href="#building-interpolation-splines" title="Link to this heading">#</a></h2>
<p>Before using a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> to interpolate genetic position data, an interpolation spline must be constructed. Often, the constructor or an import method for a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> object will automatically construct a spline from the provided data. Occasionally, it may be necessary to manually construct an interpolation spline. The <code class="docutils literal notranslate"><span class="pre">build_spline</span></code> method can be used to construct an interpolation spline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct a linear spline to interpolate genetic map positions</span>
<span class="n">gmap</span><span class="o">.</span><span class="n">build_spline</span><span class="p">()</span>
</pre></div>
</div>
<p>To test whether a <code class="docutils literal notranslate"><span class="pre">GeneticMap</span></code> has an interpolation spline, the <code class="docutils literal notranslate"><span class="pre">has_spline</span></code> method can be used:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># determine whether a GeneticMap has an interpolation spline using the</span>
<span class="c1"># ``has_spline`` method</span>
<span class="n">value</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">has_spline</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="interpolating-genetic-positions">
<h2>Interpolating Genetic Positions<a class="headerlink" href="#interpolating-genetic-positions" title="Link to this heading">#</a></h2>
<p>Interpolating genetic map positions from physical positions can be done using the <code class="docutils literal notranslate"><span class="pre">interp_genpos</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">### create new positions to interpolate</span>
<span class="c1"># construct linkage group array: everything is on chromosome 1</span>
<span class="n">chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>

<span class="c1"># construct physical position array</span>
<span class="n">phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">18203210</span><span class="p">,</span><span class="mi">19293034</span><span class="p">,</span><span class="mi">20110347</span><span class="p">,</span><span class="mi">20474722</span><span class="p">,</span><span class="mi">21398386</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>

<span class="c1"># interpolate new gentic map positions</span>
<span class="n">genpos</span> <span class="o">=</span> <span class="n">gmap</span><span class="o">.</span><span class="n">interp_genpos</span><span class="p">(</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">phypos</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../core_infrastructure/random_number_generation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Random Number Generation</p>
      </div>
    </a>
    <a class="right-next"
       href="genetic_map_functions.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genetic Map Functions</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#class-family-overview">Class Family Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-genetic-map-classes">Summary of Genetic Map Classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genetic-map-modules">Loading Genetic Map Modules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-genetic-maps">Creating Genetic Maps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-genetic-maps-from-numpy-arrays">Creating genetic maps from NumPy arrays</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-genetic-maps-from-a-file">Reading genetic maps from a file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-map-properties">Genetic Map Properties</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marker-variant-properties">Marker variant properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#spline-properties">Spline properties</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copying-genetic-maps">Copying Genetic Maps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#shallow-copying">Shallow copying</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deep-copying">Deep copying</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sorting-and-grouping-genetic-maps">Sorting and Grouping Genetic Maps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reordering-map-elements">Reordering map elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lexsorting-map-elements">Lexsorting map elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sorting-map-elements">Sorting map elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#grouping-map-elements">Grouping map elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-map-congruency">Genetic Map Congruency</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#checking-for-congruency">Checking for congruency</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-map-discrepancies">Removing map discrepancies</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#building-interpolation-splines">Building Interpolation Splines</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpolating-genetic-positions">Interpolating Genetic Positions</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>