
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genotype Matrices &#8212; pybrops 1.0.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/population_genetics/genotype_matrices';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Phased Genotype Matrices" href="phased_genotype_matrices.html" />
    <link rel="prev" title="Genetic Map Functions" href="genetic_map_functions.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.2 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/examples.html">Examples</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/germplasm_panel_reduction.html">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_single_trait_phenotypic_selection.html">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/multiobjective_genomic_selection.html">Bi-Trait Genotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/breeding_program_simulation/constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../standard_usage.html" class="nav-link">Standard PyBrOpS Usage</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Genotype Matrices</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="genotype-matrices">
<h1>Genotype Matrices<a class="headerlink" href="#genotype-matrices" title="Link to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Link to this heading">#</a></h2>
<p>Perhaps the most important family of classes in PyBrOpS is the <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> object family. The purpose of this family of classes is to store and represent genotypic data. <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> classes can be used in the estimation of genomic prediction models, the estimation of breeding values, the calculation of genomic relationship matrices, and to make selection decisions. Within PyBrOpS it is possible to read genotypic data from VCF files to create <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> objects, allowing for real-world data to be used in breeding program simulations.</p>
</section>
<section id="summary-of-genotype-matrix-classes">
<h2>Summary of Genotype Matrix Classes<a class="headerlink" href="#summary-of-genotype-matrix-classes" title="Link to this heading">#</a></h2>
<p>Genotype matrix classes in PyBrOpS are found in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmat</span></code> module. Contained in this module are several <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> class type definitions which are summarized in the table below.</p>
<table class="table" id="id1">
<caption><span class="caption-text">Summary of genotype matrix classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmat</span></code> module</span><a class="headerlink" href="#id1" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 27.8%" />
<col style="width: 16.7%" />
<col style="width: 55.6%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all genotype matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing dense, genotype matrices.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="loading-class-modules">
<h2>Loading Class Modules<a class="headerlink" href="#loading-class-modules" title="Link to this heading">#</a></h2>
<p>Genotype matrix classes can be imported as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import the GenotypeMatrix class (an abstract interface class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.GenotypeMatrix</span> <span class="kn">import</span> <span class="n">GenotypeMatrix</span>

<span class="c1"># import the DenseGenotypeMatrix class (a concrete implemented class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DenseGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DenseGenotypeMatrix</span>
</pre></div>
</div>
</section>
<section id="creating-genotype-matrices">
<h2>Creating Genotype Matrices<a class="headerlink" href="#creating-genotype-matrices" title="Link to this heading">#</a></h2>
<p>Genotype matrices can be created using several methods. Genotype matrices may be constructed from raw NumPy arrays, by reading data from VCF files, or by reading data from an HDF5 file containing a saved genotype matrix.</p>
<section id="creating-genotype-matrices-from-numpy-arrays">
<h3>Creating genotype matrices from NumPy arrays<a class="headerlink" href="#creating-genotype-matrices-from-numpy-arrays" title="Link to this heading">#</a></h3>
<p>Genotype matrices can be created from raw NumPy arrays. The example below illustrates creating a <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code>. The only requirement to construct a <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code> is a <code class="docutils literal notranslate"><span class="pre">(n,p)</span></code> numerical matrix containing genotypic codings. Here, <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of taxa and <code class="docutils literal notranslate"><span class="pre">p</span></code> is the number of marker variants. The matrix must be coded in the <code class="docutils literal notranslate"><span class="pre">{0,1,2}</span></code> (number of alternative alleles) format and must have an <code class="docutils literal notranslate"><span class="pre">int8</span></code> data type.</p>
<p>Additional optional metadata may be stored along with a <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code> including taxa names (<code class="docutils literal notranslate"><span class="pre">taxa</span></code>), taxa groups (<code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code>), marker variant chromosome assignments (<code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp</span></code>), marker variant chromosome physical positions (<code class="docutils literal notranslate"><span class="pre">vrnt_phypos</span></code>), marker variant names (<code class="docutils literal notranslate"><span class="pre">vrnt_name</span></code>), marker variant genetic map positions (<code class="docutils literal notranslate"><span class="pre">vrnt_genpos</span></code>), sequential recombination probabilities between markers (<code class="docutils literal notranslate"><span class="pre">vrnt_xoprob</span></code>), marker variant haplotype group assignments (<code class="docutils literal notranslate"><span class="pre">vrnt_hapgrp</span></code>), reference haplotype (<code class="docutils literal notranslate"><span class="pre">vrnt_hapref</span></code>), alternative haplotype (<code class="docutils literal notranslate"><span class="pre">vrnt_hapalt</span></code>), a variant mask (<code class="docutils literal notranslate"><span class="pre">vrnt_mask</span></code>), and finally the ploidy of the genotypes (<code class="docutils literal notranslate"><span class="pre">ploid</span></code>).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create random genotypes</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">nvrnt</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span>
    <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
<span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant chromsome assignments</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nchrom</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker physical positions</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span>
    <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span>
<span class="p">)</span>

<span class="c1"># create a genotype matrix from scratch using NumPy arrays</span>
<span class="n">gmat</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">vrnt_genpos</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_xoprob</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapgrp</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapalt</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_hapref</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vrnt_mask</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-genotype-matrices-from-vcf-files">
<h3>Loading genotype matrices from VCF files<a class="headerlink" href="#loading-genotype-matrices-from-vcf-files" title="Link to this heading">#</a></h3>
<p>VCF files can be loaded into Python scopes as genotype matrices using the <code class="docutils literal notranslate"><span class="pre">from_vcf</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read a genotype matrix from file</span>
<span class="n">gmat</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s2">&quot;widiv_2000SNPs.vcf.gz&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-genotype-matrices-from-hdf5-files">
<h3>Loading genotype matrices from HDF5 files<a class="headerlink" href="#loading-genotype-matrices-from-hdf5-files" title="Link to this heading">#</a></h3>
<p>Genotype matrices in PyBrOpS can be exported to HDF5 files via the <code class="docutils literal notranslate"><span class="pre">to_hdf5</span></code> method. These files can later be read into PyBrOpS via the <code class="docutils literal notranslate"><span class="pre">from_hdf5</span></code> method. The example below illustrates loading a <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code> into memory from an HDF5 file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read a genotype matrix from HDF5 file</span>
<span class="n">gmat</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;widiv_2000SNPs.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="genotype-matrix-properties">
<h2>Genotype Matrix Properties<a class="headerlink" href="#genotype-matrix-properties" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> objects share a set of common object properties which can be grouped into three groups: general properties, taxa properties, and marker variant properties.</p>
<section id="general-properties">
<h3>General properties<a class="headerlink" href="#general-properties" title="Link to this heading">#</a></h3>
<p>General properties of a genotype matrix primarily relate to the shape and format of the genotype matrix itself. These properties are summarized in the table below.</p>
<table class="table" id="id2">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> general properties</span><a class="headerlink" href="#id2" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat</span></code></p></td>
<td><p>Pointer to the raw genotype matrix pointer</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_ndim</span></code></p></td>
<td><p>The number of dimensions for the genotype matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat_shape</span></code></p></td>
<td><p>Genotype matrix shape</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_format</span></code></p></td>
<td><p>Genotype matrix format</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ploidy</span></code></p></td>
<td><p>The ploidy of the taxa represented by the genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">nphase</span></code></p></td>
<td><p>The number of chromosome phases represented by the genotype matrix</p></td>
</tr>
</tbody>
</table>
</section>
<section id="taxa-properties">
<h3>Taxa properties<a class="headerlink" href="#taxa-properties" title="Link to this heading">#</a></h3>
<p>Genotype matrices can store basic information on the taxa represented by the matrix. These data include taxa names and any grouping labels. Grouping labels may be used to indicate a family of individuals. Taxa and taxa grouping data can be accessed using the object properties summarized in the table below.</p>
<table class="table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> taxa properties</span><a class="headerlink" href="#id3" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntaxa</span></code></p></td>
<td><p>The number of taxa represented by the genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></td>
<td><p>The names of the taxa</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_axis</span></code></p></td>
<td><p>The matrix axis along which taxa are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code></p></td>
<td><p>Taxa group label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_name</span></code></p></td>
<td><p>The names of the taxa groups</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_stix</span></code></p></td>
<td><p>The start indices (inclusive) for each taxa group, post sorting and grouping</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_spix</span></code></p></td>
<td><p>The stop indices (exclusive) for each taxa group, post sorting and grouping</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_len</span></code></p></td>
<td><p>The length of each taxa group, post sorting and grouping</p></td>
</tr>
</tbody>
</table>
</section>
<section id="marker-variant-properties">
<h3>Marker variant properties<a class="headerlink" href="#marker-variant-properties" title="Link to this heading">#</a></h3>
<p>Genotype matrices can also store basic information about the genetic markers which are represented by the matrix. Key information about a marker variant’s name, chromosome assignment, physical and genetic positions, and sequential crossover probabilities can be stored. The table below summarizes marker variant properties in genotype matrices.</p>
<table class="table" id="id4">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> marker variant properties</span><a class="headerlink" href="#id4" title="Link to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nvrnt</span></code></p></td>
<td><p>The number of genotype variants represented by the genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_name</span></code></p></td>
<td><p>The names of the marker variants</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_axis</span></code></p></td>
<td><p>The axis along which marker variants are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp</span></code></p></td>
<td><p>The chromosome to which a marker variant belongs</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_phypos</span></code></p></td>
<td><p>The physical position of a marker variant</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_genpos</span></code></p></td>
<td><p>The genetic position of a marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_xoprob</span></code></p></td>
<td><p>The crossover probability between the previous marker</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapref</span></code></p></td>
<td><p>The reference haplotype for the marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapalt</span></code></p></td>
<td><p>The alternative haplotype for the marker variant</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapgrp</span></code></p></td>
<td><p>The haplotype grouping for the marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_mask</span></code></p></td>
<td><p>A mask associated with the marker variants</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_name</span></code></p></td>
<td><p>The names of the chromosomes</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_stix</span></code></p></td>
<td><p>The start indices (inclusive) for each chromosome, post sorting and grouping</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_spix</span></code></p></td>
<td><p>The stop indices (exclusive) for each chromosome, post sorting and grouping</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_len</span></code></p></td>
<td><p>The length of each chromosome, post sorting and grouping</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="copying-genotype-matrices">
<h2>Copying Genotype Matrices<a class="headerlink" href="#copying-genotype-matrices" title="Link to this heading">#</a></h2>
<p>Genotype matrices can be copied. There are two methods that can be used to copy a genotype matrix: shallow copying and deep copying.</p>
<section id="shallow-copying">
<h3>Shallow copying<a class="headerlink" href="#shallow-copying" title="Link to this heading">#</a></h3>
<p>In shallow copying, only references to a <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code>’s data are copied to a new genotype matrix. Copying is only one level deep so any changes to the data in the original object will be reflected in the copied object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy a genotype matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="deep-copying">
<h3>Deep copying<a class="headerlink" href="#deep-copying" title="Link to this heading">#</a></h3>
<p>In deep copying, a <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code>’s data is recursively copied to the deepest level. Changes to data values in the original genotype matrix will not affect the data values in the copy.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy a genotype matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="copy-on-element-manipulation">
<h2>Copy-On Element Manipulation<a class="headerlink" href="#copy-on-element-manipulation" title="Link to this heading">#</a></h2>
<p>Genotype matrices have several methods which create a modified copy of the original matrix, leaving the original genotype matrix unaltered.</p>
<section id="adjoin-elements">
<h3>Adjoin elements<a class="headerlink" href="#adjoin-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">adjoin</span></code> family of methods allows for a genotype matrix to be adjoined to another genotype matrix, creating a new matrix in the process. Use of the <code class="docutils literal notranslate"><span class="pre">adjoin</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new genotype matrix to demonstrate</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># adjoin genotype matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>

<span class="c1"># adjoin genotype matrices along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">adjoin_vrnt</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="delete-elements">
<h3>Delete elements<a class="headerlink" href="#delete-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">delete</span></code> family of methods allows for rows (taxa) and columns (marker variants) of the genotype matrix to be removed in a copy of the original. Use of the <code class="docutils literal notranslate"><span class="pre">delete</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># delete taxa examples</span>
<span class="c1">#</span>

<span class="c1"># delete first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1">#</span>
<span class="c1"># delete marker variants examples</span>
<span class="c1">#</span>

<span class="c1"># delete first marker variant using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five marker variants using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="insert-elements">
<h3>Insert elements<a class="headerlink" href="#insert-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">insert</span></code> family of methods allows for rows (taxa) and columns (marker variants) of the genotype matrix to be inserted into a copy of the original matrix. Use of the <code class="docutils literal notranslate"><span class="pre">insert</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new genotype matrix to demonstrate</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># insert genotype matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>

<span class="c1"># insert genotype matrix along the variant axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">insert_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="select-elements">
<h3>Select elements<a class="headerlink" href="#select-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">select</span></code> family of methods allows for rows (taxa) and columns (marker variants) of the genotype matrix to be extracted to a copy of the original matrix. The <code class="docutils literal notranslate"><span class="pre">select</span></code> family is the inverse of the <code class="docutils literal notranslate"><span class="pre">delete</span></code> family. Use of the <code class="docutils literal notranslate"><span class="pre">select</span></code> method family is demonstrated in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># select first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1"># select first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">select_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="in-place-element-manipulation">
<h2>In-Place Element Manipulation<a class="headerlink" href="#in-place-element-manipulation" title="Link to this heading">#</a></h2>
<p>Genotype matrices have several methods which create a modify a matrix in-place.</p>
<section id="append-elements">
<h3>Append elements<a class="headerlink" href="#append-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">append</span></code> family of methods allows for new rows (taxa) and columns (marker variants) to be appended to the genotype matrix. The code segment below demonstrates their use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># append genotype matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>

<span class="c1"># append genotype matrices along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_vrnt</span><span class="p">(</span><span class="n">gmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>
</pre></div>
</div>
</section>
<section id="remove-elements">
<h3>Remove elements<a class="headerlink" href="#remove-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">remove</span></code> family of methods allows for rows (taxa) and columns (marker variants) to be removed from a genotype matrix. A demonstration of their use can be seen below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># remove taxa examples</span>
<span class="c1">#</span>

<span class="c1"># remove first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>

<span class="c1">#</span>
<span class="c1"># remove marker variants examples</span>
<span class="c1">#</span>

<span class="c1"># remove first marker variant using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                              <span class="c1"># remove from copy</span>

<span class="c1"># remove first five marker variants using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                     <span class="c1"># remove from copy</span>

<span class="c1"># remove first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                    <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
<section id="incorporate-elements">
<h3>Incorporate elements<a class="headerlink" href="#incorporate-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">incorp</span></code> family of methods allows for new rows (taxa) and columns (marker variants) to be inserted at specific locations a genotype matrix. Use of the <code class="docutils literal notranslate"><span class="pre">incorp</span></code> family is demonstrated in the code segment below below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># incorp genotype matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>      <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>

<span class="c1"># incorp genotype matrix along the variant axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>      <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>
</pre></div>
</div>
</section>
<section id="concatenate-elements">
<h3>Concatenate elements<a class="headerlink" href="#concatenate-elements" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">concat</span></code> family of methods allows for multiple genotype matrices to be concatenated to each other. The code segment below demonstrates their use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># concatenate along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gmat</span><span class="p">,</span> <span class="n">gmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">gmat</span><span class="p">,</span> <span class="n">gmat</span><span class="p">])</span>

<span class="c1"># concatenate along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gmat</span><span class="p">,</span> <span class="n">gmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">concat_vrnt</span><span class="p">([</span><span class="n">gmat</span><span class="p">,</span> <span class="n">gmat</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping-and-sorting">
<h2>Grouping and Sorting<a class="headerlink" href="#grouping-and-sorting" title="Link to this heading">#</a></h2>
<p>Genotype matrices in PyBrOpS have several sorting and grouping focused methods. Sorting methods can be used to sort taxa alphanumerically and marker variants according to their chromosome positions. Grouping methods sort taxa and marker variants, and calculate metadata to allow for indexing of taxa groups and chromosomes groups.</p>
<section id="reordering-elements">
<h3>Reordering elements<a class="headerlink" href="#reordering-elements" title="Link to this heading">#</a></h3>
<p>Taxa and marker variants in a genotype matrix can be reordered using the <code class="docutils literal notranslate"><span class="pre">reorder</span></code> family of methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># taxa reordering example</span>
<span class="c1">#</span>

<span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values along the taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>

<span class="c1">#</span>
<span class="c1"># marker variant reordering example</span>
<span class="c1">#</span>

<span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values along the marker variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_vrnt</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="lexsorting-elements">
<h3>Lexsorting elements<a class="headerlink" href="#lexsorting-elements" title="Link to this heading">#</a></h3>
<p>An indirect sort for the taxa and marker variants axes can be performed using the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> family of methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#</span>
<span class="c1"># taxa lexsort example</span>
<span class="c1">#</span>

<span class="c1"># create lexsort keys for taxa</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the taxa axis</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>

<span class="c1">#</span>
<span class="c1"># marker variant lexsort example</span>
<span class="c1">#</span>

<span class="c1"># create lexsort keys for marker variants</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">gmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the marker variant axis</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">lexsort_vrnt</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="sorting-elements">
<h3>Sorting elements<a class="headerlink" href="#sorting-elements" title="Link to this heading">#</a></h3>
<p>Sorting along taxa and marker variant axes can be done using the <code class="docutils literal notranslate"><span class="pre">sort</span></code> family of methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1">#</span>
<span class="c1"># taxa sorting example</span>
<span class="c1">#</span>

<span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>

<span class="c1">#</span>
<span class="c1"># marker variant sorting example</span>
<span class="c1">#</span>

<span class="c1"># sort along marker variant axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_vrnt</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="grouping-elements">
<h3>Grouping elements<a class="headerlink" href="#grouping-elements" title="Link to this heading">#</a></h3>
<p>Grouping along taxa and marker variant axes can be done using the <code class="docutils literal notranslate"><span class="pre">group</span></code> family of methods.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make copy</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1">#</span>
<span class="c1"># taxa grouping example</span>
<span class="c1">#</span>

<span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>

<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>

<span class="c1">#</span>
<span class="c1"># marker variant grouping example</span>
<span class="c1">#</span>

<span class="c1"># sort along vrnt axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>

<span class="c1"># determine whether grouping has occurred along the vrnt axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_vrnt</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="summary-statistics">
<h2>Summary Statistics<a class="headerlink" href="#summary-statistics" title="Link to this heading">#</a></h2>
<p>Various summary statistics can be calculated from genotype matrices. PyBrOpS offers several common routines which are summarized in the sections below.</p>
<section id="population-allele-counts">
<h3>Population allele counts<a class="headerlink" href="#population-allele-counts" title="Link to this heading">#</a></h3>
<p>Allele counts of the dominant allele (the allele coded as 1) at each locus may be calculated using the <code class="docutils literal notranslate"><span class="pre">acount</span></code> method. The code below demonstrates this method’s use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># count the number of major alleles across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">acount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">acount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-allele-frequencies">
<h3>Population allele frequencies<a class="headerlink" href="#population-allele-frequencies" title="Link to this heading">#</a></h3>
<p>Allele frequencies of the dominant allele (the allele coded as 1) at each locus may be calculated using the <code class="docutils literal notranslate"><span class="pre">afreq</span></code> method. The code below demonstrates this method’s use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate the allele frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">afreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-allele-polymorphism-presence">
<h3>Population allele polymorphism presence<a class="headerlink" href="#population-allele-polymorphism-presence" title="Link to this heading">#</a></h3>
<p>The presense of allele polymorphisms at each locus can be determined using the <code class="docutils literal notranslate"><span class="pre">apoly</span></code> method. The code below demonstrates this method’s use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate whether a locus is polymorphic across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">apoly</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">apoly</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-genotype-counts">
<h3>Population genotype counts<a class="headerlink" href="#population-genotype-counts" title="Link to this heading">#</a></h3>
<p>Genotype counts at each locus may be calculated using the <code class="docutils literal notranslate"><span class="pre">gtcount</span></code> method. This method counts the number of homozygous recessive, heterozygous, and homozygous dominant individuals. The code below demonstrates this method’s use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># count the number of genotypes across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">gtcount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">gtcount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-genotype-frequencies">
<h3>Population genotype frequencies<a class="headerlink" href="#population-genotype-frequencies" title="Link to this heading">#</a></h3>
<p>Genotype frequencies at each locus may be calculated using the <code class="docutils literal notranslate"><span class="pre">gtfreq</span></code> method. The code below demonstrates this method’s use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate the genotype frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">gtfreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">gtfreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-minor-allele-frequencies">
<h3>Population minor allele frequencies<a class="headerlink" href="#population-minor-allele-frequencies" title="Link to this heading">#</a></h3>
<p>The minor allele frequency at each locus may be calculated using the <code class="docutils literal notranslate"><span class="pre">maf</span></code> method. The code below demonstrates the use of this method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate the minor allele frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">maf</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">maf</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="population-mean-expected-heterozygosity">
<h3>Population mean expected heterozygosity<a class="headerlink" href="#population-mean-expected-heterozygosity" title="Link to this heading">#</a></h3>
<p>The mean expected heterozygosity of the genotype matrix as a whole may be calculated using the <code class="docutils literal notranslate"><span class="pre">meh</span></code> method. Use of the <code class="docutils literal notranslate"><span class="pre">meh</span></code> method can be seen in the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate the mean expected heterozygosity for the population</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">meh</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">meh</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="taxa-allele-counts">
<h3>Taxa allele counts<a class="headerlink" href="#taxa-allele-counts" title="Link to this heading">#</a></h3>
<p>Allele counts of the dominant allele within each individual taxon may be calculated using the <code class="docutils literal notranslate"><span class="pre">tacount</span></code> method. A demonstration of this method’s use is below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># count the number of major alleles individually within taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">tacount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">tacount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="taxa-allele-frequencies">
<h3>Taxa allele frequencies<a class="headerlink" href="#taxa-allele-frequencies" title="Link to this heading">#</a></h3>
<p>Allele frequencies of the dominant allele within each individual taxon may be calculated using the <code class="docutils literal notranslate"><span class="pre">tafreq</span></code> method. The <code class="docutils literal notranslate"><span class="pre">tafreq</span></code> method can be used like so.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate the allele frequency individually within taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">tafreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">tafreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="saving-genotype-matrices">
<h2>Saving Genotype Matrices<a class="headerlink" href="#saving-genotype-matrices" title="Link to this heading">#</a></h2>
<p>Genotype matrices can be exported and saved to disk. Currently, only one export format is available: HDF5.</p>
<section id="exporting-to-hdf5">
<h3>Exporting to HDF5<a class="headerlink" href="#exporting-to-hdf5" title="Link to this heading">#</a></h3>
<p>Genotype matrices can be exported to the <a class="reference external" href="https://www.hdfgroup.org/">HDF5 format</a>. The code below demonstrates how to export a genotype matrix to an HDF5 file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write a breeding value matrix to an HDF5 file</span>
<span class="n">gmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_genotypes.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="genetic_map_functions.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genetic Map Functions</p>
      </div>
    </a>
    <a class="right-next"
       href="phased_genotype_matrices.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Phased Genotype Matrices</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#class-family-overview">Class Family Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-genotype-matrix-classes">Summary of Genotype Matrix Classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-class-modules">Loading Class Modules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-genotype-matrices">Creating Genotype Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-genotype-matrices-from-numpy-arrays">Creating genotype matrices from NumPy arrays</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genotype-matrices-from-vcf-files">Loading genotype matrices from VCF files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genotype-matrices-from-hdf5-files">Loading genotype matrices from HDF5 files</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-matrix-properties">Genotype Matrix Properties</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#general-properties">General properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#taxa-properties">Taxa properties</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marker-variant-properties">Marker variant properties</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copying-genotype-matrices">Copying Genotype Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#shallow-copying">Shallow copying</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deep-copying">Deep copying</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copy-on-element-manipulation">Copy-On Element Manipulation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adjoin-elements">Adjoin elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#delete-elements">Delete elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#insert-elements">Insert elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#select-elements">Select elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-place-element-manipulation">In-Place Element Manipulation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#append-elements">Append elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-elements">Remove elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#incorporate-elements">Incorporate elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#concatenate-elements">Concatenate elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#grouping-and-sorting">Grouping and Sorting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reordering-elements">Reordering elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lexsorting-elements">Lexsorting elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sorting-elements">Sorting elements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#grouping-elements">Grouping elements</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-statistics">Summary Statistics</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-allele-counts">Population allele counts</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-allele-frequencies">Population allele frequencies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-allele-polymorphism-presence">Population allele polymorphism presence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genotype-counts">Population genotype counts</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genotype-frequencies">Population genotype frequencies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-minor-allele-frequencies">Population minor allele frequencies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-mean-expected-heterozygosity">Population mean expected heterozygosity</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#taxa-allele-counts">Taxa allele counts</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#taxa-allele-frequencies">Taxa allele frequencies</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-genotype-matrices">Saving Genotype Matrices</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-to-hdf5">Exporting to HDF5</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>