
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Phased Genotype Matrices &#8212; pybrops 0.9.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/wider_borders.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/population_genetics/phased_genotype_matrices';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Breeding Value Matrices" href="breeding_value_matrices.html" />
    <link rel="prev" title="Genotype Matrices" href="genotype_matrices.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">pybrops 0.9.0 documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item"><nav class="bd-links" id="bd-docs-nav" aria-label="Section navigation">
  <p class="bd-links__title" role="heading" aria-level="1">
    Section Navigation
  </p>
  <div class="bd-toc-item navbar-nav">
    <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/examples.html">Examples</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
</ul>
</li>
</ul>

  </div>
</nav>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="phased-genotype-matrices">
<h1>Phased Genotype Matrices<a class="headerlink" href="#phased-genotype-matrices" title="Permalink to this heading">#</a></h1>
<section id="class-family-overview">
<h2>Class Family Overview<a class="headerlink" href="#class-family-overview" title="Permalink to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> object family is a derivative of the <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> object family. The <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> family of objects has all the same purposes and functionality as its parent family except that it is used to represent phased genotypes. Phased genotype matrix representation is essential to the representation of genomes in PyBrOpS.</p>
</section>
<section id="summary-of-phased-genotype-matrix-classes">
<h2>Summary of Phased Genotype Matrix Classes<a class="headerlink" href="#summary-of-phased-genotype-matrix-classes" title="Permalink to this heading">#</a></h2>
<p>Phased genotype matrix classes in PyBrOpS are found in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmat</span></code> module. Contained in this module are several <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> class type definitions which are summarized in the table below.</p>
<table class="colwidths-given table" id="id1">
<caption><span class="caption-text">Summary of phased genotype matrix classes in the <code class="docutils literal notranslate"><span class="pre">pybrops.popgen.gmat</span></code> module</span><a class="headerlink" href="#id1" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 28%" />
<col style="width: 17%" />
<col style="width: 56%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class Name</p></th>
<th class="head"><p>Class Type</p></th>
<th class="head"><p>Class Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code></p></td>
<td><p>Abstract</p></td>
<td><p>Interface for all phased genotype matrix child classes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DensePhasedGenotypeMatrix</span></code></p></td>
<td><p>Concrete</p></td>
<td><p>Class representing dense, phased genotype matrices.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="loading-class-modules">
<h2>Loading Class Modules<a class="headerlink" href="#loading-class-modules" title="Permalink to this heading">#</a></h2>
<p>Phased genotype matrix classes can be imported as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import the PhasedGenotypeMatrix class (an abstract interface class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.PhasedGenotypeMatrix</span> <span class="kn">import</span> <span class="n">PhasedGenotypeMatrix</span>

<span class="c1"># import the DensePhasedGenotypeMatrix class (a concrete implemented class)</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DensePhasedGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DensePhasedGenotypeMatrix</span>
</pre></div>
</div>
</section>
<section id="creating-phased-genotype-matrices">
<h2>Creating Phased Genotype Matrices<a class="headerlink" href="#creating-phased-genotype-matrices" title="Permalink to this heading">#</a></h2>
<p>Like the <code class="docutils literal notranslate"><span class="pre">GenotypeMatrix</span></code> family of classes, members of the <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> family can be constructed from raw NumPy arrays, by reading data from VCF files, or by reading data from an HDF5 file containing a saved phased genotype matrix.</p>
<section id="creating-phased-genotype-matrices-from-numpy-arrays">
<h3>Creating phased genotype matrices from NumPy arrays<a class="headerlink" href="#creating-phased-genotype-matrices-from-numpy-arrays" title="Permalink to this heading">#</a></h3>
<p>Phased genotype matrices can be constructed from raw NumPy arrays using the constructor. The example below demonstrates the construction of a phased genotype matrix using the <code class="docutils literal notranslate"><span class="pre">DensePhasedGenotypeMatrix</span></code> class. Numerical matrix inputs containing genotypic codings must have a shape of <code class="docutils literal notranslate"><span class="pre">(m,n,p)</span></code>, where <code class="docutils literal notranslate"><span class="pre">m</span></code> is the number of phases, <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of taxa, and <code class="docutils literal notranslate"><span class="pre">p</span></code> is the number of marker variants. The genotype code matrix must be binary in nature (<code class="docutils literal notranslate"><span class="pre">{0,1}</span></code>) and must have an <code class="docutils literal notranslate"><span class="pre">int8</span></code> data type.</p>
<p>Like the <code class="docutils literal notranslate"><span class="pre">DenseGenotypeMatrix</span></code> class, additional optional metadata may be stored along with a <code class="docutils literal notranslate"><span class="pre">DensePhasedGenotypeMatrix</span></code> including taxa names (<code class="docutils literal notranslate"><span class="pre">taxa</span></code>), taxa groups (<code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code>), marker variant chromosome assignments (<code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp</span></code>), marker variant chromosome physical positions (<code class="docutils literal notranslate"><span class="pre">vrnt_phypos</span></code>), marker variant names (<code class="docutils literal notranslate"><span class="pre">vrnt_name</span></code>), marker variant genetic map positions (<code class="docutils literal notranslate"><span class="pre">vrnt_genpos</span></code>), sequential recombination probabilities between markers (<code class="docutils literal notranslate"><span class="pre">vrnt_xoprob</span></code>), marker variant haplotype group assignments (<code class="docutils literal notranslate"><span class="pre">vrnt_hapgrp</span></code>), reference haplotype (<code class="docutils literal notranslate"><span class="pre">vrnt_hapref</span></code>), alternative haplotype (<code class="docutils literal notranslate"><span class="pre">vrnt_hapalt</span></code>), and a variant mask (<code class="docutils literal notranslate"><span class="pre">vrnt_mask</span></code>).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># shape parameters for random genotypes</span>
<span class="n">ntaxa</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">nvrnt</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">ngroup</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">nchrom</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">nphase</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># create random genotypes</span>
<span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">nphase</span><span class="p">,</span><span class="n">ntaxa</span><span class="p">,</span><span class="n">nvrnt</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>

<span class="c1"># create taxa names</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;taxon&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create taxa groups</span>
<span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ngroup</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">)</span>
<span class="n">taxa_grp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant chromsome assignments</span>
<span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nchrom</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvrnt</span><span class="p">)</span>
<span class="n">vrnt_chrgrp</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker physical positions</span>
<span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">nvrnt</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">vrnt_phypos</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># create marker variant names</span>
<span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;SNP&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvrnt</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create a phased genotype matrix from scratch using NumPy arrays</span>
<span class="n">pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">(</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">,</span>
    <span class="n">taxa</span> <span class="o">=</span> <span class="n">taxa</span><span class="p">,</span>
    <span class="n">taxa_grp</span> <span class="o">=</span> <span class="n">taxa_grp</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp</span> <span class="o">=</span> <span class="n">vrnt_chrgrp</span><span class="p">,</span>
    <span class="n">vrnt_phypos</span> <span class="o">=</span> <span class="n">vrnt_phypos</span><span class="p">,</span>
    <span class="n">vrnt_name</span> <span class="o">=</span> <span class="n">vrnt_name</span><span class="p">,</span>
    <span class="n">ploidy</span> <span class="o">=</span> <span class="n">nphase</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-phased-genotype-matrices-from-vcf-files">
<h3>Loading phased genotype matrices from VCF files<a class="headerlink" href="#loading-phased-genotype-matrices-from-vcf-files" title="Permalink to this heading">#</a></h3>
<p>Data from VCF files can be loaded using the <code class="docutils literal notranslate"><span class="pre">from_vcf</span></code> method. This import method assumes that the provided VCF file has been previously phased. If an input file is unphased, it will be loaded as if it were correctly phased, which will be problematic for non-homozygous loci.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read a phased genotype matrix from VCF file</span>
<span class="n">pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s2">&quot;widiv_2000SNPs.vcf.gz&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-phased-genotype-matrices-from-hdf5-files">
<h3>Loading phased genotype matrices from HDF5 files<a class="headerlink" href="#loading-phased-genotype-matrices-from-hdf5-files" title="Permalink to this heading">#</a></h3>
<p>Like regular genotype matrices, phased genotype matrices can be exported to HDF5 files via the <code class="docutils literal notranslate"><span class="pre">to_hdf5</span></code> method. These files can later be read into PyBrOpS using the <code class="docutils literal notranslate"><span class="pre">from_hdf5</span></code> method. The example below illustrates loading a <code class="docutils literal notranslate"><span class="pre">DensePhasedGenotypeMatrix</span></code> into memory from an HDF5 file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read a genotype matrix from HDF5 file</span>
<span class="n">pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">from_hdf5</span><span class="p">(</span><span class="s2">&quot;widiv_2000SNPs.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="phased-genotype-matrix-properties">
<h2>Phased Genotype Matrix Properties<a class="headerlink" href="#phased-genotype-matrix-properties" title="Permalink to this heading">#</a></h2>
<section id="general-properties">
<h3>General properties<a class="headerlink" href="#general-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id2">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> general properties</span><a class="headerlink" href="#id2" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat</span></code></p></td>
<td><p>Pointer to the raw phased genotype matrix pointer</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_ndim</span></code></p></td>
<td><p>The number of dimensions for the phased genotype matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mat_shape</span></code></p></td>
<td><p>Genotype matrix shape</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mat_format</span></code></p></td>
<td><p>Genotype matrix format</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ploidy</span></code></p></td>
<td><p>The ploidy of the taxa represented by the phased genotype matrix</p></td>
</tr>
</tbody>
</table>
</section>
<section id="phase-properties">
<h3>Phase properties<a class="headerlink" href="#phase-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id3">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> phase properties</span><a class="headerlink" href="#id3" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nphase</span></code></p></td>
<td><p>The number of chromosome phases represented by the phased genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">phase_axis</span></code></p></td>
<td><p>The matrix axis along which phases are stored</p></td>
</tr>
</tbody>
</table>
</section>
<section id="taxa-properties">
<h3>Taxa properties<a class="headerlink" href="#taxa-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id4">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> taxa properties</span><a class="headerlink" href="#id4" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ntaxa</span></code></p></td>
<td><p>The number of taxa represented by the phased genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></td>
<td><p>The names of the taxa</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_axis</span></code></p></td>
<td><p>The matrix axis along which taxa are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp</span></code></p></td>
<td><p>Taxa group label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_name</span></code></p></td>
<td><p>The names of the taxa groups</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_stix</span></code></p></td>
<td><p>The start indices (inclusive) for each taxa group, post sorting and grouping</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_spix</span></code></p></td>
<td><p>The stop indices (exclusive) for each taxa group, post sorting and grouping</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">taxa_grp_len</span></code></p></td>
<td><p>The length of each taxa group, post sorting and grouping</p></td>
</tr>
</tbody>
</table>
</section>
<section id="marker-variant-properties">
<h3>Marker variant properties<a class="headerlink" href="#marker-variant-properties" title="Permalink to this heading">#</a></h3>
<table class="colwidths-given table" id="id5">
<caption><span class="caption-text">Summary of <code class="docutils literal notranslate"><span class="pre">PhasedGenotypeMatrix</span></code> marker variant properties</span><a class="headerlink" href="#id5" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nvrnt</span></code></p></td>
<td><p>The number of genotype variants represented by the phased genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_name</span></code></p></td>
<td><p>The names of the marker variants</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_axis</span></code></p></td>
<td><p>The axis along which marker variants are stored</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp</span></code></p></td>
<td><p>The chromosome to which a marker variant belongs</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_phypos</span></code></p></td>
<td><p>The physical position of a marker variant</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_genpos</span></code></p></td>
<td><p>The genetic position of a marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_xoprob</span></code></p></td>
<td><p>The crossover probability between the previous marker</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapref</span></code></p></td>
<td><p>The reference haplotype for the marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapalt</span></code></p></td>
<td><p>The alternative haplotype for the marker variant</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_hapgrp</span></code></p></td>
<td><p>The haplotype grouping for the marker variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_mask</span></code></p></td>
<td><p>A mask associated with the marker variants</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_name</span></code></p></td>
<td><p>The names of the chromosomes</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_stix</span></code></p></td>
<td><p>The start indices (inclusive) for each chromosome, post sorting and grouping</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_spix</span></code></p></td>
<td><p>The stop indices (exclusive) for each chromosome, post sorting and grouping</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">vrnt_chrgrp_len</span></code></p></td>
<td><p>The length of each chromosome, post sorting and grouping</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="copying-phased-genotype-matrices">
<h2>Copying Phased Genotype Matrices<a class="headerlink" href="#copying-phased-genotype-matrices" title="Permalink to this heading">#</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># copy a phased genotype matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">pgmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># deep copy a phased genotype matrix</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">pgmat</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="phased-genotype-matrix-element-copy-on-manipulation">
<h2>Phased Genotype Matrix Element Copy-On-Manipulation<a class="headerlink" href="#phased-genotype-matrix-element-copy-on-manipulation" title="Permalink to this heading">#</a></h2>
<section id="adjoining-elements">
<h3>Adjoining elements<a class="headerlink" href="#adjoining-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new genotype matrix to demonstrate</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># adjoin genotype matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">adjoin_taxa</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>

<span class="c1"># adjoin genotype matrices along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">adjoin</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">adjoin_vrnt</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="deleting-elements">
<h3>Deleting elements<a class="headerlink" href="#deleting-elements" title="Permalink to this heading">#</a></h3>
<section id="delete-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">delete</span></code> taxa<a class="headerlink" href="#delete-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># delete first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="delete-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">delete</span></code> marker variants<a class="headerlink" href="#delete-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># delete first marker variant using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># delete first five marker variants using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># delete first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">delete_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="inserting-elements">
<h3>Inserting elements<a class="headerlink" href="#inserting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new genotype matrix to demonstrate</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># insert genotype matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">insert_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>

<span class="c1"># insert genotype matrix along the variant axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">insert_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="selecting-elements">
<h3>Selecting elements<a class="headerlink" href="#selecting-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># select first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>

<span class="c1"># select first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">select_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="phased-genotype-matrix-element-in-place-manipulation">
<h2>Phased Genotype Matrix Element In-Place-Manipulation<a class="headerlink" href="#phased-genotype-matrix-element-in-place-manipulation" title="Permalink to this heading">#</a></h2>
<section id="appending-elements">
<h3>Appending elements<a class="headerlink" href="#appending-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># append genotype matrices along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_taxa</span><span class="p">(</span><span class="n">pgmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>

<span class="c1"># append genotype matrices along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>  <span class="c1"># append original to copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                   <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">append_vrnt</span><span class="p">(</span><span class="n">pgmat</span><span class="p">)</span>                   <span class="c1"># append original to copy</span>
</pre></div>
</div>
</section>
<section id="removing-elements">
<h3>Removing elements<a class="headerlink" href="#removing-elements" title="Permalink to this heading">#</a></h3>
<section id="remove-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> taxa<a class="headerlink" href="#remove-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first taxon using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                               <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                      <span class="c1"># remove from copy</span>

<span class="c1"># remove first five taxa using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                     <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
<section id="remove-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">remove</span></code> marker variants<a class="headerlink" href="#remove-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove first marker variant using an integer</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>            <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                               <span class="c1"># remove from copy</span>

<span class="c1"># remove first five marker variants using a slice</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>   <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>                      <span class="c1"># remove from copy</span>

<span class="c1"># remove first five marker variants using a Sequence</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>  <span class="c1"># remove from copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">remove_vrnt</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>                     <span class="c1"># remove from copy</span>
</pre></div>
</div>
</section>
</section>
<section id="incorporating-elements">
<h3>Incorporating elements<a class="headerlink" href="#incorporating-elements" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># incorp genotype matrix along the taxa axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>     <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_taxa</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>

<span class="c1"># incorp genotype matrix along the variant axis before index 0</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>     <span class="c1"># incorporate into copy</span>

<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>                           <span class="c1"># copy original</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">incorp_vrnt</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">)</span>                        <span class="c1"># incorporate into copy</span>
</pre></div>
</div>
</section>
<section id="concatenating-matrices">
<h3>Concatenating matrices<a class="headerlink" href="#concatenating-matrices" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># concatenate along the taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">([</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">])</span>

<span class="c1"># concatenate along the variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">concat_vrnt</span><span class="p">([</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">pgmat</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping-and-sorting">
<h2>Grouping and Sorting<a class="headerlink" href="#grouping-and-sorting" title="Permalink to this heading">#</a></h2>
<section id="reordering">
<h3>Reordering<a class="headerlink" href="#reordering" title="Permalink to this heading">#</a></h3>
<section id="reorder-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> taxa<a class="headerlink" href="#reorder-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values along the taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_taxa</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reorder-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">reorder</span></code> marker variants<a class="headerlink" href="#reorder-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create reordering indices</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>

<span class="c1"># reorder values along the marker variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">reorder_vrnt</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="lexsorting">
<h3>Lexsorting<a class="headerlink" href="#lexsorting" title="Permalink to this heading">#</a></h3>
<section id="lexsort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> taxa<a class="headerlink" href="#lexsort-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for taxa</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the taxa axis</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">lexsort_taxa</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="lexsort-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">lexsort</span></code> marker variants<a class="headerlink" href="#lexsort-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create lexsort keys for marker variants</span>
<span class="n">key1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">key2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">key2</span><span class="p">)</span>

<span class="c1"># lexsort along the marker variant axis</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">lexsort_vrnt</span><span class="p">((</span><span class="n">key2</span><span class="p">,</span><span class="n">key1</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="sorting">
<h3>Sorting<a class="headerlink" href="#sorting" title="Permalink to this heading">#</a></h3>
<section id="sort-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> taxa<a class="headerlink" href="#sort-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="sort-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">sort</span></code> marker variants<a class="headerlink" href="#sort-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along marker variant axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">sort_vrnt</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="grouping">
<h3>Grouping<a class="headerlink" href="#grouping" title="Permalink to this heading">#</a></h3>
<section id="group-taxa">
<h4><code class="docutils literal notranslate"><span class="pre">group</span></code> taxa<a class="headerlink" href="#group-taxa" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along taxa axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_taxa</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the taxa axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">taxa_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_taxa</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="group-marker-variants">
<h4><code class="docutils literal notranslate"><span class="pre">group</span></code> marker variants<a class="headerlink" href="#group-marker-variants" title="Permalink to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort along vrnt axis</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">()</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>
<span class="c1"># determine whether grouping has occurred along the vrnt axis</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">vrnt_axis</span><span class="p">)</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">is_grouped_vrnt</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="summary-statistics">
<h2>Summary Statistics<a class="headerlink" href="#summary-statistics" title="Permalink to this heading">#</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># count the number of major alleles across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">acount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">acount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>

<span class="c1"># calculate the allele frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">afreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">afreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>

<span class="c1"># calculate whether a locus is polymorphic across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">apoly</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">apoly</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>

<span class="c1"># count the number of genotypes across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">gtcount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">gtcount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>

<span class="c1"># calculate the genotype frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">gtfreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">gtfreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>

<span class="c1"># calculate the minor allele frequency across all taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">maf</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">maf</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>

<span class="c1"># calculate the mean expected heterozygosity for the population</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">meh</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">meh</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>

<span class="c1"># count the number of major alleles individually within taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">tacount</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">tacount</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int32&quot;</span><span class="p">)</span>

<span class="c1"># calculate the allele frequency individually within taxa</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">tafreq</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pgmat</span><span class="o">.</span><span class="n">tafreq</span><span class="p">(</span><span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="saving-genotype-matrices">
<h2>Saving Genotype Matrices<a class="headerlink" href="#saving-genotype-matrices" title="Permalink to this heading">#</a></h2>
<section id="write-to-hdf5">
<h3>Write to HDF5<a class="headerlink" href="#write-to-hdf5" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write a breeding value matrix to an HDF5 file</span>
<span class="n">pgmat</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="s2">&quot;saved_genotypes.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="genotype_matrices.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Genotype Matrices</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="breeding_value_matrices.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Breeding Value Matrices</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#class-family-overview">
   Class Family Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-of-phased-genotype-matrix-classes">
   Summary of Phased Genotype Matrix Classes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-class-modules">
   Loading Class Modules
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-phased-genotype-matrices">
   Creating Phased Genotype Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-phased-genotype-matrices-from-numpy-arrays">
     Creating phased genotype matrices from NumPy arrays
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-phased-genotype-matrices-from-vcf-files">
     Loading phased genotype matrices from VCF files
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-phased-genotype-matrices-from-hdf5-files">
     Loading phased genotype matrices from HDF5 files
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#phased-genotype-matrix-properties">
   Phased Genotype Matrix Properties
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#general-properties">
     General properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#phase-properties">
     Phase properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#taxa-properties">
     Taxa properties
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#marker-variant-properties">
     Marker variant properties
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#copying-phased-genotype-matrices">
   Copying Phased Genotype Matrices
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#phased-genotype-matrix-element-copy-on-manipulation">
   Phased Genotype Matrix Element Copy-On-Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#adjoining-elements">
     Adjoining elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#deleting-elements">
     Deleting elements
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#delete-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         delete
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#delete-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         delete
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inserting-elements">
     Inserting elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#selecting-elements">
     Selecting elements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#phased-genotype-matrix-element-in-place-manipulation">
   Phased Genotype Matrix Element In-Place-Manipulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#appending-elements">
     Appending elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#removing-elements">
     Removing elements
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#remove-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         remove
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#remove-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         remove
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#incorporating-elements">
     Incorporating elements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#concatenating-matrices">
     Concatenating matrices
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-and-sorting">
   Grouping and Sorting
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reordering">
     Reordering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reorder-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         reorder
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reorder-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         reorder
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#lexsorting">
     Lexsorting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#lexsort-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         lexsort
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#lexsort-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         lexsort
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sorting">
     Sorting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sort-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         sort
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sort-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         sort
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#grouping">
     Grouping
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#group-taxa">
       <code class="docutils literal notranslate">
        <span class="pre">
         group
        </span>
       </code>
       taxa
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#group-marker-variants">
       <code class="docutils literal notranslate">
        <span class="pre">
         group
        </span>
       </code>
       marker variants
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-statistics">
   Summary Statistics
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#saving-genotype-matrices">
   Saving Genotype Matrices
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#write-to-hdf5">
     Write to HDF5
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2023, Robert Shrote.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>