
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Constrained Single-Trait Phenotypic Selection &#8212; pybrops 0.9.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../_static/documentation_options.js?v=39bb1c6d"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/examples/breeding_program_simulation/constrained_single_trait_phenotypic_selection';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Constrained Bi-Trait Phenotypic Selection" href="constrained_bitrait_phenotypic_selection.html" />
    <link rel="prev" title="Single-Trait Phenotypic Selection" href="single_trait_phenotypic_selection.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 0.9.0 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../examples.html">Examples</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/germplasm_panel_reduction.html">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../standard_usage.html" class="nav-link">Standard PyBrOpS Usage</a></li>
    
    
    <li class="breadcrumb-item"><a href="../examples.html" class="nav-link">Examples</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Constrained...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="constrained-single-trait-phenotypic-selection">
<h1>Constrained Single-Trait Phenotypic Selection<a class="headerlink" href="#constrained-single-trait-phenotypic-selection" title="Link to this heading">#</a></h1>
<p>In the <a class="reference internal" href="single_trait_phenotypic_selection.html"><span class="doc">Single-Trait Phenotypic Selection example</span></a>, we conducted a simulation where we selected the top individuals based on their estimated breeding values. Unfortunately, this “greedy” selection strategy quickly depletes genetic diversity, leading to poorer genetic gains in the long run. One method to combat the depletion of genetic diversity is to add an inbreeding constraint. Instead of selecting the top individuals, we select the top  individuals given that their mean relatedness does not exceed a specific value. In the example below, we perform both constrained and unconstrained selection and compare their long-term effects using simulations.</p>
<section id="simulation-preliminaries">
<h2>Simulation Preliminaries<a class="headerlink" href="#simulation-preliminaries" title="Link to this heading">#</a></h2>
<section id="loading-required-modules-and-seeding-the-global-prng">
<h3>Loading Required Modules and Seeding the global PRNG<a class="headerlink" href="#loading-required-modules-and-seeding-the-global-prng" title="Link to this heading">#</a></h3>
<p>To begin, we import the various modules we will be used into the Python namespace. We also set the seed for our simulation so that we can replicate the results at a later time.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import libraries</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Real</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">pybrops</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.bv.MeanPhenotypicBreedingValue</span> <span class="kn">import</span> <span class="n">MeanPhenotypicBreedingValue</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.mate.TwoWayCross</span> <span class="kn">import</span> <span class="n">TwoWayCross</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.mate.TwoWayDHCross</span> <span class="kn">import</span> <span class="n">TwoWayDHCross</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.pt.G_E_Phenotyping</span> <span class="kn">import</span> <span class="n">G_E_Phenotyping</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.sel.EstimatedBreedingValueSelection</span> <span class="kn">import</span> <span class="n">EstimatedBreedingValueSubsetSelection</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.sel.OptimalContributionSelection</span> <span class="kn">import</span> <span class="n">OptimalContributionSubsetSelection</span>
<span class="kn">from</span> <span class="nn">pybrops.model.gmod.DenseAdditiveLinearGenomicModel</span> <span class="kn">import</span> <span class="n">DenseAdditiveLinearGenomicModel</span>
<span class="kn">from</span> <span class="nn">pybrops.opt.algo.SortingSubsetOptimizationAlgorithm</span> <span class="kn">import</span> <span class="n">SortingSubsetOptimizationAlgorithm</span>
<span class="kn">from</span> <span class="nn">pybrops.opt.algo.SteepestDescentSubsetHillClimber</span> <span class="kn">import</span> <span class="n">SteepestDescentSubsetHillClimber</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.bvmat.DenseBreedingValueMatrix</span> <span class="kn">import</span> <span class="n">DenseBreedingValueMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.cmat.fcty.DenseMolecularCoancestryMatrixFactory</span> <span class="kn">import</span> <span class="n">DenseMolecularCoancestryMatrixFactory</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.HaldaneMapFunction</span> <span class="kn">import</span> <span class="n">HaldaneMapFunction</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.StandardGeneticMap</span> <span class="kn">import</span> <span class="n">StandardGeneticMap</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DensePhasedGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DensePhasedGenotypeMatrix</span>

<span class="c1"># seed python random and numpy random</span>
<span class="n">pybrops</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">prng</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">52347529</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="simulation-parameters">
<h3>Simulation Parameters<a class="headerlink" href="#simulation-parameters" title="Link to this heading">#</a></h3>
<p>Next, we define a couple of simulation parameter constants.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">nfndr</span> <span class="o">=</span> <span class="mi">40</span>      <span class="c1"># number of founder individuals</span>
<span class="n">nqtl</span> <span class="o">=</span> <span class="mi">1000</span>     <span class="c1"># number of QTL</span>
<span class="n">qlen</span> <span class="o">=</span> <span class="mi">6</span>        <span class="c1"># length of the queue</span>
<span class="n">ncross</span> <span class="o">=</span> <span class="mi">20</span>     <span class="c1"># number of cross configurations</span>
<span class="n">nparent</span> <span class="o">=</span> <span class="mi">2</span>     <span class="c1"># number of parents per cross configuration</span>
<span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span>     <span class="c1"># number of times to perform cross configuration</span>
<span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">80</span>   <span class="c1"># number of progenies per cross attempt</span>
<span class="n">nrandmate</span> <span class="o">=</span> <span class="mi">20</span>  <span class="c1"># number of random intermatings</span>
<span class="n">nsimul</span> <span class="o">=</span> <span class="mi">60</span>     <span class="c1"># number of simulation generations</span>
</pre></div>
</div>
</section>
<section id="loading-genetic-map-data-from-a-text-file">
<h3>Loading Genetic Map Data from a Text File<a class="headerlink" href="#loading-genetic-map-data-from-a-text-file" title="Link to this heading">#</a></h3>
<p>Then, we load genetic map data from a CSV-like file. In this example, we are using the US NAM genetic map constructed by McMullen et al. in 2009. The code below demonstrates how these data are read into a <code class="docutils literal notranslate"><span class="pre">StandardGeneticMap</span></code> object using the <code class="docutils literal notranslate"><span class="pre">from_csv</span></code> class method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read genetic map</span>
<span class="n">gmap</span> <span class="o">=</span> <span class="n">StandardGeneticMap</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
    <span class="s2">&quot;McMullen_2009_US_NAM.gmap&quot;</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp_col</span>     <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span>
    <span class="n">vrnt_phypos_col</span>     <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span>
    <span class="n">vrnt_genpos_col</span>     <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>
    <span class="n">vrnt_genpos_units</span>   <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>
    <span class="n">auto_group</span>          <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">auto_build_spline</span>   <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">sep</span>                 <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">header</span>              <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="creating-a-genetic-map-function">
<h3>Creating a Genetic Map Function<a class="headerlink" href="#creating-a-genetic-map-function" title="Link to this heading">#</a></h3>
<p>After loading our genetic map data, we want to create a genetic map function object which will be used to calculate recombination probabilities for our simulations. We create a simple Haldane genetic map function using the <code class="docutils literal notranslate"><span class="pre">HaldaneMapFunction</span></code> class.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># use Haldane map function to calculate crossover probabilities</span>
<span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="loading-genome-data-from-a-vcf-file">
<h3>Loading Genome Data from a VCF File<a class="headerlink" href="#loading-genome-data-from-a-vcf-file" title="Link to this heading">#</a></h3>
<p>Next, we load phased genetic markers from a VCF file. In this example, we are using a subset of genetic markers from the Wisconsin Maize Diversity Panel, which is composed of 942 individuals. 2000 SNPs with a minor allele frequency greater than 0.2 have been randomly selected to keep the dataset small. From the original 942 individuals and 2000 SNPs, we randomly select 40 founders and 1000 SNPs to create a founding synthetic population.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read phased genetic markers from a vcf file</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s2">&quot;widiv_2000SNPs.vcf.gz&quot;</span><span class="p">,</span> <span class="n">auto_group_vrnt</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># randomly select ``nfndr`` from which to create a founding population</span>
<span class="n">ix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">nfndr</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>

<span class="c1"># randomly select ``nqtl`` markers from founders</span>
<span class="n">ix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span> <span class="n">nqtl</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">select_vrnt</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>

<span class="c1"># sort and group variants</span>
<span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">group_vrnt</span><span class="p">()</span>
</pre></div>
</div>
<p>After loading the genetic markers, we interpolate the genetic map positions and the sequential marker crossover probabilities using the <code class="docutils literal notranslate"><span class="pre">interp_xoprob</span></code> method. We interpolate using the genetic map and genetic map function we have just created.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># interpolate genetic map positions</span>
<span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">interp_xoprob</span><span class="p">(</span><span class="n">gmap</span><span class="p">,</span> <span class="n">gmapfn</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="constructing-a-single-trait-genomic-model">
<h3>Constructing a Single-Trait Genomic Model<a class="headerlink" href="#constructing-a-single-trait-genomic-model" title="Link to this heading">#</a></h3>
<p>Next, we want to construct a true genomic model that will model a single trait with a strictly additive genetic architecture. We construct a <code class="docutils literal notranslate"><span class="pre">DenseAdditiveLinearGenomicModel</span></code> object to do this.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># model intercepts: (1,ntrait)</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.0</span><span class="p">]],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>

<span class="c1"># marker effects: (nvrnt,1)</span>
<span class="n">mkreffect</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span>
    <span class="n">loc</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="n">scale</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span>
    <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># trait names: (ntrait,)</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;Syn1&quot;</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create an additive linear genomic model to model traits</span>
<span class="n">algmod</span> <span class="o">=</span> <span class="n">DenseAdditiveLinearGenomicModel</span><span class="p">(</span>
    <span class="n">beta</span>        <span class="o">=</span> <span class="n">beta</span><span class="p">,</span>                 <span class="c1"># model intercepts</span>
    <span class="n">u_misc</span>      <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>                 <span class="c1"># miscellaneous random effects</span>
    <span class="n">u_a</span>         <span class="o">=</span> <span class="n">mkreffect</span><span class="p">,</span>            <span class="c1"># random marker effects</span>
    <span class="n">trait</span>       <span class="o">=</span> <span class="n">trait</span><span class="p">,</span>                <span class="c1"># trait names</span>
    <span class="n">model_name</span>  <span class="o">=</span> <span class="s2">&quot;synthetic_model&quot;</span><span class="p">,</span>    <span class="c1"># name of the model</span>
    <span class="n">hyperparams</span> <span class="o">=</span> <span class="kc">None</span>                  <span class="c1"># model parameters</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="build-founder-populations-run-breeding-program-burn-in">
<h2>Build Founder Populations &amp; Run Breeding Program Burn-In<a class="headerlink" href="#build-founder-populations-run-breeding-program-burn-in" title="Link to this heading">#</a></h2>
<p>In this next series of steps, we’ll create a founder population by randomly intermating our 40 selected founders for 20 generations. Following random intermating, we’ll create 6 cohorts of doubled haploid individuals, each cohort representing progenies produced in a single year. In this simulation, the oldest 3 cohorts will be aggregated into a <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> population to serve as selection candidates. The youngest 3 cohorts will be “in the pipeline” and will represent doubled haploid progenies which are being made. From the <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> population, the top 5% of individuals from each family will be selected to form a <code class="docutils literal notranslate"><span class="pre">&quot;cand&quot;</span></code> population. It is from this <code class="docutils literal notranslate"><span class="pre">&quot;cand&quot;</span></code> population that parents will be selected. The burn-in segment of the breeding program continues until the mean expected heterozygosity decreases to just under 0.3. After this point is reached, the simulated populations are ready for use in the main simulation.</p>
<section id="randomly-intermate-for-nrandmate-generations">
<h3>Randomly Intermate for <code class="docutils literal notranslate"><span class="pre">nrandmate</span></code> Generations<a class="headerlink" href="#randomly-intermate-for-nrandmate-generations" title="Link to this heading">#</a></h3>
<p>To start our random mating scheme, we first randomly pair up our 40 founders and mate them to create hybrids. Each mating event creates 80 progenies. This creates a population of size 1600.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create 2-way cross object</span>
<span class="n">mate2way</span> <span class="o">=</span> <span class="n">TwoWayCross</span><span class="p">()</span>

<span class="c1"># randomly select and pair founders</span>
<span class="n">xconfig</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">nfndr</span><span class="p">,</span> <span class="n">nfndr</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">nfndr</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

<span class="c1"># randomly intermate ``nfndr`` founders to create initial hybrids</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="p">,</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="p">,</span>
    <span class="n">nmating</span> <span class="o">=</span> <span class="n">nmating</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">nprogeny</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Then, we take our hybrid progenies, randomly pair them up with each other, and mate them. Each mating event produces 1 progeny. We randomly intermate progenies for 20 generations to create a founding population.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># randomly intermate for ``nrandmate`` generations</span>
<span class="c1"># each individual in the population is randomly mated with another individual</span>
<span class="c1"># and creates a single progeny so that the population size is held constant</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">nrandmate</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
    <span class="c1"># randomly select and pair ``ntaxa`` parents</span>
    <span class="n">ntaxa</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">ntaxa</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">ntaxa</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>
    <span class="n">xconfig</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">xconfig</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="c1"># randomly intermate ``ntaxa`` parents</span>
    <span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
        <span class="n">pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="p">,</span>
        <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="p">,</span>
        <span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Random Intermating:&quot;</span><span class="p">,</span> <span class="n">gen</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-breeding-protocols-for-burn-in">
<h3>Create Breeding Protocols for Burn-In<a class="headerlink" href="#create-breeding-protocols-for-burn-in" title="Link to this heading">#</a></h3>
<p>Next, we make a 2-way DH cross protocol for our burn-in stage.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a 2-way DH cross object, use the counters from the 2-way cross object</span>
<span class="n">mate2waydh</span> <span class="o">=</span> <span class="n">TwoWayDHCross</span><span class="p">(</span>
    <span class="n">progeny_counter</span> <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">progeny_counter</span><span class="p">,</span>
    <span class="n">family_counter</span>  <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">family_counter</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>We also create a genotyping protocol that converts phased genotypes to unphased genotypes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a genotyping protocol</span>
<span class="n">gtprot</span> <span class="o">=</span> <span class="n">DenseUnphasedGenotyping</span><span class="p">()</span>
</pre></div>
</div>
<p>To simulate phenotypes, we create a simple <span class="math notranslate nohighlight">\(G+E\)</span> phenotyping protocol that phenotypes individuals in 4 locations, 1 replication each location. We use the founding population we have just created with random mating to set the narrow sense heritability at the single plot level to 0.4. With 4 replications, this increases the heritability to approximately 0.7 to begin.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a phenotyping protocol</span>
<span class="n">ptprot</span> <span class="o">=</span> <span class="n">G_E_Phenotyping</span><span class="p">(</span><span class="n">algmod</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ptprot</span><span class="o">.</span><span class="n">set_h2</span><span class="p">(</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">fndr_pgmat</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, we create a breeding value estimation protocol that simply uses means to estimate breeding value.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a breeding value estimation protocol</span>
<span class="n">bvprot</span> <span class="o">=</span> <span class="n">MeanPhenotypicBreedingValue</span><span class="p">(</span><span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa_grp&quot;</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-a-within-family-selection-function">
<h3>Create a Within-Family Selection Function<a class="headerlink" href="#create-a-within-family-selection-function" title="Link to this heading">#</a></h3>
<p>Next, we create a custom function to manually select individuals within families based on their breeding values. This function selects the top <code class="docutils literal notranslate"><span class="pre">nindiv</span></code> individuals within each family and outputs a set of indices corresponding to these individuals.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define function to do within family selection based on yield</span>
<span class="k">def</span> <span class="nf">within_family_selection</span><span class="p">(</span><span class="n">bvmat</span><span class="p">:</span> <span class="n">DenseBreedingValueMatrix</span><span class="p">,</span> <span class="n">nindiv</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
    <span class="n">order</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">bvmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
    <span class="n">value</span> <span class="o">=</span> <span class="n">bvmat</span><span class="o">.</span><span class="n">mat</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># get yield breeding values</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bvmat</span><span class="o">.</span><span class="n">taxa_grp</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">bvmat</span><span class="o">.</span><span class="n">taxa_grp</span> <span class="o">==</span> <span class="n">group</span>
        <span class="n">tmp_order</span> <span class="o">=</span> <span class="n">order</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">tmp_value</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">value_argsort</span> <span class="o">=</span> <span class="n">tmp_value</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>
        <span class="n">ix</span> <span class="o">=</span> <span class="n">value_argsort</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="n">nindiv</span><span class="p">]</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_order</span><span class="p">[</span><span class="n">ix</span><span class="p">])</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">indices</span>
</pre></div>
</div>
</section>
<section id="create-cohort-structure">
<h3>Create Cohort Structure<a class="headerlink" href="#create-cohort-structure" title="Link to this heading">#</a></h3>
<p>In the next step, we create a cohort structure as described at the beginning of this section. We utilize dictionaries to keep track of our cohorts and populations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a dictionary to store founder individuals</span>
<span class="n">fndr_genome</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;queue&quot;</span><span class="p">:[]}</span>
<span class="n">fndr_geno</span>   <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;queue&quot;</span><span class="p">:[]}</span>
<span class="n">fndr_pheno</span>  <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">}</span>
<span class="n">fndr_bval</span>   <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>   <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">}</span>
<span class="n">fndr_gmod</span>   <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="n">algmod</span><span class="p">,</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="n">algmod</span><span class="p">,</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span><span class="n">algmod</span><span class="p">}</span>
</pre></div>
</div>
<p>In the code block below, we define a helper function to help us create initial cohorts of individuals from our starting randomly mated population.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define a helper function to help make cohorts of individuals</span>
<span class="k">def</span> <span class="nf">cohort</span><span class="p">(</span>
        <span class="n">mate2waydh</span><span class="p">:</span> <span class="n">MatingProtocol</span><span class="p">,</span>
        <span class="n">pgmat</span><span class="p">:</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">,</span>
        <span class="n">ncross</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nparent</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nmating</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nprogeny</span><span class="p">:</span> <span class="nb">int</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">:</span>
    <span class="c1"># sample indicies of individuals and reshape for input into mating protocol</span>
    <span class="n">xconfix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ncross</span> <span class="o">*</span> <span class="n">nparent</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfix</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">ncross</span><span class="p">,</span> <span class="n">nparent</span><span class="p">)</span>
    <span class="c1"># mate individuals</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span><span class="n">pgmat</span><span class="p">,</span> <span class="n">xconfig</span><span class="p">,</span> <span class="n">nmating</span><span class="p">,</span> <span class="n">nprogeny</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span>
</pre></div>
</div>
<p>Next, we fill our cohort queue and construct the <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;cand&quot;</span></code> populations from the queue.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># fill queue with cohort genomes derived from randomly mating the founders</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cohort</span><span class="p">(</span><span class="n">mate2waydh</span><span class="p">,</span> <span class="n">fndr_pgmat</span><span class="p">,</span> <span class="n">ncross</span><span class="p">,</span> <span class="n">nparent</span><span class="p">,</span> <span class="n">nmating</span><span class="p">,</span> <span class="n">nprogeny</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">qlen</span><span class="p">)]</span>

<span class="c1"># construct the main population genomes from the first three cohorts in the queue</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>

<span class="c1"># genotype individuals to fill the genotyping queue</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]]</span>

<span class="c1"># construct the main population genotypes from the first three cohorts in the queue</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>

<span class="c1"># phenotype the main population</span>
<span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>

<span class="c1"># calculate breeding values for the main population</span>
<span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>

<span class="c1"># calculate indices for within family selection to get parental candidates</span>
<span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span> <span class="c1"># select top 5%</span>

<span class="c1"># select parental candidates</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
<span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-a-burn-in-selection-protocol-object">
<h3>Create a Burn-In Selection Protocol Object<a class="headerlink" href="#create-a-burn-in-selection-protocol-object" title="Link to this heading">#</a></h3>
<p>Next, we create a selection protocol for our burn-in loop. This selects individuals with the best breeding values for our synthetic trait.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># use a hillclimber for the single-objective optimization algorithm</span>
<span class="n">soalgo</span> <span class="o">=</span> <span class="n">SortingSubsetOptimizationAlgorithm</span><span class="p">()</span>

<span class="c1"># create a selection protocol that selects based on EBVs with an inbreeding constraint</span>
<span class="n">burnin_selprot</span> <span class="o">=</span> <span class="n">EstimatedBreedingValueSubsetSelection</span><span class="p">(</span>
    <span class="n">ntrait</span>       <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of expected traits</span>
    <span class="n">unscale</span>      <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>         <span class="c1"># unscale breeding values to human-readable format</span>
    <span class="n">ncross</span>       <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>           <span class="c1"># number of cross configurations</span>
    <span class="n">nparent</span>      <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>            <span class="c1"># number of parents per cross configuration</span>
    <span class="n">nmating</span>      <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of matings per cross configuration</span>
    <span class="n">nprogeny</span>     <span class="o">=</span> <span class="mi">80</span><span class="p">,</span>           <span class="c1"># number of progeny per mating event</span>
    <span class="n">nobj</span>         <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of objectives == 1 == yield</span>
    <span class="n">soalgo</span>       <span class="o">=</span> <span class="n">soalgo</span><span class="p">,</span>       <span class="c1"># use sorting algorithm to solve single-objective problem</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="running-a-population-burn-in-until-meh-is-slightly-less-than-0-30">
<h3>Running a Population Burn-in until MEH is slightly less than 0.30<a class="headerlink" href="#running-a-population-burn-in-until-meh-is-slightly-less-than-0-30" title="Link to this heading">#</a></h3>
<p>Finally, we perform selection on our populations until the mean expected heterozygosity of the <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> population reaches slightly less than 0.30. This in effect sets the starting genetic diversity for our simulations so that they are not as dependent on the set of founders that were initially randomly selected.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">while</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mf">0.30</span><span class="p">:</span>
    <span class="c1"># parental selection: select parents from parental candidates</span>
    <span class="n">selcfg</span> <span class="o">=</span> <span class="n">burnin_selprot</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">pgmat</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gmat</span>  <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">ptdf</span>  <span class="o">=</span> <span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">bvmat</span> <span class="o">=</span> <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gpmod</span> <span class="o">=</span> <span class="n">fndr_gmod</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">t_cur</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># mate: create new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_genome</span> <span class="o">=</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
        <span class="n">pgmat</span>    <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">pgmat</span><span class="p">,</span>
        <span class="n">xconfig</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">xconfig</span><span class="p">,</span>
        <span class="n">nmating</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nmating</span><span class="p">,</span>
        <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nprogeny</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: genotype new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_geno</span> <span class="o">=</span> <span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_geno</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: phenotype main population</span>
    <span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># evaluate: calculate breeding values for the main population</span>
    <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># survivor selection: select parental candidate indices from main population</span>
    <span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span> <span class="c1"># select top 5%</span>
    <span class="c1"># survivor selection: select parental candidates from main population</span>
    <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Burn-in:&quot;</span><span class="p">,</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Starting MEH:&quot;</span><span class="p">,</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
</pre></div>
</div>
</section>
</section>
<section id="simulation-setup">
<h2>Simulation Setup<a class="headerlink" href="#simulation-setup" title="Link to this heading">#</a></h2>
<p>In the sections above, we created the starting point for our simulations. In these next sections, we’ll test two selection strategies: one constrained by inbreeding, and another unconstrained by inbreeding. We’ll plot and compare the results of these simulations.</p>
<section id="create-a-constrained-selection-protocol-object">
<h3>Create a Constrained Selection Protocol Object<a class="headerlink" href="#create-a-constrained-selection-protocol-object" title="Link to this heading">#</a></h3>
<p>First, we’ll define a selection protocol where individuals are selected on their breeding values subject to an inbreeding constraint specified by the user. We’ll use the <code class="docutils literal notranslate"><span class="pre">OptimalContributionSubsetSelection</span></code> class to accomplish this task and provide objective and inequality constraint violation transformation functions to create such a scenario.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a dense molecular coancestry matrix factory</span>
<span class="n">cmatfcty</span> <span class="o">=</span> <span class="n">DenseMolecularCoancestryMatrixFactory</span><span class="p">()</span>

<span class="c1"># define an objective transformation function</span>
<span class="k">def</span> <span class="nf">obj_trans</span><span class="p">(</span>
        <span class="n">decnvec</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">latentvec</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Receive an incoming vector of [MGR,BV1,...,BVn] and transform it to</span>
<span class="sd">    [BV1,...,BVn].</span>

<span class="sd">    Where::</span>

<span class="sd">        - MGR is the mean genomic relationship (kinship; in range [0,1]).</span>
<span class="sd">        - BVn is the nth mean breeding value for the subset.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    decnvec : numpy.ndarray</span>
<span class="sd">        A decision space vector of shape (ndecn,)</span>
<span class="sd">    latentvec : numpy.ndarray</span>
<span class="sd">        A latent space function vector of shape (1+ntrait,)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : numpy.ndarray</span>
<span class="sd">        A vector of shape (ntrait,).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># extract trait(s) as objective(s)</span>
    <span class="k">return</span> <span class="n">latentvec</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

<span class="c1"># define an inequality constraint violation function</span>
<span class="k">def</span> <span class="nf">ineqcv_trans</span><span class="p">(</span>
        <span class="n">decnvec</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">latentvec</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">maxinb</span><span class="p">:</span> <span class="n">Real</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">dict</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A custom inequality constraint violation function.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    decnvec : numpy.ndarray</span>
<span class="sd">        A decision space vector of shape (ndecn,)</span>
<span class="sd">    latentvec : numpy.ndarray</span>
<span class="sd">        A latent space function vector of shape (1+ntrait,)</span>
<span class="sd">    minvec : numpy.ndarray</span>
<span class="sd">        Vector of minimum values for which the latent vector can take.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : numpy.ndarray</span>
<span class="sd">        An inequality constraint violation vector of shape (1,).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># calculate constraint violation for inbreeding</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">max</span><span class="p">(</span><span class="n">latentvec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">maxinb</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>
    <span class="c1"># return inequality constraint violation array</span>
    <span class="k">return</span> <span class="n">out</span>

<span class="c1"># use a hillclimber for the single-objective optimization algorithm</span>
<span class="n">soalgo</span> <span class="o">=</span> <span class="n">SteepestDescentSubsetHillClimber</span><span class="p">()</span>

<span class="c1"># create a selection protocol that selects based on EBVs with an inbreeding constraint</span>
<span class="n">const_selprot</span> <span class="o">=</span> <span class="n">OptimalContributionSubsetSelection</span><span class="p">(</span>
    <span class="n">ntrait</span>       <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of expected traits</span>
    <span class="n">cmatfcty</span>     <span class="o">=</span> <span class="n">cmatfcty</span><span class="p">,</span>     <span class="c1"># coancestry/kinship matrix factory</span>
    <span class="n">unscale</span>      <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>         <span class="c1"># unscale breeding values to human-readable format</span>
    <span class="n">ncross</span>       <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>           <span class="c1"># number of cross configurations</span>
    <span class="n">nparent</span>      <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>            <span class="c1"># number of parents per cross configuration</span>
    <span class="n">nmating</span>      <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of matings per cross configuration</span>
    <span class="n">nprogeny</span>     <span class="o">=</span> <span class="mi">80</span><span class="p">,</span>           <span class="c1"># number of progeny per mating event</span>
    <span class="n">nobj</span>         <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of objectives == ntrait</span>
    <span class="n">obj_trans</span>    <span class="o">=</span> <span class="n">obj_trans</span><span class="p">,</span>    <span class="c1"># latent vector transformation to create objective function</span>
    <span class="n">nineqcv</span>      <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of inequality constraint violations</span>
    <span class="n">ineqcv_trans</span> <span class="o">=</span> <span class="n">ineqcv_trans</span><span class="p">,</span> <span class="c1"># latent vector transformation to create inequality constraints</span>
    <span class="n">ineqcv_trans_kwargs</span> <span class="o">=</span> <span class="p">{</span>      <span class="c1"># keyword arguments</span>
        <span class="s2">&quot;maxinb&quot;</span><span class="p">:</span> <span class="mf">1.0</span>
    <span class="p">},</span>
    <span class="n">soalgo</span>       <span class="o">=</span> <span class="n">soalgo</span><span class="p">,</span>       <span class="c1"># use hillclimber to solve single-objective problem</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-an-unconstrained-selection-protocol-object">
<h3>Create an Unconstrained Selection Protocol Object<a class="headerlink" href="#create-an-unconstrained-selection-protocol-object" title="Link to this heading">#</a></h3>
<p>Second, we’ll define a selection protocol where individuals are selected solely on their breeding values. We’ll use the <code class="docutils literal notranslate"><span class="pre">EstimatedBreedingValueSubsetSelection</span></code> to accomplish this task.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># use a sorting algorithm for the single-objective optimization algorithm</span>
<span class="n">soalgo</span> <span class="o">=</span> <span class="n">SortingSubsetOptimizationAlgorithm</span><span class="p">()</span>

<span class="c1"># create a selection protocol that selects based on EBVs with an inbreeding constraint</span>
<span class="n">unconst_selprot</span> <span class="o">=</span> <span class="n">EstimatedBreedingValueSubsetSelection</span><span class="p">(</span>
    <span class="n">ntrait</span>       <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of expected traits</span>
    <span class="n">unscale</span>      <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>         <span class="c1"># unscale breeding values to human-readable format</span>
    <span class="n">ncross</span>       <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>           <span class="c1"># number of cross configurations</span>
    <span class="n">nparent</span>      <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>            <span class="c1"># number of parents per cross configuration</span>
    <span class="n">nmating</span>      <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of matings per cross configuration</span>
    <span class="n">nprogeny</span>     <span class="o">=</span> <span class="mi">80</span><span class="p">,</span>           <span class="c1"># number of progeny per mating event</span>
    <span class="n">nobj</span>         <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>            <span class="c1"># number of objectives == ntrait</span>
    <span class="n">soalgo</span>       <span class="o">=</span> <span class="n">soalgo</span><span class="p">,</span>       <span class="c1"># use sorting algorithm to solve single-objective problem</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="make-a-statistics-recording-helper-function">
<h3>Make a Statistics Recording Helper Function<a class="headerlink" href="#make-a-statistics-recording-helper-function" title="Link to this heading">#</a></h3>
<p>To assist in record keeping, we’ll create a function to help us record simulation metrics and store them into a dictionary.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make recording helper function</span>
<span class="k">def</span> <span class="nf">record</span><span class="p">(</span><span class="n">lbook</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">gen</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">geno</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">pheno</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">bval</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">gmod</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gen</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_meh&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
    <span class="c1">################### main true lower selection limits ###################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lsl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_lsl&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">################### main true lower selection limits ###################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">usl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_usl&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">###################### main true breeding values #######################</span>
    <span class="n">tbv</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">gebv</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1">################## main true breeding value minimums ###################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">tbv</span><span class="o">.</span><span class="n">tmin</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_min&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">#################### main true breeding value means ####################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">tbv</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_mean&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">################## main true breeding value maximums ###################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">tbv</span><span class="o">.</span><span class="n">tmax</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_max&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">############# main true breeding value standard deviations #############</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">tbv</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">##################### main true genetic variances ######################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">var_A</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_var_A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">###################### main true genic variances #######################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">var_a</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_var_a&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">#################### main estimated breeding values ####################</span>
    <span class="n">ebv</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span>
    <span class="c1">################ main estimated breeding value minimums ################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">ebv</span><span class="o">.</span><span class="n">tmin</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_min&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">#################### main true breeding value means ####################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">ebv</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_mean&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">################## main true breeding value maximums ###################</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">ebv</span><span class="o">.</span><span class="n">tmax</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_max&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1">############# main true breeding value standard deviations #############</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">ebv</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">lbook</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_std&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</section>
</section>
<section id="simulate-constrained-phenotypic-selection-for-60-generations">
<h2>Simulate Constrained Phenotypic Selection for 60 Generations<a class="headerlink" href="#simulate-constrained-phenotypic-selection-for-60-generations" title="Link to this heading">#</a></h2>
<p>The next subsections detail simulations in the constrained scenario.</p>
<section id="copy-founders">
<h3>Copy Founders<a class="headerlink" href="#copy-founders" title="Link to this heading">#</a></h3>
<p>First we’ll copy our founder populations so that we don’t modify them and can use them for repeated simulations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy founder populations, bvals, etc. so we can replicate if needed</span>
<span class="n">simul_genome</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">)</span>
<span class="n">simul_geno</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_geno</span><span class="p">)</span>
<span class="n">simul_pheno</span>  <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_pheno</span><span class="p">)</span>
<span class="n">simul_bval</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_bval</span><span class="p">)</span>
<span class="n">simul_gmod</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_gmod</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="rudimentary-logbooks">
<h3>Rudimentary Logbooks<a class="headerlink" href="#rudimentary-logbooks" title="Link to this heading">#</a></h3>
<p>Then, we’ll make a dictionary logbook to store simulation metrics.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make a dictionary logbook</span>
<span class="n">const_lbook</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;gen&quot;</span>                   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_meh&quot;</span>              <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_lsl&quot;</span>         <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_usl&quot;</span>         <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_min&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_mean&quot;</span>    <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_max&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_std&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_var_A&quot;</span>   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_var_a&quot;</span>   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_min&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_mean&quot;</span>    <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_max&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_std&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="simulation-main-loop">
<h3>Simulation Main Loop<a class="headerlink" href="#simulation-main-loop" title="Link to this heading">#</a></h3>
<p>Next, we’ll simulate our breeding program for 60 generations. We’ll start at an inbreeding constraint of 0.84 or less to begin and gradually increase the maximum allowed inbreeding to 1.0 (complete homozygosity; zero genetic diversity) by the end of the simulation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># record initial statistics</span>
<span class="n">record</span><span class="p">(</span><span class="n">const_lbook</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">simul_genome</span><span class="p">,</span> <span class="n">simul_geno</span><span class="p">,</span> <span class="n">simul_pheno</span><span class="p">,</span> <span class="n">simul_bval</span><span class="p">,</span> <span class="n">simul_gmod</span><span class="p">)</span>

<span class="c1"># create constraint change over time</span>
<span class="c1"># 0.84 manually determined from Pareto frontier examination</span>
<span class="n">maxinb</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.84</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">nsimul</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># main simulation loop</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">nsimul</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
    <span class="c1"># set the inbreeding constraint</span>
    <span class="n">const_selprot</span><span class="o">.</span><span class="n">ineqcv_trans_kwargs</span><span class="p">[</span><span class="s2">&quot;maxinb&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">maxinb</span><span class="p">[</span><span class="n">gen</span><span class="p">]</span>
    <span class="c1"># parental selection: select parents from parental candidates</span>
    <span class="n">selcfg</span> <span class="o">=</span> <span class="n">const_selprot</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">pgmat</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gmat</span>  <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">ptdf</span>  <span class="o">=</span> <span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">bvmat</span> <span class="o">=</span> <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gpmod</span> <span class="o">=</span> <span class="n">simul_gmod</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">t_cur</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># mate: create new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_genome</span> <span class="o">=</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
        <span class="n">pgmat</span>    <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">pgmat</span><span class="p">,</span>
        <span class="n">xconfig</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">xconfig</span><span class="p">,</span>
        <span class="n">nmating</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nmating</span><span class="p">,</span>
        <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nprogeny</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: genotype new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_geno</span> <span class="o">=</span> <span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_geno</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: phenotype main population</span>
    <span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># evaluate: calculate breeding values for the main population</span>
    <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># survivor selection: select parental candidate indices from main population</span>
    <span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span> <span class="c1"># select top 5%</span>
    <span class="c1"># survivor selection: select parental candidates from main population</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="c1"># record statistics</span>
    <span class="n">record</span><span class="p">(</span><span class="n">const_lbook</span><span class="p">,</span> <span class="n">gen</span><span class="p">,</span> <span class="n">simul_genome</span><span class="p">,</span> <span class="n">simul_geno</span><span class="p">,</span> <span class="n">simul_pheno</span><span class="p">,</span> <span class="n">simul_bval</span><span class="p">,</span> <span class="n">simul_gmod</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generation:&quot;</span><span class="p">,</span> <span class="n">gen</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="saving-results-to-a-file">
<h3>Saving Results to a File<a class="headerlink" href="#saving-results-to-a-file" title="Link to this heading">#</a></h3>
<p>Finally, we’ll save the results of our simulation to a CSV file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create output dataframe and save</span>
<span class="n">const_lbook_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">const_lbook</span><span class="p">)</span>
<span class="n">const_lbook_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;const_lbook.csv&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="simulate-unconstrained-phenotypic-selection-for-60-generations">
<h2>Simulate Unconstrained Phenotypic Selection for 60 Generations<a class="headerlink" href="#simulate-unconstrained-phenotypic-selection-for-60-generations" title="Link to this heading">#</a></h2>
<p>The next subsections detail simulations in the unconstrained scenario.</p>
<section id="id1">
<h3>Copy Founders<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<p>As before, we’ll copy our founder populations so that we don’t modify them and can use them for repeated simulations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># deep copy founder populations, bvals, etc. so we can replicate if needed</span>
<span class="n">simul_genome</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">)</span>
<span class="n">simul_geno</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_geno</span><span class="p">)</span>
<span class="n">simul_pheno</span>  <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_pheno</span><span class="p">)</span>
<span class="n">simul_bval</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_bval</span><span class="p">)</span>
<span class="n">simul_gmod</span>   <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">fndr_gmod</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id2">
<h3>Rudimentary Logbooks<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>Then, we’ll make a dictionary logbook to store simulation metrics.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># make a dictionary logbook</span>
<span class="n">unconst_lbook</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;gen&quot;</span>                   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_meh&quot;</span>              <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_lsl&quot;</span>         <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_usl&quot;</span>         <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_min&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_mean&quot;</span>    <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_max&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_std&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_var_A&quot;</span>   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_tbv_var_a&quot;</span>   <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_min&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_mean&quot;</span>    <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_max&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;main_Syn1_ebv_std&quot;</span>     <span class="p">:</span> <span class="p">[],</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="id3">
<h3>Simulation Main Loop<a class="headerlink" href="#id3" title="Link to this heading">#</a></h3>
<p>Next, we’ll simulate our breeding program for 60 generations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># record initial statistics</span>
<span class="n">record</span><span class="p">(</span><span class="n">unconst_lbook</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">simul_genome</span><span class="p">,</span> <span class="n">simul_geno</span><span class="p">,</span> <span class="n">simul_pheno</span><span class="p">,</span> <span class="n">simul_bval</span><span class="p">,</span> <span class="n">simul_gmod</span><span class="p">)</span>

<span class="c1"># main simulation loop</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">nsimul</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
    <span class="c1"># parental selection: select parents from parental candidates</span>
    <span class="n">selcfg</span> <span class="o">=</span> <span class="n">unconst_selprot</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">pgmat</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gmat</span>  <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">ptdf</span>  <span class="o">=</span> <span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">bvmat</span> <span class="o">=</span> <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">gpmod</span> <span class="o">=</span> <span class="n">simul_gmod</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
        <span class="n">t_cur</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># mate: create new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_genome</span> <span class="o">=</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
        <span class="n">pgmat</span>    <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">pgmat</span><span class="p">,</span>
        <span class="n">xconfig</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">xconfig</span><span class="p">,</span>
        <span class="n">nmating</span>  <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nmating</span><span class="p">,</span>
        <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">selcfg</span><span class="o">.</span><span class="n">nprogeny</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: genotype new genomes; discard oldest cohort; concat new main population</span>
    <span class="n">new_geno</span> <span class="o">=</span> <span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_geno</span><span class="p">)</span>
    <span class="n">discard</span> <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
    <span class="c1"># evaluate: phenotype main population</span>
    <span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># evaluate: calculate breeding values for the main population</span>
    <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">simul_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
    <span class="c1"># survivor selection: select parental candidate indices from main population</span>
    <span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span> <span class="c1"># select top 5%</span>
    <span class="c1"># survivor selection: select parental candidates from main population</span>
    <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simul_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
    <span class="c1"># record statistics</span>
    <span class="n">record</span><span class="p">(</span><span class="n">unconst_lbook</span><span class="p">,</span> <span class="n">gen</span><span class="p">,</span> <span class="n">simul_genome</span><span class="p">,</span> <span class="n">simul_geno</span><span class="p">,</span> <span class="n">simul_pheno</span><span class="p">,</span> <span class="n">simul_bval</span><span class="p">,</span> <span class="n">simul_gmod</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generation:&quot;</span><span class="p">,</span> <span class="n">gen</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id4">
<h3>Saving Results to a File<a class="headerlink" href="#id4" title="Link to this heading">#</a></h3>
<p>Finally, we’ll save the results of our simulation to a CSV file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create output dataframe and save</span>
<span class="n">unconst_lbook_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">unconst_lbook</span><span class="p">)</span>
<span class="n">unconst_lbook_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;unconst_lbook.csv&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="visualizing-breeding-program-simulation-results-with-matplotlib">
<h2>Visualizing Breeding Program Simulation Results with <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code><a class="headerlink" href="#visualizing-breeding-program-simulation-results-with-matplotlib" title="Link to this heading">#</a></h2>
<section id="visualizing-true-breeding-values-tbvs">
<h3>Visualizing True Breeding Values (TBVs)<a class="headerlink" href="#visualizing-true-breeding-values-tbvs" title="Link to this heading">#</a></h3>
<p>Using the results, we’ll visualize the population mean true breeding values for the constrained and unconstrained scenarios. Furthermore, we’ll plot the upper and lower selection limits to depict the narrowing of the genetic diverity over time.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_mean&quot;</span><span class="p">],</span>   <span class="s1">&#39;-b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: Mean Pop. TBV&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_lsl&quot;</span><span class="p">],</span>   <span class="s1">&#39;:b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: LSL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_usl&quot;</span><span class="p">],</span>   <span class="s1">&#39;--b&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: USL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_mean&quot;</span><span class="p">],</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: Mean Pop. TBV&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_lsl&quot;</span><span class="p">],</span> <span class="s1">&#39;:r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: LSL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_usl&quot;</span><span class="p">],</span> <span class="s1">&#39;--r&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: USL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Single-Trait Recurrent Phenotypic Selection&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait Breeding Value&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;constrained_single_trait_phenotypic_selection_true_breeding_values.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/constrained_single_trait_phenotypic_selection_true_breeding_values.png" src="../../../_images/constrained_single_trait_phenotypic_selection_true_breeding_values.png" />
</section>
<section id="visualizing-estimated-breeding-values-tbvs">
<h3>Visualizing Estimated Breeding Values (TBVs)<a class="headerlink" href="#visualizing-estimated-breeding-values-tbvs" title="Link to this heading">#</a></h3>
<p>Next, we’ll visualize the estimated breeding values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_mean&quot;</span><span class="p">],</span>   <span class="s1">&#39;-b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: Mean Pop. EBV&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_lsl&quot;</span><span class="p">],</span>   <span class="s1">&#39;:b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: LSL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_usl&quot;</span><span class="p">],</span>   <span class="s1">&#39;--b&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: USL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_mean&quot;</span><span class="p">],</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: Mean Pop. EBV&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_lsl&quot;</span><span class="p">],</span> <span class="s1">&#39;:r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: LSL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_usl&quot;</span><span class="p">],</span> <span class="s1">&#39;--r&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: USL&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Single-Trait Recurrent Phenotypic Selection&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait Breeding Value&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;constrained_single_trait_phenotypic_selection_estimated_breeding_values.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/constrained_single_trait_phenotypic_selection_estimated_breeding_values.png" src="../../../_images/constrained_single_trait_phenotypic_selection_estimated_breeding_values.png" />
</section>
<section id="visualizing-mean-expected-heterozygosity-meh">
<h3>Visualizing Mean Expected Heterozygosity (MEH)<a class="headerlink" href="#visualizing-mean-expected-heterozygosity-meh" title="Link to this heading">#</a></h3>
<p>Mean expected heterozygosity is a valuable diversity metric. We’ll plot the change in mean expected diversity over time for both the constrained and unconstrained selection scenarios.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_meh&quot;</span><span class="p">],</span>   <span class="s1">&#39;-b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: Pop. MEH&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_meh&quot;</span><span class="p">],</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: Pop. MEH&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Single-Trait Recurrent Phenotypic Selection&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Mean Expected Heterozygosity&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;constrained_single_trait_phenotypic_selection_mean_expected_heterozygosity.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/constrained_single_trait_phenotypic_selection_mean_expected_heterozygosity.png" src="../../../_images/constrained_single_trait_phenotypic_selection_mean_expected_heterozygosity.png" />
</section>
<section id="visualizing-true-breeding-value-standard-deviations">
<h3>Visualizing True Breeding Value Standard Deviations<a class="headerlink" href="#visualizing-true-breeding-value-standard-deviations" title="Link to this heading">#</a></h3>
<p>Plotting the change in population breeding value variance is another important metric for examining the change in genetic diverity over time. We’ll plot the constrained and unconstrained population true breeding value standard deviations over time.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_std&quot;</span><span class="p">],</span>   <span class="s1">&#39;-b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: Pop. TBV SD&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_tbv_std&quot;</span><span class="p">],</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: Pop. TBV SD&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Single-Trait Recurrent Phenotypic Selection&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait Breeding Value Standard Deviation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;constrained_single_trait_phenotypic_selection_true_breeding_value_standard_deviation.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/constrained_single_trait_phenotypic_selection_true_breeding_value_standard_deviation.png" src="../../../_images/constrained_single_trait_phenotypic_selection_true_breeding_value_standard_deviation.png" />
</section>
<section id="visualizing-estimated-breeding-value-standard-deviations">
<h3>Visualizing Estimated Breeding Value Standard Deviations<a class="headerlink" href="#visualizing-estimated-breeding-value-standard-deviations" title="Link to this heading">#</a></h3>
<p>We’ll do the same thing as above but with estimated breeding values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create static figure</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">axes</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span>   <span class="n">const_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_std&quot;</span><span class="p">],</span>   <span class="s1">&#39;-b&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Const. Sel.: Pop. EBV SD&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">],</span> <span class="n">unconst_lbook_df</span><span class="p">[</span><span class="s2">&quot;main_Syn1_ebv_std&quot;</span><span class="p">],</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span>  <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Unconst. Sel.: Pop. EBV SD&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Single-Trait Recurrent Phenotypic Selection&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait Breeding Value Standard Deviation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;constrained_single_trait_phenotypic_selection_estimated_breeding_value_standard_deviation.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/constrained_single_trait_phenotypic_selection_estimated_breeding_value_standard_deviation.png" src="../../../_images/constrained_single_trait_phenotypic_selection_estimated_breeding_value_standard_deviation.png" />
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="single_trait_phenotypic_selection.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Single-Trait Phenotypic Selection</p>
      </div>
    </a>
    <a class="right-next"
       href="constrained_bitrait_phenotypic_selection.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Constrained Bi-Trait Phenotypic Selection</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-preliminaries">Simulation Preliminaries</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-required-modules-and-seeding-the-global-prng">Loading Required Modules and Seeding the global PRNG</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-parameters">Simulation Parameters</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genetic-map-data-from-a-text-file">Loading Genetic Map Data from a Text File</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-genetic-map-function">Creating a Genetic Map Function</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genome-data-from-a-vcf-file">Loading Genome Data from a VCF File</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constructing-a-single-trait-genomic-model">Constructing a Single-Trait Genomic Model</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#build-founder-populations-run-breeding-program-burn-in">Build Founder Populations &amp; Run Breeding Program Burn-In</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#randomly-intermate-for-nrandmate-generations">Randomly Intermate for <code class="docutils literal notranslate"><span class="pre">nrandmate</span></code> Generations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-breeding-protocols-for-burn-in">Create Breeding Protocols for Burn-In</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-within-family-selection-function">Create a Within-Family Selection Function</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-cohort-structure">Create Cohort Structure</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-burn-in-selection-protocol-object">Create a Burn-In Selection Protocol Object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#running-a-population-burn-in-until-meh-is-slightly-less-than-0-30">Running a Population Burn-in until MEH is slightly less than 0.30</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-setup">Simulation Setup</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-constrained-selection-protocol-object">Create a Constrained Selection Protocol Object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-an-unconstrained-selection-protocol-object">Create an Unconstrained Selection Protocol Object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#make-a-statistics-recording-helper-function">Make a Statistics Recording Helper Function</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulate-constrained-phenotypic-selection-for-60-generations">Simulate Constrained Phenotypic Selection for 60 Generations</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#copy-founders">Copy Founders</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rudimentary-logbooks">Rudimentary Logbooks</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-main-loop">Simulation Main Loop</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-results-to-a-file">Saving Results to a File</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulate-unconstrained-phenotypic-selection-for-60-generations">Simulate Unconstrained Phenotypic Selection for 60 Generations</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Copy Founders</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Rudimentary Logbooks</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Simulation Main Loop</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Saving Results to a File</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-breeding-program-simulation-results-with-matplotlib">Visualizing Breeding Program Simulation Results with <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-true-breeding-values-tbvs">Visualizing True Breeding Values (TBVs)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-estimated-breeding-values-tbvs">Visualizing Estimated Breeding Values (TBVs)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-mean-expected-heterozygosity-meh">Visualizing Mean Expected Heterozygosity (MEH)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-true-breeding-value-standard-deviations">Visualizing True Breeding Value Standard Deviations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-estimated-breeding-value-standard-deviations">Visualizing Estimated Breeding Value Standard Deviations</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, Robert Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>