
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Multi-objective Genomic Selection &#8212; pybrops 1.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'standard_usage/examples/breeding_program_simulation/multiobjective_genomic_selection';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Constrained Bi-Trait Phenotypic Selection" href="constrained_bitrait_phenotypic_selection.html" />
    <link rel="prev" title="Constrained Single-Trait Phenotypic Selection" href="constrained_single_trait_phenotypic_selection.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Infrastructure Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../core_infrastructure/random_number_generation.html">Random Number Generation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Population Genetics Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_maps.html">Genetic Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genetic_map_functions.html">Genetic Map Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/genotype_matrices.html">Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/phased_genotype_matrices.html">Phased Genotype Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/breeding_value_matrices.html">Breeding Value Matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../population_genetics/coancestry_matrices.html">Coancestry Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Optimization Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_problems.html">Optimization Problems</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_solutions.html">Optimization Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization/optimization_algorithms.html">Optimization Algorithms</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Genomic Modeling Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/genomic_models.html">Genomic Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../genomic_modeling/variance_matrices.html">Variance Matrices</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Protocol Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/genotyping_protocols.html">Genotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/phenotyping_protocols.html">Phenotyping Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/breeding_value_protocols.html">Breeding Value Estimation Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/mating_protocols.html">Mating Protocols</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_protocols/selection_protocols.html">Selection Protocols</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Operator Modules Usage:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/initialization_operators.html">Initialization Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/parental_selection_operators.html">Parental Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/mating_operators.html">Mating Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/evaluation_operators.html">Evaluation Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/survivor_selection_operators.html">Survivor Selection Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../breeding_operators/logbooks.html">Logbooks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Breeding Program Simulation Examples:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../examples.html">Examples</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/biobjective_GEBV_pareto_frontier.html">Bi-Objective GEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/biobjective_wGEBV_pareto_frontier.html">Bi-Objective wGEBV selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/triobjective_OCS_pareto_frontier.html">Tri-Objective Optimal Contribution selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pareto_frontier_visualization/germplasm_panel_reduction.html">Germplasm panel reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="single_trait_phenotypic_selection.html">Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="constrained_single_trait_phenotypic_selection.html">Constrained Single-Trait Phenotypic Selection</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Bi-Trait Genotypic Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="constrained_bitrait_phenotypic_selection.html">Constrained Bi-Trait Phenotypic Selection</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../standard_usage.html" class="nav-link">Standard PyBrOpS Usage</a></li>
    
    
    <li class="breadcrumb-item"><a href="../examples.html" class="nav-link">Examples</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Multi-object...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="multi-objective-genomic-selection">
<h1>Multi-objective Genomic Selection<a class="headerlink" href="#multi-objective-genomic-selection" title="Link to this heading">#</a></h1>
<p>In this example, we demonstrate a simulation where we select individuals based on two traits. Instead of selecting individuals using an index-type selection method, we approximate the Pareto frontier for the selection of two traits simultaneously and select a single solution from this Pareto frontier to determine our final selection decision. The advantage of approximating a Pareto frontier is that it allows us to see our selection options and it allows us to identify selection solutions which lie along non-convex frontiers. The existence of non-convex frontiers is especially important in discrete optimization problems, such as where we desire to select a discrete subset of individuals. In non-convex scenarios, index-type selection methods such as the weight-sum method will fail to identify individuals which lie along the concave portions of a non-convex Pareto frontier.</p>
<section id="simulation-preliminaries">
<h2>Simulation Preliminaries<a class="headerlink" href="#simulation-preliminaries" title="Link to this heading">#</a></h2>
<section id="loading-required-modules-and-seeding-the-global-prng">
<h3>Loading Required Modules and Seeding the global PRNG<a class="headerlink" href="#loading-required-modules-and-seeding-the-global-prng" title="Link to this heading">#</a></h3>
<p>To begin, we import the various modules we will be used into the Python namespace. We also set the seed for our simulation so that we can replicate the results at a later time.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">pybrops</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.arch.RecurrentSelectionBreedingProgram</span> <span class="kn">import</span> <span class="n">RecurrentSelectionBreedingProgram</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.eval.EvaluationOperator</span> <span class="kn">import</span> <span class="n">EvaluationOperator</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.init.InitializationOperator</span> <span class="kn">import</span> <span class="n">InitializationOperator</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.log.Logbook</span> <span class="kn">import</span> <span class="n">Logbook</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.mate.MatingOperator</span> <span class="kn">import</span> <span class="n">MatingOperator</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.psel.ParentSelectionOperator</span> <span class="kn">import</span> <span class="n">ParentSelectionOperator</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.op.ssel.SurvivorSelectionOperator</span> <span class="kn">import</span> <span class="n">SurvivorSelectionOperator</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.bv.MeanPhenotypicBreedingValue</span> <span class="kn">import</span> <span class="n">MeanPhenotypicBreedingValue</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.gt.DenseUnphasedGenotyping</span> <span class="kn">import</span> <span class="n">DenseUnphasedGenotyping</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.mate.MatingProtocol</span> <span class="kn">import</span> <span class="n">MatingProtocol</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.mate.TwoWayCross</span> <span class="kn">import</span> <span class="n">TwoWayCross</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.mate.TwoWayDHCross</span> <span class="kn">import</span> <span class="n">TwoWayDHCross</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.pt.G_E_Phenotyping</span> <span class="kn">import</span> <span class="n">G_E_Phenotyping</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.sel.EstimatedBreedingValueSelection</span> <span class="kn">import</span> <span class="n">EstimatedBreedingValueSubsetSelection</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.sel.GenomicEstimatedBreedingValueSelection</span> <span class="kn">import</span> <span class="n">GenomicEstimatedBreedingValueSubsetSelection</span>
<span class="kn">from</span> <span class="nn">pybrops.breed.prot.sel.transfn</span> <span class="kn">import</span> <span class="n">trans_ndpt_to_vec_dist</span>
<span class="kn">from</span> <span class="nn">pybrops.model.gmod.DenseAdditiveLinearGenomicModel</span> <span class="kn">import</span> <span class="n">DenseAdditiveLinearGenomicModel</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.bvmat.DenseBreedingValueMatrix</span> <span class="kn">import</span> <span class="n">DenseBreedingValueMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.HaldaneMapFunction</span> <span class="kn">import</span> <span class="n">HaldaneMapFunction</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmap.StandardGeneticMap</span> <span class="kn">import</span> <span class="n">StandardGeneticMap</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DenseGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DenseGenotypeMatrix</span>
<span class="kn">from</span> <span class="nn">pybrops.popgen.gmat.DensePhasedGenotypeMatrix</span> <span class="kn">import</span> <span class="n">DensePhasedGenotypeMatrix</span>

<span class="c1"># seed python ``random`` and ``numpy.random`` simultaneously for reproducability</span>
<span class="n">pybrops</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">prng</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">17320132226</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="simulation-parameters">
<h3>Simulation Parameters<a class="headerlink" href="#simulation-parameters" title="Link to this heading">#</a></h3>
<p>Next, we define a couple of simulation parameter constants. These constants include the number of founders, number of QTL to simulate, number of burn-in generations, and number of generations for which to conduct simulations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">nfndr</span> <span class="o">=</span> <span class="mi">40</span>          <span class="c1"># number of random founders to select</span>
<span class="n">nqtl</span> <span class="o">=</span> <span class="mi">1000</span>         <span class="c1"># number of QTL</span>
<span class="n">qlen</span> <span class="o">=</span> <span class="mi">6</span>            <span class="c1"># breeding pipeline queue length</span>
<span class="n">ncross</span> <span class="o">=</span> <span class="mi">20</span>         <span class="c1"># number of cross configurations (20, 2-way crosses)</span>
<span class="n">nparent</span> <span class="o">=</span> <span class="mi">2</span>         <span class="c1"># number of parents per cross configuration (2-way crosses)</span>
<span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span>         <span class="c1"># number of times to perform cross configuration</span>
<span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">80</span>       <span class="c1"># number of progenies per cross attempt</span>
<span class="n">nrandmate</span> <span class="o">=</span> <span class="mi">10</span>      <span class="c1"># number of random intermatings prior to burn-in step</span>
<span class="n">nburnin</span> <span class="o">=</span> <span class="mi">10</span>        <span class="c1"># number of burnin generations</span>
<span class="n">ngen</span> <span class="o">=</span> <span class="mi">10</span>           <span class="c1"># number of simulation generations</span>
<span class="n">nrep</span> <span class="o">=</span> <span class="mi">1</span>            <span class="c1"># number of simulation replications</span>
<span class="n">fndr_heritability</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">])</span> <span class="c1"># heritability of founder lines</span>
<span class="n">t_cur</span> <span class="o">=</span> <span class="o">-</span><span class="n">nburnin</span>    <span class="c1"># current time for burn-in</span>
<span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span>           <span class="c1"># maximum time for burn-in</span>
<span class="n">nindiv_fam</span> <span class="o">=</span> <span class="mi">4</span>      <span class="c1"># number of individuals to select per family for within-family selection</span>
<span class="n">nmating_fam</span> <span class="o">=</span> <span class="mi">1</span>     <span class="c1"># number of times to perform cross configuration for within-family selection</span>
<span class="n">nprogeny_fam</span> <span class="o">=</span> <span class="mi">80</span>   <span class="c1"># number of progenies to generate for within-family selection</span>
</pre></div>
</div>
</section>
<section id="loading-genetic-map-data-from-a-text-file">
<h3>Loading Genetic Map Data from a Text File<a class="headerlink" href="#loading-genetic-map-data-from-a-text-file" title="Link to this heading">#</a></h3>
<p>Then, we load genetic map data from a CSV-like file. In this example, we are using the US NAM genetic map constructed by McMullen et al. in 2009. We create a <code class="docutils literal notranslate"><span class="pre">StandardGeneticMap</span></code> object using the <code class="docutils literal notranslate"><span class="pre">from_csv</span></code> class method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read genetic map from a CSV-like file</span>
<span class="n">gmap</span> <span class="o">=</span> <span class="n">StandardGeneticMap</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
    <span class="s2">&quot;McMullen_2009_US_NAM.gmap&quot;</span><span class="p">,</span>
    <span class="n">vrnt_chrgrp_col</span>     <span class="o">=</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="c1"># name of marker chromosome number column</span>
    <span class="n">vrnt_phypos_col</span>     <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="c1"># name of marker physical position column</span>
    <span class="n">vrnt_genpos_col</span>     <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>  <span class="c1"># name of marker genetic position column</span>
    <span class="n">vrnt_genpos_units</span>   <span class="o">=</span> <span class="s2">&quot;cM&quot;</span><span class="p">,</span>  <span class="c1"># units of the genetic position column (cM, M)</span>
    <span class="n">auto_group</span>          <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># whether to auto group chromosomes on load</span>
    <span class="n">auto_build_spline</span>   <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># whether to auto build interpolation spline on load</span>
    <span class="n">sep</span>                 <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>  <span class="c1"># separator for file format</span>
    <span class="n">header</span>              <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>     <span class="c1"># index of the header row</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="creating-a-genetic-map-function">
<h3>Creating a Genetic Map Function<a class="headerlink" href="#creating-a-genetic-map-function" title="Link to this heading">#</a></h3>
<p>After loading our genetic map data, we create a genetic map function object which will be used to calculate recombination probabilities for our simulations. We create a simple Haldane genetic map function using the <code class="docutils literal notranslate"><span class="pre">HaldaneMapFunction</span></code> class.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># use Haldane map function to calculate crossover probabilities</span>
<span class="n">gmapfn</span> <span class="o">=</span> <span class="n">HaldaneMapFunction</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="loading-genome-data-from-a-vcf-file">
<h3>Loading Genome Data from a VCF File<a class="headerlink" href="#loading-genome-data-from-a-vcf-file" title="Link to this heading">#</a></h3>
<p>Next, we load genetic marker data into PyBrOpS from a VCF file. We treat the file as being phased and create a <code class="docutils literal notranslate"><span class="pre">DensePhasedGenotypeMatrix</span></code> object from the VCF. In this example, we are using a subset of genetic markers from the Wisconsin Maize Diversity Panel, which is composed of 942 individuals. 2000 SNPs with a minor allele frequency greater than 0.2 have been randomly selected to keep the dataset small.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># read phased genetic markers from a vcf file</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span>
    <span class="s2">&quot;widiv_2000SNPs_q0.2_Q0.8.vcf.gz&quot;</span><span class="p">,</span> <span class="c1"># file name to load</span>
    <span class="n">auto_group_vrnt</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>            <span class="c1"># automatically sort and group variants</span>
<span class="p">)</span>
</pre></div>
</div>
<p>After loading the genetic markers, we interpolate the genetic map positions and the sequential marker crossover probabilities using the <code class="docutils literal notranslate"><span class="pre">interp_xoprob</span></code> method. We interpolate using the genetic map and genetic map function we have just created.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># interpolate genetic map positions</span>
<span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">interp_xoprob</span><span class="p">(</span><span class="n">gmap</span><span class="p">,</span> <span class="n">gmapfn</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="sampling-founders-and-markers-from-the-loaded-vcf-file">
<h3>Sampling founders and markers from the loaded VCF file<a class="headerlink" href="#sampling-founders-and-markers-from-the-loaded-vcf-file" title="Link to this heading">#</a></h3>
<p>Then, we’ll randomly sample 40 founders and 1000 markers from the genotypic data we have loaded.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># get indices of random selections</span>
<span class="c1"># randomly choose ``nfounder`` indices from ``ntaxa``</span>
<span class="n">sel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">nfndr</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># select founder individuals</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">sel</span><span class="p">)</span>

<span class="c1"># get indices of random markers</span>
<span class="c1"># randomly choose ``nqtl`` indices from ``nvrnt``</span>
<span class="n">sel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">nvrnt</span><span class="p">,</span> <span class="n">nqtl</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># select markers</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">select_vrnt</span><span class="p">(</span><span class="n">sel</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="constructing-a-bi-trait-genomic-model">
<h3>Constructing a Bi-Trait Genomic Model<a class="headerlink" href="#constructing-a-bi-trait-genomic-model" title="Link to this heading">#</a></h3>
<p>Next, we want to construct a true genomic model that will model two traits in negative covariance (pleiotropy) with a strictly additive genetic architecture. We construct a <code class="docutils literal notranslate"><span class="pre">DenseAdditiveLinearGenomicModel</span></code> object to do this.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create trait means (intercepts) for model</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">10.0</span><span class="p">,</span> <span class="mf">25.0</span><span class="p">]],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>

<span class="c1"># create marker effects from MVN distribution for model:</span>
<span class="c1"># 1) marker effects have mean zero</span>
<span class="c1"># 2) marker effects have a covariance structure with negative marker covariance</span>
<span class="c1">#    meaning that markers are pleiotropic and competing in nature.</span>
<span class="c1"># marker effect matrix is shape (nvrnt,2)</span>
<span class="n">mkreffect</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multivariate_normal</span><span class="p">(</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]),</span>
    <span class="n">cov</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
        <span class="p">[</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.4</span><span class="p">],</span>
        <span class="p">[</span><span class="o">-</span><span class="mf">0.4</span><span class="p">,</span>  <span class="mf">1.0</span><span class="p">]</span>
    <span class="p">]),</span>
    <span class="n">size</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">nvrnt</span>
<span class="p">)</span>

<span class="c1"># create trait names for model:</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;syn1&quot;</span><span class="p">,</span><span class="s2">&quot;syn2&quot;</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">object</span><span class="p">)</span>

<span class="c1"># create the true genomic model</span>
<span class="n">algmod_true</span> <span class="o">=</span> <span class="n">DenseAdditiveLinearGenomicModel</span><span class="p">(</span>
    <span class="n">beta</span>   <span class="o">=</span> <span class="n">beta</span><span class="p">,</span>      <span class="c1"># model intercepts</span>
    <span class="n">u_misc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>      <span class="c1"># miscellaneous random effects</span>
    <span class="n">u_a</span>    <span class="o">=</span> <span class="n">mkreffect</span><span class="p">,</span> <span class="c1"># random marker effects</span>
    <span class="n">trait</span>  <span class="o">=</span> <span class="n">trait</span><span class="p">,</span>     <span class="c1"># trait names</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="constructing-the-founder-population">
<h2>Constructing the Founder Population<a class="headerlink" href="#constructing-the-founder-population" title="Link to this heading">#</a></h2>
<p>In this next series of steps, we’ll create a founder population by randomly intermating 40 selected founders for 10 generations. Following random intermating, we’ll create 6 cohorts of doubled haploid individuals, each cohort representing progenies produced in a single year. In this simulation, the oldest 3 cohorts will be aggregated into a <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> population to serve as selection candidates. The youngest 3 cohorts will be “in the pipeline” and will represent doubled haploid progenies which are being made. From the <code class="docutils literal notranslate"><span class="pre">&quot;main&quot;</span></code> population, the top 5% of individuals from each family will be selected to form a <code class="docutils literal notranslate"><span class="pre">&quot;cand&quot;</span></code> population. It is from this <code class="docutils literal notranslate"><span class="pre">&quot;cand&quot;</span></code> population that parents will be selected. The burn-in segment of the breeding program will proceed for 10 generations. After this point is reached, the simulated populations are ready for use in the main simulation.</p>
<section id="randomly-intermate-founders-for-nrandmate-generations">
<h3>Randomly Intermate founders for <code class="docutils literal notranslate"><span class="pre">nrandmate</span></code> Generations<a class="headerlink" href="#randomly-intermate-founders-for-nrandmate-generations" title="Link to this heading">#</a></h3>
<p>First, we start by randomly intermating our 40 founders for 10 generations. We randomly intermate in the heterozygous state, generating one progeny per cross to maximize the number of recombinations between founder chromosomes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create 2-way cross object</span>
<span class="n">mate2way</span> <span class="o">=</span> <span class="n">TwoWayCross</span><span class="p">()</span>

<span class="c1"># randomly select and pair founders</span>
<span class="n">xconfig</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">nfndr</span><span class="p">,</span> <span class="n">nfndr</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">nfndr</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

<span class="c1"># randomly intermate ``nfndr`` founders to create initial hybrids</span>
<span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
    <span class="n">pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="p">,</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="p">,</span>
    <span class="n">nmating</span> <span class="o">=</span> <span class="n">nmating</span><span class="p">,</span>
    <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">nprogeny</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># randomly intermate for ``nrandmate`` generations</span>
<span class="c1"># each individual in the population is randomly mated with another individual</span>
<span class="c1"># and creates a single progeny so that the population size is held constant</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">nrandmate</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
    <span class="c1"># randomly select and pair ``ntaxa`` parents</span>
    <span class="n">ntaxa</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">ntaxa</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">ntaxa</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">)</span>
    <span class="n">xconfig</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">xconfig</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="c1"># randomly intermate ``ntaxa`` parents</span>
    <span class="n">fndr_pgmat</span> <span class="o">=</span> <span class="n">mate2way</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
        <span class="n">pgmat</span> <span class="o">=</span> <span class="n">fndr_pgmat</span><span class="p">,</span>
        <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfig</span><span class="p">,</span>
        <span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="construct-bootstrap-cohort-structure-to-begin-burnin">
<h2>Construct bootstrap cohort structure to begin burnin<a class="headerlink" href="#construct-bootstrap-cohort-structure-to-begin-burnin" title="Link to this heading">#</a></h2>
<section id="create-mating-protocols-for-burn-in">
<h3>Create Mating Protocols for Burn-In<a class="headerlink" href="#create-mating-protocols-for-burn-in" title="Link to this heading">#</a></h3>
<p>Next, we make a 2-way DH cross protocol for our burn-in stage.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a 2-way DH cross object</span>
<span class="n">mate2waydh</span> <span class="o">=</span> <span class="n">TwoWayDHCross</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="create-genotyping-protocols-for-burn-in">
<h3>Create Genotyping Protocols for Burn-In<a class="headerlink" href="#create-genotyping-protocols-for-burn-in" title="Link to this heading">#</a></h3>
<p>We also create a genotyping protocol that converts phased genotypes to unphased genotypes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a genotyping protocol</span>
<span class="n">gtprot</span> <span class="o">=</span> <span class="n">DenseUnphasedGenotyping</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="create-phenotyping-protocols-for-burn-in">
<h3>Create Phenotyping Protocols for Burn-In<a class="headerlink" href="#create-phenotyping-protocols-for-burn-in" title="Link to this heading">#</a></h3>
<p>To simulate phenotypes, we create a simple <span class="math notranslate nohighlight">\(G+E\)</span> phenotyping protocol that phenotypes individuals in 4 locations, 1 replication each location. We use the founding population we have just created with random mating to set the narrow sense heritability at the single plot level to 0.4 and 0.6 for our traits.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a phenotyping protocol using the true genomic model</span>
<span class="n">ptprot</span> <span class="o">=</span> <span class="n">G_E_Phenotyping</span><span class="p">(</span><span class="n">algmod_true</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># set the heritability using the founder population</span>
<span class="n">ptprot</span><span class="o">.</span><span class="n">set_h2</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">]),</span> <span class="n">fndr_pgmat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-breeing-value-estimation-protocols-for-burn-in">
<h3>Create Breeing Value Estimation Protocols for Burn-In<a class="headerlink" href="#create-breeing-value-estimation-protocols-for-burn-in" title="Link to this heading">#</a></h3>
<p>Next, we create a breeding value estimation protocol that uses simple means to estimate breeding value.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a breeding value estimation protocol</span>
<span class="n">bvprot</span> <span class="o">=</span> <span class="n">MeanPhenotypicBreedingValue</span><span class="p">(</span><span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa_grp&quot;</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-a-within-family-selection-helper-function">
<h3>Create a Within-Family Selection Helper Function<a class="headerlink" href="#create-a-within-family-selection-helper-function" title="Link to this heading">#</a></h3>
<p>We create a function to the best select individuals within their respective families. Since we have two traits we are selecting on, we select on a simple sum of their breeding values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define function to do within family selection based on yield</span>
<span class="k">def</span> <span class="nf">within_family_selection</span><span class="p">(</span><span class="n">bvmat</span><span class="p">:</span> <span class="n">DenseBreedingValueMatrix</span><span class="p">,</span> <span class="n">nindiv</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Select individuals within a family based on the sum of their breeding values.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bvmat : DenseBreedingValueMatrix</span>
<span class="sd">        Input breeding value matrix from which to calculate values.</span>
<span class="sd">    nindiv : int</span>
<span class="sd">        Number of individuals to select from each family.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    indices : numpy.ndarray</span>
<span class="sd">        An array of array indices indicating which individuals are to be selected.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">order</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">bvmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">)</span>
    <span class="n">value</span> <span class="o">=</span> <span class="n">bvmat</span><span class="o">.</span><span class="n">mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bvmat</span><span class="o">.</span><span class="n">taxa_grp</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">bvmat</span><span class="o">.</span><span class="n">taxa_grp</span> <span class="o">==</span> <span class="n">group</span>
        <span class="n">tmp_order</span> <span class="o">=</span> <span class="n">order</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">tmp_value</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="n">value_argsort</span> <span class="o">=</span> <span class="n">tmp_value</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>
        <span class="n">ix</span> <span class="o">=</span> <span class="n">value_argsort</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="n">nindiv</span><span class="p">]</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_order</span><span class="p">[</span><span class="n">ix</span><span class="p">])</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">indices</span>
</pre></div>
</div>
</section>
<section id="create-a-cohort-construction-helper-function">
<h3>Create a Cohort Construction Helper function<a class="headerlink" href="#create-a-cohort-construction-helper-function" title="Link to this heading">#</a></h3>
<p>We define a function to help us construct population cohorts for our simulation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># define a helper function to help make cohorts of individuals</span>
<span class="k">def</span> <span class="nf">cohort</span><span class="p">(</span>
        <span class="n">mateprot</span><span class="p">:</span> <span class="n">MatingProtocol</span><span class="p">,</span>
        <span class="n">fndr_pgmat</span><span class="p">:</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">,</span>
        <span class="n">ncross</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nparent</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nmating</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">nprogeny</span><span class="p">:</span> <span class="nb">int</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Randomly sample individuals from a founder population</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># sample indicies of individuals and reshape for input into mating protocol</span>
    <span class="n">xconfix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="o">.</span><span class="n">ntaxa</span><span class="p">,</span> <span class="n">ncross</span> <span class="o">*</span> <span class="n">nparent</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">xconfig</span> <span class="o">=</span> <span class="n">xconfix</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">ncross</span><span class="p">,</span> <span class="n">nparent</span><span class="p">)</span>
    <span class="c1"># mate individuals</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">mateprot</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span><span class="n">fndr_pgmat</span><span class="p">,</span> <span class="n">xconfig</span><span class="p">,</span> <span class="n">nmating</span><span class="p">,</span> <span class="n">nprogeny</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span>
</pre></div>
</div>
</section>
<section id="create-cohort-structure">
<h3>Create Cohort Structure<a class="headerlink" href="#create-cohort-structure" title="Link to this heading">#</a></h3>
<p>In this next step, we create a bootstrap cohort structure to satisfy the simulation description above.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">################ Build founder populations #################</span>
<span class="n">fndr_genome</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;queue&quot;</span><span class="p">:[]}</span>
<span class="n">fndr_geno</span> <span class="o">=</span>   <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;queue&quot;</span><span class="p">:[]}</span>
<span class="n">fndr_pheno</span> <span class="o">=</span>  <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">}</span>
<span class="n">fndr_bval</span> <span class="o">=</span>   <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span>      <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">}</span>
<span class="n">fndr_gmod</span> <span class="o">=</span>   <span class="p">{</span><span class="s2">&quot;cand&quot;</span><span class="p">:</span><span class="n">algmod_true</span><span class="p">,</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span><span class="n">algmod_true</span><span class="p">,</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span><span class="n">algmod_true</span><span class="p">}</span>

<span class="c1"># fill queue with cohort genomes derived from randomly mating the founders</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cohort</span><span class="p">(</span><span class="n">mate2waydh</span><span class="p">,</span> <span class="n">fndr_pgmat</span><span class="p">,</span> <span class="n">ncross</span><span class="p">,</span> <span class="n">nparent</span><span class="p">,</span> <span class="n">nmating</span><span class="p">,</span> <span class="n">nprogeny</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">qlen</span><span class="p">)]</span>

<span class="c1"># construct the main population genomes from the first three cohorts in the queue</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DensePhasedGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>

<span class="c1"># genotype individuals to fill the genotyping queue</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]]</span>

<span class="c1"># construct the main population genotypes from the first three cohorts in the queue</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">DenseGenotypeMatrix</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>

<span class="c1"># phenotype the main population</span>
<span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>

<span class="c1"># calculate breeding values for the main population</span>
<span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">fndr_pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>

<span class="c1"># # calculate indices for within family selection to get parental candidates</span>
<span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">nindiv_fam</span><span class="p">)</span> <span class="c1"># select top 5%</span>

<span class="c1"># # select parental candidates</span>
<span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
<span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
<span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">fndr_bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span> <span class="c1"># breeding values have been recentered and rescaled</span>
</pre></div>
</div>
</section>
</section>
<section id="define-breeding-program-operators">
<h2>Define breeding program operators<a class="headerlink" href="#define-breeding-program-operators" title="Link to this heading">#</a></h2>
<p>In this section, we define several breeding program operators which will be used in the universal breeding program algorithm.</p>
<section id="define-parent-selection-operator-for-initialization-operator">
<h3>Define parent selection operator for initialization operator<a class="headerlink" href="#define-parent-selection-operator-for-initialization-operator" title="Link to this heading">#</a></h3>
<p>We define the parent selection operator for the initialization operator. This operator selects parents using a 50/50 weighting of both traits based on the normalized Pareto frontier.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyInitParentSelectionOperator</span><span class="p">(</span><span class="n">ParentSelectionOperator</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Custom Parent Selection Operator class for our custom Initialization Operator</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># create parental selection protocol that selects individuals based on</span>
        <span class="c1"># their estimated breeding values; this is phenotypic selection</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pselprot</span> <span class="o">=</span> <span class="n">EstimatedBreedingValueSubsetSelection</span><span class="p">(</span>
            <span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ncross</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
            <span class="n">nparent</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">80</span><span class="p">,</span>
            <span class="n">nobj</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">ndset_wt</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
            <span class="n">ndset_trans</span> <span class="o">=</span> <span class="n">trans_ndpt_to_vec_dist</span><span class="p">,</span>
            <span class="n">ndset_trans_kwargs</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;objfn_wt&quot;</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]),</span>    <span class="c1"># all objectives maximizing</span>
                <span class="s2">&quot;wt&quot;</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>           <span class="c1"># 1/2; equal weight to all</span>
            <span class="p">},</span>
        <span class="p">)</span>
    <span class="k">def</span> <span class="nf">pselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">mcfg</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pselprot</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">pgmat</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">gmat</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">ptdf</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">bvmat</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">gpmod</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">t_cur</span> <span class="o">=</span> <span class="n">t_cur</span><span class="p">,</span>
            <span class="n">t_max</span> <span class="o">=</span> <span class="n">t_max</span><span class="p">,</span>
            <span class="n">miscout</span> <span class="o">=</span> <span class="n">miscout</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">mcfg</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-mating-operator-for-initialization-operator">
<h3>Define mating operator for initialization operator<a class="headerlink" href="#define-mating-operator-for-initialization-operator" title="Link to this heading">#</a></h3>
<p>We define the mating operator for the initialization operator. This operator simply creates DH progenies from 2-way crosses.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyInitMatingOperator</span><span class="p">(</span><span class="n">MatingOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pcnt</span><span class="p">,</span> <span class="n">fcnt</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">MyInitMatingOperator</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mprot</span> <span class="o">=</span> <span class="n">TwoWayDHCross</span><span class="p">(</span>
            <span class="n">progeny_counter</span> <span class="o">=</span> <span class="n">pcnt</span><span class="p">,</span>
            <span class="n">family_counter</span> <span class="o">=</span> <span class="n">fcnt</span>
        <span class="p">)</span>
    <span class="k">def</span> <span class="nf">mate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mcfg</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">progeny</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mprot</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
            <span class="n">pgmat</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pgmat</span><span class="p">,</span>
            <span class="n">xconfig</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">xconfig</span><span class="p">,</span>
            <span class="n">nmating</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nmating</span><span class="p">,</span>
            <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nprogeny</span><span class="p">,</span>
            <span class="n">miscout</span> <span class="o">=</span> <span class="n">miscout</span><span class="p">,</span>
            <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">progeny</span><span class="p">)</span>                 <span class="c1"># add progeny to queue in genome dict</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-evaluation-operator-for-initialization-operator">
<h3>Define evaluation operator for initialization operator<a class="headerlink" href="#define-evaluation-operator-for-initialization-operator" title="Link to this heading">#</a></h3>
<p>We define the evaluation operator for the initialization operator. Genotypes individuals, tests them in 4 locations, and calculates breeding values using the mean.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyInitEvaluationOperator</span><span class="p">(</span><span class="n">EvaluationOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gpmod</span><span class="p">,</span> <span class="n">var_err</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gtprot</span> <span class="o">=</span> <span class="n">DenseUnphasedGenotyping</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ptprot</span> <span class="o">=</span> <span class="n">G_E_Phenotyping</span><span class="p">(</span><span class="n">gpmod</span> <span class="o">=</span> <span class="n">gpmod</span><span class="p">,</span> <span class="n">nenv</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">var_err</span> <span class="o">=</span> <span class="n">var_err</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bvprot</span> <span class="o">=</span> <span class="n">MeanPhenotypicBreedingValue</span><span class="p">(</span><span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa_grp&quot;</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>         <span class="c1"># genotype incoming inbreds</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                                                  <span class="c1"># remove oldest inbreds</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                                                    <span class="c1"># remove oldest inbreds</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>   <span class="c1"># construct main population using 3 oldest cohorts</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>         <span class="c1"># construct main population using 3 oldest cohorts</span>
        <span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>                   <span class="c1"># phenotype main population</span>
        <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>        <span class="c1"># estimate breeding values after phenotyping</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-survivor-selection-operator-for-initialization-operator">
<h3>Define survivor selection operator for initialization operator<a class="headerlink" href="#define-survivor-selection-operator-for-initialization-operator" title="Link to this heading">#</a></h3>
<p>We define the surivor selection operator for the initialization operator. This operator selects the top 5% of individuals within families based on the sum of their breeding values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyInitSurvivorSelectionOperator</span><span class="p">(</span><span class="n">SurvivorSelectionOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nindiv_fam</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nindiv_fam</span> <span class="o">=</span> <span class="n">nindiv_fam</span>
    <span class="k">def</span> <span class="nf">sselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="c1"># calculate indices for within family selection to get parental candidates</span>
        <span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">nindiv_fam</span><span class="p">)</span> <span class="c1"># select top 5%</span>
        <span class="c1"># select parental candidates</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
        <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span> <span class="c1"># breeding values have been recentered and rescaled</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-initialization-operator-for-universal-breeding-algorithm">
<h3>Define initialization operator for universal breeding algorithm<a class="headerlink" href="#define-initialization-operator-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>We define the initialization operator which conducts the simulation burn-in using the operators we defined above.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyInitializationOperator</span><span class="p">(</span><span class="n">InitializationOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fndr_genome</span><span class="p">,</span> <span class="n">fndr_geno</span><span class="p">,</span> <span class="n">fndr_pheno</span><span class="p">,</span> <span class="n">fndr_bval</span><span class="p">,</span> <span class="n">fndr_gmod</span><span class="p">,</span> <span class="n">pselop</span><span class="p">,</span> <span class="n">mateop</span><span class="p">,</span> <span class="n">evalop</span><span class="p">,</span> <span class="n">sselop</span><span class="p">,</span> <span class="n">burnin</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">MyInitializationOperator</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">fndr_genome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">geno</span> <span class="o">=</span> <span class="n">fndr_geno</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span> <span class="o">=</span> <span class="n">fndr_pheno</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bval</span> <span class="o">=</span> <span class="n">fndr_bval</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span> <span class="o">=</span> <span class="n">fndr_gmod</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pselop</span> <span class="o">=</span> <span class="n">pselop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mateop</span> <span class="o">=</span> <span class="n">mateop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">evalop</span> <span class="o">=</span> <span class="n">evalop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sselop</span> <span class="o">=</span> <span class="n">sselop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">burnin</span> <span class="o">=</span> <span class="n">burnin</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span> <span class="o">=</span> <span class="o">-</span><span class="n">burnin</span>
    <span class="k">def</span> <span class="nf">initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">burnin</span><span class="p">):</span> <span class="c1"># iterate through main breeding loop for burnin generations</span>
            <span class="n">mcfg</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pselop</span><span class="o">.</span><span class="n">pselect</span><span class="p">(</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
                <span class="n">geno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span>
                <span class="n">pheno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">bval</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span>
                <span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span><span class="p">,</span>
                <span class="n">t_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span><span class="p">,</span>
                <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mateop</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
                <span class="n">mcfg</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">,</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
                <span class="n">geno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span>
                <span class="n">pheno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">bval</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span>
                <span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span><span class="p">,</span>
                <span class="n">t_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span><span class="p">,</span>
                <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">evalop</span><span class="o">.</span><span class="n">evaluate</span><span class="p">(</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
                <span class="n">geno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span>
                <span class="n">pheno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">bval</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span>
                <span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span><span class="p">,</span>
                <span class="n">t_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span><span class="p">,</span>
                <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sselop</span><span class="o">.</span><span class="n">sselect</span><span class="p">(</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
                <span class="n">geno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span>
                <span class="n">pheno</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">bval</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span>
                <span class="n">gmod</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span><span class="p">,</span>
                <span class="n">t_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span><span class="p">,</span>
                <span class="n">t_max</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">t_cur</span> <span class="o">+=</span> <span class="mi">1</span>     <span class="c1"># increment time variables</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Burn-in generation </span><span class="si">{0}</span><span class="s2"> of </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">_</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">burnin</span><span class="p">))</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bval</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-parent-selection-operator-for-universal-breeding-algorithm">
<h3>Define parent selection operator for universal breeding algorithm<a class="headerlink" href="#define-parent-selection-operator-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>We define the parent selection operator for the main simulation. This operator selects the best individuals based on their GEVBs using a 50/50 weighting from the scaled Pareto frontier of possible selection decisions for both traits.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyParentSelectionOperator</span><span class="p">(</span><span class="n">ParentSelectionOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pselprot</span> <span class="o">=</span> <span class="n">GenomicEstimatedBreedingValueSubsetSelection</span><span class="p">(</span>
            <span class="n">ntrait</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ncross</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
            <span class="n">nparent</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">nmating</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">nprogeny</span> <span class="o">=</span> <span class="mi">80</span><span class="p">,</span>
            <span class="n">nobj</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">ndset_wt</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
            <span class="n">ndset_trans</span> <span class="o">=</span> <span class="n">trans_ndpt_to_vec_dist</span><span class="p">,</span>
            <span class="n">ndset_trans_kwargs</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;objfn_wt&quot;</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]),</span>    <span class="c1"># all objectives maximizing</span>
                <span class="s2">&quot;wt&quot;</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>           <span class="c1"># 1/2; equal weight to all</span>
            <span class="p">},</span>
        <span class="p">)</span>
    <span class="k">def</span> <span class="nf">pselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">mcfg</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pselprot</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">pgmat</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">gmat</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">ptdf</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">bvmat</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">gpmod</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span>
            <span class="n">t_cur</span> <span class="o">=</span> <span class="n">t_cur</span><span class="p">,</span>
            <span class="n">t_max</span> <span class="o">=</span> <span class="n">t_max</span><span class="p">,</span>
            <span class="n">miscout</span> <span class="o">=</span> <span class="n">miscout</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">mcfg</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-mating-operator-for-universal-breeding-algorithm">
<h3>Define mating operator for universal breeding algorithm<a class="headerlink" href="#define-mating-operator-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>We define the mating operator for the main simulation. This operator simply mates individuals in 2-way crosses and derives DH progenies from the crosses.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyMatingOperator</span><span class="p">(</span><span class="n">MatingOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pcnt</span><span class="p">,</span> <span class="n">fcnt</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mprot</span> <span class="o">=</span> <span class="n">TwoWayDHCross</span><span class="p">(</span>
            <span class="n">progeny_counter</span> <span class="o">=</span> <span class="n">pcnt</span><span class="p">,</span>
            <span class="n">family_counter</span> <span class="o">=</span> <span class="n">fcnt</span>
        <span class="p">)</span>
    <span class="k">def</span> <span class="nf">mate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mcfg</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="c1"># mate parents</span>
        <span class="n">progeny</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mprot</span><span class="o">.</span><span class="n">mate</span><span class="p">(</span>
            <span class="n">pgmat</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pgmat</span><span class="p">,</span>
            <span class="n">xconfig</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">xconfig</span><span class="p">,</span>
            <span class="n">nmating</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nmating</span><span class="p">,</span>
            <span class="n">nprogeny</span> <span class="o">=</span> <span class="n">mcfg</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nprogeny</span><span class="p">,</span>
            <span class="n">miscout</span> <span class="o">=</span> <span class="n">miscout</span><span class="p">,</span>
            <span class="n">nself</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">progeny</span><span class="p">)</span>                 <span class="c1"># add progeny to queue in genome dict</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-evaluation-operator-for-universal-breeding-algorithm">
<h3>Define evaluation operator for universal breeding algorithm<a class="headerlink" href="#define-evaluation-operator-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>We define the evaluation operator for the main simulation. This operator genotypes individuals, tests them in 4 locations, and calculates their breeding values using simple means.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyEvaluationOperator</span><span class="p">(</span><span class="n">EvaluationOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gpmod</span><span class="p">,</span> <span class="n">var_err</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gtprot</span> <span class="o">=</span> <span class="n">DenseUnphasedGenotyping</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ptprot</span> <span class="o">=</span> <span class="n">G_E_Phenotyping</span><span class="p">(</span><span class="n">gpmod</span> <span class="o">=</span> <span class="n">gpmod</span><span class="p">,</span> <span class="n">nenv</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">var_err</span> <span class="o">=</span> <span class="n">var_err</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bvprot</span> <span class="o">=</span> <span class="n">MeanPhenotypicBreedingValue</span><span class="p">(</span><span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa_grp&quot;</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gtprot</span><span class="o">.</span><span class="n">genotype</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>         <span class="c1"># genotype incoming inbreds</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                                                  <span class="c1"># remove oldest inbreds</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>                                                    <span class="c1"># remove oldest inbreds</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>   <span class="c1"># construct main population using 3 oldest cohorts</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">concat_taxa</span><span class="p">(</span><span class="n">geno</span><span class="p">[</span><span class="s2">&quot;queue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>         <span class="c1"># construct main population using 3 oldest cohorts</span>
        <span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptprot</span><span class="o">.</span><span class="n">phenotype</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>                   <span class="c1"># phenotype main population</span>
        <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bvprot</span><span class="o">.</span><span class="n">estimate</span><span class="p">(</span><span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>      <span class="c1"># estimate breeding values after phenotyping</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-survivor-selection-operator-for-universal-breeding-algorithm">
<h3>Define survivor selection operator for universal breeding algorithm<a class="headerlink" href="#define-survivor-selection-operator-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>We define the survivor selection operator for the main simulation. This operator selects the top 5% of individuals within each family from the main breeding population and puts these selected individuals into the parental candidate breeding population.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MySurvivorSelectionOperator</span><span class="p">(</span><span class="n">SurvivorSelectionOperator</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nindiv_fam</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nindiv_fam</span> <span class="o">=</span> <span class="n">nindiv_fam</span>
    <span class="k">def</span> <span class="nf">sselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">miscout</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="c1"># calculate indices for within family selection to get parental candidates</span>
        <span class="n">ix</span> <span class="o">=</span> <span class="n">within_family_selection</span><span class="p">(</span><span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">nindiv_fam</span><span class="p">)</span> <span class="c1"># select top 5%</span>
        <span class="c1"># select parental candidates</span>
        <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
        <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span>
        <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">select_taxa</span><span class="p">(</span><span class="n">ix</span><span class="p">)</span> <span class="c1"># breeding values have been recentered and rescaled</span>
        <span class="k">return</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span>
</pre></div>
</div>
</section>
<section id="define-logbook-to-collect-simulation-data">
<h3>Define logbook to collect simulation data<a class="headerlink" href="#define-logbook-to-collect-simulation-data" title="Link to this heading">#</a></h3>
<p>Finally, we define a logbook class to record numerous simulation metrics for our analysis.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyLogbook</span><span class="p">(</span><span class="n">Logbook</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">MyLogbook</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reset</span><span class="p">()</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Logbook data.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span>
    <span class="nd">@data</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set logbook data.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">data_frontier</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Logbook frontier data.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data_frontier</span>
    <span class="nd">@data_frontier</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">data_frontier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set logbook frontier data.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data_frontier</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">rep</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Replication number.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_rep</span>
    <span class="nd">@rep</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">rep</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set replication number.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rep</span> <span class="o">=</span> <span class="n">value</span>
    <span class="k">def</span> <span class="nf">log_initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">gpmod</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span>
        <span class="n">cand_bval_true</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">gebv</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="n">main_bval_true</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">gebv</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rep</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;t_cur&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_cur</span><span class="p">)</span>
        <span class="c1">################ candidate mean expected heterozygosity ################</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_meh&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
        <span class="c1">########################### candidate means ############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">######################### candidate true means #########################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">cand_bval_true</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#################### candidate standard deviations #####################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################## candidate true standard deviations ##################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">cand_bval_true</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">############### candidate true additive genetic variance ###############</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_A</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_A_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_A_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################ candidate true additive genic variance ################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_a</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_a_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_a_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">##################### candidate true bulmer ratio ######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">bulmer</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_bulmer_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_bulmer_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################# candidate true upper selection limit #################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">usl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_usl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_usl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################# candidate true lower selection limit #################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">lsl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_lsl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_lsl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">########################################################################</span>
        <span class="c1">################## main mean expected heterozygosity ###################</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_meh&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
        <span class="c1">############################## main means ##############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">########################### main true means ############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">main_bval_true</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">####################### main standard deviations #######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#################### main true standard deviations #####################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">main_bval_true</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">##################### main true genetic variances ######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_A</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_A_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_A_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">###################### main true genic variances #######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_a</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_a_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_a_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">####################### main true bulmer ratios ########################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">bulmer</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_bulmer_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_bulmer_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################### main true lower selection limits ###################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">usl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_usl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_usl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################### main true lower selection limits ###################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">lsl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_lsl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_lsl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">def</span> <span class="nf">log_pselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mcfg</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rep</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;t_cur&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_cur</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;mosoln&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;frontier&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mosoln&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;frontier&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">log_mate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">pass</span>
    <span class="k">def</span> <span class="nf">log_evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">pass</span>
    <span class="k">def</span> <span class="nf">log_sselect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">geno</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">bval</span><span class="p">,</span> <span class="n">gmod</span><span class="p">,</span> <span class="n">t_cur</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">gpmod</span> <span class="o">=</span> <span class="n">gmod</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">]</span>
        <span class="n">cand_bval_true</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">gebv</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="n">main_bval_true</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">gebv</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rep</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;t_cur&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_cur</span><span class="p">)</span>
        <span class="c1">################ candidate mean expected heterozygosity ################</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_meh&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
        <span class="c1">########################### candidate means ############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">######################### candidate true means #########################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">cand_bval_true</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#################### candidate standard deviations #####################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################## candidate true standard deviations ##################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">cand_bval_true</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">############### candidate true additive genetic variance ###############</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_A</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_A_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_A_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################ candidate true additive genic variance ################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_a</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_a_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_var_a_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">##################### candidate true bulmer ratio ######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">bulmer</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_bulmer_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_bulmer_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################# candidate true upper selection limit #################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">usl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_usl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_usl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################# candidate true lower selection limit #################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">lsl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;cand&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_lsl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cand_true_lsl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">########################################################################</span>
        <span class="c1">################## main mean expected heterozygosity ###################</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_meh&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">meh</span><span class="p">())</span>
        <span class="c1">############################## main means ##############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">########################### main true means ############################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">main_bval_true</span><span class="o">.</span><span class="n">tmean</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">####################### main standard deviations #######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">bval</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#################### main true standard deviations #####################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">main_bval_true</span><span class="o">.</span><span class="n">tstd</span><span class="p">(</span><span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_std_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_std_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">##################### main true genetic variances ######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_A</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_A_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_A_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">###################### main true genic variances #######################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">var_a</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_a_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_var_a_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">####################### main true bulmer ratios ########################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">bulmer</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_bulmer_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_bulmer_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################### main true lower selection limits ###################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">usl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_usl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_usl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">################### main true lower selection limits ###################</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">gpmod</span><span class="o">.</span><span class="n">lsl</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">],</span> <span class="n">unscale</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_lsl_syn1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;main_true_lsl_syn2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">def</span> <span class="nf">reset</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;rep&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;t_cur&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_meh&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_mean_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_mean_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_mean_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_mean_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_std_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_std_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_std_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_std_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_var_A_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_var_A_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_var_a_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_var_a_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_bulmer_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_bulmer_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_usl_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_usl_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_lsl_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;cand_true_lsl_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_meh&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_mean_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_mean_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_mean_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_mean_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_std_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_std_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_std_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_std_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_var_A_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_var_A_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_var_a_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_var_a_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_bulmer_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_bulmer_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_usl_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_usl_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_lsl_syn1&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;main_true_lsl_syn2&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;rep&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;t_cur&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;frontier&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rep</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
        <span class="n">out_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
        <span class="n">out_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">write_frontier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
        <span class="n">out_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_pandas_frontier</span><span class="p">()</span>
        <span class="n">out_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">to_pandas</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">out_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out_df</span>
    <span class="k">def</span> <span class="nf">to_pandas_frontier</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tmp_df_ls</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;frontier&quot;</span><span class="p">])):</span>
            <span class="n">tmp_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_frontier</span><span class="p">[</span><span class="s2">&quot;frontier&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">soln_obj</span><span class="p">,</span>
                <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;syn1&quot;</span><span class="p">,</span> <span class="s2">&quot;syn2&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="n">tmp_df</span><span class="p">[</span><span class="s2">&quot;t_cur&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;t_cur&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>
            <span class="n">tmp_df_ls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_df</span><span class="p">)</span>
        <span class="n">out_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">tmp_df_ls</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out_df</span>
</pre></div>
</div>
</section>
</section>
<section id="simulation-setup">
<h2>Simulation Setup<a class="headerlink" href="#simulation-setup" title="Link to this heading">#</a></h2>
<section id="create-operators-used-by-initialization-operator">
<h3>Create operators used by initialization operator<a class="headerlink" href="#create-operators-used-by-initialization-operator" title="Link to this heading">#</a></h3>
<p>After defining our operator classes, we construct operators for our initialization operator to use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">init_pselop</span> <span class="o">=</span> <span class="n">MyInitParentSelectionOperator</span><span class="p">()</span>
<span class="n">init_mateop</span> <span class="o">=</span> <span class="n">MyInitMatingOperator</span><span class="p">(</span><span class="n">mate2waydh</span><span class="o">.</span><span class="n">progeny_counter</span><span class="p">,</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">family_counter</span><span class="p">)</span>
<span class="n">init_evalop</span> <span class="o">=</span> <span class="n">MyInitEvaluationOperator</span><span class="p">(</span><span class="n">algmod_true</span><span class="p">,</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">var_err</span><span class="p">)</span>
<span class="n">init_sselop</span> <span class="o">=</span> <span class="n">MyInitSurvivorSelectionOperator</span><span class="p">(</span><span class="n">nindiv_fam</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-main-operators-for-universal-breeding-algorithm">
<h3>Create main operators for universal breeding algorithm<a class="headerlink" href="#create-main-operators-for-universal-breeding-algorithm" title="Link to this heading">#</a></h3>
<p>Next, we construct initialization, parent selection, mating, evaluation, survivor selection, and logbook objects for the universal breeding program algorithm.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># construct operators</span>
<span class="n">initop</span> <span class="o">=</span> <span class="n">MyInitializationOperator</span><span class="p">(</span>
    <span class="n">fndr_genome</span> <span class="o">=</span> <span class="n">fndr_genome</span><span class="p">,</span>
    <span class="n">fndr_geno</span> <span class="o">=</span> <span class="n">fndr_geno</span><span class="p">,</span>
    <span class="n">fndr_pheno</span> <span class="o">=</span> <span class="n">fndr_pheno</span><span class="p">,</span>
    <span class="n">fndr_bval</span> <span class="o">=</span> <span class="n">fndr_bval</span><span class="p">,</span>
    <span class="n">fndr_gmod</span> <span class="o">=</span> <span class="n">fndr_gmod</span><span class="p">,</span>
    <span class="n">pselop</span> <span class="o">=</span> <span class="n">init_pselop</span><span class="p">,</span>
    <span class="n">mateop</span> <span class="o">=</span> <span class="n">init_mateop</span><span class="p">,</span>
    <span class="n">evalop</span> <span class="o">=</span> <span class="n">init_evalop</span><span class="p">,</span>
    <span class="n">sselop</span> <span class="o">=</span> <span class="n">init_sselop</span><span class="p">,</span>
    <span class="n">burnin</span> <span class="o">=</span> <span class="n">nburnin</span>
<span class="p">)</span>
<span class="n">pselop</span> <span class="o">=</span> <span class="n">MyParentSelectionOperator</span><span class="p">()</span>
<span class="n">mateop</span> <span class="o">=</span> <span class="n">MyMatingOperator</span><span class="p">(</span><span class="n">mate2waydh</span><span class="o">.</span><span class="n">progeny_counter</span><span class="p">,</span> <span class="n">mate2waydh</span><span class="o">.</span><span class="n">family_counter</span><span class="p">)</span>
<span class="n">evalop</span> <span class="o">=</span> <span class="n">MyEvaluationOperator</span><span class="p">(</span><span class="n">algmod_true</span><span class="p">,</span> <span class="n">ptprot</span><span class="o">.</span><span class="n">var_err</span><span class="p">)</span>
<span class="n">sselop</span> <span class="o">=</span> <span class="n">MySurvivorSelectionOperator</span><span class="p">(</span><span class="n">nindiv_fam</span><span class="p">)</span>
<span class="n">lbook</span> <span class="o">=</span> <span class="n">MyLogbook</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="assemble-universal-breeding-algorithm-from-operators">
<h3>Assemble universal breeding algorithm from operators<a class="headerlink" href="#assemble-universal-breeding-algorithm-from-operators" title="Link to this heading">#</a></h3>
<p>Then we assemble the operators into a universal breeding algorithm, which is a recurrent selection program.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a recurrent selection object using constructed operators</span>
<span class="n">rsprog</span> <span class="o">=</span> <span class="n">RecurrentSelectionBreedingProgram</span><span class="p">(</span>
    <span class="n">initop</span> <span class="o">=</span> <span class="n">initop</span><span class="p">,</span>
    <span class="n">pselop</span> <span class="o">=</span> <span class="n">pselop</span><span class="p">,</span>
    <span class="n">mateop</span> <span class="o">=</span> <span class="n">mateop</span><span class="p">,</span>
    <span class="n">evalop</span> <span class="o">=</span> <span class="n">evalop</span><span class="p">,</span>
    <span class="n">sselop</span> <span class="o">=</span> <span class="n">sselop</span><span class="p">,</span>
    <span class="n">t_max</span> <span class="o">=</span> <span class="mi">20</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="evolve-the-breeding-program">
<h3>Evolve the breeding program<a class="headerlink" href="#evolve-the-breeding-program" title="Link to this heading">#</a></h3>
<p>Finally, we evolve our simulation for 10 generations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># evolve the population</span>
<span class="n">rsprog</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">nrep</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ngen</span> <span class="o">=</span> <span class="n">ngen</span><span class="p">,</span> <span class="n">lbook</span> <span class="o">=</span> <span class="n">lbook</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="export-simulation-results-to-a-pandas-dataframe">
<h3>Export simulation results to a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code><a class="headerlink" href="#export-simulation-results-to-a-pandas-dataframe" title="Link to this heading">#</a></h3>
<p>After evolving our population, we export our simulation results to csv files for our reference.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># export results as pandas dataframe</span>
<span class="n">program_df</span> <span class="o">=</span> <span class="n">lbook</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="n">program_frontier_df</span> <span class="o">=</span> <span class="n">lbook</span><span class="o">.</span><span class="n">to_pandas_frontier</span><span class="p">()</span>

<span class="c1"># write dataframes to files</span>
<span class="n">program_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_program.csv&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">program_frontier_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_program_frontier.csv&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="plot-results">
<h2>Plot results<a class="headerlink" href="#plot-results" title="Link to this heading">#</a></h2>
<section id="import-libraries-and-set-global-variables">
<h3>Import libraries and set global variables<a class="headerlink" href="#import-libraries-and-set-global-variables" title="Link to this heading">#</a></h3>
<p>Here, we import various plotting libraries and apply some font settings.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># import data and plotting libraries</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span>
<span class="kn">import</span> <span class="nn">seaborn</span>

<span class="c1"># set default font to use in plots to something that isn&#39;t completely ugly</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">rcParams</span>
<span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Liberation Serif&#39;</span>
</pre></div>
</div>
</section>
<section id="import-data-from-breakpoint">
<h3>Import data from breakpoint<a class="headerlink" href="#import-data-from-breakpoint" title="Link to this heading">#</a></h3>
<p>We load the data we will be plotting and negate our trait values, since our multi-objective optimization values are negated GEBVs.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># breakpoint for testing: read from file</span>
<span class="n">program_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_program.csv&quot;</span><span class="p">)</span>
<span class="n">program_frontier_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_program_frontier.csv&quot;</span><span class="p">)</span>

<span class="c1"># negate frontiers to reflect mean GEVBs (since objectives are -GEBV)</span>
<span class="n">program_frontier_df</span><span class="p">[</span><span class="s2">&quot;syn1&quot;</span><span class="p">]</span> <span class="o">*=</span> <span class="o">-</span><span class="mf">1.0</span>
<span class="n">program_frontier_df</span><span class="p">[</span><span class="s2">&quot;syn2&quot;</span><span class="p">]</span> <span class="o">*=</span> <span class="o">-</span><span class="mf">1.0</span>
</pre></div>
</div>
</section>
<section id="plot-pareto-frontiers-over-time">
<h3>Plot Pareto frontiers over time<a class="headerlink" href="#plot-pareto-frontiers-over-time" title="Link to this heading">#</a></h3>
<p>We plot pareto frontiers over time below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot frontiers over time</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">))</span>
<span class="n">seaborn</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">program_frontier_df</span><span class="p">,</span>
    <span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;syn1&quot;</span><span class="p">,</span>
    <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;syn2&quot;</span><span class="p">,</span>
    <span class="n">hue</span> <span class="o">=</span> <span class="s2">&quot;t_cur&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Pareto Frontier for Multi-Objective Conventional Genomic Selection Over Time&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait 1 Breeding Value: Selection Mean (h² = 0.4)&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait 2 Breeding Value: Selection Mean (h² = 0.6)&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_program_frontier.png&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/multiobjective_genomic_selection_program_frontier.png" src="../../../_images/multiobjective_genomic_selection_program_frontier.png" />
</section>
<section id="plot-mean-expected-heterozygosity-over-time">
<h3>Plot mean expected heterozygosity over time<a class="headerlink" href="#plot-mean-expected-heterozygosity-over-time" title="Link to this heading">#</a></h3>
<p>We plot the change in mean expected heterozygosity (genetic diversity) over time using the code below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot MEH over time</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">))</span>
<span class="n">seaborn</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">program_df</span><span class="p">,</span>
    <span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;t_cur&quot;</span><span class="p">,</span>
    <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;main_meh&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Multi-Objective Conventional Genomic Selection Over Time&quot;</span><span class="p">)</span>
<span class="c1"># pyplot.suptitle(&quot;Mean Expected Heterozygosity&quot;)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Mean Expected Heterozygosity&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_mean_expected_heterozygosity.png&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/multiobjective_genomic_selection_mean_expected_heterozygosity.png" src="../../../_images/multiobjective_genomic_selection_mean_expected_heterozygosity.png" />
</section>
<section id="plot-population-mean-breeding-values-over-time">
<h3>Plot population mean breeding values over time<a class="headerlink" href="#plot-population-mean-breeding-values-over-time" title="Link to this heading">#</a></h3>
<p>We plot the change in mean population GEBVs over the simulation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot population mean over time</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">))</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
    <span class="n">program_df</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn1&quot;</span><span class="p">],</span>
    <span class="n">program_df</span><span class="p">[</span><span class="s2">&quot;main_true_mean_syn2&quot;</span><span class="p">],</span>
    <span class="n">color</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;black&quot;</span><span class="p">,</span><span class="mf">0.4</span><span class="p">),</span>
<span class="p">)</span>
<span class="n">seaborn</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">program_df</span><span class="p">,</span>
    <span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;main_true_mean_syn1&quot;</span><span class="p">,</span>
    <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;main_true_mean_syn2&quot;</span><span class="p">,</span>
    <span class="n">hue</span> <span class="o">=</span> <span class="s2">&quot;t_cur&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Multi-Objective Conventional Genomic Selection Over Time&quot;</span><span class="p">)</span>
<span class="c1"># pyplot.suptitle(&quot;Mean Expected Heterozygosity&quot;)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait 1 Breeding Value: Population Mean (h² = 0.4)&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Synthetic Trait 2 Breeding Value: Population Mean (h² = 0.6)&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Generation&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;multiobjective_genomic_selection_breeding_values.png&quot;</span><span class="p">)</span>
<span class="n">pyplot</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>The figure below is the results of the code above.</p>
<img alt="../../../_images/multiobjective_genomic_selection_breeding_values.png" src="../../../_images/multiobjective_genomic_selection_breeding_values.png" />
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="constrained_single_trait_phenotypic_selection.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Constrained Single-Trait Phenotypic Selection</p>
      </div>
    </a>
    <a class="right-next"
       href="constrained_bitrait_phenotypic_selection.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Constrained Bi-Trait Phenotypic Selection</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-preliminaries">Simulation Preliminaries</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-required-modules-and-seeding-the-global-prng">Loading Required Modules and Seeding the global PRNG</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-parameters">Simulation Parameters</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genetic-map-data-from-a-text-file">Loading Genetic Map Data from a Text File</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-genetic-map-function">Creating a Genetic Map Function</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-genome-data-from-a-vcf-file">Loading Genome Data from a VCF File</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sampling-founders-and-markers-from-the-loaded-vcf-file">Sampling founders and markers from the loaded VCF file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constructing-a-bi-trait-genomic-model">Constructing a Bi-Trait Genomic Model</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#constructing-the-founder-population">Constructing the Founder Population</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#randomly-intermate-founders-for-nrandmate-generations">Randomly Intermate founders for <code class="docutils literal notranslate"><span class="pre">nrandmate</span></code> Generations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#construct-bootstrap-cohort-structure-to-begin-burnin">Construct bootstrap cohort structure to begin burnin</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-mating-protocols-for-burn-in">Create Mating Protocols for Burn-In</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-genotyping-protocols-for-burn-in">Create Genotyping Protocols for Burn-In</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-phenotyping-protocols-for-burn-in">Create Phenotyping Protocols for Burn-In</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-breeing-value-estimation-protocols-for-burn-in">Create Breeing Value Estimation Protocols for Burn-In</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-within-family-selection-helper-function">Create a Within-Family Selection Helper Function</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-cohort-construction-helper-function">Create a Cohort Construction Helper function</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-cohort-structure">Create Cohort Structure</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#define-breeding-program-operators">Define breeding program operators</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-parent-selection-operator-for-initialization-operator">Define parent selection operator for initialization operator</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-mating-operator-for-initialization-operator">Define mating operator for initialization operator</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-evaluation-operator-for-initialization-operator">Define evaluation operator for initialization operator</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-survivor-selection-operator-for-initialization-operator">Define survivor selection operator for initialization operator</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-initialization-operator-for-universal-breeding-algorithm">Define initialization operator for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-parent-selection-operator-for-universal-breeding-algorithm">Define parent selection operator for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-mating-operator-for-universal-breeding-algorithm">Define mating operator for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-evaluation-operator-for-universal-breeding-algorithm">Define evaluation operator for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-survivor-selection-operator-for-universal-breeding-algorithm">Define survivor selection operator for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-logbook-to-collect-simulation-data">Define logbook to collect simulation data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-setup">Simulation Setup</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-operators-used-by-initialization-operator">Create operators used by initialization operator</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-main-operators-for-universal-breeding-algorithm">Create main operators for universal breeding algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assemble-universal-breeding-algorithm-from-operators">Assemble universal breeding algorithm from operators</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evolve-the-breeding-program">Evolve the breeding program</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#export-simulation-results-to-a-pandas-dataframe">Export simulation results to a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-results">Plot results</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#import-libraries-and-set-global-variables">Import libraries and set global variables</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#import-data-from-breakpoint">Import data from breakpoint</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-pareto-frontiers-over-time">Plot Pareto frontiers over time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-mean-expected-heterozygosity-over-time">Plot mean expected heterozygosity over time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-population-mean-breeding-values-over-time">Plot population mean breeding values over time</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>