Search.setIndex({"docnames": ["advanced_usage/advanced_usage", "api", "contributing/contributing", "faq/faq", "fundamentals/architecture/architecture", "fundamentals/architecture/philosophy", "fundamentals/core_infrastructure/core_infrastructure", "fundamentals/data_types/breeding_simulation", "fundamentals/data_types/data_types", "fundamentals/data_types/genomic_modeling", "fundamentals/data_types/optimization", "fundamentals/data_types/population_genetics", "fundamentals/fundamentals", "fundamentals/simulation/philosophy", "getting_started/getting_started", "index", "standard_usage/breeding_operators/evaluation_operators", "standard_usage/breeding_operators/initialization_operators", "standard_usage/breeding_operators/logbooks", "standard_usage/breeding_operators/mating_operators", "standard_usage/breeding_operators/parental_selection_operators", "standard_usage/breeding_operators/survivor_selection_operators", "standard_usage/breeding_protocols/breeding_value_protocols", "standard_usage/breeding_protocols/genotyping_protocols", "standard_usage/breeding_protocols/mating_protocols", "standard_usage/breeding_protocols/phenotyping_protocols", "standard_usage/breeding_protocols/selection_protocols", "standard_usage/core_infrastructure/random_number_generation", "standard_usage/examples/examples", "standard_usage/genomic_modeling/genomic_models", "standard_usage/genomic_modeling/variance_matrices", "standard_usage/optimization/optimization_algorithms", "standard_usage/optimization/optimization_problems", "standard_usage/optimization/optimization_solutions", "standard_usage/population_genetics/breeding_value_matrices", "standard_usage/population_genetics/coancestry_matrices", "standard_usage/population_genetics/genetic_map_functions", "standard_usage/population_genetics/genetic_maps", "standard_usage/population_genetics/genotype_matrices", "standard_usage/population_genetics/phased_genotype_matrices", "standard_usage/standard_usage"], "filenames": ["advanced_usage/advanced_usage.rst", "api.rst", "contributing/contributing.rst", "faq/faq.rst", "fundamentals/architecture/architecture.rst", "fundamentals/architecture/philosophy.rst", "fundamentals/core_infrastructure/core_infrastructure.rst", "fundamentals/data_types/breeding_simulation.rst", "fundamentals/data_types/data_types.rst", "fundamentals/data_types/genomic_modeling.rst", "fundamentals/data_types/optimization.rst", "fundamentals/data_types/population_genetics.rst", "fundamentals/fundamentals.rst", "fundamentals/simulation/philosophy.rst", "getting_started/getting_started.rst", "index.rst", "standard_usage/breeding_operators/evaluation_operators.rst", "standard_usage/breeding_operators/initialization_operators.rst", "standard_usage/breeding_operators/logbooks.rst", "standard_usage/breeding_operators/mating_operators.rst", "standard_usage/breeding_operators/parental_selection_operators.rst", "standard_usage/breeding_operators/survivor_selection_operators.rst", "standard_usage/breeding_protocols/breeding_value_protocols.rst", "standard_usage/breeding_protocols/genotyping_protocols.rst", "standard_usage/breeding_protocols/mating_protocols.rst", "standard_usage/breeding_protocols/phenotyping_protocols.rst", "standard_usage/breeding_protocols/selection_protocols.rst", "standard_usage/core_infrastructure/random_number_generation.rst", "standard_usage/examples/examples.rst", "standard_usage/genomic_modeling/genomic_models.rst", "standard_usage/genomic_modeling/variance_matrices.rst", "standard_usage/optimization/optimization_algorithms.rst", "standard_usage/optimization/optimization_problems.rst", "standard_usage/optimization/optimization_solutions.rst", "standard_usage/population_genetics/breeding_value_matrices.rst", "standard_usage/population_genetics/coancestry_matrices.rst", "standard_usage/population_genetics/genetic_map_functions.rst", "standard_usage/population_genetics/genetic_maps.rst", "standard_usage/population_genetics/genotype_matrices.rst", "standard_usage/population_genetics/phased_genotype_matrices.rst", "standard_usage/standard_usage.rst"], "titles": ["Advanced PyBrOpS Usage", "&lt;no title&gt;", "Contributing to PyBrOpS", "Frequently Asked Questions", "PyBrOpS Implemented Software Architecture", "PyBrOpS Software Architecture Philosophy", "Core PyBrOpS Infrastructure", "Breeding Simulation Data Types", "Data Types in PyBrOpS", "Genomic Modeling Data Types", "Optimization Data Types", "Population Genetics Data Types", "PyBrOpS Fundamentals", "PyBrOpS Simulation Philosophy", "Getting Started", "PyBrOpS Main Page", "Evaluation Operators", "Initialization Operators", "Logbooks", "Mating Operators", "Parental Selection Operators", "Survivor Selection Operators", "Breeding Value Estimation Protocols", "Genotyping Protocols", "Mating Protocols", "Phenotyping Protocols", "Selection Protocols", "Random Number Generation", "Examples", "Genomic Models", "Variance Matrices", "Optimization Algorithms", "Optimization Problems", "Optimization Solutions", "Breeding Value Matrices", "Coancestry Matrices", "Genetic Map Functions", "Genetic Maps", "Genotype Matrices", "Phased Genotype Matrices", "Standard PyBrOpS Usage"], "terms": {"under": [0, 20, 21], "construct": [0, 13, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 34, 35, 37, 38, 39], "guid": [5, 14], "its": [5, 10, 11, 14, 15, 18, 22, 23, 25, 32, 34, 36, 39], "By": 5, "i": [5, 8, 9, 10, 11, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 32, 33, 34, 36, 37, 38, 39], "easili": 5, "comprehens": [5, 8, 15], "extent": 5, "maintain": 5, "mani": [5, 8, 9, 29], "sourc": [5, 14, 25, 26, 30, 34, 35, 40], "commonli": 5, "plant": [5, 15], "anim": [5, 15], "breed": [5, 8, 9, 10, 14, 15, 37, 38, 39], "commun": 5, "develop": [5, 14], "robust": 5, "set": [5, 12, 13, 15, 17, 18, 22, 23, 24, 26, 27, 29, 30, 33, 37, 38], "make": [5, 11, 14, 15, 24, 26, 34, 35, 37, 38], "them": [5, 15, 26, 32], "difficult": 5, "understand": [5, 12, 14], "extend": [5, 15, 16, 17, 18, 19, 20, 21], "The": [5, 8, 9, 10, 11, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "aim": 5, "remedi": 5, "thi": [5, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 23, 24, 25, 26, 29, 30, 31, 32, 34, 35, 36, 37, 38, 39], "maxim": [5, 26], "user": [5, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 25, 26, 32, 37], "briefli": [5, 9], "ar": [5, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 29, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], "summar": [5, 9, 22, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "descript": [5, 9, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "A": [5, 16, 17, 19, 20, 21, 23, 24, 25, 26, 29, 30, 32, 34, 35, 38, 39], "modul": [5, 9, 10, 11, 15, 27], "should": [5, 31], "respons": [5, 13, 26, 32], "onli": [5, 10, 11, 23, 24, 26, 30, 31, 34, 35, 37, 38], "one": [5, 13, 18, 20, 21, 24, 25, 26, 27, 30, 32, 34, 35, 37, 38], "task": [5, 13, 26, 30, 35], "have": [5, 9, 11, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "reason": 5, "chang": [5, 26, 30, 34, 35, 37, 38], "exens": 5, "modif": [5, 30, 34, 35], "deriv": [5, 9, 10, 11, 26, 29, 30, 33, 35, 39], "substitut": 5, "base": [5, 13, 14, 15, 16, 17, 18, 19, 20, 21, 26, 27, 29, 30, 32, 34, 35, 37], "narrow": [5, 22, 25], "focu": 5, "so": [5, 18, 23, 26, 30, 34, 35, 36, 37, 38], "prevent": 5, "unnecessari": 5, "abstract": [5, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "concret": [5, 22, 23, 24, 25, 30, 31, 33, 34, 35, 36, 37, 38, 39], "To": [5, 14, 16, 17, 18, 19, 20, 21, 26, 30, 32, 34, 35, 37], "adher": [5, 29], "heavi": 5, "class": [5, 9, 10, 11], "python": [5, 15, 27, 29, 30, 34, 35, 37, 38, 40], "": [5, 10, 13, 14, 15, 17, 18, 26, 29, 30, 32, 33, 34, 35, 37, 38], "sinc": [5, 10, 11, 16, 17, 18, 19, 20, 21, 26, 29, 30, 31, 32, 34, 35], "duck": 5, "type": [5, 12, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "manual": [5, 14, 26, 37], "enforc": 5, "seek": 5, "creat": [5, 9, 11, 13, 14, 31, 32, 33, 36], "new": [5, 15, 18, 27, 29, 30, 32, 34, 35, 37, 38, 39], "implement": [5, 9, 10, 12, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 29, 30, 31, 32, 33, 34, 35, 38, 39], "must": [5, 10, 14, 16, 17, 18, 19, 20, 21, 32, 37, 38, 39], "ensur": [5, 15], "function": [5, 9, 10, 13, 15, 26, 27, 29, 32, 33, 34, 35, 37, 39, 40], "requir": [5, 10, 13, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 31, 32, 36, 38], "everi": 5, "famili": [5, 9, 10, 11], "ha": [5, 23, 26, 30, 34, 35, 37, 38, 39, 40], "singular": 5, "focus": [5, 30, 34, 35, 38], "purpos": [5, 8, 9, 10, 11, 16, 17, 18, 19, 20, 21, 29, 30, 33, 37, 38, 39], "repres": [5, 10, 11, 13, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "fundament": [5, 14, 15], "entiti": 5, "data": [5, 12, 14, 17, 18, 24, 29, 30, 34, 35, 37, 38, 39], "structur": [5, 12, 13], "program": [5, 10, 11, 13, 15, 29, 38], "almost": 5, "all": [5, 10, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 31, 32, 33, 34, 35, 36, 37, 38, 39], "least": 5, "most": [5, 11, 13, 24, 26, 29, 30, 34, 36, 38], "method": [5, 10, 13, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 30, 31, 32, 34, 36, 37, 38, 39], "These": [5, 13, 18, 24, 26, 27, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "two": [5, 9, 16, 18, 20, 21, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38], "decis": [5, 10, 11, 13, 15, 26, 31, 32, 33, 34, 35, 38], "allow": [5, 9, 11, 13, 15, 29, 30, 34, 35, 36, 37, 38], "If": [5, 14, 16, 17, 19, 20, 21, 26, 30, 34, 35, 37, 39], "featur": [5, 10, 15, 24], "desir": [5, 13, 14, 21, 23, 26, 27, 37], "can": [5, 8, 9, 11, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], "simpli": 5, "instead": 5, "modifi": [5, 13, 38], "current": [5, 16, 18, 19, 20, 21, 23, 38], "which": [5, 9, 10, 11, 13, 16, 17, 18, 19, 20, 21, 22, 23, 25, 26, 27, 29, 30, 32, 34, 35, 36, 37, 38, 39, 40], "an": [5, 8, 9, 10, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "do": [5, 14, 16, 17, 19, 20, 21, 22, 23, 24, 26, 34, 37], "signatur": 5, "ident": [5, 30, 34, 35], "those": [5, 8, 37], "specifi": [5, 15, 24, 26, 34], "within": [5, 11, 13, 16, 19, 22, 25, 26, 27, 29, 31, 32, 33, 34, 35, 36, 38, 39], "same": [5, 11, 37, 39], "each": [5, 22, 25, 26, 29, 30, 35, 37, 38, 39, 40], "other": [5, 10, 13, 15, 25, 26, 29, 30, 34, 35, 38], "provid": [5, 8, 9, 11, 13, 15, 22, 24, 25, 26, 27, 29, 30, 34, 35, 36, 37, 39], "altern": [5, 14, 38, 39], "lean": 5, "from": [5, 9, 10, 11, 13, 15, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 27, 31, 32, 33, 36], "As": [5, 9, 29, 30, 35], "mention": 5, "section": [5, 8, 14, 29, 30, 34, 35, 38], "decoupl": 5, "improv": [5, 15], "modular": [5, 15], "larg": 8, "softwar": [8, 10, 12, 15], "packag": [8, 13, 15], "differ": [8, 10, 11, 13, 26, 29, 34], "goal": [8, 15], "brief": 8, "overview": [8, 15], "essenti": [8, 10, 13, 14, 26, 27, 33, 39], "group": [8, 22, 23, 24, 25, 26, 29, 36], "four": [8, 9, 24, 29, 30, 36], "gener": [8, 10, 13, 17, 25, 31, 40], "categori": [8, 26, 29, 30, 37], "associ": [8, 15, 38, 39], "popul": [8, 9, 13, 17, 25, 26, 37, 39], "genet": [8, 9, 13, 14, 24, 31, 38, 39], "genom": [8, 11, 13, 16, 17, 18, 19, 20, 21, 22, 24, 26, 34, 35, 37, 38, 39], "model": [8, 11, 13, 16, 17, 18, 19, 20, 21, 22, 26, 34, 35, 37, 38], "optim": [8, 13, 15, 35], "final": [8, 13, 29, 38], "simul": [8, 10, 11, 12, 14, 15, 16, 17, 18, 19, 20, 21, 23, 25, 27, 29, 36, 37, 38], "pybrop": [9, 10, 11], "gmod": [9, 16, 17, 18, 19, 20, 21, 30], "hous": [9, 11, 13], "submodul": [9, 10, 32, 33, 36], "relat": [9, 11, 13, 29, 30, 34, 35, 38], "In": [9, 14, 26, 27, 29, 32, 36, 37], "context": 9, "refer": [9, 22, 29, 30, 34, 35, 37, 38, 39], "anyth": 9, "take": [9, 13, 22, 23, 24, 26, 32], "inform": [9, 11, 13, 15, 18, 26, 32, 34, 37, 38], "input": [9, 13, 23, 26, 29, 34, 39], "output": [9, 13, 16, 17, 19, 20, 21, 23, 26, 32], "genomicmodel": [9, 30], "represent": [9, 11, 13, 29, 30, 35, 36, 39], "ani": [9, 16, 17, 18, 19, 20, 21, 22, 23, 24, 29, 38], "primarili": [9, 38], "fit": [9, 29], "predict": [9, 11, 13, 25, 26, 34, 35, 38], "estim": [9, 11, 13, 26, 34, 35, 36, 37, 38, 40], "valu": [9, 13, 16, 17, 18, 19, 20, 21, 24, 25, 26, 30, 31, 32, 33, 37, 38, 39, 40], "also": [9, 10, 11, 14, 25, 26, 27, 29, 30, 32, 34, 35, 37, 38], "genic": 9, "allel": [9, 11, 39], "metric": [9, 18], "upper": [9, 32, 33], "lower": [9, 32, 33], "select": [9, 11, 13, 16, 19, 40], "limit": 9, "There": [9, 16, 17, 18, 19, 20, 21, 24, 37, 38], "interfac": [9, 10, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "main": [9, 34, 35, 40], "divid": 9, "lineargenomicmodel": 9, "nonlineargenomicmodel": [9, 29], "name": [9, 11, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "suggest": [9, 11, 29], "former": [9, 11, 29, 37], "defin": [9, 10, 13, 22, 23, 26, 29, 32, 35, 37], "linear": [9, 22, 25, 26, 30, 37], "natur": [9, 10, 13, 26, 29, 30, 31, 35, 39], "while": [9, 11, 26, 29, 32, 37], "latter": [9, 11, 29, 37], "non": [9, 24, 26, 29, 32, 39], "mai": [9, 10, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 32, 34, 35, 36, 37, 38, 39], "us": [9, 10, 11, 13, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "machin": [9, 29], "learn": [9, 14, 29], "sever": [9, 13, 14, 15, 18, 22, 25, 29, 30, 31, 32, 34, 35, 36, 37, 38, 39], "subtyp": [9, 26, 29], "below": [9, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], "additivelineargenomicmodel": [9, 30], "assum": [9, 26, 29, 30, 39], "strictli": [9, 29], "addit": [9, 14, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 30, 32, 34, 35, 37, 38, 39], "effect": [9, 13, 15, 22, 25, 26, 30], "coancestrylineargenomicmodel": [9, 29], "calcul": [9, 11, 13, 16, 17, 18, 19, 20, 21, 22, 26, 32, 34, 38, 39], "coancestri": [9, 13, 29, 40], "relationship": [9, 11, 29, 35, 38], "additivedominancelineargenomicmodel": 9, "domin": [9, 26, 38], "additivedominanceepistaticlineargenomicmodel": 9, "epistat": 9, "geneticvariancematrix": 9, "genicvariancematrix": 9, "expect": [9, 11, 26, 30, 36, 39], "progeni": [9, 11, 13, 21, 26, 30, 36], "trait": [9, 13, 20, 21, 22, 26], "cross": [9, 13, 19, 24, 26, 30], "between": [9, 11, 13, 24, 30, 34, 35, 36, 38, 39], "individu": [9, 11, 13, 19, 20, 22, 26, 30, 32, 34, 35, 38, 39], "linkag": [9, 29, 30, 37], "both": [9, 15, 26, 27, 29, 30, 34, 35, 36], "util": [9, 10, 13, 30, 32, 37], "design": [9, 12, 13, 14, 15, 30], "agnost": [9, 30], "For": [9, 16, 17, 18, 19, 20, 21, 26, 29, 30, 32, 35], "exist": [9, 15, 30], "determinist": [9, 30], "equat": [9, 30], "three": [9, 13, 24, 26, 29, 30, 32, 34, 35, 38], "wai": [9, 24, 26, 30], "additivegeneticvariancematrix": [9, 30], "additivegenicvariancematrix": [9, 30], "respectfulli": [9, 30], "thei": [9, 11, 13, 14, 24, 30, 35], "densetwowaydhadditivegeneticvariancematrix": [9, 30], "densethreewaydhadditivegeneticvariancematrix": [9, 30], "densefourwaydhadditivegeneticvariancematrix": [9, 30], "densedihybriddhadditivegeneticvariancematrix": [9, 30], "dihybrid": [9, 30], "One": 10, "kei": [10, 13, 30, 34, 35, 37, 38, 39], "separ": 10, "abil": [10, 11, 15, 37], "nativ": 10, "perform": [10, 13, 15, 16, 19, 26, 30, 32, 34, 35, 37, 38], "singl": [10, 13, 15, 23, 27], "multi": [10, 13, 15], "object": [10, 11, 13, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 29, 30, 34, 37, 38, 39], "encapul": 10, "appropri": [10, 26], "approach": [10, 13, 15], "framework": [10, 13, 14, 15], "centric": 10, "strategi": [10, 13, 15, 19, 26], "conduct": [10, 13, 29, 37], "opitm": 10, "infrastructur": [10, 12], "found": [10, 16, 17, 18, 19, 20, 21, 22, 23, 25, 26, 30, 32, 33, 34, 35, 36, 37, 38, 39, 40], "opt": 10, "prob": [10, 31], "contain": [10, 13, 17, 29, 32, 33, 36, 37, 38, 39], "metadata": [10, 30, 32, 33, 34, 35, 37, 38, 39], "pertain": [10, 29, 32, 33], "space": [10, 26, 31, 32, 33], "inequ": [10, 26, 32, 33], "equal": [10, 26, 32, 33], "constraint": [10, 26, 32, 33], "evalu": [10, 13, 15, 26, 40], "candid": 10, "express": [10, 32], "minim": [10, 26, 32], "soln": [10, 26], "identifi": [10, 26, 33, 34], "store": [10, 11, 13, 16, 17, 18, 19, 20, 21, 26, 30, 33, 34, 35, 38, 39], "result": [10, 13, 19, 26, 33], "regardless": [10, 31], "optimizationalgorithm": [10, 31], "veri": [10, 29, 31], "divers": [10, 13, 16, 17, 19, 20, 21, 29, 31], "popgen": 11, "serv": [11, 12, 20, 34, 35], "foundat": 11, "marker": [11, 13, 22, 23, 24, 25, 26, 29, 30, 35, 36], "wise": [11, 37], "posit": [11, 22, 23, 24, 25, 26, 30, 36, 38, 39], "mate": [11, 13, 20, 26, 37, 40], "read": [11, 14, 17, 29, 30, 34, 35, 38, 39], "write": [11, 15, 16, 17, 19, 20, 21, 29, 30, 34, 35, 37, 38], "interpol": 11, "via": [11, 25, 27, 35, 37, 38, 39], "geneticmap": 11, "file": [11, 14], "real": [11, 17, 26, 31, 32, 33, 37, 38], "recombin": [11, 13, 24, 37, 38, 39], "ad": [11, 25, 29, 37], "realism": [11, 25, 37], "geneticmapfunct": [11, 36], "convert": [11, 26, 32, 36], "probabl": [11, 13, 24, 38, 39], "varianc": [11, 13, 36, 40], "popular": [11, 36], "like": [11, 13, 30, 35, 36, 38, 39], "haldan": [11, 36], "kosambi": [11, 36], "out": [11, 16, 17, 19, 20, 21, 23, 26, 29, 30, 32, 34, 35, 36, 37, 38, 39], "box": [11, 36], "perhap": [11, 32, 38], "import": [11, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "genotypematrix": [11, 23, 29, 39], "possibl": [11, 15, 38], "vcf": [11, 14], "world": [11, 17, 38], "phasedgenotypematrix": [11, 23, 30], "includ": [11, 13, 26, 29, 30, 34, 35, 38, 39], "state": [11, 13], "chromosom": [11, 36, 37, 38, 39], "regular": [11, 39], "breedingvaluematrix": 11, "impli": [11, 34], "coancestrymatrix": 11, "document": 12, "describ": [12, 26, 34], "crucial": [12, 13], "concept": 12, "architectur": [12, 15], "start": [12, 13, 15, 17, 30, 34, 35, 36, 37, 38, 39], "point": [12, 13, 26], "philosophi": [12, 14], "core": [12, 27], "adopt": [13, 15], "script": [13, 14, 15, 40], "maximum": [13, 15, 26, 29], "flexibl": [13, 15], "accomplish": [13, 25, 26, 29, 30, 31, 32, 33, 34, 35, 37], "compon": [13, 32], "equip": 13, "rich": 13, "arrai": [13, 14, 22, 23, 24, 25, 26, 32, 36], "plai": 13, "role": 13, "basic": [13, 25, 29, 38], "build": [13, 14], "block": [13, 26, 30, 34, 35], "unit": [13, 14, 34], "e": [13, 29], "g": [13, 29], "genotyp": [13, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 30, 35, 40], "compos": 13, "offer": [13, 15, 34, 38], "list": [13, 32, 37], "map": [13, 38, 39, 40], "covari": 13, "work": [13, 14], "tandem": 13, "enabl": [13, 15], "matric": [13, 29, 40], "pivot": 13, "merit": 13, "vital": 13, "parent": [13, 16, 19, 24, 26, 30, 39, 40], "quantifi": [13, 25], "related": 13, "indispens": 13, "manag": 13, "avoid": 13, "inbreed": 13, "control": [13, 15], "more": [13, 14, 15, 20, 21, 22, 25, 32], "interest": 13, "support": [13, 14, 36, 37], "variou": [13, 15, 29, 34, 37, 38], "mode": 13, "help": [13, 26, 34, 35], "assess": [13, 29], "among": 13, "offspr": [13, 19, 24], "some": [13, 14, 24, 27, 29, 35, 36, 37], "techniqu": [13, 29], "problem": [13, 40], "facilit": 13, "formul": [13, 26], "involv": [13, 16, 17, 20, 21], "scheme": 13, "solut": [13, 26, 31, 40], "after": [13, 25, 26, 35, 37], "solv": 13, "record": [13, 18], "breeder": [13, 15], "decid": 13, "algorithm": [13, 14, 16, 17, 18, 19, 20, 21, 32, 40], "rang": [13, 15, 22, 23, 24, 25, 26, 29, 30, 35, 37, 38, 39], "tackl": 13, "specif": [13, 15, 24, 30, 34, 37, 38], "face": 13, "small": [13, 35], "process": [13, 17, 30, 34, 35, 38], "special": [13, 18], "return": [13, 16, 17, 18, 19, 20, 21, 23, 25, 26, 32, 35], "five": [13, 27, 30, 34, 35, 38, 39], "phenotyp": [13, 16, 17, 18, 19, 20, 21, 22, 29, 40], "routin": [13, 14, 34, 38], "potenti": 13, "extern": 13, "lme4": 13, "handl": 13, "configur": [13, 18, 19, 20, 24, 26], "field": [13, 16, 17, 19, 20, 21, 29], "trial": 13, "lastli": 13, "entir": [13, 16, 17, 18, 19, 20, 21, 22, 23, 35], "transform": [13, 26, 34], "versatil": [13, 15], "encompass": 13, "assembl": 13, "univers": [13, 16, 17, 18, 19, 20, 21], "On": [13, 14, 37], "hand": 13, "multipl": [13, 15, 18, 24, 29, 30, 34, 35, 38], "produc": [13, 23, 24], "intend": [13, 29], "initi": [13, 24, 27, 40], "survivor": [13, 40], "It": [13, 14, 16, 20, 21], "load": 13, "burn": [13, 17], "measur": [13, 29], "determin": [13, 26, 34, 37, 38, 39], "reproduct": 13, "mimick": 13, "meiosi": 13, "mutat": 13, "choos": [13, 37], "next": [13, 14, 24, 26], "cycl": 13, "merg": 13, "subpopul": 13, "need": [13, 14, 15, 25, 26, 29, 37], "3": [14, 22, 23, 24, 25, 26, 30, 34, 35, 37, 38, 39], "8": [14, 36], "0": [14, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "download": 14, "see": [14, 32], "websit": 14, "reli": [14, 25], "numpi": [14, 22, 23, 24, 25, 26, 27, 32, 33, 36], "1": [14, 18, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "22": 14, "2": [14, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39], "matrix": [14, 22, 23, 24, 25, 26, 33], "storag": 14, "algebra": 14, "panda": [14, 25], "4": [14, 22, 23, 24, 25, 26, 30, 34, 35, 36, 38, 39], "frame": 14, "scipi": 14, "miscellan": [14, 16, 17, 19, 20, 21, 22, 25, 26, 29, 30], "numer": [14, 26, 34, 35, 38, 39], "matplotlib": 14, "5": [14, 16, 17, 19, 21, 24, 27, 30, 32, 33, 34, 35, 36, 38, 39], "graphic": 14, "pymoo": 14, "6": [14, 20], "cyvcf2": 14, "30": 14, "14": 14, "h5py": 14, "hdf5": 14, "follow": [14, 18, 22, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 38, 39], "pytest": 14, "7": [14, 22, 25, 30], "test": [14, 26, 29, 37], "datadir": 14, "setuptool": 14, "wheel": 14, "pip3": 14, "assumpt": 14, "tabl": [14, 22, 24, 25, 26, 29, 30, 33, 34, 35, 36, 37, 38, 39], "link": [14, 40], "refresh": 14, "cours": 14, "you": 14, "languag": 14, "tutori": 14, "datafram": [14, 25], "10": [14, 22, 23, 24, 25, 26, 30, 32, 33, 34, 35, 37, 38, 39], "minut": 14, "option": [14, 16, 17, 19, 20, 21, 24, 30, 34, 35, 37, 38, 39], "org": 14, "how": [14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 26, 29, 30, 31, 34, 35, 38], "custom": [14, 15, 16, 17, 18, 19, 20, 21, 22, 25, 26, 32], "your": 14, "own": [14, 22, 23, 25, 37], "immedi": 14, "want": [14, 26], "complet": [14, 18, 29], "exampl": [14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "gain": 14, "depth": 14, "about": [14, 26, 37, 38], "befor": [14, 30, 34, 35, 37, 38, 39], "embark": 14, "standard": [14, 15, 26], "usag": [14, 15, 25, 26, 30, 35, 36], "detail": [14, 26, 29, 30, 34, 35], "instruct": 14, "system": 14, "header": 14, "even": 14, "though": 14, "written": [14, 30], "pure": [14, 25], "compil": 14, "distro": 14, "fedora": 14, "sudo": 14, "dnf": 14, "python3": 14, "devel": 14, "ubuntu": 14, "apt": 14, "dev": 14, "first": [14, 26, 30, 34, 35, 38, 39], "clone": 14, "repositori": [14, 40], "git": 14, "http": 14, "github": 14, "com": 14, "rzshrote": 14, "best": [14, 26], "practic": 14, "virtual": 14, "environ": [14, 22, 25], "where": [14, 16, 17, 19, 20, 21, 26, 29, 32, 37, 39], "makefil": 14, "done": [14, 25, 30, 34, 35, 37, 38], "m": [14, 39], "venv": 14, "env": 14, "bin": 14, "activ": [14, 16], "pip": 14, "edit": 14, "either": [14, 29], "sh": 14, "bash": 14, "now": 14, "access": [14, 25, 29, 30, 34, 35, 38], "through": [14, 15], "line": [14, 24], "prompt": 14, "execut": [14, 30, 34, 35], "capabl": 15, "stochast": 15, "empow": 15, "research": [15, 27], "explor": 15, "silico": 15, "studi": 15, "novel": 15, "establish": 15, "methodologi": [15, 16], "drive": 15, "signific": 15, "although": 15, "extant": 15, "apart": 15, "tool": 15, "wide": 15, "applic": 15, "simultan": 15, "visual": 15, "complex": [15, 22, 25], "tradeoff": 15, "inher": [15, 30], "lead": [15, 25], "built": [15, 24, 31, 37], "solid": 15, "principl": 15, "extens": [15, 18, 26], "adapt": 15, "meet": 15, "expans": [15, 26], "seamlessli": 15, "integr": 15, "ones": [15, 29], "evolv": 15, "aspect": 15, "easi": 15, "precis": 15, "fine": 15, "tune": 15, "explain": 15, "get": [15, 23, 30, 31, 34, 35], "advanc": [15, 32], "contribut": [15, 26, 30], "frequent": 15, "ask": 15, "question": 15, "index": [15, 30, 34, 35, 38, 39], "search": [15, 26], "evaluationoper": [16, 18], "step": [16, 18, 19, 20, 21, 24], "extrem": [16, 19, 20, 21, 29], "across": [16, 25, 29, 35, 38, 39], "speci": [16, 19], "combin": [16, 20, 21, 30], "job": [16, 17, 19, 20, 21], "evalut": 16, "evalutationoper": 16, "pybropo": [16, 17, 18, 19, 20, 21], "demonstr": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 34, 35, 36, 37, 38, 39], "code": [16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 29, 30, 31, 34, 35, 36, 37, 38, 39], "incredibli": [16, 17, 19, 20, 21], "hi": [16, 17, 19, 20, 21, 22, 25], "her": [16, 17, 19, 20, 21, 22, 25], "overrid": [16, 17, 18, 19, 20, 21, 30, 32, 34, 35], "left": [16, 17, 18, 19, 20, 21, 23], "empti": [16, 17, 18, 19, 20, 21], "myevaluationoper": 16, "def": [16, 17, 18, 19, 20, 21, 26, 32], "__init__": [16, 17, 18, 19, 20, 21], "self": [16, 17, 18, 19, 20, 21, 24, 32], "arg": [16, 17, 18, 19, 20, 21, 32], "tupl": [16, 17, 18, 19, 20, 21, 32], "kwarg": [16, 17, 18, 19, 20, 21, 26, 32], "dict": [16, 17, 18, 19, 20, 21, 26, 32], "none": [16, 17, 18, 19, 20, 21, 22, 25, 26, 29, 30, 33, 35, 37, 38], "constructor": [16, 17, 18, 19, 20, 21, 25, 26, 29, 30, 32, 33, 34, 35, 37, 39], "paramet": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 39], "argument": [16, 17, 18, 19, 20, 21, 22, 23, 24, 26, 29, 32, 34, 35, 36], "keyword": [16, 17, 18, 19, 20, 21, 26, 32], "pass": [16, 17, 18, 19, 20, 21, 26], "geno": [16, 17, 18, 19, 20, 21], "pheno": [16, 17, 18, 19, 20, 21], "bval": [16, 17, 18, 19, 20, 21], "t_cur": [16, 18, 19, 20, 21, 26], "int": [16, 18, 19, 20, 21, 37, 38, 39], "t_max": [16, 18, 19, 20, 21, 26], "miscout": [16, 17, 19, 20, 21], "dictionari": [16, 17, 18, 19, 20, 21], "time": [16, 18, 19, 20, 21, 37], "deadlin": [16, 18, 19, 20, 21], "pointer": [16, 17, 19, 20, 21, 30, 34, 35, 38, 39], "overwrit": [16, 17, 19, 20, 21], "previous": [16, 17, 19, 20, 21, 36, 39], "length": [16, 17, 19, 20, 21, 32, 34, 35, 38, 39], "depend": [16, 17, 18, 19, 20, 21, 22, 23, 25, 26, 29, 35], "restrict": [16, 17, 18, 19, 20, 21], "number": [16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 29, 30, 32, 33, 34, 35, 37, 38, 39, 40], "abov": [16, 17, 18, 19, 20, 21, 26, 32, 37], "evalop": 16, "initializationoper": [17, 18], "scratch": [17, 22, 23, 24, 25, 26, 30, 35, 38, 39], "typic": [17, 23, 25, 34], "seri": 17, "random": [17, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 37, 38, 39, 40], "prepar": 17, "myinitializationoper": 17, "default": [17, 26, 30, 32, 34, 35, 37], "initop": 17, "examin": [18, 26], "save": [18, 29, 30, 34, 35], "rep": 18, "replic": [18, 22, 25], "given": [18, 26, 29], "could": 18, "collect": [18, 26], "mylogbook": 18, "_data": 18, "setter": 18, "_rep": 18, "log_initi": 18, "directli": 18, "call": [18, 30, 34, 35], "log_pselect": 18, "mcfg": [18, 19, 20], "parentselectionoper": [18, 20], "pselect": [18, 20], "log_mat": 18, "matingoper": [18, 19], "log_evalu": 18, "log_sselect": 18, "survivorselectionoper": [18, 21], "sselect": [18, 21], "intern": [18, 26, 29, 32, 35], "filenam": [18, 29, 30, 34, 35], "str": [18, 22, 23, 24, 25, 26, 29, 30, 34, 35, 37, 38, 39], "correspond": [18, 26, 30, 32, 33, 35, 40], "oper": [18, 24, 29], "lbook": 18, "statist": [18, 29], "gather": 18, "eras": 18, "export": [18, 39], "csv": 18, "mymatingoper": 19, "mateop": 19, "myparentselectionoper": 20, "pselop": 20, "discard": 21, "undesir": [21, 29], "addition": 21, "surivor": 21, "mysurvivorselectionoper": 21, "sselop": 21, "breedingvalueprotocol": 22, "simpl": [22, 25], "scenario": [22, 25], "truebreedingvalu": 22, "true": [22, 25, 26, 32, 37], "meanphenotypicbreedingvalu": 22, "mean": [22, 30, 39], "statement": [22, 29, 30, 31, 32, 33, 35], "taxa": [22, 23, 24, 25, 26], "column": [22, 29, 30, 34, 35, 38], "bvprot": 22, "taxa_col": [22, 34, 35], "trait_col": [22, 30, 34], "trait01": 22, "trait02": 22, "nfix": [22, 25, 26, 29, 30], "fix": [22, 25, 26, 29, 30], "ntrait": [22, 25, 26, 29, 30, 34], "nmisc": [22, 25, 26, 29, 30], "naddit": [22, 25, 26, 29, 30], "50": [22, 24, 25, 26, 29, 30], "dummi": [22, 25, 26, 29, 30, 34], "beta": [22, 25, 26, 29, 30], "u_misc": [22, 25, 26, 29, 30], "u_a": [22, 25, 26, 29, 30], "zfill": [22, 23, 24, 25, 26, 29, 30, 34, 35, 37, 38, 39], "dtype": [22, 23, 24, 25, 26, 29, 30, 32, 34, 35, 37, 38, 39], "algmod": [22, 25, 26, 29, 30], "denseadditivelineargenomicmodel": [22, 25, 26, 29, 30], "model_nam": [22, 25, 26, 29, 30], "param": [22, 25, 26, 29, 30], "shape": [22, 23, 24, 25, 26, 29, 30, 32, 34, 35, 38, 39], "ntaxa": [22, 23, 24, 25, 26, 29, 30, 34, 35, 38, 39], "100": [22, 23, 24, 25, 26, 30, 31, 34, 35, 37, 39], "nvrnt": [22, 23, 24, 25, 26, 29, 30, 35, 37, 38, 39], "ngroup": [22, 23, 24, 25, 26, 30, 34, 35, 38, 39], "20": [22, 23, 24, 25, 26, 30, 34, 35, 37, 39], "nchrom": [22, 23, 24, 25, 26, 30, 35, 38, 39], "nphase": [22, 23, 24, 25, 26, 38, 39], "mat": [22, 23, 24, 25, 26, 29, 30, 34, 35, 38, 39], "randint": [22, 23, 24, 25, 26, 29, 30, 34, 35, 37, 38, 39], "size": [22, 23, 24, 25, 26, 29, 30, 34, 35, 38, 39], "astyp": [22, 23, 24, 25, 26, 29, 30, 35, 37, 38, 39], "int8": [22, 23, 24, 25, 26, 29, 30, 35, 38, 39], "taxon": [22, 23, 24, 25, 26, 30, 34, 35, 38, 39], "taxa_grp": [22, 23, 24, 25, 26, 30, 34, 35, 38, 39], "sort": [22, 23, 24, 25, 26], "variant": [22, 23, 24, 25, 26, 30, 35], "chromsom": [22, 23, 24, 25, 26, 30, 35, 38, 39], "assign": [22, 23, 24, 25, 26, 29, 30, 35, 36, 37, 38, 39], "vrnt_chrgrp": [22, 23, 24, 25, 26, 30, 35, 36, 37, 38, 39], "physic": [22, 23, 24, 25, 26, 30, 35, 36, 37, 38, 39], "vrnt_phypo": [22, 23, 24, 25, 26, 30, 35, 36, 37, 38, 39], "choic": [22, 23, 24, 25, 26, 30, 35, 37, 38, 39], "1000000": [22, 23, 24, 25, 26, 30, 35, 38, 39], "replac": [22, 23, 24, 25, 26, 30, 35, 38, 39], "fals": [22, 23, 24, 25, 26, 30, 35, 37, 38, 39], "vrnt_name": [22, 23, 24, 25, 26, 30, 35, 37, 38, 39], "snp": [22, 23, 24, 25, 26, 30, 35, 37, 38, 39], "phase": [22, 23, 24, 25, 26, 38, 40], "pgmat": [22, 23, 24, 25, 26, 30, 39], "densephasedgenotypematrix": [22, 23, 24, 25, 26, 30, 39], "ploidi": [22, 23, 24, 25, 26, 29, 30, 35, 38, 39], "nenv": [22, 25], "nrep": [22, 25], "ptprot": [22, 25], "g_e_phenotyp": [22, 25], "gpmod": [22, 25, 26], "sens": [22, 25], "herit": 22, "set_h2": [22, 25], "h2": [22, 25], "pheno_df": [22, 25], "without": [22, 25], "align": 22, "bvmat1": 22, "bvmat2": 22, "genotypingprotocol": 23, "genotypeprotocol": 23, "unphas": [23, 39], "denseunphasedgenotyp": 23, "flawless": 23, "whole": [23, 35, 37, 38], "deliber": 23, "open": 23, "dens": [23, 29, 30, 34, 35, 38, 39], "illustr": [23, 24, 27, 30, 31, 32, 33, 34, 35, 37, 38, 39], "creation": [23, 25, 27, 29, 30, 34, 35], "gtprot": 23, "shown": 23, "1000": [23, 24, 35, 39], "gmat": [23, 26, 29, 35], "check": 23, "isinst": 23, "densegenotypematrix": [23, 26, 35, 38, 39], "pre": [24, 31], "avail": [24, 26, 29, 38], "child": [24, 25, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "selfcross": 24, "fertil": 24, "twowaycross": 24, "twowaydhcross": 24, "doubl": 24, "haploid": [24, 26], "threewaycross": 24, "threewaydhcross": 24, "fourwaycross": 24, "fourwaydhcross": 24, "share": [24, 26, 27, 29, 30, 34, 35, 37, 38], "common": [24, 29, 30, 33, 34, 35, 38], "subsect": [24, 25, 26, 29, 30, 34, 35], "nparent": [24, 26], "counter": 24, "automat": [24, 34, 37], "straightforward": 24, "zero": [24, 26, 32, 34], "mate_self": 24, "mate_2w": 24, "mate_2wdh": 24, "mate_3w": 24, "mate_3wdh": 24, "mate_4w": 24, "mate_4wdh": 24, "progeny_count": 24, "family_count": 24, "rest": 24, "mtprot": 24, "accept": [24, 25, 29, 37], "indic": [24, 30, 34, 35, 37, 38, 39], "vrnt_genpo": [24, 30, 35, 36, 37, 38, 39], "uniform": [24, 30, 32, 33, 35], "crossov": [24, 38, 39], "vrnt_xoprob": [24, 35, 38, 39], "configu": 24, "xconfig": [24, 26], "nmate": [24, 26], "nprogeni": [24, 26, 30], "nself": [24, 30], "induc": 25, "increas": 25, "truephenotyp": 25, "environment": 25, "gxe": 25, "interact": 25, "coupl": 25, "amount": 25, "var_err": 25, "add": [25, 29], "definit": [25, 26, 32, 34, 36, 37, 38, 39], "implment": 25, "broad": [25, 29], "respect": [25, 26, 29, 30, 34, 36], "founder": 25, "said": [26, 29], "pair": [26, 36], "doe": [26, 29], "matter": 26, "broken": 26, "major": [26, 38, 39], "mateselectionprotocol": 26, "binari": [26, 31, 32, 33, 39], "integ": [26, 30, 31, 32, 33, 34, 35, 38, 39], "subset": [26, 31, 32, 33], "binaryselectionprotocol": 26, "integerselectionprotocol": 26, "realselectionprotocol": 26, "subsetselectionprotocol": 26, "binarymateselectionprotocol": 26, "integermateselectionprotocol": 26, "realmateselectionprotocol": 26, "subsetmateselectionprotocol": 26, "estimatedbreedingvaluebinaryselect": 26, "estimatedbreedingvalueintegerselect": 26, "estimatedbreedingvaluerealselect": 26, "estimatedbreedingvaluesubsetselect": 26, "expectedmaximumbreedingvaluebinaryselect": 26, "expectedmaximumbreedingvalueintegerselect": 26, "expectedmaximumbreedingvaluerealselect": 26, "expectedmaximumbreedingvaluesubsetselect": 26, "familyestimatedbreedingvaluebinaryselect": 26, "familyestimatedbreedingvalueintegerselect": 26, "familyestimatedbreedingvaluerealselect": 26, "familyestimatedbreedingvaluesubsetselect": 26, "genomicestimatedbreedingvaluebinaryselect": 26, "genomicestimatedbreedingvalueintegerselect": 26, "genomicestimatedbreedingvaluerealselect": 26, "genomicestimatedbreedingvaluesubsetselect": 26, "genotypebuildersubsetselect": 26, "builder": 26, "optimalcontributionbinaryselect": 26, "optimalcontributionintegerselect": 26, "optimalcontributionrealselect": 26, "optimalcontributionsubsetselect": 26, "optimalhaploidvaluebinaryselect": 26, "optimalhaploidvalueintegerselect": 26, "optimalhaploidvaluerealselect": 26, "optimalhaploidvaluesubsetselect": 26, "optimalpopulationvaluesubsetselect": 26, "randombinaryselect": 26, "randomintegerselect": 26, "randomrealselect": 26, "randomsubsetselect": 26, "usefulnesscriterionbinaryselect": 26, "usefulnesscriterionintegerselect": 26, "usefulnesscriterionrealselect": 26, "usefulnesscriterionsubsetselect": 26, "criterion": 26, "weightedgenomicbinaryselect": 26, "weightedgenomicintegerselect": 26, "weightedgenomicrealselect": 26, "weightedgenomicsubsetselect": 26, "weight": [26, 32, 33], "ebv": 26, "estimatedbreedingvalueselect": 26, "embv": 26, "expectedmaximumbreedingvalueselect": 26, "familyestimatedbreedingvalueselect": 26, "gebv": [26, 29], "genomicestimatedbreedingvalueselect": 26, "gb": 26, "genotypebuilderselect": 26, "optimalcontributionselect": 26, "ohv": 26, "optimalhaploidvalueselect": 26, "opv": 26, "optimalpopulationvalueselect": 26, "randomselect": 26, "uc": 26, "usefulnesscriterionselect": 26, "weightedgenomicselect": 26, "nselindiv": 26, "ncross": [26, 30], "consid": 26, "per": 26, "event": [26, 35], "later": [26, 29, 38, 39], "nobj": [26, 32, 33], "obj_wt": [26, 32, 33], "obj_tran": 26, "latentfn": 26, "obj_trans_kwarg": 26, "latent": 26, "nineqcv": [26, 32, 33], "violat": [26, 32, 33], "ineqcv_wt": [26, 32, 33], "ineqcv_tran": 26, "ineqcv_trans_kwarg": 26, "neqcv": [26, 32, 33], "eqcv_wt": [26, 32, 33], "eqcv_tran": 26, "eqcv_trans_kwarg": 26, "ndset_wt": 26, "nondomin": 26, "ndset_tran": 26, "ndset_trans_kwarg": 26, "rng": 26, "soalgo": [26, 31], "moalgo": [26, 31], "opim": 26, "selprot": 26, "40": 26, "here": [26, 38], "we": 26, "selprot_unconstrain": 26, "ten": 26, "total": 26, "suppos": 26, "wish": 26, "negat": 26, "sum": [26, 32], "equival": [26, 35], "our": 26, "second": 26, "greater": 26, "than": 26, "decnvec": 26, "ndarrai": [26, 32], "latentvec": 26, "maskvec": 26, "vector": [26, 32, 33], "ndecn": [26, 32, 33], "mask": [26, 29, 38, 39], "extract": [26, 29, 30, 34, 35, 38], "minvec": 26, "minimum": [26, 29], "neg": 26, "bool": 26, "don": 26, "t": 26, "care": 26, "inf": 26, "float": [26, 32], "selprot_constrain": 26, "prob_unconstrain": 26, "ptdf": 26, "bvmat": 26, "prob_constrain": 26, "decn_spac": [26, 32, 33], "eval_unconstrain": 26, "evalfn": [26, 32], "eval_constrain": 26, "sosolv": 26, "soln_constrain": 26, "soln_decn": [26, 33], "soln_obj": [26, 33], "soln_ineqcv": [26, 33], "soltuion": 26, "soln_eqcv": [26, 33], "mosolv": 26, "soln_unconstrain": 26, "score": 26, "distanc": 26, "weigh": 26, "invari": 26, "scale": [26, 34], "selcfg_constrain": 26, "view": 26, "pareto": 26, "frontier": 26, "closest": 26, "project": 26, "selcfg_unconstrain": 26, "bulk": 27, "synchron": 27, "underneath": 27, "begin": 27, "prng": 27, "reproduc": 27, "remain": 27, "independ": 27, "despit": 27, "mechan": [27, 29], "48823": 27, "instanc": [27, 36, 37], "spawn": 27, "stream": 27, "prng_list": 27, "central": 29, "behavior": 29, "epistasi": 29, "addititivelineargenomicmodel": 29, "exhibit": 29, "present": 29, "additivedomin": 29, "additivedominanceepistat": 29, "denselineargenomicmodel": 29, "categor": 29, "extra": 29, "case": 29, "been": [29, 34, 37, 39], "hyperparamet": 29, "hyperparam": 29, "dimension": 29, "explanatori": 29, "variabl": [29, 32, 33], "predictor": 29, "intension": 29, "becaus": 29, "nexplan": 29, "nparam": 29, "top": 29, "regress": 29, "mix": 29, "nexplan_beta": 29, "nparam_beta": 29, "nexplan_u": 29, "nparam_u": 29, "u": [29, 37], "subdivid": 29, "subdivis": 29, "nexplan_u_misc": 29, "nparam_u_misc": 29, "nexplan_u_a": 29, "nparam_u_a": 29, "nexplan_u_d": 29, "nparam_u_d": 29, "u_d": 29, "nexplan_u_i": 29, "nparam_u_i": 29, "u_i": 29, "aforement": 29, "lowest": 29, "denomin": 29, "norm": 29, "freedom": 29, "impract": 29, "from_pandas_dict": 29, "df_dict": 29, "mod": 29, "similar": [29, 30, 35], "manner": [29, 30, 35], "dic": 29, "from_csv_dict": 29, "reload": 29, "from_hdf5": [29, 30, 34, 35, 38, 39], "h5": [29, 30, 34, 35, 38, 39], "tmp": [29, 30, 34, 35, 37, 38, 39], "deepcopi": [29, 30, 34, 35, 37, 38, 39], "recurs": [29, 30, 34, 35, 37, 38], "made": [29, 30, 34, 35, 37], "predict_numpi": 29, "x": [29, 32], "z": 29, "cvobj": 29, "gtobj": 29, "b": [29, 30], "y": 29, "normal": [29, 34], "score_numpi": 29, "ptobj": 29, "gebv_numpi": 29, "gegv": 29, "gegv_numpi": 29, "var_g_numpi": 29, "var_g": 29, "var": 29, "var_a_numpi": 29, "var_a": 29, "equilibrium": [29, 30], "p": [29, 38, 39], "afreq": [29, 38, 39], "bulmer_numpi": 29, "much": 29, "valuabl": 29, "variat": 29, "usl": 29, "usl_numpi": 29, "achiev": [29, 32], "fixat": 29, "lsl_numpi": 29, "lsl": 29, "when": 29, "count": [29, 39], "locu": [29, 38, 39], "facount": 29, "frequenc": [29, 39], "fafreq": 29, "faavail": 29, "fafix": 29, "loci": [29, 37, 39], "dacount": 29, "dafreq": 29, "daavail": 29, "dafix": 29, "anoth": [29, 38], "to_pandas_dict": 29, "to_csv_dict": 29, "to_hdf5": [29, 30, 34, 35, 38, 39], "saved_algmod": 29, "disequilibrium": 30, "densegeneticvariancematrix": 30, "semi": [30, 32, 33, 35], "denseadditivegeneticvariancematrix": 30, "densegenicvariancematrix": 30, "denseadditivegenicvariancematrix": 30, "densetwowaydhadditivegenicvariancematrix": 30, "densethreewaydhadditivegenicvariancematrix": 30, "densefourwaydhadditivegenicvariancematrix": 30, "densedihybriddhadditivegenicvariancematrix": 30, "epgc": 30, "axi": [30, 32, 34, 35, 38, 39], "raw": [30, 34, 35, 37, 38, 39], "mat_ndim": [30, 34, 35, 38, 39], "dimens": [30, 34, 35, 38, 39], "mat_shap": [30, 34, 35, 38, 39], "quick": [30, 34, 35], "taxa_axi": [30, 34, 35, 38, 39], "along": [30, 34, 38, 39], "label": [30, 34, 35, 37, 38, 39], "taxa_grp_nam": [30, 34, 35, 38, 39], "taxa_grp_stix": [30, 34, 35, 38, 39], "inclus": [30, 34, 35, 38, 39], "taxa_grp_spix": [30, 34, 35, 38, 39], "stop": [30, 34, 35, 37, 38, 39], "exclus": [30, 34, 35, 38, 39], "taxa_grp_len": [30, 34, 35, 38, 39], "trait_axi": [30, 34], "regard": [30, 32, 35], "nsquar": [30, 35], "square_ax": [30, 35], "square_axes_len": [30, 35], "gvmat": 30, "from_gmod": 30, "from_algmod": 30, "group_vrnt": [30, 35, 38, 39], "gmapfn": 30, "haldanemapfunct": [30, 36], "from_panda": [30, 34, 35], "df": [30, 34, 35], "femal": 30, "male": 30, "yield": 30, "tmp_vmat": 30, "female_col": 30, "female_grp_col": 30, "male_col": 30, "male_grp_col": 30, "variance_col": 30, "tmp_gvmat": 30, "from_csv": [30, 34, 35], "to_csv": [30, 34, 35], "saved_df": 30, "format": [30, 32, 34, 38, 39], "saved_vmat": 30, "saved_gvmat": 30, "level": [30, 34, 35, 37, 38], "origin": [30, 34, 35, 37, 38, 39], "affect": [30, 34, 35, 37, 38], "bound": [30, 34, 35], "occur": [30, 34, 35, 37, 38, 39], "down": [30, 34, 35, 37], "deepest": [30, 34, 35, 37, 38], "togeth": [30, 34, 35], "newvmat": 30, "newgvmat": 30, "adjoin_taxa": [30, 34, 35, 38, 39], "adjoin_trait": [30, 34], "delete_taxa": [30, 34, 35, 38, 39], "slice": [30, 34, 35, 38, 39], "sequenc": [30, 34, 35, 38, 39], "delete_trait": [30, 34], "insert_taxa": [30, 34, 38, 39], "insert_trait": [30, 34], "select_taxa": [30, 34, 35, 38, 39], "select_trait": [30, 34], "segment": [30, 34, 35, 36, 38], "append_taxa": [30, 34, 35, 38, 39], "append_trait": [30, 34], "seen": [30, 34, 35, 36, 38], "remove_taxa": [30, 34, 35, 38, 39], "remove_trait": [30, 34], "incorp": [30, 34, 35, 38, 39], "locat": [30, 34, 35, 38], "incorp_taxa": [30, 34, 35, 38, 39], "incorp_trait": [30, 34], "concat": [30, 34, 35, 38, 39], "concat_taxa": [30, 34, 38, 39], "concat_trait": [30, 34], "alphanumer": [30, 34, 35, 38], "arang": [30, 34, 35, 37, 38, 39], "shuffl": [30, 34, 35, 37, 38, 39], "reorder_taxa": [30, 34, 35, 38, 39], "reorder_trait": [30, 34], "indirect": [30, 34, 35, 37, 38], "stabl": [30, 32, 35, 37], "key1": [30, 34, 35, 37, 38, 39], "key2": [30, 34, 35, 37, 38, 39], "lexsort_taxa": [30, 34, 35, 38, 39], "sort_taxa": [30, 34, 35, 38, 39], "sort_trait": [30, 34], "group_taxa": [30, 34, 35, 38, 39], "is_group": [30, 34, 35, 37, 38, 39], "is_grouped_taxa": [30, 34, 35, 38, 39], "deal": 30, "is_squar": 30, "boolean": 30, "to_panda": [30, 34, 35], "out_df": 30, "binaryoptimizationalgorithm": 31, "integeroptimizationalgorithm": 31, "realoptimizationalgorithm": 31, "subsetoptimizationalgorithm": 31, "binarygeneticalgorithm": 31, "integergeneticalgorithm": 31, "realgeneticalgorithm": 31, "subsetgeneticalgorithm": 31, "nsga2binarygeneticalgorithm": 31, "nsga": 31, "ii": 31, "nsga2integergeneticalgorithm": 31, "nsga2realgeneticalgorithm": 31, "nsga2subsetgeneticalgorithm": 31, "steepestdescentsubsethillclimb": 31, "steepest": 31, "descent": 31, "hill": 31, "climb": 31, "climber": 31, "ngen": 31, "pop_siz": 31, "intuit": 31, "sosoln": [31, 33], "soprob": [31, 32], "mosoln": [31, 33], "moprob": [31, 32], "binaryproblem": 32, "integerproblem": 32, "realproblem": 32, "subsetproblem": 32, "inherit": 32, "webpag": [32, 40], "overlai": 32, "boundari": [32, 33, 36], "decn_space_low": [32, 33], "decn_space_upp": [32, 33], "_evalu": 32, "overriden": 32, "compat": 32, "better": 32, "sphere": 32, "min_": 32, "mathbf": 32, "f": 32, "sum_": 32, "n_": 32, "decn": 32, "x_": 32, "dummysingleobjectiverealproblem": 32, "init": 32, "unchang": 32, "math": 32, "v_": 32, "obj": 32, "w_": 32, "odot": 32, "f_": 32, "ineqcv": 32, "g_": 32, "eqcv": 32, "h_": 32, "keepdim": 32, "dual": 32, "dummymultiobjectiverealproblem": 32, "repeat": [32, 33, 34], "stack": [32, 33], "cand": 32, "obj1": 32, "ineqcv1": 32, "eqcv1": 32, "obj2": 32, "ineqcv2": 32, "eqcv2": 32, "unpack": 32, "zip": 32, "2d": 32, "binarysolut": 33, "integersolut": 33, "realsolut": 33, "subsetsolut": 33, "nsoln": 33, "sometim": [34, 37], "reconstitut": 34, "un": 34, "row": [34, 35, 38], "densebreedingvaluematrix": 34, "around": 34, "chunk": 34, "from_numpi": 34, "101": 34, "trait1": 34, "trait2": 34, "trait3": 34, "explicit": 34, "taxa_grp_col": [34, 35], "sample_breeding_valu": 34, "ax": [34, 35, 38], "whether": [34, 35, 37, 38, 39], "tmax": 34, "largest": 34, "targmax": 34, "tmin": 34, "smallest": 34, "targmin": 34, "tmean": 34, "trang": 34, "tstd": 34, "tvar": 34, "saved_breeding_valu": 34, "maodul": 35, "densecoancestrymatrix": 35, "densemolecularcoancestrymatrix": 35, "molecular": 35, "densevanradencoancestrymatrix": 35, "vanraden": 35, "2008": 35, "denseyangcoancestrymatrix": 35, "yang": 35, "fiel": 35, "from_gmat": 35, "vrnt_hapgrp": [35, 38, 39], "vrnt_hapalt": [35, 38, 39], "vrnt_hapref": [35, 38, 39], "vrnt_mask": [35, 38, 39], "saved_coancestry_matrix": 35, "sample_coancestry_matrix": 35, "guarante": 35, "correct": [35, 37], "coordin": 35, "NOT": 35, "bare": 35, "bone": 35, "mat_asformat": 35, "necessari": [35, 37], "is_positive_semidefinit": 35, "convex": 35, "apply_jitt": 35, "try": 35, "vari": 35, "particular": 35, "max_inbreed": 35, "min_inbreed": 35, "max": 35, "min": 35, "kosambimapfunct": 36, "basal": 36, "arbitrari": 36, "mapfn": 36, "morgan": 36, "d": 36, "r_haldan": 36, "r_kosambi": 36, "invmapfn": 36, "invers": [36, 38], "invert": 36, "d_haldan": 36, "d_kosambi": 36, "rprob1g": 36, "rprob1p": 36, "rprob2g": 36, "rprob2p": 36, "success": 36, "alwai": 36, "xoprob": 36, "inter": 36, "extendedgeneticmap": 37, "scope": [37, 38], "9": 37, "chrom": 37, "123456789": 37, "28": 37, "vrnt_stop": 37, "vrnt_fncode": 37, "who": 37, "empir": 37, "organ": 37, "maiz": 37, "nam": 37, "publish": 37, "mcmullen": 37, "et": 37, "al": 37, "2009": 37, "from_egmap": 37, "mcmullen_2009_us_nam": 37, "egmap": 37, "auto_group": 37, "auto_build_splin": 37, "constitut": 37, "vrnt_chrgrp_name": [37, 38, 39], "vrnt_chrgrp_stix": [37, 38, 39], "vrnt_chrgrp_spix": [37, 38, 39], "vrnt_chrgrp_len": [37, 38, 39], "spline_kind": 37, "spline_fill_valu": 37, "fill": 37, "At": 37, "order": 37, "prioriti": 37, "place": 37, "often": 37, "occasion": 37, "disagr": 37, "caus": 37, "error": 37, "assembli": 37, "unfortun": 37, "issu": 37, "elementwis": 37, "congruent": 37, "is_congru": 37, "agreement": 37, "remove_discrep": 37, "Be": 37, "mind": 37, "inspect": 37, "superior": 37, "build_splin": 37, "has_splin": 37, "interp_genpo": 37, "everyth": 37, "chrgrp": 37, "phypo": 37, "18203210": 37, "19293034": 37, "20110347": 37, "20474722": 37, "21398386": 37, "gentic": 37, "genpo": 37, "n": [38, 39], "sequenti": [38, 39], "haplotyp": [38, 39], "ploid": 38, "from_vcf": [38, 39], "widiv_2000snp": [38, 39], "gz": [38, 39], "memori": [38, 39], "itself": 38, "mat_format": [38, 39], "post": [38, 39], "vrnt_axi": [38, 39], "belong": [38, 39], "previou": [38, 39], "reflect": 38, "leav": 38, "unalt": 38, "adjoin_vrnt": [38, 39], "delete_vrnt": [38, 39], "insert_vrnt": [38, 39], "select_vrnt": [38, 39], "append_vrnt": [38, 39], "remove_vrnt": [38, 39], "incorp_vrnt": [38, 39], "concat_vrnt": [38, 39], "accord": 38, "reorder_vrnt": [38, 39], "lexsort_vrnt": [38, 39], "sort_vrnt": [38, 39], "vrnt": [38, 39], "is_grouped_vrnt": [38, 39], "acount": [38, 39], "int32": [38, 39], "float32": [38, 39], "presens": 38, "apoli": [38, 39], "gtcount": [38, 39], "homozyg": [38, 39], "recess": 38, "heterozyg": 38, "gtfreq": [38, 39], "maf": [38, 39], "meh": [38, 39], "tacount": [38, 39], "tafreq": [38, 39], "disk": 38, "saved_genotyp": [38, 39], "except": 39, "member": 39, "were": 39, "correctli": 39, "problemat": 39, "phase_axi": 39, "deep": 39, "polymorph": 39, "minor": 39, "heterozygos": 39, "directori": 40, "logbook": 40}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"advanc": [0, 14], "pybrop": [0, 2, 4, 5, 6, 8, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], "usag": [0, 40], "contribut": 2, "frequent": 3, "ask": 3, "question": 3, "implement": [4, 26], "softwar": [4, 5], "architectur": [4, 5, 7], "philosophi": [5, 13], "solid": 5, "design": 5, "principl": 5, "how": 5, "us": [5, 14], "singl": [5, 26, 31, 32, 33], "responsibl": 5, "srp": 5, "open": 5, "close": 5, "ocp": 5, "liskov": 5, "substitutuion": 5, "lsp": 5, "interfac": [5, 26], "segreg": 5, "isp": 5, "depend": [5, 14], "invers": [5, 35], "dip": 5, "core": [6, 40], "infrastructur": [6, 40], "random": [6, 27], "number": [6, 27], "gener": [6, 24, 26, 27, 29, 30, 34, 35, 38, 39], "error": [6, 25], "subroutin": 6, "breed": [7, 11, 13, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 34, 40], "simul": [7, 13, 24, 40], "data": [7, 8, 9, 10, 11, 13], "type": [7, 8, 9, 10, 11, 13], "introduct": 7, "univers": 7, "program": [7, 14, 16, 17, 18, 19, 20, 21, 40], "algorithm": [7, 10, 26, 31], "protocol": [7, 13, 22, 23, 24, 25, 26, 40], "genotyp": [7, 11, 23, 29, 38, 39], "phenotyp": [7, 25], "valu": [7, 11, 22, 29, 34, 35], "estim": [7, 22, 29], "mate": [7, 18, 19, 24], "select": [7, 18, 20, 21, 26, 29, 30, 34, 35, 38, 39], "problem": [7, 10, 26, 31, 32, 33], "solut": [7, 10, 32, 33], "oper": [7, 13, 16, 17, 19, 20, 21, 40], "initi": [7, 17, 18], "parent": [7, 18, 20], "evalut": 7, "survivor": [7, 18, 21], "logbook": [7, 18], "larg": 7, "scale": 7, "genom": [9, 23, 25, 29, 30, 40], "model": [9, 25, 29, 30, 40], "varianc": [9, 25, 29, 30, 34], "matric": [9, 11, 30, 34, 35, 38, 39], "optim": [10, 26, 31, 32, 33, 40], "popul": [11, 29, 38, 40], "genet": [11, 29, 30, 36, 37, 40], "map": [11, 36, 37], "function": [11, 30, 36], "phase": [11, 39], "coancestri": [11, 35], "fundament": [12, 13], "get": 14, "start": 14, "basic": 14, "instal": 14, "python": 14, "requir": 14, "packag": 14, "releas": 14, "version": 14, "knowledg": 14, "prerequisit": 14, "linux": 14, "development": 14, "librari": 14, "command": 14, "main": 15, "page": 15, "what": 15, "i": [15, 35], "user": 15, "guid": 15, "content": 15, "indic": 15, "tabl": 15, "evalu": [16, 18, 32], "class": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "famili": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "overview": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "summari": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "op": [16, 17, 18, 19, 20, 21], "eval": 16, "modul": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], "properti": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 32, 33, 34, 35, 37, 38, 39], "load": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "defin": [16, 17, 18, 19, 20, 21], "creat": [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 34, 35, 37, 38, 39], "individu": [16, 23, 25], "init": 17, "log": 18, "state": 18, "after": 18, "reset": 18, "write": [18, 39], "file": [18, 29, 30, 34, 35, 37, 38, 39], "individiu": 19, "psel": 20, "ssel": 21, "prot": [22, 23, 24, 25, 26], "bv": 22, "gt": 23, "from": [23, 29, 30, 34, 35, 37, 38, 39], "matingprotocol": 24, "progeni": 24, "pt": 25, "phenotypingprotocol": 25, "set": 25, "trait": [25, 29, 30, 34], "herit": 25, "abstract": [26, 29], "sel": 26, "concret": [26, 29], "selectionprotocol": 26, "setup": 26, "unconstrain": 26, "constrain": 26, "object": [26, 31, 32, 33, 35, 36], "multi": [26, 31, 32, 33], "seed": 27, "multipl": 27, "pseudorandom": 27, "exampl": [28, 40], "doubl": 28, "haploid": 28, "gmod": 29, "genomicmodel": 29, "linear": 29, "addit": 29, "lineargenomicmodel": 29, "additivelineargenomicmodel": 29, "domin": 29, "additivedominancelineargenomicmodel": 29, "epistat": 29, "coeffici": 29, "additivedominanceepistaticlineargenomicmodel": 29, "raw": 29, "numpi": [29, 30, 34, 35, 37, 38, 39], "arrai": [29, 30, 34, 35, 37, 38, 39], "dictionari": 29, "panda": [29, 30, 34, 35], "datafram": [29, 30, 34, 35], "csv": [29, 30, 34, 35], "name": 29, "hdf5": [29, 30, 34, 35, 38, 39], "copi": [29, 30, 34, 35, 37, 38, 39], "shallow": [29, 30, 34, 35, 37, 38], "deep": [29, 30, 34, 35, 37, 38], "predict": 29, "method": [29, 35], "score": 29, "accuraci": 29, "calcul": [29, 30, 35, 36], "term": 29, "genic": [29, 30], "bulmer": 29, "effect": 29, "limit": 29, "upper": 29, "lower": 29, "favor": 29, "allel": [29, 38], "metric": 29, "deleteri": 29, "export": [29, 30, 34, 35, 38], "matrix": [30, 34, 35, 38, 39], "vmat": 30, "geneticvariancematrix": 30, "genicvariancematrix": 30, "taxa": [30, 34, 35, 38, 39], "squar": [30, 35], "On": [30, 34, 35, 38, 39], "element": [30, 34, 35, 37, 38, 39], "manipul": [30, 34, 35, 38, 39], "adjoin": [30, 34, 35, 38, 39], "delet": [30, 34, 35, 38, 39], "insert": [30, 34, 35, 38, 39], "In": [30, 34, 35, 38, 39], "place": [30, 34, 35, 38, 39], "append": [30, 34, 35, 38, 39], "remov": [30, 34, 35, 37, 38, 39], "incorpor": [30, 34, 35, 38, 39], "concaten": [30, 34, 35, 38, 39], "group": [30, 34, 35, 37, 38, 39], "sort": [30, 34, 35, 37, 38, 39], "reorder": [30, 34, 35, 37, 38, 39], "lexsort": [30, 34, 35, 37, 38, 39], "determin": [30, 35], "whether": 30, "all": 30, "ax": 30, "ar": 30, "equal": 30, "length": 30, "opt": [31, 32, 33], "algo": 31, "construct": [31, 32, 33, 36], "minim": 31, "prob": 32, "pymoo": 32, "specif": [32, 35], "deriv": 32, "candid": 32, "soln": 33, "popgen": [34, 35, 36, 37, 38, 39], "bvmat": 34, "breedingvaluematrix": 34, "statist": [34, 35, 38, 39], "maximum": [34, 35], "each": 34, "minimum": [34, 35], "mean": [34, 35, 38], "rang": 34, "standard": [34, 40], "deviat": 34, "unscal": 34, "de": 34, "center": 34, "cmat": 35, "coancestrymatrix": 35, "genotypematrix": [35, 38], "kinship": 35, "retriev": 35, "format": 35, "posit": [35, 37], "semidefinit": 35, "appli": 35, "jitter": 35, "along": 35, "diagon": 35, "attain": 35, "inbreed": 35, "gmap": [36, 37], "recombin": 36, "probabl": 36, "distanc": 36, "sequenti": 36, "pairwis": 36, "read": 37, "marker": [37, 38, 39], "variant": [37, 38, 39], "geneticmap": 37, "spline": 37, "congruenc": 37, "check": 37, "discrep": 37, "build": 37, "interpol": 37, "gmat": [38, 39], "vcf": [38, 39], "count": 38, "frequenc": 38, "polymorph": 38, "presenc": 38, "minor": 38, "expect": 38, "heterozygos": 38, "save": [38, 39], "phasedgenotypematrix": 39}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"Advanced PyBrOpS Usage": [[0, "advanced-pybrops-usage"]], "Contributing to PyBrOpS": [[2, "contributing-to-pybrops"]], "Frequently Asked Questions": [[3, "frequently-asked-questions"]], "PyBrOpS Implemented Software Architecture": [[4, "pybrops-implemented-software-architecture"]], "PyBrOpS Software Architecture Philosophy": [[5, "pybrops-software-architecture-philosophy"]], "SOLID Design Principles": [[5, "id1"]], "How PyBrOpS uses SOLID Design Principles": [[5, "how-pybrops-uses-solid-design-principles"]], "Use of the Single Responsiblity Principle (SRP)": [[5, "use-of-the-single-responsiblity-principle-srp"]], "Use of the Open-Closed Principle (OCP)": [[5, "use-of-the-open-closed-principle-ocp"]], "Use of the Liskov Substitutuion Principle (LSP)": [[5, "use-of-the-liskov-substitutuion-principle-lsp"]], "Use of the Interface Segregation Principle (ISP)": [[5, "use-of-the-interface-segregation-principle-isp"]], "Use of the Dependency Inversion Principle (DIP)": [[5, "use-of-the-dependency-inversion-principle-dip"]], "Core PyBrOpS Infrastructure": [[6, "core-pybrops-infrastructure"]], "Random Number Generation": [[6, "random-number-generation"], [27, "random-number-generation"]], "Error Subroutines": [[6, "error-subroutines"]], "Breeding Simulation Data Types": [[7, "breeding-simulation-data-types"]], "Introduction to the Universal Breeding Program Algorithm": [[7, "introduction-to-the-universal-breeding-program-algorithm"]], "Breeding Protocols": [[7, "breeding-protocols"], [13, "breeding-protocols"]], "Genotyping Protocols": [[7, "genotyping-protocols"], [23, "genotyping-protocols"]], "Phenotyping Protocols": [[7, "phenotyping-protocols"], [25, "phenotyping-protocols"]], "Breeding Value Estimation Protocols": [[7, "breeding-value-estimation-protocols"], [22, "breeding-value-estimation-protocols"]], "Mating Protocols": [[7, "mating-protocols"], [24, "mating-protocols"]], "Selection Protocols": [[7, "selection-protocols"], [26, "selection-protocols"]], "Selection Problems for Selection Protocols": [[7, "selection-problems-for-selection-protocols"]], "Selection Solutions for Selection Protocols": [[7, "selection-solutions-for-selection-protocols"]], "Breeding Operators": [[7, "breeding-operators"], [13, "breeding-operators"]], "Initialization Operators": [[7, "initialization-operators"], [17, "initialization-operators"]], "Parental Selection Operators": [[7, "parental-selection-operators"], [20, "parental-selection-operators"]], "Mating Operators": [[7, "mating-operators"], [19, "mating-operators"]], "Evalution Operators": [[7, "evalution-operators"]], "Survivor Selection Operators": [[7, "survivor-selection-operators"], [21, "survivor-selection-operators"]], "Logbooks": [[7, "logbooks"], [18, "logbooks"]], "Large-Scale Architecture Operators": [[7, "large-scale-architecture-operators"]], "Data Types in PyBrOpS": [[8, "data-types-in-pybrops"]], "PyBrOpS Data Types:": [[8, null]], "Genomic Modeling Data Types": [[9, "genomic-modeling-data-types"]], "Genomic Models": [[9, "genomic-models"], [29, "genomic-models"]], "Variance Matrices": [[9, "variance-matrices"], [30, "variance-matrices"]], "Optimization Data Types": [[10, "optimization-data-types"]], "Optimization Problems": [[10, "optimization-problems"], [32, "optimization-problems"]], "Optimization Solutions": [[10, "optimization-solutions"], [33, "optimization-solutions"]], "Optimization Algorithms": [[10, "optimization-algorithms"], [31, "optimization-algorithms"]], "Population Genetics Data Types": [[11, "population-genetics-data-types"]], "Genetic Maps": [[11, "genetic-maps"], [37, "genetic-maps"]], "Genetic Map Functions": [[11, "genetic-map-functions"], [36, "genetic-map-functions"]], "Genotype Matrices": [[11, "genotype-matrices"], [38, "genotype-matrices"]], "Phased Genotype Matrices": [[11, "phased-genotype-matrices"], [39, "phased-genotype-matrices"]], "Breeding Value Matrices": [[11, "breeding-value-matrices"], [34, "breeding-value-matrices"]], "Coancestry Matrices": [[11, "coancestry-matrices"], [35, "coancestry-matrices"]], "PyBrOpS Fundamentals": [[12, "pybrops-fundamentals"]], "PyBrOpS Fundamentals:": [[12, null]], "PyBrOpS Simulation Philosophy": [[13, "pybrops-simulation-philosophy"]], "Fundamental Data Types": [[13, "fundamental-data-types"]], "Getting Started": [[14, "getting-started"]], "Basic Installation": [[14, "basic-installation"]], "Python installation requirements": [[14, "python-installation-requirements"]], "Python package dependencies": [[14, "python-package-dependencies"]], "Release version installation": [[14, "release-version-installation"]], "Basic Programming Knowledge Prerequisites": [[14, "basic-programming-knowledge-prerequisites"]], "PyBrOpS Programming Knowledge Prerequisites": [[14, "id1"]], "Using PyBrOpS": [[14, "using-pybrops"]], "Advanced Installation": [[14, "advanced-installation"]], "Linux installation requirements": [[14, "linux-installation-requirements"]], "Linux Developmental Library Installation Commands": [[14, "id2"]], "Developmental version installation": [[14, "developmental-version-installation"]], "PyBrOpS Main Page": [[15, "pybrops-main-page"]], "What is PyBrOpS?": [[15, "what-is-pybrops"]], "PyBrOpS User Guide": [[15, "pybrops-user-guide"]], "Contents:": [[15, null]], "Indices and tables": [[15, "indices-and-tables"]], "Evaluation Operators": [[16, "evaluation-operators"]], "Class Family Overview": [[16, "class-family-overview"], [17, "class-family-overview"], [18, "class-family-overview"], [19, "class-family-overview"], [20, "class-family-overview"], [21, "class-family-overview"], [22, "class-family-overview"], [23, "class-family-overview"], [24, "class-family-overview"], [25, "class-family-overview"], [26, "class-family-overview"], [29, "class-family-overview"], [30, "class-family-overview"], [31, "class-family-overview"], [32, "class-family-overview"], [33, "class-family-overview"], [34, "class-family-overview"], [35, "class-family-overview"], [36, "class-family-overview"], [37, "class-family-overview"], [38, "class-family-overview"], [39, "class-family-overview"]], "Summary of Evaluation Operators": [[16, "summary-of-evaluation-operators"]], "Summary of classes in the pybrops.breed.op.eval module": [[16, "id1"]], "Evaluation Operator Properties": [[16, "evaluation-operator-properties"]], "Loading Class Modules": [[16, "loading-class-modules"], [17, "loading-class-modules"], [18, "loading-class-modules"], [19, "loading-class-modules"], [20, "loading-class-modules"], [21, "loading-class-modules"], [22, "loading-class-modules"], [23, "loading-class-modules"], [24, "loading-class-modules"], [25, "loading-class-modules"], [26, "loading-class-modules"], [38, "loading-class-modules"], [39, "loading-class-modules"]], "Defining Evaluation Operators": [[16, "defining-evaluation-operators"]], "Creating Evaluation Operators": [[16, "creating-evaluation-operators"]], "Evaluation of Individuals in a Breeding Program": [[16, "evaluation-of-individuals-in-a-breeding-program"]], "Summary of Initialization Operators": [[17, "summary-of-initialization-operators"]], "Summary of classes in the pybrops.breed.op.init module": [[17, "id1"]], "Initialization Operators Properties": [[17, "initialization-operators-properties"]], "Defining Initialization Operators": [[17, "defining-initialization-operators"]], "Creating Initialization Operators": [[17, "creating-initialization-operators"]], "Initialization of a Breeding Program": [[17, "initialization-of-a-breeding-program"]], "Summary of Logbook Classes": [[18, "summary-of-logbook-classes"]], "Summary of classes in the pybrops.breed.op.log module": [[18, "id1"]], "Logbook Properties": [[18, "logbook-properties"]], "Summary of Logbook properties": [[18, "id2"]], "Defining Logbooks": [[18, "defining-logbooks"]], "Creating Logbooks": [[18, "creating-logbooks"]], "Logging States in a Breeding Program": [[18, "logging-states-in-a-breeding-program"]], "Logging after breeding program initialization": [[18, "logging-after-breeding-program-initialization"]], "Logging after breeding program parent selection": [[18, "logging-after-breeding-program-parent-selection"]], "Logging after breeding program mating": [[18, "logging-after-breeding-program-mating"]], "Logging after breeding program evaluation": [[18, "logging-after-breeding-program-evaluation"]], "Logging after breeding program survivor selection": [[18, "logging-after-breeding-program-survivor-selection"]], "Resetting a Logbook": [[18, "resetting-a-logbook"]], "Writing a Logbook to a File": [[18, "writing-a-logbook-to-a-file"]], "Summary of Mating Operators": [[19, "summary-of-mating-operators"]], "Summary of classes in the pybrops.breed.op.mate module": [[19, "id1"]], "Mating Operator Properties": [[19, "mating-operator-properties"]], "Defining Mating Operators": [[19, "defining-mating-operators"]], "Creating Mating Operators": [[19, "creating-mating-operators"]], "Mating Individiuals in a Breeding Program": [[19, "mating-individiuals-in-a-breeding-program"]], "Summary of Parent Selection Operators": [[20, "summary-of-parent-selection-operators"]], "Summary of classes in the pybrops.breed.op.psel module": [[20, "id1"]], "Parent Selection Operator Properties": [[20, "parent-selection-operator-properties"]], "Defining Parent Selection Operators": [[20, "defining-parent-selection-operators"]], "Creating Parent Selection Operators": [[20, "creating-parent-selection-operators"]], "Parental Selection for a Breeding Program": [[20, "parental-selection-for-a-breeding-program"]], "Summary of Survivor Selection Operators": [[21, "summary-of-survivor-selection-operators"]], "Summary of classes in the pybrops.breed.op.ssel module": [[21, "id1"]], "Survivor Selection Operator Properties": [[21, "survivor-selection-operator-properties"]], "Defining Survivor Selection Operators": [[21, "defining-survivor-selection-operators"]], "Creating Survivor Selection Operators": [[21, "creating-survivor-selection-operators"]], "Survivor Selection for a Breeding Program": [[21, "survivor-selection-for-a-breeding-program"]], "Summary of Breeding Value Protocol Classes": [[22, "summary-of-breeding-value-protocol-classes"]], "Summary of classes in the pybrops.breed.prot.bv module": [[22, "id1"]], "Breeding Value Protocol Properties": [[22, "breeding-value-protocol-properties"]], "Creating Breeding Value Estimation Protocols": [[22, "creating-breeding-value-estimation-protocols"]], "Estimating Breeding Values": [[22, "estimating-breeding-values"]], "Summary of Genotyping Protocol Classes": [[23, "summary-of-genotyping-protocol-classes"]], "Summary of classes in the pybrops.breed.prot.gt module": [[23, "id1"]], "Genotyping Protocol Properties": [[23, "genotyping-protocol-properties"]], "Creating Genotyping Protocols": [[23, "creating-genotyping-protocols"]], "Genotyping Individuals from their Genomes": [[23, "genotyping-individuals-from-their-genomes"]], "Summary of Mating Protocol Classes": [[24, "summary-of-mating-protocol-classes"]], "Summary of classes in the pybrops.breed.prot.mate module": [[24, "id1"]], "Mating Protocol Properties": [[24, "mating-protocol-properties"]], "General properties": [[24, "general-properties"], [26, "general-properties"], [30, "general-properties"], [35, "general-properties"], [38, "general-properties"], [39, "general-properties"]], "Summary of MatingProtocol general properties": [[24, "id2"]], "Creating Mating Protocols": [[24, "creating-mating-protocols"]], "Simulating Mating and Progeny Generation": [[24, "simulating-mating-and-progeny-generation"]], "Summary of Phenotyping Protocol Classes": [[25, "summary-of-phenotyping-protocol-classes"]], "Summary of classes in the pybrops.breed.prot.pt module": [[25, "id1"]], "Phenotyping Protocol Properties": [[25, "phenotyping-protocol-properties"]], "Genomic model properties": [[25, "genomic-model-properties"]], "Summary of PhenotypingProtocol genomic model properties": [[25, "id2"]], "Error variance properties": [[25, "error-variance-properties"]], "Summary of PhenotypingProtocol error variance properties": [[25, "id3"]], "Creating Phenotyping Protocols": [[25, "creating-phenotyping-protocols"]], "Setting Trait Heritabilities": [[25, "setting-trait-heritabilities"]], "Phenotyping Individuals": [[25, "phenotyping-individuals"]], "Summary of Selection Protocol Classes": [[26, "summary-of-selection-protocol-classes"]], "Abstract selection protocol classes": [[26, "abstract-selection-protocol-classes"]], "Summary of abstract classes in the pybrops.breed.prot.sel module": [[26, "id1"]], "Concrete selection protocol classes": [[26, "concrete-selection-protocol-classes"]], "Summary of concrete classes in the pybrops.breed.prot.sel module": [[26, "id2"]], "Abstract interface classes": [[26, "abstract-interface-classes"]], "Concrete implementation classes": [[26, "concrete-implementation-classes"]], "Selection Protocol Properties": [[26, "selection-protocol-properties"]], "Summary of SelectionProtocol general properties": [[26, "id3"]], "Optimization problem properties": [[26, "optimization-problem-properties"]], "Summary of SelectionProtocol optimization problem properties": [[26, "id4"]], "Optimization algorithm properties": [[26, "optimization-algorithm-properties"]], "Summary of SelectionProtocol optimization algorithm properties": [[26, "id5"]], "Creating Selection Protocol Classes": [[26, "creating-selection-protocol-classes"]], "Generating Selection Problems for Optimization": [[26, "generating-selection-problems-for-optimization"]], "Setup for unconstrained optimization": [[26, "setup-for-unconstrained-optimization"]], "Setup for constrained optimization": [[26, "setup-for-constrained-optimization"]], "Generating the selection problem": [[26, "generating-the-selection-problem"]], "Single-Objective Optimization": [[26, "single-objective-optimization"]], "Multi-Objective Optimization": [[26, "multi-objective-optimization"]], "Selection": [[26, "selection"]], "Seeding PyBrOpS": [[27, "seeding-pybrops"]], "Creating Multiple Pseudorandom Number Generators": [[27, "creating-multiple-pseudorandom-number-generators"]], "Examples": [[28, "examples"]], "Doubled haploid example": [[28, "doubled-haploid-example"]], "Summary of Genomic Model Classes": [[29, "summary-of-genomic-model-classes"]], "Summary of abstract classes in the pybrops.model.gmod module": [[29, "id1"]], "Summary of concrete classes in the pybrops.model.gmod module": [[29, "id2"]], "Genomic Model Properties": [[29, "genomic-model-properties"]], "Genomic Models: General properties": [[29, "genomic-models-general-properties"]], "Summary of GenomicModel general properties": [[29, "id3"]], "Genomic Models: Model properties": [[29, "genomic-models-model-properties"]], "Summary of GenomicModel model properties": [[29, "id4"]], "Genomic Models: Trait properties": [[29, "genomic-models-trait-properties"]], "Summary of GenomicModel trait properties": [[29, "id5"]], "Linear Genomic Models: Additional model properties": [[29, "linear-genomic-models-additional-model-properties"]], "Summary of LinearGenomicModel additional model properties": [[29, "id6"]], "Additive Linear Genomic Models: Additional model properties": [[29, "additive-linear-genomic-models-additional-model-properties"]], "Summary of AdditiveLinearGenomicModel additional model properties": [[29, "id7"]], "Additive Dominance Linear Genomic Models: Additional model properties": [[29, "additive-dominance-linear-genomic-models-additional-model-properties"]], "Summary of AdditiveDominanceLinearGenomicModel additional model properties": [[29, "id8"]], "Additive Dominance Epistatic Linear Genomic Models: Model coefficient properties": [[29, "additive-dominance-epistatic-linear-genomic-models-model-coefficient-properties"]], "Summary of AdditiveDominanceEpistaticLinearGenomicModel model coefficient properties": [[29, "id9"]], "Loading Genomic Model Modules": [[29, "loading-genomic-model-modules"]], "Creating Genomic Models": [[29, "creating-genomic-models"]], "Creating genomic models from raw NumPy arrays": [[29, "creating-genomic-models-from-raw-numpy-arrays"]], "Loading linear genomic models from dictionaries of Pandas DataFrames": [[29, "loading-linear-genomic-models-from-dictionaries-of-pandas-dataframes"]], "Loading linear genomic models from dictionaries of CSV file names": [[29, "loading-linear-genomic-models-from-dictionaries-of-csv-file-names"]], "Loading genomic models HDF5 files": [[29, "loading-genomic-models-hdf5-files"]], "Copying Genomic Models": [[29, "copying-genomic-models"]], "Shallow copying": [[29, "shallow-copying"], [30, "shallow-copying"], [34, "shallow-copying"], [35, "shallow-copying"], [37, "shallow-copying"], [38, "shallow-copying"]], "Deep copying": [[29, "deep-copying"], [30, "deep-copying"], [34, "deep-copying"], [35, "deep-copying"], [37, "deep-copying"], [38, "deep-copying"]], "Model prediction methods": [[29, "model-prediction-methods"]], "Score model prediction accuracy": [[29, "score-model-prediction-accuracy"]], "Predicting genomic estimated breeding values": [[29, "predicting-genomic-estimated-breeding-values"]], "Predicting genomic estimated genotypic values": [[29, "predicting-genomic-estimated-genotypic-values"]], "Calculating population genetic variance terms": [[29, "calculating-population-genetic-variance-terms"]], "Predicting genetic variance": [[29, "predicting-genetic-variance"]], "Predicting additive genetic variance": [[29, "predicting-additive-genetic-variance"]], "Predicting additive genic variance": [[29, "predicting-additive-genic-variance"]], "Predicting the Bulmer effect": [[29, "predicting-the-bulmer-effect"]], "Calculating population selection limits": [[29, "calculating-population-selection-limits"]], "Upper selection limit": [[29, "upper-selection-limit"]], "Lower selection limit": [[29, "lower-selection-limit"]], "Calculating favorable allele metrics": [[29, "calculating-favorable-allele-metrics"]], "Calculating deleterious allele metrics": [[29, "calculating-deleterious-allele-metrics"]], "Exporting Genomic Models": [[29, "exporting-genomic-models"]], "Exporting to dictionaries of Pandas DataFrames": [[29, "exporting-to-dictionaries-of-pandas-dataframes"]], "Exporting to CSV files": [[29, "exporting-to-csv-files"]], "Exporting to HDF5": [[29, "exporting-to-hdf5"], [30, "exporting-to-hdf5"], [34, "exporting-to-hdf5"], [35, "exporting-to-hdf5"], [38, "exporting-to-hdf5"]], "Summary of Variance Matrix Classes": [[30, "summary-of-variance-matrix-classes"]], "Genetic variance matrix classes": [[30, "genetic-variance-matrix-classes"]], "Summary of genetic variance classes in the pybrops.model.vmat module": [[30, "id1"]], "Genic variance matrix classes": [[30, "genic-variance-matrix-classes"]], "Summary of genic variance classes in the pybrops.model.vmat module": [[30, "id2"]], "Variance Matrix Properties": [[30, "variance-matrix-properties"]], "Summary of GeneticVarianceMatrix and GenicVarianceMatrix general properties": [[30, "id3"]], "Taxa properties": [[30, "taxa-properties"], [35, "taxa-properties"], [38, "taxa-properties"], [39, "taxa-properties"]], "Summary of GeneticVarianceMatrix and GenicVarianceMatrix taxa properties": [[30, "id4"]], "Trait properties": [[30, "trait-properties"]], "Summary of GeneticVarianceMatrix and GenicVarianceMatrix trait properties": [[30, "id5"]], "Square matrix properties": [[30, "square-matrix-properties"], [35, "square-matrix-properties"]], "Summary of GeneticVarianceMatrix and GenicVarianceMatrix square matrix properties": [[30, "id6"]], "Loading Variance Matrix Modules": [[30, "loading-variance-matrix-modules"]], "Loading genetic variance matrix modules": [[30, "loading-genetic-variance-matrix-modules"]], "Loading genic variance matrix modules": [[30, "loading-genic-variance-matrix-modules"]], "Creating Variance Matrices": [[30, "creating-variance-matrices"]], "Creating variance matrices from NumPy arrays": [[30, "creating-variance-matrices-from-numpy-arrays"]], "Calculating variance matrices from genomic models": [[30, "calculating-variance-matrices-from-genomic-models"]], "Creating variance matrices from Pandas DataFrames": [[30, "creating-variance-matrices-from-pandas-dataframes"]], "Loading variance matrices from CSV files": [[30, "loading-variance-matrices-from-csv-files"]], "Loading variance matrices from HDF5 files": [[30, "loading-variance-matrices-from-hdf5-files"]], "Copying Variance Matrices": [[30, "copying-variance-matrices"]], "Copy-On Element Manipulation": [[30, "copy-on-element-manipulation"], [34, "copy-on-element-manipulation"], [35, "copy-on-element-manipulation"], [38, "copy-on-element-manipulation"]], "Adjoining elements": [[30, "adjoining-elements"], [39, "adjoining-elements"]], "Deleting elements": [[30, "deleting-elements"], [39, "deleting-elements"]], "delete taxa": [[30, "delete-taxa"], [39, "delete-taxa"]], "delete traits": [[30, "delete-traits"]], "Inserting elements": [[30, "inserting-elements"], [39, "inserting-elements"]], "Selecting elements": [[30, "selecting-elements"], [39, "selecting-elements"]], "In-Place Element Manipulation": [[30, "in-place-element-manipulation"], [34, "in-place-element-manipulation"], [35, "in-place-element-manipulation"], [38, "in-place-element-manipulation"]], "Appending elements": [[30, "appending-elements"], [39, "appending-elements"]], "Removing elements": [[30, "removing-elements"], [39, "removing-elements"]], "remove taxa": [[30, "remove-taxa"], [39, "remove-taxa"]], "remove traits": [[30, "remove-traits"]], "Incorporating elements": [[30, "incorporating-elements"], [39, "incorporating-elements"]], "Concatenating matrices": [[30, "concatenating-matrices"], [39, "concatenating-matrices"]], "Grouping and Sorting": [[30, "grouping-and-sorting"], [34, "grouping-and-sorting"], [35, "grouping-and-sorting"], [38, "grouping-and-sorting"], [39, "grouping-and-sorting"]], "Reordering elements": [[30, "reordering-elements"], [34, "reordering-elements"], [35, "reordering-elements"], [38, "reordering-elements"]], "reorder taxa": [[30, "reorder-taxa"], [39, "reorder-taxa"]], "reorder traits": [[30, "reorder-traits"]], "Lexsorting elements": [[30, "lexsorting-elements"], [34, "lexsorting-elements"], [35, "lexsorting-elements"], [38, "lexsorting-elements"]], "lexsort taxa": [[30, "lexsort-taxa"], [39, "lexsort-taxa"]], "lexsort traits": [[30, "lexsort-traits"]], "Sorting elements": [[30, "sorting-elements"], [34, "sorting-elements"], [35, "sorting-elements"], [38, "sorting-elements"]], "sort taxa": [[30, "sort-taxa"], [39, "sort-taxa"]], "sort traits": [[30, "sort-traits"]], "Grouping elements": [[30, "grouping-elements"], [34, "grouping-elements"], [35, "grouping-elements"], [38, "grouping-elements"]], "Square matrix functions": [[30, "square-matrix-functions"]], "Determine whether all square axes are of equal length": [[30, "determine-whether-all-square-axes-are-of-equal-length"]], "Exporting Variance Matrices": [[30, "exporting-variance-matrices"]], "Exporting to Pandas DataFrame": [[30, "exporting-to-pandas-dataframe"], [34, "exporting-to-pandas-dataframe"], [35, "exporting-to-pandas-dataframe"]], "Exporting to CSV": [[30, "exporting-to-csv"], [34, "exporting-to-csv"], [35, "exporting-to-csv"]], "Summary of Optimization Algorithm Classes": [[31, "summary-of-optimization-algorithm-classes"]], "Summary of classes in the pybrops.opt.algo module": [[31, "id1"]], "Loading Optimization Algorithm Modules": [[31, "loading-optimization-algorithm-modules"]], "Constructing Optimization Algorithms": [[31, "constructing-optimization-algorithms"]], "Single-objective optimization algorithms": [[31, "single-objective-optimization-algorithms"]], "Multi-objective optimization algorithms": [[31, "multi-objective-optimization-algorithms"]], "Minimizing Optimization Problems": [[31, "minimizing-optimization-problems"]], "Optimization of single-objective problems": [[31, "optimization-of-single-objective-problems"]], "Optimization of multi-objective problems": [[31, "optimization-of-multi-objective-problems"]], "Summary of Optimization Problem Classes": [[32, "summary-of-optimization-problem-classes"]], "Summary of classes in the pybrops.opt.prob module": [[32, "id1"]], "Loading Optimization Problem Modules": [[32, "loading-optimization-problem-modules"]], "Optimization Problem Properties": [[32, "optimization-problem-properties"], [33, "optimization-problem-properties"]], "PyMOO specific properties": [[32, "pymoo-specific-properties"]], "PyBrOpS specific properties": [[32, "pybrops-specific-properties"]], "Summary of PyBrOpS specific Problem properties": [[32, "id2"]], "Deriving Problem Classes": [[32, "deriving-problem-classes"]], "Single objective problem specification": [[32, "single-objective-problem-specification"]], "Multi-objective problem specification": [[32, "multi-objective-problem-specification"]], "Constructing Problems": [[32, "constructing-problems"]], "Construct a single-objective problem": [[32, "construct-a-single-objective-problem"]], "Construct a multi-objective problem": [[32, "construct-a-multi-objective-problem"]], "Evaluating candidate solutions": [[32, "evaluating-candidate-solutions"]], "Summary of Optimization Solution Classes": [[33, "summary-of-optimization-solution-classes"]], "Summary of classes in the pybrops.opt.soln module": [[33, "id1"]], "Loading Optimization Solution Modules": [[33, "loading-optimization-solution-modules"]], "Constructing Solutions": [[33, "constructing-solutions"]], "Construct a single-objective solution": [[33, "construct-a-single-objective-solution"]], "Construct a multi-objective solution": [[33, "construct-a-multi-objective-solution"]], "Summary of Solution properties": [[33, "id2"]], "Summary of Breeding Value Matrix Classes": [[34, "summary-of-breeding-value-matrix-classes"]], "Summary of breeding value matrix classes in the pybrops.popgen.bvmat module": [[34, "id1"]], "Breeding Value Matrix Properties": [[34, "breeding-value-matrix-properties"]], "Breeding value matrix general properties": [[34, "breeding-value-matrix-general-properties"]], "Summary of BreedingValueMatrix general properties": [[34, "id2"]], "Breeding value matrix taxa properties": [[34, "breeding-value-matrix-taxa-properties"]], "Summary of BreedingValueMatrix taxa properties": [[34, "id3"]], "Breeding value matrix trait properties": [[34, "breeding-value-matrix-trait-properties"]], "Summary of BreedingValueMatrix trait properties": [[34, "id4"]], "Loading Breeding Value Matrix Modules": [[34, "loading-breeding-value-matrix-modules"]], "Creating Breeding Value Matrices": [[34, "creating-breeding-value-matrices"]], "Creating breeding value matrices from NumPy arrays": [[34, "creating-breeding-value-matrices-from-numpy-arrays"]], "Creating breeding value matrices from Pandas DataFrames": [[34, "creating-breeding-value-matrices-from-pandas-dataframes"]], "Loading breeding value matrices from CSV files": [[34, "loading-breeding-value-matrices-from-csv-files"]], "Loading breeding value matrices from HDF5 files": [[34, "loading-breeding-value-matrices-from-hdf5-files"]], "Copying Breeding Value Matrices": [[34, "copying-breeding-value-matrices"]], "Adjoin elements": [[34, "adjoin-elements"], [35, "adjoin-elements"], [38, "adjoin-elements"]], "Delete elements": [[34, "delete-elements"], [35, "delete-elements"], [38, "delete-elements"]], "Insert elements": [[34, "insert-elements"], [35, "insert-elements"], [38, "insert-elements"]], "Select elements": [[34, "select-elements"], [35, "select-elements"], [38, "select-elements"]], "Append elements": [[34, "append-elements"], [35, "append-elements"], [38, "append-elements"]], "Remove elements": [[34, "remove-elements"], [35, "remove-elements"], [38, "remove-elements"]], "Incorporate elements": [[34, "incorporate-elements"], [35, "incorporate-elements"], [38, "incorporate-elements"]], "Concatenate elements": [[34, "concatenate-elements"], [35, "concatenate-elements"], [38, "concatenate-elements"]], "Summary Statistics": [[34, "summary-statistics"], [35, "summary-statistics"], [38, "summary-statistics"], [39, "summary-statistics"]], "Maximum breeding values for each trait": [[34, "maximum-breeding-values-for-each-trait"]], "Minimum breeding values for each trait": [[34, "minimum-breeding-values-for-each-trait"]], "Mean breeding values for each trait": [[34, "mean-breeding-values-for-each-trait"]], "Breeding value ranges for each trait": [[34, "breeding-value-ranges-for-each-trait"]], "Breeding value standard deviations for each trait": [[34, "breeding-value-standard-deviations-for-each-trait"]], "Breeding value variances for each trait": [[34, "breeding-value-variances-for-each-trait"]], "Unscaling and de-centering breeding values": [[34, "unscaling-and-de-centering-breeding-values"]], "Exporting Breeding Value Matrices": [[34, "exporting-breeding-value-matrices"]], "Summary of Coancestry Matrix Classes": [[35, "summary-of-coancestry-matrix-classes"]], "Summary of classes in the pybrops.popgen.cmat module": [[35, "id1"]], "Coancestry Matrix Properties": [[35, "coancestry-matrix-properties"]], "Summary of CoancestryMatrix general properties": [[35, "id2"]], "Summary of CoancestryMatrix taxa properties": [[35, "id3"]], "Summary of CoancestryMatrix square matrix properties": [[35, "id4"]], "Loading Coancestry Matrix Modules": [[35, "loading-coancestry-matrix-modules"]], "Creating Coancestry Matrices": [[35, "creating-coancestry-matrices"]], "Creating coancestry matrices from NumPy arrays": [[35, "creating-coancestry-matrices-from-numpy-arrays"]], "Creating coancestry matrices from GenotypeMatrix objects": [[35, "creating-coancestry-matrices-from-genotypematrix-objects"]], "Creating coancestry matrices from Pandas DataFrames": [[35, "creating-coancestry-matrices-from-pandas-dataframes"]], "Loading coancestry matrices from CSV files": [[35, "loading-coancestry-matrices-from-csv-files"]], "Loading coancestry matrices from HDF5 files": [[35, "loading-coancestry-matrices-from-hdf5-files"]], "Copying Coancestry Matrices": [[35, "copying-coancestry-matrices"]], "Coancestry and Kinship Methods": [[35, "coancestry-and-kinship-methods"]], "Retrieving coancestry values": [[35, "retrieving-coancestry-values"]], "Retrieving kinship values": [[35, "retrieving-kinship-values"]], "Retrieving the coancestry matrix as a specific format": [[35, "retrieving-the-coancestry-matrix-as-a-specific-format"]], "Determining if the coancestry matrix is positive semidefinite": [[35, "determining-if-the-coancestry-matrix-is-positive-semidefinite"]], "Applying jitter values along the diagonal": [[35, "applying-jitter-values-along-the-diagonal"]], "Calculating the matrix inverse": [[35, "calculating-the-matrix-inverse"]], "Calculating maximum attainable inbreeding": [[35, "calculating-maximum-attainable-inbreeding"]], "Calculating minimum attainable inbreeding": [[35, "calculating-minimum-attainable-inbreeding"]], "Maximum coancestry": [[35, "maximum-coancestry"]], "Mean coancestry": [[35, "mean-coancestry"]], "Minimum coancestry": [[35, "minimum-coancestry"]], "Exporting Coancestry Matrices": [[35, "exporting-coancestry-matrices"]], "Summary of Genetic Map Function Classes": [[36, "summary-of-genetic-map-function-classes"]], "Summary of genetic map function classes in the pybrops.popgen.gmap module": [[36, "id1"]], "Loading Genetic Map Function Modules": [[36, "loading-genetic-map-function-modules"]], "Constructing Genetic Map Function Objects": [[36, "constructing-genetic-map-function-objects"]], "Calculating Recombination Probabilities": [[36, "calculating-recombination-probabilities"]], "Calculating Genetic Map Distances": [[36, "calculating-genetic-map-distances"]], "Calculating Sequential and Pairwise Recombination Probabilities": [[36, "calculating-sequential-and-pairwise-recombination-probabilities"]], "Calculating sequential recombination probabilities": [[36, "calculating-sequential-recombination-probabilities"]], "Calculating pairwise recombination probabilities": [[36, "calculating-pairwise-recombination-probabilities"]], "Summary of Genetic Map Classes": [[37, "summary-of-genetic-map-classes"]], "Summary of genetic map classes in the pybrops.popgen.gmap module": [[37, "id1"]], "Loading Genetic Map Modules": [[37, "loading-genetic-map-modules"]], "Creating Genetic Maps": [[37, "creating-genetic-maps"]], "Creating genetic maps from NumPy arrays": [[37, "creating-genetic-maps-from-numpy-arrays"]], "Reading genetic maps from a file": [[37, "reading-genetic-maps-from-a-file"]], "Genetic Map Properties": [[37, "genetic-map-properties"]], "Marker variant properties": [[37, "marker-variant-properties"], [38, "marker-variant-properties"], [39, "marker-variant-properties"]], "Summary of GeneticMap marker variant properties": [[37, "id2"]], "Spline properties": [[37, "spline-properties"]], "Summary of GeneticMap spline properties": [[37, "id3"]], "Copying Genetic Maps": [[37, "copying-genetic-maps"]], "Sorting and Grouping Genetic Maps": [[37, "sorting-and-grouping-genetic-maps"]], "Reordering map elements": [[37, "reordering-map-elements"]], "Lexsorting map elements": [[37, "lexsorting-map-elements"]], "Sorting map elements": [[37, "sorting-map-elements"]], "Grouping map elements": [[37, "grouping-map-elements"]], "Genetic Map Congruency": [[37, "genetic-map-congruency"]], "Checking for congruency": [[37, "checking-for-congruency"]], "Removing map discrepancies": [[37, "removing-map-discrepancies"]], "Building Interpolation Splines": [[37, "building-interpolation-splines"]], "Interpolating Genetic Positions": [[37, "interpolating-genetic-positions"]], "Summary of Genotype Matrix Classes": [[38, "summary-of-genotype-matrix-classes"]], "Summary of genotype matrix classes in the pybrops.popgen.gmat module": [[38, "id1"]], "Creating Genotype Matrices": [[38, "creating-genotype-matrices"]], "Creating genotype matrices from NumPy arrays": [[38, "creating-genotype-matrices-from-numpy-arrays"]], "Loading genotype matrices from VCF files": [[38, "loading-genotype-matrices-from-vcf-files"]], "Loading genotype matrices from HDF5 files": [[38, "loading-genotype-matrices-from-hdf5-files"]], "Genotype Matrix Properties": [[38, "genotype-matrix-properties"]], "Summary of GenotypeMatrix general properties": [[38, "id2"]], "Summary of GenotypeMatrix taxa properties": [[38, "id3"]], "Summary of GenotypeMatrix marker variant properties": [[38, "id4"]], "Copying Genotype Matrices": [[38, "copying-genotype-matrices"]], "Population allele counts": [[38, "population-allele-counts"]], "Population allele frequencies": [[38, "population-allele-frequencies"]], "Population allele polymorphism presence": [[38, "population-allele-polymorphism-presence"]], "Population genotype counts": [[38, "population-genotype-counts"]], "Population genotype frequencies": [[38, "population-genotype-frequencies"]], "Population minor allele frequencies": [[38, "population-minor-allele-frequencies"]], "Population mean expected heterozygosity": [[38, "population-mean-expected-heterozygosity"]], "Taxa allele counts": [[38, "taxa-allele-counts"]], "Taxa allele frequencies": [[38, "taxa-allele-frequencies"]], "Saving Genotype Matrices": [[38, "saving-genotype-matrices"], [39, "saving-genotype-matrices"]], "Summary of Phased Genotype Matrix Classes": [[39, "summary-of-phased-genotype-matrix-classes"]], "Summary of phased genotype matrix classes in the pybrops.popgen.gmat module": [[39, "id1"]], "Creating Phased Genotype Matrices": [[39, "creating-phased-genotype-matrices"]], "Creating phased genotype matrices from NumPy arrays": [[39, "creating-phased-genotype-matrices-from-numpy-arrays"]], "Loading phased genotype matrices from VCF files": [[39, "loading-phased-genotype-matrices-from-vcf-files"]], "Loading phased genotype matrices from HDF5 files": [[39, "loading-phased-genotype-matrices-from-hdf5-files"]], "Phased Genotype Matrix Properties": [[39, "phased-genotype-matrix-properties"]], "Summary of PhasedGenotypeMatrix general properties": [[39, "id2"]], "Phase properties": [[39, "phase-properties"]], "Summary of PhasedGenotypeMatrix phase properties": [[39, "id3"]], "Summary of PhasedGenotypeMatrix taxa properties": [[39, "id4"]], "Summary of PhasedGenotypeMatrix marker variant properties": [[39, "id5"]], "Copying Phased Genotype Matrices": [[39, "copying-phased-genotype-matrices"]], "Phased Genotype Matrix Element Copy-On-Manipulation": [[39, "phased-genotype-matrix-element-copy-on-manipulation"]], "delete marker variants": [[39, "delete-marker-variants"]], "Phased Genotype Matrix Element In-Place-Manipulation": [[39, "phased-genotype-matrix-element-in-place-manipulation"]], "remove marker variants": [[39, "remove-marker-variants"]], "Reordering": [[39, "reordering"]], "reorder marker variants": [[39, "reorder-marker-variants"]], "Lexsorting": [[39, "lexsorting"]], "lexsort marker variants": [[39, "lexsort-marker-variants"]], "Sorting": [[39, "sorting"]], "sort marker variants": [[39, "sort-marker-variants"]], "Grouping": [[39, "grouping"]], "group taxa": [[39, "group-taxa"]], "group marker variants": [[39, "group-marker-variants"]], "Write to HDF5": [[39, "write-to-hdf5"]], "Standard PyBrOpS Usage": [[40, "standard-pybrops-usage"]], "Core Infrastructure Usage:": [[40, null]], "Population Genetics Modules Usage:": [[40, null]], "Optimization Modules Usage:": [[40, null]], "Genomic Modeling Modules Usage:": [[40, null]], "Breeding Protocol Modules Usage:": [[40, null]], "Breeding Operator Modules Usage:": [[40, null]], "Breeding Program Simulation Examples:": [[40, null]]}, "indexentries": {}})