
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pybrops.core.util.haplo &#8212; pybrops 1.0.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/wider_borders.css?v=ec38ee01" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../../_static/documentation_options.js?v=292eb321"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pybrops/core/util/haplo';</script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pybrops 1.0.2 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../getting_started/getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../fundamentals/fundamentals.html">
                        Fundamentals
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/standard_usage.html">
                        Standard Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../advanced_usage/advanced_usage.html">
                        Advanced Usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../../standard_usage/examples/examples.html">
                        Examples
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../contributing/contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../faq/faq.html">
                        Frequently Asked Questions
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../_autosummary/pybrops.html">
                        API reference
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pybrops.core...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pybrops.core.util.haplo</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module containing haplotype calculation utility functions.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Integral</span>
<span class="kn">import</span> <span class="nn">numpy</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;nhaploblk_chrom&quot;</span><span class="p">,</span> 
    <span class="s2">&quot;haplobin&quot;</span><span class="p">,</span> 
    <span class="s2">&quot;haplobin_bounds&quot;</span><span class="p">,</span>
    <span class="s2">&quot;haplomat&quot;</span><span class="p">,</span>
<span class="p">]</span>

<div class="viewcode-block" id="nhaploblk_chrom">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.core.util.haplo.nhaploblk_chrom.html#pybrops.core.util.haplo.nhaploblk_chrom">[docs]</a>
<span class="k">def</span> <span class="nf">nhaploblk_chrom</span><span class="p">(</span>
        <span class="n">nhaploblk</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
        <span class="n">genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
        <span class="n">chrgrp_stix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
        <span class="n">chrgrp_spix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a total number of haplotype blocks to assign across the genome,</span>
<span class="sd">    determine the number of haplotype blocks to give to each chromosome.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    nhaploblk : int</span>
<span class="sd">        Total number of haplotype blocks to assign across the genome.</span>
<span class="sd">    genpos : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,)`` containing genetic positions of markers across</span>
<span class="sd">        the genome.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>

<span class="sd">        Input contraints:</span>

<span class="sd">        - Must be sorted in ascending order.</span>
<span class="sd">        - Must be grouped based on chromosome. The chromosome boundary start</span>
<span class="sd">          and stop indices must be provided in ``chrgrp_stix`` and</span>
<span class="sd">          ``chrgrp_spix``, respectively.</span>
<span class="sd">    chrgrp_stix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary start indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    chrgrp_spix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary stop indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : numpy.ndarray</span>
<span class="sd">        Array of shape ``(c,)`` containing the number of haplotype blocks</span>
<span class="sd">        assigned to each chromosome.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># get number of chromosomes</span>
    <span class="n">nchr</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chrgrp_stix</span><span class="p">)</span>

    <span class="c1"># make sure that number of haplotype blocks &gt;= number of chromosomes</span>
    <span class="k">if</span> <span class="n">nhaploblk</span> <span class="o">&lt;</span> <span class="n">nchr</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Number of haplotype blocks (nhaploblk = </span><span class="si">{0}</span><span class="s2">) &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nhaploblk</span><span class="p">)</span> <span class="o">+</span>
            <span class="s2">&quot;is less than the number of chromosomes (nchr = </span><span class="si">{1}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nchr</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># calculate genetic lengths of each chromosome in Morgans/cM/other</span>
    <span class="n">genlen</span> <span class="o">=</span> <span class="n">genpos</span><span class="p">[</span><span class="n">chrgrp_spix</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">genpos</span><span class="p">[</span><span class="n">chrgrp_stix</span><span class="p">]</span>

    <span class="c1"># calculate ideal number of markers per chromosome</span>
    <span class="n">nhaploblk_ideal</span> <span class="o">=</span> <span class="p">(</span><span class="n">nhaploblk</span> <span class="o">/</span> <span class="n">genlen</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">*</span> <span class="n">genlen</span>

    <span class="c1"># calculate number of chromosome markers, assuming at least one per chromosome</span>
    <span class="n">nhaploblk_chrom</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">nchr</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int&quot;</span><span class="p">)</span>    <span class="c1"># start with min of one</span>

    <span class="c1"># greedily increment number of haplotype blocks assigned to chromosomes</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nhaploblk</span> <span class="o">-</span> <span class="n">nchr</span><span class="p">):</span>               <span class="c1"># forces conformance to self.nhaploblk</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">nhaploblk_chrom</span> <span class="o">-</span> <span class="n">nhaploblk_ideal</span>    <span class="c1"># take actual - ideal</span>
        <span class="n">ix</span> <span class="o">=</span> <span class="n">diff</span><span class="o">.</span><span class="n">argmin</span><span class="p">()</span>                          <span class="c1"># get index of lowest difference</span>
        <span class="n">nhaploblk_chrom</span><span class="p">[</span><span class="n">ix</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>                    <span class="c1"># increment at lowest index</span>

    <span class="k">return</span> <span class="n">nhaploblk_chrom</span></div>


<div class="viewcode-block" id="haplobin">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.core.util.haplo.haplobin.html#pybrops.core.util.haplo.haplobin">[docs]</a>
<span class="k">def</span> <span class="nf">haplobin</span><span class="p">(</span>
        <span class="n">nhaploblk_chrom</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
        <span class="n">genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
        <span class="n">chrgrp_stix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
        <span class="n">chrgrp_spix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the number of haplotype blocks to give to each chromosome across the</span>
<span class="sd">    genome, assign bins for each marker</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    nhaploblk_chrom : numpy.ndarray</span>
<span class="sd">        Array of shape ``(c,)`` containing the total number of haplotype blocks</span>
<span class="sd">        to assign to each chromosome across the genome.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    genpos : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,)`` containing genetic positions of markers across</span>
<span class="sd">        the genome.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>

<span class="sd">        Input contraints:</span>

<span class="sd">        - Must be sorted in ascending order.</span>
<span class="sd">        - Must be grouped based on chromosome. The chromosome boundary start</span>
<span class="sd">          and stop indices must be provided in ``chrgrp_stix`` and</span>
<span class="sd">          ``chrgrp_spix``, respectively.</span>
<span class="sd">    chrgrp_stix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary start indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    chrgrp_spix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary stop indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,)`` containing haplotype bin assignments for each</span>
<span class="sd">        marker across the genome. Bins are assigned starting at ``0`` and</span>
<span class="sd">        strictly increase as the index in the array increases due to constraints</span>
<span class="sd">        placed on ``genpos``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>

<span class="sd">        Example output::</span>

<span class="sd">            [0,0,0,0,0,1,1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># initialize output array</span>
    <span class="n">haplobin</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span>     <span class="c1"># create empty array</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">genpos</span><span class="p">),</span>            <span class="c1"># same length as number of markers</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="s2">&quot;int&quot;</span>           <span class="c1"># integer array</span>
    <span class="p">)</span>

    <span class="c1"># get number of chromosomes</span>
    <span class="n">nchr</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chrgrp_stix</span><span class="p">)</span>

    <span class="n">k</span> <span class="o">=</span> <span class="mi">0</span>                           <span class="c1"># haplotype block counter</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nchr</span><span class="p">):</span>           <span class="c1"># for each chromosome</span>
        <span class="n">nhap</span> <span class="o">=</span> <span class="n">nhaploblk_chrom</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>   <span class="c1"># get number of haplotype blocks on chromosome</span>
        <span class="n">stix</span> <span class="o">=</span> <span class="n">chrgrp_stix</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>       <span class="c1"># get chromosome start index</span>
        <span class="n">spix</span> <span class="o">=</span> <span class="n">chrgrp_spix</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>       <span class="c1"># get chromosome stop index</span>
        <span class="n">hbound</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span>    <span class="c1"># calculate haplotype boundaries</span>
            <span class="n">genpos</span><span class="p">[</span><span class="n">stix</span><span class="p">],</span>           <span class="c1"># start genetic position on chromosome</span>
            <span class="n">genpos</span><span class="p">[</span><span class="n">spix</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>         <span class="c1"># stop genetic position on chromosome</span>
            <span class="n">nhap</span><span class="o">+</span><span class="mi">1</span>                  <span class="c1"># nhap + 1 for chromosome tips</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nhap</span><span class="p">):</span>               <span class="c1"># for each haplotype block</span>
            <span class="n">chrmap</span> <span class="o">=</span> <span class="n">genpos</span><span class="p">[</span><span class="n">stix</span><span class="p">:</span><span class="n">spix</span><span class="p">]</span>      <span class="c1"># get chromosome genetic positions</span>
            <span class="n">lmask</span> <span class="o">=</span> <span class="n">chrmap</span> <span class="o">&gt;=</span> <span class="n">hbound</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>     <span class="c1"># select all &gt;= lower haplotype block bound</span>
            <span class="n">umask</span> <span class="o">=</span> <span class="n">chrmap</span> <span class="o">&lt;=</span> <span class="n">hbound</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>   <span class="c1"># select all &lt;= upper haplotype block bound</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="n">lmask</span> <span class="o">&amp;</span> <span class="n">umask</span>            <span class="c1"># select all: lower &lt;= markers &lt;= upper</span>
            <span class="n">haplobin</span><span class="p">[</span><span class="n">stix</span><span class="p">:</span><span class="n">spix</span><span class="p">][</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>   <span class="c1"># set haplotype bin to &#39;k&#39;</span>
            <span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>                          <span class="c1"># increment bin count</span>

    <span class="k">return</span> <span class="n">haplobin</span></div>


<div class="viewcode-block" id="haplobin_bounds">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.core.util.haplo.haplobin_bounds.html#pybrops.core.util.haplo.haplobin_bounds">[docs]</a>
<span class="k">def</span> <span class="nf">haplobin_bounds</span><span class="p">(</span>
        <span class="n">haplobin</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate haplotype bin boundaries and lengths.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    haplobin : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,)`` containing haplotype bin assignments for each</span>
<span class="sd">        marker across the genome. Bins are assigned starting at ``0`` and</span>
<span class="sd">        strictly increase as the index in the array increases.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>

<span class="sd">        Example input::</span>

<span class="sd">            [0,0,0,0,0,1,1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3]</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : tuple</span>
<span class="sd">        A tuple of length 3 containing ``(hstix, hspix, hlen)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``hstix`` is a ``numpy.ndarray`` containing haplotype bin boundary</span>
<span class="sd">          start indices.</span>
<span class="sd">        - ``hspix`` is a ``numpy.ndarray`` containing haplotype bin boundary</span>
<span class="sd">          stop indices.</span>
<span class="sd">        - ``hlen`` is a ``numpy.ndarray`` containing haplotype bin lengths.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hstix</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>                         <span class="c1"># starting indices</span>
    <span class="n">hspix</span> <span class="o">=</span> <span class="p">[]</span>                          <span class="c1"># stopping indices</span>
    <span class="n">prev</span> <span class="o">=</span> <span class="n">haplobin</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>                  <span class="c1"># get first group label</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">haplobin</span><span class="p">)):</span>   <span class="c1"># for each group label</span>
        <span class="k">if</span> <span class="n">haplobin</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">prev</span><span class="p">:</span>         <span class="c1"># if the label is different</span>
            <span class="n">hspix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>             <span class="c1"># add the stop index for prev</span>
            <span class="n">prev</span> <span class="o">=</span> <span class="n">haplobin</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>          <span class="c1"># get next label</span>
            <span class="n">hstix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>             <span class="c1"># add the start index for next prev</span>
    <span class="n">hspix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">haplobin</span><span class="p">))</span>         <span class="c1"># append last stop index</span>
    <span class="n">hstix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int_</span><span class="p">(</span><span class="n">hstix</span><span class="p">)</span>           <span class="c1"># convert to ndarray</span>
    <span class="n">hspix</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int_</span><span class="p">(</span><span class="n">hspix</span><span class="p">)</span>           <span class="c1"># convert to ndarray</span>
    <span class="n">hlen</span> <span class="o">=</span> <span class="n">hspix</span> <span class="o">-</span> <span class="n">hstix</span>                <span class="c1"># get lengths of each haplotype group</span>

    <span class="k">return</span> <span class="n">hstix</span><span class="p">,</span> <span class="n">hspix</span><span class="p">,</span> <span class="n">hlen</span></div>


<div class="viewcode-block" id="haplomat">
<a class="viewcode-back" href="../../../../_autosummary/pybrops.core.util.haplo.haplomat.html#pybrops.core.util.haplo.haplomat">[docs]</a>
<span class="k">def</span> <span class="nf">haplomat</span><span class="p">(</span>
        <span class="n">nhaploblk</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">genomemat</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">genpos</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">chrgrp_stix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">chrgrp_spix</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">chrgrp_len</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">u_a</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate a haplotype matrix from a genome matrix and model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    nhaploblk : Integral</span>
<span class="sd">        Number of haplotype blocks</span>
<span class="sd">    genomemat : numpy.ndarray</span>
<span class="sd">        A genome matrix of shape ``(m,n,p)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``m`` is the number of chromsome phases.</span>
<span class="sd">        - ``n`` is the number of individuals.</span>
<span class="sd">        - ``p`` is the number of markers.</span>
<span class="sd">    genpos : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,)`` containing genetic positions of markers across</span>
<span class="sd">        the genome.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>

<span class="sd">        Input contraints:</span>

<span class="sd">        - Must be sorted in ascending order.</span>
<span class="sd">        - Must be grouped based on chromosome. The chromosome boundary start</span>
<span class="sd">          and stop indices must be provided in ``chrgrp_stix`` and</span>
<span class="sd">          ``chrgrp_spix``, respectively.</span>
<span class="sd">    chrgrp_stix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary start indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    chrgrp_spix : numpy.ndarray</span>
<span class="sd">        Chromosome boundary stop indices array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    chrgrp_len : numpy.ndarray</span>
<span class="sd">        Chromosome length array of shape ``(c,)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``c`` is the number of chromosomes.</span>
<span class="sd">    u_a : numpy.ndarray</span>
<span class="sd">        Array of shape ``(p,t)`` containing additive marker effect estimates.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``p`` is the number of markers.</span>
<span class="sd">        - ``t`` is the number of traits.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    out : numpy.ndarray</span>
<span class="sd">        A haplotype effect matrix of shape ``(m,n,b,t)``.</span>

<span class="sd">        Where:</span>

<span class="sd">        - ``m`` is the number of chromsome phases.</span>
<span class="sd">        - ``n`` is the number of individuals.</span>
<span class="sd">        - ``b`` is the number of haplotype blocks.</span>
<span class="sd">        - ``t`` is the number of traits.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">chrgrp_stix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">chrgrp_spix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;markers are not sorted by chromosome position&quot;</span><span class="p">)</span>

    <span class="c1"># get number of chromosomes</span>
    <span class="n">nchr</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chrgrp_stix</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">nhaploblk</span> <span class="o">&lt;</span> <span class="n">nchr</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;number of haplotype blocks is less than the number of chromosomes&quot;</span><span class="p">)</span>

    <span class="c1"># calculate number of marker blocks to assign to each chromosome</span>
    <span class="n">nblk</span> <span class="o">=</span> <span class="n">nhaploblk_chrom</span><span class="p">(</span><span class="n">nhaploblk</span><span class="p">,</span> <span class="n">genpos</span><span class="p">,</span> <span class="n">chrgrp_stix</span><span class="p">,</span> <span class="n">chrgrp_spix</span><span class="p">)</span>

    <span class="c1"># ensure there are enough markers per chromosome</span>
    <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">nblk</span> <span class="o">&gt;</span> <span class="n">chrgrp_len</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="s2">&quot;number of haplotype blocks assigned to a chromosome greater than number of available markers&quot;</span>
        <span class="p">)</span>

    <span class="c1"># calculate haplotype bins</span>
    <span class="n">hbin</span> <span class="o">=</span> <span class="n">haplobin</span><span class="p">(</span><span class="n">nblk</span><span class="p">,</span> <span class="n">genpos</span><span class="p">,</span> <span class="n">chrgrp_stix</span><span class="p">,</span> <span class="n">chrgrp_spix</span><span class="p">)</span>

    <span class="c1"># define shape</span>
    <span class="n">s</span> <span class="o">=</span> <span class="p">(</span><span class="n">genomemat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">genomemat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">nhaploblk</span><span class="p">,</span> <span class="n">u_a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># (m,n,b,t)</span>

    <span class="c1"># allocate haplotype matrix</span>
    <span class="n">hmat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">u_a</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>   <span class="c1"># (m,n,b,t)</span>

    <span class="c1"># get boundary indices</span>
    <span class="n">hstix</span><span class="p">,</span> <span class="n">hspix</span><span class="p">,</span> <span class="n">hlen</span> <span class="o">=</span> <span class="n">haplobin_bounds</span><span class="p">(</span><span class="n">hbin</span><span class="p">)</span>

    <span class="c1"># OPTIMIZE: perhaps eliminate one loop using dot function</span>
    <span class="c1"># fill haplotype matrix</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">hmat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">3</span><span class="p">]):</span>                          <span class="c1"># for each trait</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,(</span><span class="n">st</span><span class="p">,</span><span class="n">sp</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">hstix</span><span class="p">,</span><span class="n">hspix</span><span class="p">)):</span>       <span class="c1"># for each haplotype block</span>
            <span class="n">hmat</span><span class="p">[:,:,</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">genomemat</span><span class="p">[:,:,</span><span class="n">st</span><span class="p">:</span><span class="n">sp</span><span class="p">]</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">u_a</span><span class="p">[</span><span class="n">st</span><span class="p">:</span><span class="n">sp</span><span class="p">,</span><span class="n">i</span><span class="p">])</span>  <span class="c1"># take dot product and fill</span>

    <span class="k">return</span> <span class="n">hmat</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      Â© Copyright 2024, Robert Z. Shrote.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>